diff --git a/ui/aura/client/drag_drop_client.cc b/ui/aura/client/drag_drop_client.cc
index c3549b665dccd02a084f99ae081e88bd160f6e36..32739a1ae0ba222d0714688a966083ad13d6b3cb 100644
--- a/ui/aura/client/drag_drop_client.cc
+++ b/ui/aura/client/drag_drop_client.cc
@@ -27,5 +27,29 @@ DragDropClient* GetDragDropClient(Window* root_window) {
       root_window->GetProperty(kRootWindowDragDropClientKey) : NULL;
 }
 
+int DragDropClient::StartDragAndDrop(
+    const ui::OSExchangeData& data,
+    aura::Window* root_window,
+    aura::Window* source_window,
+    const gfx::Point& screen_location,
+    int operation,
+    ui::DragDropTypes::DragEventSource source) {
+  bool cancelled;
+  return StartDragAndDrop(data, root_window, source_window, screen_location,
+                          operation, source, cancelled);
+}
+
+int DragDropClient::StartDragAndDrop(const ui::OSExchangeData& data,
+                                     aura::Window* root_window,
+                                     aura::Window* source_window,
+                                     const gfx::Point& screen_location,
+                                     int operation,
+                                     ui::DragDropTypes::DragEventSource source,
+                                     bool& cancelled) {
+  cancelled = false;
+  return StartDragAndDrop(data, root_window, source_window, screen_location,
+                          operation, source);
+}
+
 }  // namespace client
 }  // namespace aura
diff --git a/ui/aura/client/drag_drop_client.h b/ui/aura/client/drag_drop_client.h
index 8e9634807d3fcf3e29d8324b6b8f3f8ed774dd72..e7766b04c395dc6c77ea3f184f979abd36d581d3 100644
--- a/ui/aura/client/drag_drop_client.h
+++ b/ui/aura/client/drag_drop_client.h
@@ -37,7 +37,15 @@ class AURA_EXPORT DragDropClient {
                                aura::Window* source_window,
                                const gfx::Point& screen_location,
                                int operation,
-                               ui::DragDropTypes::DragEventSource source) = 0;
+                               ui::DragDropTypes::DragEventSource source);
+
+  virtual int StartDragAndDrop(const ui::OSExchangeData& data,
+                               aura::Window* root_window,
+                               aura::Window* source_window,
+                               const gfx::Point& screen_location,
+                               int operation,
+                               ui::DragDropTypes::DragEventSource source,
+                               bool& cancelled);
 
   // Called when a drag and drop session is cancelled.
   virtual void DragCancel() = 0;
diff --git a/ui/base/BUILD.gn b/ui/base/BUILD.gn
index 14011a657ff3fbcf43f41eb16231eb45e545fad4..dd8b82143865ea1d0e2a8c54e10c5ca51003516d 100644
--- a/ui/base/BUILD.gn
+++ b/ui/base/BUILD.gn
@@ -141,6 +141,8 @@ component("base") {
     "cocoa/remote_layer_api.mm",
     "cocoa/scoped_cg_context_smooth_fonts.h",
     "cocoa/scoped_cg_context_smooth_fonts.mm",
+    "cocoa/secure_password_input.h",
+    "cocoa/secure_password_input.mm",
     "cocoa/three_part_image.h",
     "cocoa/three_part_image.mm",
     "cocoa/tool_tip_base_view.h",
diff --git a/ui/base/accelerators/accelerator.cc b/ui/base/accelerators/accelerator.cc
index 0694f7433df5fe487bdfb2b2ff1643674247d13d..238477431ed91e563c06931c97d6afae75452ef7 100644
--- a/ui/base/accelerators/accelerator.cc
+++ b/ui/base/accelerators/accelerator.cc
@@ -23,6 +23,8 @@
 #include "ui/events/keycodes/keyboard_code_conversion.h"
 #endif
 
+#include "app/vivaldi_apptools.h"
+
 namespace ui {
 
 namespace {
@@ -235,6 +237,15 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut = l10n_util::GetStringUTF16(string_id);
   }
 
+  if (vivaldi::IsVivaldiRunning()) {
+    // We want to display all function keys. They do not have to be translated.
+    if (key_code_ >= ui::VKEY_F1 && key_code_ <= ui::VKEY_F24) {
+      char buf[4];
+      base::snprintf(buf, 4, "F%d", key_code_ - ui::VKEY_F1 + 1);
+      base::UTF8ToUTF16(buf, strlen(buf), &shortcut);
+    }
+  }
+
   // Checking whether the character used for the accelerator is alphanumeric.
   // If it is not, then we need to adjust the string later on if the locale is
   // right-to-left. See below for more information of why such adjustment is
@@ -247,6 +258,15 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut_rtl.assign(shortcut);
   }
 
+  if (vivaldi::IsVivaldiRunning()) {
+    // We allow all modifier combinations.
+    if (IsShiftDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_SHIFT_MODIFIER, shortcut);
+    if (IsAltDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_ALT_MODIFIER, shortcut);
+    if (IsCtrlDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_CONTROL_MODIFIER, shortcut);
+  } else  {
   if (IsShiftDown())
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_SHIFT_MODIFIER, shortcut);
 
@@ -257,6 +277,7 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_CONTROL_MODIFIER, shortcut);
   else if (IsAltDown())
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_ALT_MODIFIER, shortcut);
+  }
 
   if (IsCmdDown()) {
 #if defined(OS_MACOSX)
diff --git a/ui/base/base_window.h b/ui/base/base_window.h
index f6af762c1306f228e461fe29bd27056e5aa55069..92ac7e67c00ea4e6f74bc3dc7fe873bf12e9a57c 100644
--- a/ui/base/base_window.h
+++ b/ui/base/base_window.h
@@ -100,6 +100,9 @@ class UI_BASE_EXPORT BaseWindow {
   // If set to true, the window will stay on top of other windows which do not
   // have this flag enabled.
   virtual void SetAlwaysOnTop(bool always_on_top) = 0;
+
+  // Update the inset. Used to update when a frameless window goes to maximized.
+  virtual void UpdateEventTargeterWithInset() {};
 };
 
 }  // namespace ui
diff --git a/ui/base/cocoa/secure_password_input.h b/ui/base/cocoa/secure_password_input.h
new file mode 100644
index 0000000000000000000000000000000000000000..ca90d3964a5df4ecd1d7b56eb84b1ccf812434c8
--- /dev/null
+++ b/ui/base/cocoa/secure_password_input.h
@@ -0,0 +1,29 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef UI_BASE_COCOA_SECURE_PASSWORD_INPUT_H_
+#define UI_BASE_COCOA_SECURE_PASSWORD_INPUT_H_
+
+#include "base/macros.h"
+#include "ui/base/ui_base_export.h"
+
+namespace ui {
+
+// Enables the secure password input mode while in scope.
+class UI_BASE_EXPORT ScopedPasswordInputEnabler {
+ public:
+  ScopedPasswordInputEnabler();
+  ~ScopedPasswordInputEnabler();
+
+  // Returns true if the password input mode is currently enabled. Useful for
+  // unit tests.
+  static bool IsPasswordInputEnabled();
+
+ private:
+  DISALLOW_COPY_AND_ASSIGN(ScopedPasswordInputEnabler);
+};
+
+}  // namespace ui
+
+#endif  // UI_BASE_COCOA_SECURE_PASSWORD_INPUT_H_
diff --git a/ui/base/cocoa/secure_password_input.mm b/ui/base/cocoa/secure_password_input.mm
new file mode 100644
index 0000000000000000000000000000000000000000..d0345336d4c13f89693329d447d76ad5c70e412a
--- /dev/null
+++ b/ui/base/cocoa/secure_password_input.mm
@@ -0,0 +1,63 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "ui/base/cocoa/secure_password_input.h"
+
+#import <Carbon/Carbon.h>
+
+#include "base/logging.h"
+
+namespace {
+
+// Used to protect from out-of-order calls to enabling/disabling functions.
+int g_password_input_counter = 0;
+
+// SetPasswordInputEnabled() is copied from
+// enableSecureTextInput() and disableSecureTextInput() functions in
+// third_party/WebKit/WebCore/platform/SecureTextInput.cpp
+//
+// The following technote describes proper EnableSecureEventInput() usage:
+// https://developer.apple.com/library/content/technotes/tn2150/_index.html
+void SetPasswordInputEnabled(bool enabled) {
+  if (enabled) {
+    DCHECK(!IsSecureEventInputEnabled());
+    EnableSecureEventInput();
+
+    CFArrayRef inputSources = TISCreateASCIICapableInputSourceList();
+    TSMSetDocumentProperty(0, kTSMDocumentEnabledInputSourcesPropertyTag,
+                           sizeof(CFArrayRef), &inputSources);
+    CFRelease(inputSources);
+  } else {
+    DCHECK(IsSecureEventInputEnabled());
+    TSMRemoveDocumentProperty(0, kTSMDocumentEnabledInputSourcesPropertyTag);
+
+    DisableSecureEventInput();
+  }
+}
+
+}  // namespace
+
+namespace ui {
+
+ScopedPasswordInputEnabler::ScopedPasswordInputEnabler() {
+  if (!g_password_input_counter) {
+    SetPasswordInputEnabled(true);
+  }
+  ++g_password_input_counter;
+}
+
+ScopedPasswordInputEnabler::~ScopedPasswordInputEnabler() {
+  --g_password_input_counter;
+  DCHECK_LE(0, g_password_input_counter);
+  if (!g_password_input_counter) {
+    SetPasswordInputEnabled(false);
+  }
+}
+
+// static
+bool ScopedPasswordInputEnabler::IsPasswordInputEnabled() {
+  return g_password_input_counter > 0;
+}
+
+}  // namespace ui
diff --git a/ui/base/dragdrop/drop_target_win.cc b/ui/base/dragdrop/drop_target_win.cc
index 5dddafa3591db395e010e5daa6ec94978355c6dd..b59264aff3dae5246bce0bc1e3a56895551c3167 100644
--- a/ui/base/dragdrop/drop_target_win.cc
+++ b/ui/base/dragdrop/drop_target_win.cc
@@ -58,6 +58,12 @@ HRESULT DropTargetWin::DragEnter(IDataObject* data_object,
 HRESULT DropTargetWin::DragOver(DWORD key_state,
                                 POINTL cursor_position,
                                 DWORD* effect) {
+  // NOTE(andre@vivaldi.com): We end up here after |DropTargetWin::DragLeave|
+  // for certain drag out of patterns. Bail as the operation is done and will be
+  // resumed with an additional |DragEnter|.
+  if (!current_data_object_) {
+    return E_UNEXPECTED;
+  }
   // Tell the helper that we moved over it so it can update the drag image.
   IDropTargetHelper* drop_helper = DropHelper();
   if (drop_helper)
diff --git a/ui/base/l10n/l10n_util.cc b/ui/base/l10n/l10n_util.cc
index 35b202fd9949e6859d834852fa6e4a01895d576b..92dc896ec47caa1c31d5bd94a9ccd737be5425bc 100644
--- a/ui/base/l10n/l10n_util.cc
+++ b/ui/base/l10n/l10n_util.cc
@@ -33,6 +33,8 @@
 #include "ui/base/resource/resource_bundle.h"
 #include "ui/base/ui_base_paths.h"
 
+#include "app/vivaldi_apptools.h"
+
 #if defined(OS_ANDROID)
 #include "base/android/locale_utils.h"
 #include "ui/base/l10n/l10n_util_android.h"
@@ -428,6 +430,13 @@ bool CheckAndResolveLocale(const std::string& locale,
 
 std::string GetApplicationLocaleInternal(const std::string& pref_locale) {
 #if defined(OS_MACOSX)
+  if (vivaldi::IsVivaldiRunning()) {
+    if (!pref_locale.empty()) {
+      std::string app_locale = base::i18n::GetCanonicalLocale(pref_locale);
+      if (!app_locale.empty())
+        return app_locale;
+    }
+  }
 
   // Use any override (Cocoa for the browser), otherwise use the preference
   // passed to the function.
@@ -476,6 +485,10 @@ std::string GetApplicationLocaleInternal(const std::string& pref_locale) {
   candidates.push_back(base::android::GetDefaultLocaleString());
 
 #elif defined(USE_GLIB) && !defined(OS_CHROMEOS)
+  if (vivaldi::IsVivaldiRunning()) {
+    if (!pref_locale.empty())
+      candidates.push_back(base::i18n::GetCanonicalLocale(pref_locale));
+  }
 
   // GLib implements correct environment variable parsing with
   // the precedence order: LANGUAGE, LC_ALL, LC_MESSAGES and LANG.
diff --git a/ui/base/resource/resource_bundle.cc b/ui/base/resource/resource_bundle.cc
index 56101888f276df4a4e3cef922ea4afe9558d371b..fb7a344f921d09a15cae8275427cb786c4c1fe58 100644
--- a/ui/base/resource/resource_bundle.cc
+++ b/ui/base/resource/resource_bundle.cc
@@ -10,6 +10,7 @@
 #include <utility>
 #include <vector>
 
+#include "app/vivaldi_apptools.h"
 #include "base/big_endian.h"
 #include "base/command_line.h"
 #include "base/files/file.h"
@@ -265,7 +266,11 @@ void ResourceBundle::LoadSecondaryLocaleDataWithPakFileRegion(
 
 #if !defined(OS_ANDROID)
 bool ResourceBundle::LocaleDataPakExists(const std::string& locale) {
-  return !GetLocaleFilePath(locale, true).empty();
+  bool locale_file_path_exists = !GetLocaleFilePath(locale, true).empty();
+  if (!locale_file_path_exists) {
+    locale_file_path_exists = vivaldi::IsVivaldiExtraLocale(locale);
+  }
+  return locale_file_path_exists;
 }
 #endif  // !defined(OS_ANDROID)
 
@@ -363,10 +368,17 @@ std::string ResourceBundle::LoadLocaleResources(
   if (locale_file_path.empty())
     locale_file_path = GetLocaleFilePath(app_locale, true);
 
-  if (locale_file_path.empty()) {
+  if (locale_file_path.empty()) { // (jarle@vivaldi) load fallback locale if 'unsupported' locale
+    if (pref_locale == "nn") {
+      locale_file_path = GetLocaleFilePath("nb", true); // Bokmål as a fallback for Nynorsk
+    } else {
+      locale_file_path = GetLocaleFilePath("en-US", true);
+    }
+    if (locale_file_path.empty()) {
     // It's possible that there is no locale.pak.
     LOG(WARNING) << "locale_file_path.empty() for locale " << app_locale;
     return std::string();
+    }
   }
 
   std::unique_ptr<DataPack> data_pack(new DataPack(SCALE_FACTOR_100P));
@@ -803,12 +815,12 @@ void ResourceBundle::LoadChromeResources() {
   // scale factor to gfx::ImageSkia::AddRepresentation.
   if (IsScaleFactorSupported(SCALE_FACTOR_100P)) {
     AddDataPackFromPath(GetResourcesPakFilePath(
-        "chrome_100_percent.pak"), SCALE_FACTOR_100P);
+        "vivaldi_100_percent.pak"), SCALE_FACTOR_100P);
   }
 
   if (IsScaleFactorSupported(SCALE_FACTOR_200P)) {
     AddOptionalDataPackFromPath(GetResourcesPakFilePath(
-        "chrome_200_percent.pak"), SCALE_FACTOR_200P);
+        "vivaldi_200_percent.pak"), SCALE_FACTOR_200P);
   }
 }
 
@@ -831,7 +843,7 @@ void ResourceBundle::AddDataPackFromPathInternal(
   std::unique_ptr<DataPack> data_pack(new DataPack(scale_factor));
   if (data_pack->LoadFromPath(pack_path)) {
     AddDataPack(std::move(data_pack));
-  } else if (!optional) {
+  } else if (false) {
     LOG(ERROR) << "Failed to load " << pack_path.value()
                << "\nSome features may not be available.";
   }
diff --git a/ui/base/resource/resource_bundle_android.cc b/ui/base/resource/resource_bundle_android.cc
index 0923b6fd6a3f41211585c9dcfb46f746fb4fa1a3..30471282026a38c73b05792e075ed52c42ab67c2 100644
--- a/ui/base/resource/resource_bundle_android.cc
+++ b/ui/base/resource/resource_bundle_android.cc
@@ -83,8 +83,8 @@ std::unique_ptr<DataPack> LoadDataPackFromLocalePak(
 void ResourceBundle::LoadCommonResources() {
   base::FilePath disk_path;
   PathService::Get(ui::DIR_RESOURCE_PAKS_ANDROID, &disk_path);
-  disk_path = disk_path.AppendASCII("chrome_100_percent.pak");
-  if (LoadFromApkOrFile("assets/chrome_100_percent.pak",
+  disk_path = disk_path.AppendASCII("vivaldi_100_percent.pak");
+  if (LoadFromApkOrFile("assets/vivaldi_100_percent.pak",
                         &disk_path,
                         &g_chrome_100_percent_fd,
                         &g_chrome_100_percent_region)) {
diff --git a/ui/base/resource/resource_bundle_mac.mm b/ui/base/resource/resource_bundle_mac.mm
index 3cf1847860a56e0898aaae68cbabfc1e35536c7d..06e9f0ef0b9eff278686c81ff27efbae9fc1105d 100644
--- a/ui/base/resource/resource_bundle_mac.mm
+++ b/ui/base/resource/resource_bundle_mac.mm
@@ -48,13 +48,13 @@ base::FilePath GetResourcesPakFilePath(NSString* name, NSString* mac_locale) {
 }  // namespace
 
 void ResourceBundle::LoadCommonResources() {
-  AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_100_percent",
+  AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_100_percent",
                         nil), SCALE_FACTOR_100P);
 
   // On Mac we load 1x and 2x resources and we let the UI framework decide
   // which one to use.
   if (IsScaleFactorSupported(SCALE_FACTOR_200P)) {
-    AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_200_percent", nil),
+    AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_200_percent", nil),
                         SCALE_FACTOR_200P);
   }
 }
diff --git a/ui/events/blink/input_handler_proxy.cc b/ui/events/blink/input_handler_proxy.cc
index ba36cfb0d5d6608390175118c55ad4aeefcdbb04..581a58554c491fb0077cff524640632897bd92e2 100644
--- a/ui/events/blink/input_handler_proxy.cc
+++ b/ui/events/blink/input_handler_proxy.cc
@@ -33,6 +33,8 @@
 #include "ui/gfx/geometry/point_conversions.h"
 #include "ui/latency/latency_info.h"
 
+#include "app/vivaldi_apptools.h"
+
 using blink::WebFloatPoint;
 using blink::WebFloatSize;
 using blink::WebGestureEvent;
@@ -650,7 +652,25 @@ InputHandlerProxy::EventDisposition InputHandlerProxy::HandleMouseWheel(
           cc::EventListenerClass::kMouseWheel);
   switch (properties) {
     case cc::EventListenerProperties::kPassive:
+      // Last resort fix for 1.13 (VB-34466). Should be removed with proper
+      // handling in View layer before calling render code. ASAP.
+      if (vivaldi::IsVivaldiRunning()) {
+        int modifiers = wheel_event.GetModifiers() &
+            (WebInputEvent::kShiftKey | WebInputEvent::kControlKey |
+             WebInputEvent::kAltKey | WebInputEvent::kMetaKey |
+             WebInputEvent::kRightButtonDown);
+        if (modifiers == WebInputEvent::kAltKey ||
+            modifiers == WebInputEvent::kControlKey ||
+            modifiers == WebInputEvent::kMetaKey ||
+            modifiers == WebInputEvent::kRightButtonDown)
+        {
+          result = DID_NOT_HANDLE;
+        } else {
+          result = DID_HANDLE_NON_BLOCKING;
+        }
+      } else {
       result = DID_HANDLE_NON_BLOCKING;
+      }
       break;
     case cc::EventListenerProperties::kBlockingAndPassive:
     case cc::EventListenerProperties::kBlocking:
diff --git a/ui/message_center/views/notification_view.cc b/ui/message_center/views/notification_view.cc
index 8b34638444c5e5a4365ea49b3180ccb16f692acb..8b2043d33c5ca169b1c9b86ed89d453734ac31ad 100644
--- a/ui/message_center/views/notification_view.cc
+++ b/ui/message_center/views/notification_view.cc
@@ -280,8 +280,11 @@ void NotificationView::Layout() {
   // Top views.
   int top_height = top_view_->GetHeightForWidth(content_width);
   top_view_->SetBounds(insets.left(), insets.top(), content_width, top_height);
+// NOTE(jarle@vivaldi.com): Calling ShrinkTopmostLabel() causes a regression,
+// VB-34628 / crbug 789904. Remove #if-define when the Chromium bug is fixed.
+#if !defined(VIVALDI_BUILD)
   ShrinkTopmostLabel();
-
+#endif // VIVALDI_BUILD
   // Icon.
   icon_view_->SetBounds(insets.left(), insets.top(), kNotificationIconSize,
                         kNotificationIconSize);
diff --git a/ui/shell_dialogs/select_file_dialog_mac.mm b/ui/shell_dialogs/select_file_dialog_mac.mm
index e408e2158c98f96887bfb22fedb4f348c5369b1f..33f2661ea904870a96ae97d21a67dc9681399c57 100644
--- a/ui/shell_dialogs/select_file_dialog_mac.mm
+++ b/ui/shell_dialogs/select_file_dialog_mac.mm
@@ -429,6 +429,18 @@ SelectFileDialog* CreateSelectFileDialog(
   return [url isFileURL];
 }
 
+- (BOOL)panel:(id)sender validateURL:(NSURL*)url error:(NSError**)outError {
+  // Refuse to accept users closing the dialog with a key repeat, since the key
+  // may have been first pressed while the user was looking at insecure content.
+  // See https://crbug.com/637098.
+  if ([[NSApp currentEvent] type] == NSKeyDown &&
+      [[NSApp currentEvent] isARepeat]) {
+    return NO;
+  }
+
+  return YES;
+}
+
 @end
 
 @implementation ExtensionDropdownHandler
diff --git a/ui/shell_dialogs/select_file_dialog_win.cc b/ui/shell_dialogs/select_file_dialog_win.cc
index be73a0677dbde4a30d198a97a0491f5cc9d9ef7d..8cdea0bac9e5cb974d7e8e805e92a5568b04b4da 100644
--- a/ui/shell_dialogs/select_file_dialog_win.cc
+++ b/ui/shell_dialogs/select_file_dialog_win.cc
@@ -611,6 +611,10 @@ bool SelectFileDialogImpl::RunOpenFileDialog(
     else
       ofn.SetInitialSelection(path->DirName(), path->BaseName());
   }
+  // Vivaldi addition
+  if (!title.empty()) {
+    ofn.GetOPENFILENAME()->lpstrTitle = title.c_str();
+  }
 
   if (!filter.empty())
     ofn.GetOPENFILENAME()->lpstrFilter = filter.c_str();
diff --git a/ui/views/bubble/bubble_dialog_delegate.cc b/ui/views/bubble/bubble_dialog_delegate.cc
index aa98ca418f73b45bdfeef566b40f8c8b02c59b30..d5d49c068bad5f6d3269862200b46e50e3d8e3c4 100644
--- a/ui/views/bubble/bubble_dialog_delegate.cc
+++ b/ui/views/bubble/bubble_dialog_delegate.cc
@@ -27,6 +27,8 @@
 #include "ui/base/win/shell.h"
 #endif
 
+#include "app/vivaldi_apptools.h"
+
 namespace views {
 
 namespace {
@@ -78,8 +80,17 @@ Widget* BubbleDialogDelegateView::CreateBubble(
   // the parent frame and let DWM handle compositing.  If not, then we don't
   // want to allow the bubble to extend the frame because it will be clipped.
   bubble_delegate->set_adjust_if_offscreen(ui::win::IsAeroGlassEnabled());
-#elif (defined(OS_LINUX) && !defined(OS_CHROMEOS)) || defined(OS_MACOSX)
-  // Linux clips bubble windows that extend outside their parent window bounds.
+#elif (defined(OS_LINUX) && !defined(OS_CHROMEOS))
+  if (vivaldi::IsVivaldiRunning()) {
+    // In Vivaldi we use some of these bubbles at the bottom of the window, so
+    // we must adjust to try to fit inside the window.
+    bubble_delegate->set_adjust_if_offscreen(true);
+  } else {
+    // Linux clips bubble windows that extend outside their parent window
+    // bounds.
+    bubble_delegate->set_adjust_if_offscreen(false);
+  }
+#elif defined(OS_MACOSX)
   // Mac never adjusts.
   bubble_delegate->set_adjust_if_offscreen(false);
 #endif
diff --git a/ui/views/controls/menu/menu_controller.cc b/ui/views/controls/menu/menu_controller.cc
index 9389f25f034d131f9aab4bd521db7999c12c5a5c..96d554f52ff5c69b99d92485e07b12c1e261f768 100644
--- a/ui/views/controls/menu/menu_controller.cc
+++ b/ui/views/controls/menu/menu_controller.cc
@@ -492,6 +492,10 @@ void MenuController::Run(Widget* parent,
     ViewsDelegate::GetInstance()->AddRef();
 }
 
+void MenuController::VivaldiSelectItem(MenuItemView* item) {
+  SetSelection(item, SELECTION_OPEN_SUBMENU | SELECTION_UPDATE_IMMEDIATELY);
+}
+
 void MenuController::Cancel(ExitType type) {
   // If the menu has already been destroyed, no further cancellation is
   // needed.  We especially don't want to set the |exit_type_| to a lesser
diff --git a/ui/views/controls/menu/menu_controller.h b/ui/views/controls/menu/menu_controller.h
index 1a056eb2404d87d0fe686182a342e4878d9b9193..c8da8a3c3f34094b4ee74222c504112ab49197a1 100644
--- a/ui/views/controls/menu/menu_controller.h
+++ b/ui/views/controls/menu/menu_controller.h
@@ -195,6 +195,8 @@ class VIEWS_EXPORT MenuController
   // notifying any menu items.
   void ClearStateForTest();
 
+  void VivaldiSelectItem(MenuItemView* item);
+
   // Only used for testing.
   static void TurnOffMenuSelectionHoldForTest();
 
diff --git a/ui/views/controls/menu/menu_model_adapter.cc b/ui/views/controls/menu/menu_model_adapter.cc
index 06a9d3cfda9b6fdd3f6d315fd62e22504e9cc574..b7299cd2242487841d1d94be40c916c304df6675 100644
--- a/ui/views/controls/menu/menu_model_adapter.cc
+++ b/ui/views/controls/menu/menu_model_adapter.cc
@@ -49,6 +49,26 @@ void MenuModelAdapter::BuildMenu(MenuItemView* menu) {
   menu->ChildrenChanged();
 }
 
+void MenuModelAdapter::VivaldiUpdateMenu(MenuItemView* menu,
+                                         ui::MenuModel* model) {
+  // Clear the menu.
+  if (menu->HasSubmenu()) {
+    const int subitem_count = menu->GetSubmenu()->child_count();
+    for (int i = 0; i < subitem_count; ++i)
+      menu->RemoveMenuItemAt(0);
+  }
+
+  // Leave entries in the map if the menu is being shown.  This
+  // allows the map to find the menu model of submenus being closed
+  // so ui::MenuModel::MenuClosed() can be called.
+  if (!menu->GetMenuController())
+    menu_map_.clear();
+  menu_map_[menu] = model;
+
+  // Repopulate the menu.
+  BuildMenuImpl(menu, model);
+}
+
 MenuItemView* MenuModelAdapter::CreateMenu() {
   MenuItemView* item = new MenuItemView(this);
   BuildMenu(item);
diff --git a/ui/views/controls/menu/menu_model_adapter.h b/ui/views/controls/menu/menu_model_adapter.h
index 0ac493c3c6a0ce46afe5cfba347c413cf970301e..bf98321e181803710c51acb8d7fc5e1be894f4f6 100644
--- a/ui/views/controls/menu/menu_model_adapter.h
+++ b/ui/views/controls/menu/menu_model_adapter.h
@@ -33,6 +33,8 @@ class VIEWS_EXPORT MenuModelAdapter : public MenuDelegate {
   // (including submenus).
   virtual void BuildMenu(MenuItemView* menu);
 
+  void VivaldiUpdateMenu(MenuItemView* menu, ui::MenuModel* model);
+
   // Convenience for creating and populating a menu. The caller owns the
   // returned MenuItemView.
   MenuItemView* CreateMenu();
diff --git a/ui/views/controls/menu/menu_runner_impl.cc b/ui/views/controls/menu/menu_runner_impl.cc
index 0004fe4a546bf447419454f73d48809c3e48e512..926f895611ae7bfa7298310a01ba2375595c1df5 100644
--- a/ui/views/controls/menu/menu_runner_impl.cc
+++ b/ui/views/controls/menu/menu_runner_impl.cc
@@ -21,6 +21,8 @@
 #include "ui/events/x/events_x_utils.h"  // nogncheck
 #endif
 
+#include "app/vivaldi_apptools.h"
+
 namespace views {
 namespace internal {
 
@@ -198,6 +200,8 @@ MenuRunnerImpl::~MenuRunnerImpl() {
 bool MenuRunnerImpl::ShouldShowMnemonics(MenuButton* button) {
   // Show mnemonics if the button has focus or alt is pressed.
   bool show_mnemonics = button ? button->HasFocus() : false;
+  if (vivaldi::IsVivaldiRunning())
+    show_mnemonics = true;
 #if defined(OS_WIN)
   show_mnemonics |= ui::win::IsAltPressed();
 #elif defined(USE_X11)
diff --git a/ui/views/controls/textfield/textfield.cc b/ui/views/controls/textfield/textfield.cc
index 7be7cbfa1a61f0d40578e52d9d55c7ed1aaa3720..b1a77a28fd7e3b0ff3b548197189d377b711e69f 100644
--- a/ui/views/controls/textfield/textfield.cc
+++ b/ui/views/controls/textfield/textfield.cc
@@ -73,6 +73,10 @@
 #include "ui/wm/core/ime_util_chromeos.h"
 #endif
 
+#if defined(OS_MACOSX)
+#include "ui/base/cocoa/secure_password_input.h"
+#endif
+
 namespace views {
 
 namespace {
@@ -1030,6 +1034,11 @@ void Textfield::OnPaint(gfx::Canvas* canvas) {
 }
 
 void Textfield::OnFocus() {
+#if defined(OS_MACOSX)
+  if (text_input_type_ == ui::TEXT_INPUT_TYPE_PASSWORD)
+    password_input_enabler_.reset(new ui::ScopedPasswordInputEnabler());
+#endif  // defined(OS_MACOSX)
+
   GetRenderText()->set_focused(true);
   if (ShouldShowCursor()) {
     UpdateCursorViewPosition();
@@ -1073,6 +1082,10 @@ void Textfield::OnBlur() {
     FocusRing::Uninstall(this);
   SchedulePaint();
   View::OnBlur();
+
+#if defined(OS_MACOSX)
+  password_input_enabler_.reset();
+#endif  // defined(OS_MACOSX)
 }
 
 gfx::Point Textfield::GetKeyboardContextMenuLocation() {
diff --git a/ui/views/controls/textfield/textfield.h b/ui/views/controls/textfield/textfield.h
index 1c36080fcf68d7cb344da5fd2543ab3ad39e13ba..2c9da38ebebe7a6efa087dded47ef21fac9233f9 100644
--- a/ui/views/controls/textfield/textfield.h
+++ b/ui/views/controls/textfield/textfield.h
@@ -16,6 +16,7 @@
 #include "base/memory/weak_ptr.h"
 #include "base/strings/string16.h"
 #include "base/time/time.h"
+#include "build/build_config.h"
 #include "third_party/skia/include/core/SkColor.h"
 #include "ui/base/ime/text_input_client.h"
 #include "ui/base/ime/text_input_type.h"
@@ -38,6 +39,12 @@ namespace base {
 class TimeDelta;
 }
 
+#if defined(OS_MACOSX)
+namespace ui {
+class ScopedPasswordInputEnabler;
+}
+#endif  // defined(OS_MACOSX)
+
 namespace views {
 
 class MenuRunner;
@@ -556,6 +563,11 @@ class VIEWS_EXPORT Textfield : public View,
   // View containing the text cursor.
   View cursor_view_;
 
+#if defined(OS_MACOSX)
+  // Used to track active password input sessions.
+  std::unique_ptr<ui::ScopedPasswordInputEnabler> password_input_enabler_;
+#endif  // defined(OS_MACOSX)
+
   // Used to bind callback functions to this object.
   base::WeakPtrFactory<Textfield> weak_ptr_factory_;
 
diff --git a/ui/views/controls/textfield/textfield_unittest.cc b/ui/views/controls/textfield/textfield_unittest.cc
index 4ee16e762840792aba8c29a401c2712e61ddac75..f1f5da720583c9a8293f940f444cdfa9f0989cad 100644
--- a/ui/views/controls/textfield/textfield_unittest.cc
+++ b/ui/views/controls/textfield/textfield_unittest.cc
@@ -67,6 +67,10 @@
 #include "ui/wm/core/ime_util_chromeos.h"
 #endif
 
+#if defined(OS_MACOSX)
+#include "ui/base/cocoa/secure_password_input.h"
+#endif
+
 using base::ASCIIToUTF16;
 using base::UTF8ToUTF16;
 using base::WideToUTF16;
@@ -3400,4 +3404,24 @@ TEST_F(TextfieldTest, SendingDeletePreservesShiftFlag) {
   EXPECT_EQ(ui::EF_SHIFT_DOWN, textfield_->event_flags());
 }
 
+#if defined(OS_MACOSX)
+TEST_F(TextfieldTest, SecurePasswordInput) {
+  InitTextfield();
+  ASSERT_FALSE(ui::ScopedPasswordInputEnabler::IsPasswordInputEnabled());
+
+  // Shouldn't enable secure input if it's not a password textfield.
+  textfield_->OnFocus();
+  EXPECT_FALSE(ui::ScopedPasswordInputEnabler::IsPasswordInputEnabled());
+
+  textfield_->SetTextInputType(ui::TEXT_INPUT_TYPE_PASSWORD);
+
+  // Single matched calls immediately update IsPasswordInputEnabled().
+  textfield_->OnFocus();
+  EXPECT_TRUE(ui::ScopedPasswordInputEnabler::IsPasswordInputEnabled());
+
+  textfield_->OnBlur();
+  EXPECT_FALSE(ui::ScopedPasswordInputEnabler::IsPasswordInputEnabled());
+}
+#endif  // defined(OS_MACOSX)
+
 }  // namespace views
diff --git a/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc b/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc
index 55dcdf780ebcbb604dc9d066859756b65befa8b8..6e4013a9fcf2b882e485abc4e9c1f51177978a39 100644
--- a/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc
+++ b/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc
@@ -751,7 +751,8 @@ int DesktopDragDropClientAuraX11::StartDragAndDrop(
     aura::Window* source_window,
     const gfx::Point& screen_location,
     int operation,
-    ui::DragDropTypes::DragEventSource source) {
+    ui::DragDropTypes::DragEventSource source,
+    bool& cancelled) {
   UMA_HISTOGRAM_ENUMERATION("Event.DragDrop.Start", source,
                             ui::DragDropTypes::DRAG_EVENT_SOURCE_COUNT);
 
@@ -807,9 +808,11 @@ int DesktopDragDropClientAuraX11::StartDragAndDrop(
 
   if (alive) {
     if (negotiated_operation_ == ui::DragDropTypes::DRAG_NONE) {
+      cancelled = true;
       UMA_HISTOGRAM_ENUMERATION("Event.DragDrop.Cancel", source,
                                 ui::DragDropTypes::DRAG_EVENT_SOURCE_COUNT);
     } else {
+      cancelled = false;
       UMA_HISTOGRAM_ENUMERATION("Event.DragDrop.Drop", source,
                                 ui::DragDropTypes::DRAG_EVENT_SOURCE_COUNT);
     }
@@ -823,6 +826,7 @@ int DesktopDragDropClientAuraX11::StartDragAndDrop(
 
     return negotiated_operation_;
   }
+  cancelled = true;
   UMA_HISTOGRAM_ENUMERATION("Event.DragDrop.Cancel", source,
                             ui::DragDropTypes::DRAG_EVENT_SOURCE_COUNT);
   return ui::DragDropTypes::DRAG_NONE;
diff --git a/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h b/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h
index 37c90c1e7fe452054eff348a3301db48d8f91273..c1bfc1ba8e15151876dd84bb87c23c0b6e67175b 100644
--- a/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h
+++ b/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h
@@ -86,7 +86,10 @@ class VIEWS_EXPORT DesktopDragDropClientAuraX11
                        aura::Window* source_window,
                        const gfx::Point& screen_location,
                        int operation,
-                       ui::DragDropTypes::DragEventSource source) override;
+                       ui::DragDropTypes::DragEventSource source,
+                       bool& cancelled) override;
+  using aura::client::DragDropClient::StartDragAndDrop;
+
   void DragCancel() override;
   bool IsDragDropInProgress() override;
   void AddObserver(aura::client::DragDropClientObserver* observer) override;
diff --git a/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.cc b/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.cc
index 2800896f01fc61259ccf63107b35d14a39d98bab..7ed7f86b20b9f2db50938e50163866966f9a14cc 100644
--- a/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.cc
+++ b/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.cc
@@ -12,6 +12,9 @@
 #include "ui/views/widget/desktop_aura/desktop_drop_target_win.h"
 #include "ui/views/widget/desktop_aura/desktop_window_tree_host_win.h"
 
+#include "app/vivaldi_apptools.h"
+#include "ui/dragging/custom_drag_source_win.h"
+
 namespace views {
 
 DesktopDragDropClientWin::DesktopDragDropClientWin(
@@ -35,13 +38,19 @@ int DesktopDragDropClientWin::StartDragAndDrop(
     aura::Window* source_window,
     const gfx::Point& screen_location,
     int operation,
-    ui::DragDropTypes::DragEventSource source) {
+    ui::DragDropTypes::DragEventSource source,
+    bool& cancelled) {
   drag_drop_in_progress_ = true;
   drag_operation_ = operation;
 
   base::WeakPtr<DesktopDragDropClientWin> alive(weak_factory_.GetWeakPtr());
 
+  if (vivaldi::IsVivaldiRunning()) {
+    drag_source_ = Microsoft::WRL::Make<vivaldi::CustomDragSourceWin>(
+        vivaldi::IsTabDragInProgress());
+  } else {
   drag_source_ = ui::DragSourceWin::Create();
+  }
   Microsoft::WRL::ComPtr<ui::DragSourceWin> drag_source_copy = drag_source_;
   drag_source_copy->set_data(&data);
   ui::OSExchangeDataProviderWin::GetDataObjectImpl(data)
@@ -60,6 +69,8 @@ int DesktopDragDropClientWin::StartDragAndDrop(
   if (alive)
     drag_drop_in_progress_ = false;
 
+  cancelled = (result == DRAGDROP_S_CANCEL);
+
   if (result != DRAGDROP_S_DROP)
     effect = DROPEFFECT_NONE;
 
diff --git a/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.h b/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.h
index 25845e08f222eac50c066f4bc2bec713751166b7..76e8cf2a02c9ebef9339bfed4b068d909b49c4cc 100644
--- a/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.h
+++ b/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.h
@@ -39,7 +39,8 @@ class VIEWS_EXPORT DesktopDragDropClientWin
                        aura::Window* source_window,
                        const gfx::Point& screen_location,
                        int operation,
-                       ui::DragDropTypes::DragEventSource source) override;
+                       ui::DragDropTypes::DragEventSource source,
+                       bool& cancelled) override;
   void DragCancel() override;
   bool IsDragDropInProgress() override;
   void AddObserver(aura::client::DragDropClientObserver* observer) override;
diff --git a/ui/views/widget/desktop_aura/desktop_screen_win.cc b/ui/views/widget/desktop_aura/desktop_screen_win.cc
index f772f64d656e88cd2631e388f92e00a845f84f16..0c63ea93115b6e2e6a3d1af9023a77d69cfb3605 100644
--- a/ui/views/widget/desktop_aura/desktop_screen_win.cc
+++ b/ui/views/widget/desktop_aura/desktop_screen_win.cc
@@ -32,7 +32,7 @@ display::Display DesktopScreenWin::GetDisplayMatching(
 }
 
 HWND DesktopScreenWin::GetHWNDFromNativeView(gfx::NativeView window) const {
-  aura::WindowTreeHost* host = window->GetHost();
+  aura::WindowTreeHost* host = window ? window->GetHost() : NULL;
   return host ? host->GetAcceleratedWidget() : NULL;
 }
 
diff --git a/ui/views/widget/desktop_aura/desktop_screen_x11.cc b/ui/views/widget/desktop_aura/desktop_screen_x11.cc
index 17464aafc7829104361848118dc47909bc69e296..4fd279838d6edde87a5b417a877fbf38a111bd30 100644
--- a/ui/views/widget/desktop_aura/desktop_screen_x11.cc
+++ b/ui/views/widget/desktop_aura/desktop_screen_x11.cc
@@ -203,6 +203,7 @@ display::Display DesktopScreenX11::GetDisplayNearestWindow(
   // created before we create the aura::WindowEventDispatcher. So we ask what
   // the DRWHX11 believes the window bounds are instead of going through the
   // aura::Window's screen bounds.
+  if (window) {
   aura::WindowTreeHost* host = window->GetHost();
   if (host) {
     DesktopWindowTreeHostX11* rwh = DesktopWindowTreeHostX11::GetHostForXID(
@@ -210,6 +211,7 @@ display::Display DesktopScreenX11::GetDisplayNearestWindow(
     if (rwh)
       return GetDisplayMatching(rwh->GetX11RootWindowBounds());
   }
+  }
 
   return GetPrimaryDisplay();
 }
diff --git a/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc b/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
index deaf9cbc13ccdcb0a0dc5a04a5c5a160b27d1f5b..8000e8795f549c0540e2085c78b97973857337c9 100644
--- a/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
+++ b/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
@@ -61,6 +61,8 @@
 #include "ui/wm/core/compound_event_filter.h"
 #include "ui/wm/core/window_util.h"
 
+#include "ui/views/widget/widget_delegate.h"
+
 DEFINE_UI_CLASS_PROPERTY_TYPE(views::DesktopWindowTreeHostX11*);
 
 namespace views {
@@ -1341,6 +1343,15 @@ void DesktopWindowTreeHostX11::InitX11Window(
   swa.background_pixmap = x11::None;
   swa.bit_gravity = NorthWestGravity;
 
+  // NOTE(espen@vivaldi.com): Set override_redirect for the window we use to
+  // generate thumbnails. It will be opened offscreen. Override_redirect
+  // will disable window managers from forcing the window to open in a visible
+  // area.
+  if (params.thumbnail_window) {
+    swa.override_redirect = x11::True;
+    attribute_mask |= CWOverrideRedirect;
+  }
+
   ::Atom window_type;
   switch (params.type) {
     case Widget::InitParams::TYPE_MENU:
@@ -1753,6 +1764,22 @@ void DesktopWindowTreeHostX11::DispatchMouseEvent(ui::MouseEvent* event) {
   if (event->IsAnyButton() || event->IsMouseWheelEvent())
     FlashFrame(false);
 
+  // Support for page history navigation with mouse. Only when capture to avoid
+  // navigating when a context menu etc is open (vivaldi/espen).
+  if (event->type() == ui::ET_MOUSE_RELEASED && HasCapture()) {
+    bool back_button_pressed =
+      (event->changed_button_flags() == ui::EF_BACK_MOUSE_BUTTON);
+    bool forward_button_pressed =
+      (event->changed_button_flags() == ui::EF_FORWARD_MOUSE_BUTTON);
+    if (back_button_pressed || forward_button_pressed) {
+      Widget* widget = native_widget_delegate_->AsWidget();
+      if (widget && widget->widget_delegate()) {
+        widget->widget_delegate()->HandleKeyboardCode(back_button_pressed ?
+            ui::VKEY_BROWSER_BACK : ui::VKEY_BROWSER_FORWARD);
+      }
+    }
+  }
+
   if (!g_current_capture || g_current_capture == this) {
     SendEventToSink(event);
   } else {
diff --git a/ui/views/widget/native_widget_aura.cc b/ui/views/widget/native_widget_aura.cc
index 8797bf889fbc6529b89ccdda9a9b54dd18ed640a..2c65de71689c01b9fabd07718fb82a4adf55b6e4 100644
--- a/ui/views/widget/native_widget_aura.cc
+++ b/ui/views/widget/native_widget_aura.cc
@@ -1167,6 +1167,8 @@ void NativeWidgetPrivate::ReparentNativeView(gfx::NativeView native_view,
     // matches our previous behaviour; the global stacking client would almost
     // always reattach the window to the same RootWindow.
     aura::Window* root_window = native_view->GetRootWindow();
+    if (!root_window)
+      return;
     aura::client::ParentWindowWithContext(
         native_view, root_window, root_window->GetBoundsInScreen());
   }
diff --git a/ui/views/widget/widget.cc b/ui/views/widget/widget.cc
index d685bf4f40e39686afb94164c475f6658fa66a84..75a00364380a9fe81e85bddebc16f7b063cb22eb 100644
--- a/ui/views/widget/widget.cc
+++ b/ui/views/widget/widget.cc
@@ -135,6 +135,7 @@ Widget::InitParams::InitParams(Type type)
       layer_type(ui::LAYER_TEXTURED),
       context(nullptr),
       force_show_in_taskbar(false),
+      thumbnail_window(false),
       force_software_compositing(false) {}
 
 Widget::InitParams::InitParams(const InitParams& other) = default;
diff --git a/ui/views/widget/widget.h b/ui/views/widget/widget.h
index 242c9496b29a6c1a5a375761fc9650f0a0112a14..e995e5bd57e55455ebc5e945f97061534518d083 100644
--- a/ui/views/widget/widget.h
+++ b/ui/views/widget/widget.h
@@ -284,6 +284,9 @@ class VIEWS_EXPORT Widget : public internal::NativeWidgetDelegate,
     // If true, forces the window to be shown in the taskbar, even for window
     // types that do not appear in the taskbar by default (popup and bubble).
     bool force_show_in_taskbar;
+    // If true, this is a window used for generating off-screen thumbnails and
+    // should not appear in the taskbar.
+    bool thumbnail_window;
     // Only used by X11, for root level windows. Specifies the res_name and
     // res_class fields, respectively, of the WM_CLASS window property. Controls
     // window grouping and desktop file matching in Linux window managers.
@@ -656,6 +659,8 @@ class VIEWS_EXPORT Widget : public internal::NativeWidgetDelegate,
   void set_frame_type(FrameType frame_type) { frame_type_ = frame_type; }
   FrameType frame_type() const { return frame_type_; }
 
+  void SetSavedShowState(ui::WindowShowState saved_show_state) { saved_show_state_ = saved_show_state; }
+
   // Creates an appropriate NonClientFrameView for this widget. The
   // WidgetDelegate is given the first opportunity to create one, followed by
   // the NativeWidget implementation. If both return NULL, a default one is
diff --git a/ui/views/widget/widget_delegate.cc b/ui/views/widget/widget_delegate.cc
index a7696103df3117b7dd32cf0d8786ea80cacd7f37..9a3e5bc85adcd0b28b6eef9639a7fc49ad0ab655 100644
--- a/ui/views/widget/widget_delegate.cc
+++ b/ui/views/widget/widget_delegate.cc
@@ -120,6 +120,9 @@ bool WidgetDelegate::ExecuteWindowsCommand(int command_id) {
   return false;
 }
 
+void WidgetDelegate::HandleKeyboardCode(ui::KeyboardCode code) {
+}
+
 std::string WidgetDelegate::GetWindowName() const {
   return std::string();
 }
diff --git a/ui/views/widget/widget_delegate.h b/ui/views/widget/widget_delegate.h
index bd4d8a7a1de290efe33f61084d1640e85e1badd2..b2a9249b38d29dc54ea0a69fd4271aba2f6972b1 100644
--- a/ui/views/widget/widget_delegate.h
+++ b/ui/views/widget/widget_delegate.h
@@ -104,6 +104,9 @@ class VIEWS_EXPORT WidgetDelegate {
   // was handled, false if it was not.
   virtual bool ExecuteWindowsCommand(int command_id);
 
+  // Execute a command in the window's controller. (vivaldi/espen).
+  virtual void HandleKeyboardCode(ui::KeyboardCode code);
+
   // Returns the window's name identifier. Used to identify this window for
   // state restoration.
   virtual std::string GetWindowName() const;
diff --git a/ui/views/win/hwnd_message_handler.cc b/ui/views/win/hwnd_message_handler.cc
index 5906900d721c2b97a65e7a1c42af0e14eefb4d9a..4646993d25ea8a6da10a14b7882db1222958ac76 100644
--- a/ui/views/win/hwnd_message_handler.cc
+++ b/ui/views/win/hwnd_message_handler.cc
@@ -1225,8 +1225,12 @@ bool HWNDMessageHandler::GetClientAreaInsets(gfx::Insets* insets) const {
     // Windows automatically adds a standard width border to all sides when a
     // window is maximized.
     int border_thickness = GetSystemMetrics(SM_CXSIZEFRAME);
+// NOTE(jarle@vivaldi.com): The adjustment of the border by 1 pixel causes
+// VB-17434, VB-22957 (and many, many duplicates). Ref. Chromium bug 157728.
+#if !defined(VIVALDI_BUILD)
     if (!delegate_->HasFrame())
       border_thickness -= 1;
+#endif
     *insets = gfx::Insets(
         border_thickness, border_thickness, border_thickness, border_thickness);
     return true;
diff --git a/ui/webui/resources/js/util.js b/ui/webui/resources/js/util.js
index 27faa1b60bf992dce5d5a68c16d5e2cdfcc02220..1819166b71bd437d308268bc2ef2dd5ef8568279 100644
--- a/ui/webui/resources/js/util.js
+++ b/ui/webui/resources/js/util.js
@@ -200,6 +200,13 @@ function queryRequiredElement(selectors, opt_context) {
       element, HTMLElement, 'Missing required element: ' + selectors);
 }
 
+// We don't have access to window.parent, so we'll receive the window
+// message like this.
+var vivaldiWindow = null;
+window.addEventListener("message", function(event) {
+  vivaldiWindow = event.source;
+});
+
 // Handle click on a link. If the link points to a chrome: or file: url, then
 // call into the browser to do the navigation.
 ['click', 'auxclick'].forEach(function(eventName) {
@@ -235,6 +242,13 @@ function queryRequiredElement(selectors, opt_context) {
       return;
 
     anchor = /** @type {!HTMLAnchorElement} */ (anchor);
+
+    // Send this URL to vivaldi.
+    if (vivaldiWindow) {
+      e.preventDefault();
+      vivaldiWindow.postMessage(anchor.href, "*");
+    }
+
     if ((anchor.protocol == 'file:' || anchor.protocol == 'about:') &&
         (e.button == 0 || e.button == 1)) {
       chrome.send('navigateToUrl', [
