diff --git a/ash/ash_strings.gyp b/ash/ash_strings.gyp
index e71aa4639d3936cedb3c17979bdcf6f54ad5c5ad..200ab1040f0825edf88c3636e56522786aaba2dd 100644
--- a/ash/ash_strings.gyp
+++ b/ash/ash_strings.gyp
@@ -11,9 +11,11 @@
       'variables': {
         'grit_out_dir': '<(SHARED_INTERMEDIATE_DIR)/ash/strings',
       },
+      'dependencies': [ '<(VIVALDI)/app/vivaldi_resources.gyp:ash_strings' ],
       'actions': [
         {
           'action_name': 'generate_ash_strings',
+          'disabled': 1,
           'variables': {
             'grit_grd_file': 'ash_strings.grd',
           },
diff --git a/ui/base/accelerators/accelerator.cc b/ui/base/accelerators/accelerator.cc
index fa166bdf00915300a5979fe8c93192b5fb548357..638ab30640afa8d20ec55fef85bb9e87b027d2c7 100644
--- a/ui/base/accelerators/accelerator.cc
+++ b/ui/base/accelerators/accelerator.cc
@@ -23,6 +23,8 @@
 #include "ui/events/keycodes/keyboard_code_conversion.h"
 #endif
 
+#include "app/vivaldi_apptools.h"
+
 namespace ui {
 
 namespace {
@@ -230,6 +232,15 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut = l10n_util::GetStringUTF16(string_id);
   }
 
+  if (vivaldi::IsVivaldiRunning()) {
+    // We want to display all function keys. They do not have to be translated.
+    if (key_code_ >= ui::VKEY_F1 && key_code_ <= ui::VKEY_F24) {
+      char buf[4];
+      base::snprintf(buf, 4, "F%d", key_code_ - ui::VKEY_F1 + 1);
+      base::UTF8ToUTF16(buf, strlen(buf), &shortcut);
+    }
+  }
+
   // Checking whether the character used for the accelerator is alphanumeric.
   // If it is not, then we need to adjust the string later on if the locale is
   // right-to-left. See below for more information of why such adjustment is
@@ -242,6 +253,15 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut_rtl.assign(shortcut);
   }
 
+  if (vivaldi::IsVivaldiRunning()) {
+    // We allow all modifier combinations.
+    if (IsShiftDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_SHIFT_MODIFIER, shortcut);
+    if (IsAltDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_ALT_MODIFIER, shortcut);
+    if (IsCtrlDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_CONTROL_MODIFIER, shortcut);
+  } else  {
   if (IsShiftDown())
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_SHIFT_MODIFIER, shortcut);
 
@@ -252,6 +272,7 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_CONTROL_MODIFIER, shortcut);
   else if (IsAltDown())
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_ALT_MODIFIER, shortcut);
+  }
 
   if (IsCmdDown()) {
 #if defined(OS_MACOSX)
diff --git a/ui/base/base_window.h b/ui/base/base_window.h
index 1819405fb4b9b065689a4f1ee4dabdbc5e4f2597..b4caa6889f97b0189f19c867e01ba9fce0b7cf72 100644
--- a/ui/base/base_window.h
+++ b/ui/base/base_window.h
@@ -92,6 +92,9 @@ class UI_BASE_EXPORT BaseWindow {
   // If set to true, the window will stay on top of other windows which do not
   // have this flag enabled.
   virtual void SetAlwaysOnTop(bool always_on_top) = 0;
+
+  // Update the inset. Used to update when a frameless window goes to maximized.
+  virtual void UpdateEventTargeterWithInset() {};
 };
 
 }  // namespace ui
diff --git a/ui/base/resource/resource_bundle.cc b/ui/base/resource/resource_bundle.cc
index a2aa5824a19ead6a593c0d9865b7240eb83b4833..84122040ff7e41e6ce724abc671abeef4941f4a5 100644
--- a/ui/base/resource/resource_bundle.cc
+++ b/ui/base/resource/resource_bundle.cc
@@ -10,6 +10,7 @@
 #include <utility>
 #include <vector>
 
+#include "app/vivaldi_apptools.h"
 #include "base/big_endian.h"
 #include "base/command_line.h"
 #include "base/files/file.h"
@@ -219,7 +220,11 @@ ResourceBundle& ResourceBundle::GetSharedInstance() {
 
 #if !defined(OS_ANDROID)
 bool ResourceBundle::LocaleDataPakExists(const std::string& locale) {
-  return !GetLocaleFilePath(locale, true).empty();
+  bool locale_file_path_exists = !GetLocaleFilePath(locale, true).empty();
+  if (!locale_file_path_exists) {
+    locale_file_path_exists = vivaldi::IsVivaldiExtraLocale(locale);
+  }
+  return locale_file_path_exists;
 }
 #endif  // !defined(OS_ANDROID)
 
@@ -306,10 +311,17 @@ std::string ResourceBundle::LoadLocaleResources(
   if (locale_file_path.empty())
     locale_file_path = GetLocaleFilePath(app_locale, true);
 
-  if (locale_file_path.empty()) {
+  if (locale_file_path.empty()) { // (jarle@vivaldi) load fallback locale if 'unsupported' locale
+    if (pref_locale == "nn") {
+      locale_file_path = GetLocaleFilePath("nb", true); // Bokmål as a fallback for Nynorsk
+    } else {
+      locale_file_path = GetLocaleFilePath("en-US", true);
+    }
+    if (locale_file_path.empty()) {
     // It's possible that there is no locale.pak.
     LOG(WARNING) << "locale_file_path.empty() for locale " << app_locale;
     return std::string();
+    }
   }
 
   scoped_ptr<DataPack> data_pack(
@@ -670,13 +682,13 @@ void ResourceBundle::LoadChromeResources() {
   if (MaterialDesignController::IsModeMaterial()) {
     if (IsScaleFactorSupported(SCALE_FACTOR_100P)) {
       AddMaterialDesignDataPackFromPath(
-          GetResourcesPakFilePath("chrome_material_100_percent.pak"),
+          GetResourcesPakFilePath("vivaldi_material_100_percent.pak"),
           SCALE_FACTOR_100P);
     }
 
     if (IsScaleFactorSupported(SCALE_FACTOR_200P)) {
       AddOptionalMaterialDesignDataPackFromPath(
-          GetResourcesPakFilePath("chrome_material_200_percent.pak"),
+          GetResourcesPakFilePath("vivaldi_material_200_percent.pak"),
           SCALE_FACTOR_200P);
     }
   }
@@ -686,12 +698,12 @@ void ResourceBundle::LoadChromeResources() {
   // scale factor to gfx::ImageSkia::AddRepresentation.
   if (IsScaleFactorSupported(SCALE_FACTOR_100P)) {
     AddDataPackFromPath(GetResourcesPakFilePath(
-        "chrome_100_percent.pak"), SCALE_FACTOR_100P);
+        "vivaldi_100_percent.pak"), SCALE_FACTOR_100P);
   }
 
   if (IsScaleFactorSupported(SCALE_FACTOR_200P)) {
     AddOptionalDataPackFromPath(GetResourcesPakFilePath(
-        "chrome_200_percent.pak"), SCALE_FACTOR_200P);
+        "vivaldi_200_percent.pak"), SCALE_FACTOR_200P);
   }
 }
 
@@ -716,7 +728,7 @@ void ResourceBundle::AddDataPackFromPathInternal(
   data_pack->set_has_only_material_design_assets(has_only_material_assets);
   if (data_pack->LoadFromPath(pack_path)) {
     AddDataPack(data_pack.release());
-  } else if (!optional) {
+  } else if (false) {
     LOG(ERROR) << "Failed to load " << pack_path.value()
                << "\nSome features may not be available.";
   }
diff --git a/ui/base/resource/resource_bundle_android.cc b/ui/base/resource/resource_bundle_android.cc
index ae731d680cb4849d6fd7a78716abd52226a88196..7a31a68550d032acef4c3e86b3dd8d56daa5164f 100644
--- a/ui/base/resource/resource_bundle_android.cc
+++ b/ui/base/resource/resource_bundle_android.cc
@@ -56,8 +56,8 @@ bool LoadFromApkOrFile(const char* apk_path,
 void ResourceBundle::LoadCommonResources() {
   base::FilePath disk_path;
   PathService::Get(ui::DIR_RESOURCE_PAKS_ANDROID, &disk_path);
-  disk_path = disk_path.AppendASCII("chrome_100_percent.pak");
-  if (LoadFromApkOrFile("assets/chrome_100_percent.pak",
+  disk_path = disk_path.AppendASCII("vivaldi_100_percent.pak");
+  if (LoadFromApkOrFile("assets/vivaldi_100_percent.pak",
                         &disk_path,
                         &g_chrome_100_percent_fd,
                         &g_chrome_100_percent_region)) {
diff --git a/ui/base/resource/resource_bundle_ios.mm b/ui/base/resource/resource_bundle_ios.mm
index 88cf829c6dd8c360d7079fddb4c77fe8e351f7b4..948167abb7db62dcb6deb3d8d46e61b36ba7475e 100644
--- a/ui/base/resource/resource_bundle_ios.mm
+++ b/ui/base/resource/resource_bundle_ios.mm
@@ -44,17 +44,17 @@ base::FilePath GetResourcesPakFilePath(NSString* name, NSString* mac_locale) {
 
 void ResourceBundle::LoadCommonResources() {
   if (IsScaleFactorSupported(SCALE_FACTOR_100P)) {
-    AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_100_percent", nil),
+    AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_100_percent", nil),
                         SCALE_FACTOR_100P);
   }
 
   if (IsScaleFactorSupported(SCALE_FACTOR_200P)) {
-    AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_200_percent", nil),
+    AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_200_percent", nil),
                         SCALE_FACTOR_200P);
   }
 
   if (IsScaleFactorSupported(SCALE_FACTOR_300P)) {
-    AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_300_percent", nil),
+    AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_300_percent", nil),
                         SCALE_FACTOR_300P);
   }
 }
diff --git a/ui/base/resource/resource_bundle_mac.mm b/ui/base/resource/resource_bundle_mac.mm
index 74d8261560623953aa66ec7550346d484bd200d3..d9341587feee8c62afa4057b5f31f78fcf55fed9 100644
--- a/ui/base/resource/resource_bundle_mac.mm
+++ b/ui/base/resource/resource_bundle_mac.mm
@@ -55,21 +55,21 @@ void ResourceBundle::LoadCommonResources() {
   // made.
   if (MaterialDesignController::IsModeMaterial()) {
     AddMaterialDesignDataPackFromPath(
-        GetResourcesPakFilePath(@"chrome_material_100_percent", nil),
+        GetResourcesPakFilePath(@"vivaldi_material_100_percent", nil),
         SCALE_FACTOR_100P);
 
     AddOptionalMaterialDesignDataPackFromPath(
-        GetResourcesPakFilePath(@"chrome_material_200_percent", nil),
+        GetResourcesPakFilePath(@"vivaldi_material_200_percent", nil),
         SCALE_FACTOR_200P);
   }
 
-  AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_100_percent",
+  AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_100_percent",
                         nil), SCALE_FACTOR_100P);
 
   // On Mac we load 1x and 2x resources and we let the UI framework decide
   // which one to use.
   if (IsScaleFactorSupported(SCALE_FACTOR_200P)) {
-    AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_200_percent", nil),
+    AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_200_percent", nil),
                         SCALE_FACTOR_200P);
   }
 }
diff --git a/ui/base/ui_base_tests_bundle.gypi b/ui/base/ui_base_tests_bundle.gypi
index 2eeabe88a68f15af2e9b303ac363da4d95b3f8e7..e81c9e53b968ad5d4fd96bdac7442f258e9c11d6 100644
--- a/ui/base/ui_base_tests_bundle.gypi
+++ b/ui/base/ui_base_tests_bundle.gypi
@@ -48,7 +48,7 @@
         'ln',
         '-fns',
         'ui_test.pak',
-        '${BUILT_PRODUCTS_DIR}/${WRAPPER_NAME}/Versions/A/Resources/chrome_100_percent.pak'
+        '${BUILT_PRODUCTS_DIR}/${WRAPPER_NAME}/Versions/A/Resources/vivaldi_100_percent.pak'
       ],
     },
   ],
diff --git a/ui/resources/ui_resources.gyp b/ui/resources/ui_resources.gyp
index 60a93dbd49db768ec9c1440670f836f7e16c819f..065df786ad5fdc3ad0a3e38261c70e8e609b82e7 100644
--- a/ui/resources/ui_resources.gyp
+++ b/ui/resources/ui_resources.gyp
@@ -8,12 +8,16 @@
       # GN version: //ui/resources
       'target_name': 'ui_resources',
       'type': 'none',
+      'dependencies': [
+        '<(VIVALDI)/app/vivaldi_resources.gyp:ui_resources',
+      ],
       'variables': {
         'grit_out_dir': '<(SHARED_INTERMEDIATE_DIR)/ui/resources',
       },
       'actions': [
         {
           'action_name': 'ui_resources',
+          'disabled':1,
           'variables': {
             'grit_grd_file': 'ui_resources.grd',
           },
diff --git a/ui/shell_dialogs/select_file_dialog_win.cc b/ui/shell_dialogs/select_file_dialog_win.cc
index 40215a344fee59fc74d0d75987bfd0e974dbb43a..7429592875dc785c4a56be8a05dbc7d56a93daa9 100644
--- a/ui/shell_dialogs/select_file_dialog_win.cc
+++ b/ui/shell_dialogs/select_file_dialog_win.cc
@@ -641,6 +641,10 @@ bool SelectFileDialogImpl::RunOpenFileDialog(
     else
       ofn.SetInitialSelection(path->DirName(), path->BaseName());
   }
+  // Vivaldi addition
+  if (!title.empty()) {
+    ofn.GetOPENFILENAME()->lpstrTitle = title.c_str();
+  }
 
   if (!filter.empty())
     ofn.GetOPENFILENAME()->lpstrFilter = filter.c_str();
diff --git a/ui/strings/ui_strings.gyp b/ui/strings/ui_strings.gyp
index d0bd0c7f953c1448ae8e815d3ecb0f780e0319a8..96283fb7fe45a5b0b7e8a5b929b27b08845fd351 100644
--- a/ui/strings/ui_strings.gyp
+++ b/ui/strings/ui_strings.gyp
@@ -7,12 +7,14 @@
     {
       'target_name': 'ui_strings',
       'type': 'none',
+      'dependencies': [ '<(VIVALDI)/app/vivaldi_resources.gyp:ui_strings' ],
       'variables': {
         'grit_out_dir': '<(SHARED_INTERMEDIATE_DIR)/ui/strings',
       },
       'actions': [
         {
           'action_name': 'generate_ui_strings',
+          'disabled': 1,
           'variables': {
             'grit_grd_file': 'ui_strings.grd',
           },
diff --git a/ui/views/controls/menu/menu_controller.cc b/ui/views/controls/menu/menu_controller.cc
index 10b2cd914e768af579dec903fc6996c741838c10..06127e8ead9b3d3f3b3b295ae7d080ccf36fb979 100644
--- a/ui/views/controls/menu/menu_controller.cc
+++ b/ui/views/controls/menu/menu_controller.cc
@@ -388,6 +388,10 @@ MenuItemView* MenuController::Run(Widget* parent,
   return ExitMenuRun();
 }
 
+void MenuController::VivaldiSelectItem(MenuItemView* item) {
+  SetSelection(item, SELECTION_OPEN_SUBMENU | SELECTION_UPDATE_IMMEDIATELY);
+}
+
 void MenuController::Cancel(ExitType type) {
   // If the menu has already been destroyed, no further cancellation is
   // needed.  We especially don't want to set the |exit_type_| to a lesser
diff --git a/ui/views/controls/menu/menu_controller.h b/ui/views/controls/menu/menu_controller.h
index 12af0aa18746cde6c38d22e5239e00ee097dc491..7919c8b75e614d270241f03f27ac0b5d4c80ac96 100644
--- a/ui/views/controls/menu/menu_controller.h
+++ b/ui/views/controls/menu/menu_controller.h
@@ -195,6 +195,8 @@ class VIEWS_EXPORT MenuController : public WidgetObserver {
   // Only used for testing.
   bool IsCancelAllTimerRunningForTest();
 
+  void VivaldiSelectItem(MenuItemView* item);
+
   // Only used for testing.
   static void TurnOffMenuSelectionHoldForTest();
 
diff --git a/ui/views/controls/menu/menu_runner_impl.cc b/ui/views/controls/menu/menu_runner_impl.cc
index d02f7de1eff520e79f6e8dd96d6e8424522d3a2d..0f6ad035a33198af44f7ba31fa65272ff02c5dbc 100644
--- a/ui/views/controls/menu/menu_runner_impl.cc
+++ b/ui/views/controls/menu/menu_runner_impl.cc
@@ -17,6 +17,8 @@
 #include "ui/events/win/system_event_state_lookup.h"
 #endif
 
+#include "app/vivaldi_apptools.h"
+
 namespace views {
 namespace internal {
 
@@ -212,6 +214,8 @@ MenuRunner::RunResult MenuRunnerImpl::MenuDone(NotifyType type,
 bool MenuRunnerImpl::ShouldShowMnemonics(MenuButton* button) {
   // Show mnemonics if the button has focus or alt is pressed.
   bool show_mnemonics = button ? button->HasFocus() : false;
+  if (vivaldi::IsVivaldiRunning())
+    show_mnemonics = true;
 #if defined(OS_WIN)
   // This is only needed on Windows.
   if (!show_mnemonics)
diff --git a/ui/views/widget/desktop_aura/desktop_screen_win.cc b/ui/views/widget/desktop_aura/desktop_screen_win.cc
index a8e088c92c6bb3c36d9439ad25047dd783234a1d..7d16072de8ba8782d06bd8d8f727a1b0ac3f74c9 100644
--- a/ui/views/widget/desktop_aura/desktop_screen_win.cc
+++ b/ui/views/widget/desktop_aura/desktop_screen_win.cc
@@ -32,7 +32,7 @@ gfx::Display DesktopScreenWin::GetDisplayMatching(
 }
 
 HWND DesktopScreenWin::GetHWNDFromNativeView(gfx::NativeView window) const {
-  aura::WindowTreeHost* host = window->GetHost();
+  aura::WindowTreeHost* host = window ? window->GetHost() : NULL;
   return host ? host->GetAcceleratedWidget() : NULL;
 }
 
diff --git a/ui/views/widget/desktop_aura/desktop_screen_x11.cc b/ui/views/widget/desktop_aura/desktop_screen_x11.cc
index 07bcfca61899920bbca2f94f4301df74350af4b2..ccc012dbc0abe42f54f3a052dab6b2fdfab0162a 100644
--- a/ui/views/widget/desktop_aura/desktop_screen_x11.cc
+++ b/ui/views/widget/desktop_aura/desktop_screen_x11.cc
@@ -172,6 +172,7 @@ gfx::Display DesktopScreenX11::GetDisplayNearestWindow(
   // created before we create the aura::WindowEventDispatcher. So we ask what
   // the DRWHX11 believes the window bounds are instead of going through the
   // aura::Window's screen bounds.
+  if (window) {
   aura::WindowTreeHost* host = window->GetHost();
   if (host) {
     DesktopWindowTreeHostX11* rwh = DesktopWindowTreeHostX11::GetHostForXID(
@@ -179,6 +180,7 @@ gfx::Display DesktopScreenX11::GetDisplayNearestWindow(
     if (rwh)
       return GetDisplayMatching(rwh->GetX11RootWindowBounds());
   }
+  }
 
   return GetPrimaryDisplay();
 }
diff --git a/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc b/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
index e2fb724b9146b073cbb06b8b2d997c12ba6c4888..d586f2516abd7ff979fb9acde6ad26f6fa3ac767 100644
--- a/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
+++ b/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
@@ -44,6 +44,7 @@
 #include "ui/native_theme/native_theme_aura.h"
 #include "ui/views/corewm/tooltip_aura.h"
 #include "ui/views/linux_ui/linux_ui.h"
+#include "ui/views/widget/widget_delegate.h"
 #include "ui/views/views_delegate.h"
 #include "ui/views/views_switches.h"
 #include "ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h"
@@ -1092,6 +1093,15 @@ void DesktopWindowTreeHostX11::InitX11Window(
   swa.background_pixmap = None;
   swa.bit_gravity = NorthWestGravity;
 
+  // NOTE(espen@vivaldi.com): Set override_redirect for the window we use to
+  // generate thumbnails. It will be opened offscreen. Override_redirect
+  // will disable window managers from forcing the window to open in a visible
+  // area.
+  if (params.thumbnail_window) {
+    swa.override_redirect = True;
+    attribute_mask |= CWOverrideRedirect;
+  }
+
   ::Atom window_type;
   switch (params.type) {
     case Widget::InitParams::TYPE_MENU:
@@ -1504,6 +1514,22 @@ void DesktopWindowTreeHostX11::DispatchMouseEvent(ui::MouseEvent* event) {
   if (event->IsAnyButton() || event->IsMouseWheelEvent())
     FlashFrame(false);
 
+  // Support for page history navigation with mouse. Only when capture to avoid
+  // navigating when a context menu etc is open (vivaldi/espen).
+  if (event->type() == ui::ET_MOUSE_RELEASED && HasCapture()) {
+    bool back_button_pressed =
+      (event->changed_button_flags() == ui::EF_BACK_MOUSE_BUTTON);
+    bool forward_button_pressed =
+      (event->changed_button_flags() == ui::EF_FORWARD_MOUSE_BUTTON);
+    if (back_button_pressed || forward_button_pressed) {
+      Widget* widget = native_widget_delegate_->AsWidget();
+      if (widget && widget->widget_delegate()) {
+        widget->widget_delegate()->HandleKeyboardCode(back_button_pressed ?
+            ui::VKEY_BROWSER_BACK : ui::VKEY_BROWSER_FORWARD);
+      }
+    }
+  }
+
   if (!g_current_capture || g_current_capture == this) {
     SendEventToProcessor(event);
   } else {
diff --git a/ui/views/widget/native_widget_aura.cc b/ui/views/widget/native_widget_aura.cc
index fb2396c8399b45eb78997a79f30e97d0c65c482b..718840853d133820b10e73a5be9cea7c8ac3f89e 100644
--- a/ui/views/widget/native_widget_aura.cc
+++ b/ui/views/widget/native_widget_aura.cc
@@ -1144,6 +1144,8 @@ void NativeWidgetPrivate::ReparentNativeView(gfx::NativeView native_view,
     // matches our previous behaviour; the global stacking client would almost
     // always reattach the window to the same RootWindow.
     aura::Window* root_window = native_view->GetRootWindow();
+    if (!root_window)
+      return;
     aura::client::ParentWindowWithContext(
         native_view, root_window, root_window->GetBoundsInScreen());
   }
diff --git a/ui/views/widget/widget.cc b/ui/views/widget/widget.cc
index 02a570db6da1d350414ef1df80c626d185a186ed..955dfecbd05fd40a186ab5fd9a89e78524e7295f 100644
--- a/ui/views/widget/widget.cc
+++ b/ui/views/widget/widget.cc
@@ -132,6 +132,7 @@ Widget::InitParams::InitParams()
       layer_type(ui::LAYER_TEXTURED),
       context(nullptr),
       force_show_in_taskbar(false),
+      thumbnail_window(false),
       force_software_compositing(false) {
 }
 
@@ -157,6 +158,7 @@ Widget::InitParams::InitParams(Type type)
       layer_type(ui::LAYER_TEXTURED),
       context(nullptr),
       force_show_in_taskbar(false),
+      thumbnail_window(false),
       force_software_compositing(false) {
 }
 
diff --git a/ui/views/widget/widget.h b/ui/views/widget/widget.h
index 80aaddee1a61610b0e154cd8da5cd3398c35f69d..1397b45dda8882709af2dec3ad03bc5d8b41c66c 100644
--- a/ui/views/widget/widget.h
+++ b/ui/views/widget/widget.h
@@ -263,6 +263,9 @@ class VIEWS_EXPORT Widget : public internal::NativeWidgetDelegate,
     // If true, forces the window to be shown in the taskbar, even for window
     // types that do not appear in the taskbar by default (popup and bubble).
     bool force_show_in_taskbar;
+    // If true, this is a window used for generating off-screen thumbnails and
+    // should not appear in the taskbar.
+    bool thumbnail_window;
     // Only used by X11, for root level windows. Specifies the res_name and
     // res_class fields, respectively, of the WM_CLASS window property. Controls
     // window grouping and desktop file matching in Linux window managers.
@@ -641,6 +644,8 @@ class VIEWS_EXPORT Widget : public internal::NativeWidgetDelegate,
   void set_frame_type(FrameType frame_type) { frame_type_ = frame_type; }
   FrameType frame_type() const { return frame_type_; }
 
+  void SetSavedShowState(ui::WindowShowState saved_show_state) { saved_show_state_ = saved_show_state; }
+
   // Creates an appropriate NonClientFrameView for this widget. The
   // WidgetDelegate is given the first opportunity to create one, followed by
   // the NativeWidget implementation. If both return NULL, a default one is
diff --git a/ui/views/widget/widget_delegate.cc b/ui/views/widget/widget_delegate.cc
index e62aa92847e67feb742e9451dda687b6fd920ac3..0d7422fb634c1ead6c1a436274671e83955c51e1 100644
--- a/ui/views/widget/widget_delegate.cc
+++ b/ui/views/widget/widget_delegate.cc
@@ -109,6 +109,9 @@ bool WidgetDelegate::ExecuteWindowsCommand(int command_id) {
   return false;
 }
 
+void WidgetDelegate::HandleKeyboardCode(ui::KeyboardCode code) {
+}
+
 std::string WidgetDelegate::GetWindowName() const {
   return std::string();
 }
diff --git a/ui/views/widget/widget_delegate.h b/ui/views/widget/widget_delegate.h
index 660313d1648bb813b44454bb514f479aa0b696a6..b0252364d544baacf9e79fae8f2a463a6c338639 100644
--- a/ui/views/widget/widget_delegate.h
+++ b/ui/views/widget/widget_delegate.h
@@ -101,6 +101,9 @@ class VIEWS_EXPORT WidgetDelegate {
   // was handled, false if it was not.
   virtual bool ExecuteWindowsCommand(int command_id);
 
+  // Execute a command in the window's controller. (vivaldi/espen).
+  virtual void HandleKeyboardCode(ui::KeyboardCode code);
+
   // Returns the window's name identifier. Used to identify this window for
   // state restoration.
   virtual std::string GetWindowName() const;
diff --git a/ui/views/widget/widget_hwnd_utils.cc b/ui/views/widget/widget_hwnd_utils.cc
index ca561b6d6dc106cce9c94b5c2e38298156845633..e9c150644cd433064ccebf060b67fd9107c3460c 100644
--- a/ui/views/widget/widget_hwnd_utils.cc
+++ b/ui/views/widget/widget_hwnd_utils.cc
@@ -116,7 +116,7 @@ void CalculateWindowStylesFromInitParams(
 
       // See layered window comment above.
       if (*ex_style & WS_EX_COMPOSITED)
-        *style &= ~(WS_THICKFRAME | WS_CAPTION);
+        *style &= ~(/*WS_THICKFRAME |*/ WS_CAPTION);
       break;
     }
     case Widget::InitParams::TYPE_CONTROL:
diff --git a/ui/views/win/hwnd_message_handler.cc b/ui/views/win/hwnd_message_handler.cc
index 0cacffe4a76dfdde167943bf1f232f87736b9a76..a54e956daa12d65a00f70b4ee0a4d57bd7f68a05 100644
--- a/ui/views/win/hwnd_message_handler.cc
+++ b/ui/views/win/hwnd_message_handler.cc
@@ -345,6 +345,7 @@ HWNDMessageHandler::~HWNDMessageHandler() {
 
 void HWNDMessageHandler::Init(HWND parent, const gfx::Rect& bounds) {
   TRACE_EVENT0("views", "HWNDMessageHandler::Init");
+
   GetMonitorAndRects(bounds.ToRECT(), &last_monitor_, &last_monitor_rect_,
                      &last_work_area_);
 
@@ -459,6 +460,9 @@ gfx::Rect HWNDMessageHandler::GetClientAreaBoundsInScreen() const {
   GetClientRect(hwnd(), &r);
   POINT point = { r.left, r.top };
   ClientToScreen(hwnd(), &point);
+  if (point.x == -1 && point.y == -1) { // (jarle) VB-2493 : ClientToScreen returns negative x and y
+    point.x = 0;                        // on the main monitor in a multi-monitor setup (when maximized)
+  }
   return gfx::Rect(point.x, point.y, r.right - r.left, r.bottom - r.top);
 }
 
@@ -1678,7 +1682,9 @@ LRESULT HWNDMessageHandler::OnNCCalcSize(BOOL mode, LPARAM l_param) {
     if (autohide_edges & ViewsDelegate::EDGE_LEFT)
       client_rect->left += kAutoHideTaskbarThicknessPx;
     if (autohide_edges & ViewsDelegate::EDGE_TOP) {
-      if (!delegate_->IsUsingCustomFrame()) {
+      // Vivaldi uses a frameless window so check if the standard frame is
+      // removed. delegate_->IsUsingCustomFrame() does not give us that info.
+      if (!remove_standard_frame_ && !delegate_->IsUsingCustomFrame()) {
         // Tricky bit.  Due to a bug in DwmDefWindowProc()'s handling of
         // WM_NCHITTEST, having any nonclient area atop the window causes the
         // caption buttons to draw onscreen but not respond to mouse
diff --git a/ui/webui/resources/js/util.js b/ui/webui/resources/js/util.js
index 3907c5215aa302660dd221706cfda9e6028ea6c0..e495d8520e89451c310bc1fafae2030e1d86868f 100644
--- a/ui/webui/resources/js/util.js
+++ b/ui/webui/resources/js/util.js
@@ -285,6 +285,13 @@ function queryRequiredElement(selectors, opt_context) {
                           'Missing required element: ' + selectors);
 }
 
+// We don't have access to window.parent, so we'll receive the window
+// message like this.
+var vivaldiWindow = null;
+window.addEventListener("message", function(event) {
+  vivaldiWindow = event.source;
+});
+
 // Handle click on a link. If the link points to a chrome: or file: url, then
 // call into the browser to do the navigation.
 document.addEventListener('click', function(e) {
@@ -298,6 +305,12 @@ document.addEventListener('click', function(e) {
       el = el.parentElement;
     }
 
+    // Send this URL to vivaldi.
+    if (vivaldiWindow) {
+      e.preventDefault();
+      vivaldiWindow.postMessage(el.href, "*");
+    }
+
     if ((el.protocol == 'file:' || el.protocol == 'about:') &&
         (e.button == 0 || e.button == 1)) {
       chrome.send('navigateToUrl', [
