diff --git a/ui/app_list/test/run_all_unittests.cc b/ui/app_list/test/run_all_unittests.cc
index b7db379c1ca5420141a04a735a3534c55e0a8759..7e1a4455a4353c2ec3cc3ab6b2c98506367e66a3 100644
--- a/ui/app_list/test/run_all_unittests.cc
+++ b/ui/app_list/test/run_all_unittests.cc
@@ -38,13 +38,14 @@ class AppListTestSuite : public base::TestSuite {
     base::CommandLine* command_line = base::CommandLine::ForCurrentProcess();
     command_line->AppendSwitchASCII(kTestType, "applist");
 
+    // Vivaldi: Have to init here, because InitializeOneOff check features early
+    base::TestSuite::Initialize();
 #if defined(OS_MACOSX)
     mock_cr_app::RegisterMockCrApp();
 #endif
 #if defined(TOOLKIT_VIEWS)
     gl::GLSurfaceTestSupport::InitializeOneOff();
 #endif
-    base::TestSuite::Initialize();
     ui::RegisterPathProvider();
 
     base::FilePath ui_test_pak_path;
diff --git a/ui/aura/client/drag_drop_client.cc b/ui/aura/client/drag_drop_client.cc
index f7dc619d5907c6f615a79ffc4166d62baf7c31cf..9d7374cc2b54d79be153f31fce95b25fb048e313 100644
--- a/ui/aura/client/drag_drop_client.cc
+++ b/ui/aura/client/drag_drop_client.cc
@@ -27,5 +27,29 @@ DragDropClient* GetDragDropClient(Window* root_window) {
       root_window->GetProperty(kRootWindowDragDropClientKey) : NULL;
 }
 
+int DragDropClient::StartDragAndDrop(
+    const ui::OSExchangeData& data,
+    aura::Window* root_window,
+    aura::Window* source_window,
+    const gfx::Point& screen_location,
+    int operation,
+    ui::DragDropTypes::DragEventSource source) {
+  bool cancelled;
+  return StartDragAndDrop(data, root_window, source_window, screen_location,
+                          operation, source, cancelled);
+}
+
+int DragDropClient::StartDragAndDrop(const ui::OSExchangeData& data,
+                                     aura::Window* root_window,
+                                     aura::Window* source_window,
+                                     const gfx::Point& screen_location,
+                                     int operation,
+                                     ui::DragDropTypes::DragEventSource source,
+                                     bool& cancelled) {
+  cancelled = false;
+  return StartDragAndDrop(data, root_window, source_window, screen_location,
+                          operation, source);
+}
+
 }  // namespace client
 }  // namespace aura
diff --git a/ui/aura/client/drag_drop_client.h b/ui/aura/client/drag_drop_client.h
index 4015f0f1a93773f4073f13c433c8d1790fa2982b..5ec5bf35df192afb57f22a4963f0ea3b136f377c 100644
--- a/ui/aura/client/drag_drop_client.h
+++ b/ui/aura/client/drag_drop_client.h
@@ -34,7 +34,15 @@ class AURA_EXPORT DragDropClient {
                                aura::Window* source_window,
                                const gfx::Point& screen_location,
                                int operation,
-                               ui::DragDropTypes::DragEventSource source) = 0;
+                               ui::DragDropTypes::DragEventSource source);
+
+  virtual int StartDragAndDrop(const ui::OSExchangeData& data,
+                               aura::Window* root_window,
+                               aura::Window* source_window,
+                               const gfx::Point& screen_location,
+                               int operation,
+                               ui::DragDropTypes::DragEventSource source,
+                               bool& cancelled);
 
   // Called when a drag and drop session is cancelled.
   virtual void DragCancel() = 0;
diff --git a/ui/base/BUILD.gn b/ui/base/BUILD.gn
index 8def626ed9c8d82f971f705711273c824d40d18a..d3361b3f17f4d0bc1619b259e2e4d108938dd3ad 100644
--- a/ui/base/BUILD.gn
+++ b/ui/base/BUILD.gn
@@ -982,7 +982,7 @@ test("ui_base_unittests") {
   }
 }
 
-if (is_mac) {
+if (is_mac && !is_vivaldi) {
   mac_framework_bundle("ui_unittests_framework") {
     testonly = true
     deps = [
diff --git a/ui/base/accelerators/accelerator.cc b/ui/base/accelerators/accelerator.cc
index f338a9501c1f843224a1e3ce10f68743085e2cdb..3af4c4671a24bfc53fdeb3a18ec21fdf73c61713 100644
--- a/ui/base/accelerators/accelerator.cc
+++ b/ui/base/accelerators/accelerator.cc
@@ -23,6 +23,8 @@
 #include "ui/events/keycodes/keyboard_code_conversion.h"
 #endif
 
+#include "app/vivaldi_apptools.h"
+
 namespace ui {
 
 namespace {
@@ -226,6 +228,15 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut = l10n_util::GetStringUTF16(string_id);
   }
 
+  if (vivaldi::IsVivaldiRunning()) {
+    // We want to display all function keys. They do not have to be translated.
+    if (key_code_ >= ui::VKEY_F1 && key_code_ <= ui::VKEY_F24) {
+      char buf[4];
+      base::snprintf(buf, 4, "F%d", key_code_ - ui::VKEY_F1 + 1);
+      base::UTF8ToUTF16(buf, strlen(buf), &shortcut);
+    }
+  }
+
   // Checking whether the character used for the accelerator is alphanumeric.
   // If it is not, then we need to adjust the string later on if the locale is
   // right-to-left. See below for more information of why such adjustment is
@@ -238,6 +249,15 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut_rtl.assign(shortcut);
   }
 
+  if (vivaldi::IsVivaldiRunning()) {
+    // We allow all modifier combinations.
+    if (IsShiftDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_SHIFT_MODIFIER, shortcut);
+    if (IsAltDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_ALT_MODIFIER, shortcut);
+    if (IsCtrlDown())
+      shortcut = l10n_util::GetStringFUTF16(IDS_APP_CONTROL_MODIFIER, shortcut);
+  } else  {
   if (IsShiftDown())
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_SHIFT_MODIFIER, shortcut);
 
@@ -248,6 +268,7 @@ base::string16 Accelerator::GetShortcutText() const {
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_CONTROL_MODIFIER, shortcut);
   else if (IsAltDown())
     shortcut = l10n_util::GetStringFUTF16(IDS_APP_ALT_MODIFIER, shortcut);
+  }
 
   if (IsCmdDown()) {
 #if defined(OS_MACOSX)
diff --git a/ui/base/base_window.h b/ui/base/base_window.h
index 5ba8480da3979e0b5ad1d36680d7f0c71a3a9458..ef51d7c20a4f69b7312360872cfce32d91e334de 100644
--- a/ui/base/base_window.h
+++ b/ui/base/base_window.h
@@ -90,6 +90,9 @@ class UI_BASE_EXPORT BaseWindow {
   // If set to true, the window will stay on top of other windows which do not
   // have this flag enabled.
   virtual void SetAlwaysOnTop(bool always_on_top) = 0;
+
+  // Update the inset. Used to update when a frameless window goes to maximized.
+  virtual void UpdateEventTargeterWithInset() {};
 };
 
 }  // namespace ui
diff --git a/ui/base/l10n/l10n_util.cc b/ui/base/l10n/l10n_util.cc
index 440ab1a04d4d37c31f22bc853bb449a913a02d08..6d6ea07474c803a615a718b84e879f96e1d4cc77 100644
--- a/ui/base/l10n/l10n_util.cc
+++ b/ui/base/l10n/l10n_util.cc
@@ -33,6 +33,8 @@
 #include "ui/base/resource/resource_bundle.h"
 #include "ui/base/ui_base_paths.h"
 
+#include "app/vivaldi_apptools.h"
+
 #if defined(OS_ANDROID)
 #include "base/android/locale_utils.h"
 #include "ui/base/l10n/l10n_util_android.h"
@@ -415,6 +417,13 @@ bool CheckAndResolveLocale(const std::string& locale,
 
 std::string GetApplicationLocaleInternal(const std::string& pref_locale) {
 #if defined(OS_MACOSX)
+  if (vivaldi::IsVivaldiRunning()) {
+    if (!pref_locale.empty()) {
+      std::string app_locale = base::i18n::GetCanonicalLocale(pref_locale);
+      if (!app_locale.empty())
+        return app_locale;
+    }
+  }
 
   // Use any override (Cocoa for the browser), otherwise use the preference
   // passed to the function.
@@ -463,6 +472,10 @@ std::string GetApplicationLocaleInternal(const std::string& pref_locale) {
   candidates.push_back(base::android::GetDefaultLocaleString());
 
 #elif defined(USE_GLIB) && !defined(OS_CHROMEOS)
+  if (vivaldi::IsVivaldiRunning()) {
+    if (!pref_locale.empty())
+      candidates.push_back(base::i18n::GetCanonicalLocale(pref_locale));
+  }
 
   // GLib implements correct environment variable parsing with
   // the precedence order: LANGUAGE, LC_ALL, LC_MESSAGES and LANG.
diff --git a/ui/base/resource/resource_bundle.cc b/ui/base/resource/resource_bundle.cc
index 03ab51f52ceb656de3b75a232353bff5e4bd2e82..8157e3c76d92791177f80465e05f58bb5880fa08 100644
--- a/ui/base/resource/resource_bundle.cc
+++ b/ui/base/resource/resource_bundle.cc
@@ -10,6 +10,7 @@
 #include <utility>
 #include <vector>
 
+#include "app/vivaldi_apptools.h"
 #include "base/big_endian.h"
 #include "base/command_line.h"
 #include "base/files/file.h"
@@ -249,7 +250,11 @@ ResourceBundle& ResourceBundle::GetSharedInstance() {
 
 #if !defined(OS_ANDROID)
 bool ResourceBundle::LocaleDataPakExists(const std::string& locale) {
-  return !GetLocaleFilePath(locale, true).empty();
+  bool locale_file_path_exists = !GetLocaleFilePath(locale, true).empty();
+  if (!locale_file_path_exists) {
+    locale_file_path_exists = vivaldi::IsVivaldiExtraLocale(locale);
+  }
+  return locale_file_path_exists;
 }
 #endif  // !defined(OS_ANDROID)
 
@@ -333,10 +338,17 @@ std::string ResourceBundle::LoadLocaleResources(
   if (locale_file_path.empty())
     locale_file_path = GetLocaleFilePath(app_locale, true);
 
-  if (locale_file_path.empty()) {
+  if (locale_file_path.empty()) { // (jarle@vivaldi) load fallback locale if 'unsupported' locale
+    if (pref_locale == "nn") {
+      locale_file_path = GetLocaleFilePath("nb", true); // Bokmål as a fallback for Nynorsk
+    } else {
+      locale_file_path = GetLocaleFilePath("en-US", true);
+    }
+    if (locale_file_path.empty()) {
     // It's possible that there is no locale.pak.
     LOG(WARNING) << "locale_file_path.empty() for locale " << app_locale;
     return std::string();
+    }
   }
 
   std::unique_ptr<DataPack> data_pack(new DataPack(SCALE_FACTOR_100P));
@@ -735,12 +747,12 @@ void ResourceBundle::LoadChromeResources() {
   // scale factor to gfx::ImageSkia::AddRepresentation.
   if (IsScaleFactorSupported(SCALE_FACTOR_100P)) {
     AddDataPackFromPath(GetResourcesPakFilePath(
-        "chrome_100_percent.pak"), SCALE_FACTOR_100P);
+        "vivaldi_100_percent.pak"), SCALE_FACTOR_100P);
   }
 
   if (IsScaleFactorSupported(SCALE_FACTOR_200P)) {
     AddOptionalDataPackFromPath(GetResourcesPakFilePath(
-        "chrome_200_percent.pak"), SCALE_FACTOR_200P);
+        "vivaldi_200_percent.pak"), SCALE_FACTOR_200P);
   }
 }
 
@@ -763,7 +775,7 @@ void ResourceBundle::AddDataPackFromPathInternal(
   std::unique_ptr<DataPack> data_pack(new DataPack(scale_factor));
   if (data_pack->LoadFromPath(pack_path)) {
     AddDataPack(std::move(data_pack));
-  } else if (!optional) {
+  } else if (false) {
     LOG(ERROR) << "Failed to load " << pack_path.value()
                << "\nSome features may not be available.";
   }
diff --git a/ui/base/resource/resource_bundle_mac.mm b/ui/base/resource/resource_bundle_mac.mm
index 3b17113ddd461f0646780b5d73a11fa745a1e346..619e8496035a9b5035ae4ad90c5941a966cce3a8 100644
--- a/ui/base/resource/resource_bundle_mac.mm
+++ b/ui/base/resource/resource_bundle_mac.mm
@@ -48,13 +48,13 @@ base::FilePath GetResourcesPakFilePath(NSString* name, NSString* mac_locale) {
 }  // namespace
 
 void ResourceBundle::LoadCommonResources() {
-  AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_100_percent",
+  AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_100_percent",
                         nil), SCALE_FACTOR_100P);
 
   // On Mac we load 1x and 2x resources and we let the UI framework decide
   // which one to use.
   if (IsScaleFactorSupported(SCALE_FACTOR_200P)) {
-    AddDataPackFromPath(GetResourcesPakFilePath(@"chrome_200_percent", nil),
+    AddDataPackFromPath(GetResourcesPakFilePath(@"vivaldi_200_percent", nil),
                         SCALE_FACTOR_200P);
   }
 }
diff --git a/ui/shell_dialogs/select_file_dialog_win.cc b/ui/shell_dialogs/select_file_dialog_win.cc
index 46804d2ac696ba305b23bc68cce4e4f5ecd9697e..e9d0a8470b604838a464f1cdf850b4335aa19bff 100644
--- a/ui/shell_dialogs/select_file_dialog_win.cc
+++ b/ui/shell_dialogs/select_file_dialog_win.cc
@@ -588,6 +588,10 @@ bool SelectFileDialogImpl::RunOpenFileDialog(
     else
       ofn.SetInitialSelection(path->DirName(), path->BaseName());
   }
+  // Vivaldi addition
+  if (!title.empty()) {
+    ofn.GetOPENFILENAME()->lpstrTitle = title.c_str();
+  }
 
   if (!filter.empty())
     ofn.GetOPENFILENAME()->lpstrFilter = filter.c_str();
diff --git a/ui/views/bubble/bubble_dialog_delegate.cc b/ui/views/bubble/bubble_dialog_delegate.cc
index 735a2237bcb536fb0bb272a2fa06b6abff199ae6..f3fdb512ecb7748ab8395f49906ba55e4f424543 100644
--- a/ui/views/bubble/bubble_dialog_delegate.cc
+++ b/ui/views/bubble/bubble_dialog_delegate.cc
@@ -25,6 +25,8 @@
 #include "ui/base/win/shell.h"
 #endif
 
+#include "app/vivaldi_apptools.h"
+
 namespace views {
 
 namespace {
@@ -76,8 +78,17 @@ Widget* BubbleDialogDelegateView::CreateBubble(
   // the parent frame and let DWM handle compositing.  If not, then we don't
   // want to allow the bubble to extend the frame because it will be clipped.
   bubble_delegate->set_adjust_if_offscreen(ui::win::IsAeroGlassEnabled());
-#elif (defined(OS_LINUX) && !defined(OS_CHROMEOS)) || defined(OS_MACOSX)
-  // Linux clips bubble windows that extend outside their parent window bounds.
+#elif (defined(OS_LINUX) && !defined(OS_CHROMEOS))
+  if (vivaldi::IsVivaldiRunning()) {
+    // In Vivaldi we use some of these bubbles at the bottom of the window, so
+    // we must adjust to try to fit inside the window.
+    bubble_delegate->set_adjust_if_offscreen(true);
+  } else {
+    // Linux clips bubble windows that extend outside their parent window
+    // bounds.
+    bubble_delegate->set_adjust_if_offscreen(false);
+  }
+#elif defined(OS_MACOSX)
   // Mac never adjusts.
   bubble_delegate->set_adjust_if_offscreen(false);
 #endif
diff --git a/ui/views/controls/menu/menu_controller.cc b/ui/views/controls/menu/menu_controller.cc
index 335945f284ed025510dfe538b0f75ded3e3421d5..c4050d836faaf9f194470e648ea5eba7ae66e535 100644
--- a/ui/views/controls/menu/menu_controller.cc
+++ b/ui/views/controls/menu/menu_controller.cc
@@ -496,6 +496,10 @@ MenuItemView* MenuController::Run(Widget* parent,
   return ExitMenuRun();
 }
 
+void MenuController::VivaldiSelectItem(MenuItemView* item) {
+  SetSelection(item, SELECTION_OPEN_SUBMENU | SELECTION_UPDATE_IMMEDIATELY);
+}
+
 void MenuController::Cancel(ExitType type) {
   // If the menu has already been destroyed, no further cancellation is
   // needed.  We especially don't want to set the |exit_type_| to a lesser
diff --git a/ui/views/controls/menu/menu_controller.h b/ui/views/controls/menu/menu_controller.h
index 697e9a5781f25efd1c917c02607a39967087a4fd..598165b299f103e82c7a9e476e0ea220c2ec5bb7 100644
--- a/ui/views/controls/menu/menu_controller.h
+++ b/ui/views/controls/menu/menu_controller.h
@@ -193,6 +193,8 @@ class VIEWS_EXPORT MenuController
   // notifying any menu items.
   void ClearStateForTest();
 
+  void VivaldiSelectItem(MenuItemView* item);
+
   // Only used for testing.
   static void TurnOffMenuSelectionHoldForTest();
 
diff --git a/ui/views/controls/menu/menu_runner_impl.cc b/ui/views/controls/menu/menu_runner_impl.cc
index a247d5129bba444b463610ba9855590974ef70ae..0e68ebe30631a3ee9e44ac1e0db80d5dfb2752bf 100644
--- a/ui/views/controls/menu/menu_runner_impl.cc
+++ b/ui/views/controls/menu/menu_runner_impl.cc
@@ -17,6 +17,8 @@
 #include "ui/events/win/system_event_state_lookup.h"
 #endif
 
+#include "app/vivaldi_apptools.h"
+
 namespace views {
 namespace internal {
 
@@ -220,6 +222,8 @@ MenuRunner::RunResult MenuRunnerImpl::MenuDone(NotifyType type,
 bool MenuRunnerImpl::ShouldShowMnemonics(MenuButton* button) {
   // Show mnemonics if the button has focus or alt is pressed.
   bool show_mnemonics = button ? button->HasFocus() : false;
+  if (vivaldi::IsVivaldiRunning())
+    show_mnemonics = true;
 #if defined(OS_WIN)
   // This is only needed on Windows.
   if (!show_mnemonics)
diff --git a/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc b/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc
index dd0c1947c461fe81bd513cb972b2b63a9cacc6ec..cdbca951b57878539be02cdb64241c14513cafd4 100644
--- a/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc
+++ b/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.cc
@@ -796,7 +796,8 @@ int DesktopDragDropClientAuraX11::StartDragAndDrop(
     aura::Window* source_window,
     const gfx::Point& screen_location,
     int operation,
-    ui::DragDropTypes::DragEventSource source) {
+    ui::DragDropTypes::DragEventSource source,
+    bool& cancelled) {
   UMA_HISTOGRAM_ENUMERATION("Event.DragDrop.Start", source,
                             ui::DragDropTypes::DRAG_EVENT_SOURCE_COUNT);
 
@@ -854,9 +855,11 @@ int DesktopDragDropClientAuraX11::StartDragAndDrop(
 
   if (alive) {
     if (negotiated_operation_ == ui::DragDropTypes::DRAG_NONE) {
+      cancelled = true;
       UMA_HISTOGRAM_ENUMERATION("Event.DragDrop.Cancel", source,
                                 ui::DragDropTypes::DRAG_EVENT_SOURCE_COUNT);
     } else {
+      cancelled = false;
       UMA_HISTOGRAM_ENUMERATION("Event.DragDrop.Drop", source,
                                 ui::DragDropTypes::DRAG_EVENT_SOURCE_COUNT);
     }
@@ -870,6 +873,7 @@ int DesktopDragDropClientAuraX11::StartDragAndDrop(
 
     return negotiated_operation_;
   }
+  cancelled = true;
   UMA_HISTOGRAM_ENUMERATION("Event.DragDrop.Cancel", source,
                             ui::DragDropTypes::DRAG_EVENT_SOURCE_COUNT);
   return ui::DragDropTypes::DRAG_NONE;
diff --git a/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h b/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h
index 732d0454017cc797ae435abb983c4244e834ff55..172a7cde9d77873db2d0f2458364dde69d02f1cf 100644
--- a/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h
+++ b/ui/views/widget/desktop_aura/desktop_drag_drop_client_aurax11.h
@@ -87,7 +87,10 @@ class VIEWS_EXPORT DesktopDragDropClientAuraX11
                        aura::Window* source_window,
                        const gfx::Point& screen_location,
                        int operation,
-                       ui::DragDropTypes::DragEventSource source) override;
+                       ui::DragDropTypes::DragEventSource source,
+                       bool& cancelled) override;
+  using aura::client::DragDropClient::StartDragAndDrop;
+
   void DragCancel() override;
   bool IsDragDropInProgress() override;
 
diff --git a/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.cc b/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.cc
index 67fd2faa13f594ad7f1d0371a8408813be12b413..9c0ea8870ba19d02cfbab44c216760fe1bd1c303 100644
--- a/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.cc
+++ b/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.cc
@@ -12,6 +12,9 @@
 #include "ui/views/widget/desktop_aura/desktop_drop_target_win.h"
 #include "ui/views/widget/desktop_aura/desktop_window_tree_host_win.h"
 
+#include "app/vivaldi_apptools.h"
+#include "ui/dragging/custom_drag_source_win.h"
+
 namespace views {
 
 DesktopDragDropClientWin::DesktopDragDropClientWin(
@@ -34,13 +37,19 @@ int DesktopDragDropClientWin::StartDragAndDrop(
     aura::Window* source_window,
     const gfx::Point& screen_location,
     int operation,
-    ui::DragDropTypes::DragEventSource source) {
+    ui::DragDropTypes::DragEventSource source,
+    bool& cancelled) {
   drag_drop_in_progress_ = true;
   drag_operation_ = operation;
 
   base::WeakPtr<DesktopDragDropClientWin> alive(weak_factory_.GetWeakPtr());
 
-  drag_source_ = Microsoft::WRL::Make<ui::DragSourceWin>();
+  if (vivaldi::IsVivaldiRunning()) {
+    drag_source_ = Microsoft::WRL::Make<vivaldi::CustomDragSourceWin>(
+        vivaldi::IsTabDragInProgress());
+  } else {
+    drag_source_ = Microsoft::WRL::Make<ui::DragSourceWin>();
+  }
   Microsoft::WRL::ComPtr<ui::DragSourceWin> drag_source_copy = drag_source_;
   drag_source_copy->set_data(&data);
   ui::OSExchangeDataProviderWin::GetDataObjectImpl(data)
@@ -59,6 +68,8 @@ int DesktopDragDropClientWin::StartDragAndDrop(
   if (alive)
     drag_drop_in_progress_ = false;
 
+  cancelled = (result == DRAGDROP_S_CANCEL);
+
   if (result != DRAGDROP_S_DROP)
     effect = DROPEFFECT_NONE;
 
diff --git a/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.h b/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.h
index 44c28416eea34b215415f5d4e0c9dc6eefe16b59..073fe97e2d89715966409ea770becc1ffc2e56c6 100644
--- a/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.h
+++ b/ui/views/widget/desktop_aura/desktop_drag_drop_client_win.h
@@ -33,7 +33,8 @@ class VIEWS_EXPORT DesktopDragDropClientWin
                        aura::Window* source_window,
                        const gfx::Point& screen_location,
                        int operation,
-                       ui::DragDropTypes::DragEventSource source) override;
+                       ui::DragDropTypes::DragEventSource source,
+                       bool& cancelled) override;
   void DragCancel() override;
   bool IsDragDropInProgress() override;
 
diff --git a/ui/views/widget/desktop_aura/desktop_screen_win.cc b/ui/views/widget/desktop_aura/desktop_screen_win.cc
index f772f64d656e88cd2631e388f92e00a845f84f16..0c63ea93115b6e2e6a3d1af9023a77d69cfb3605 100644
--- a/ui/views/widget/desktop_aura/desktop_screen_win.cc
+++ b/ui/views/widget/desktop_aura/desktop_screen_win.cc
@@ -32,7 +32,7 @@ display::Display DesktopScreenWin::GetDisplayMatching(
 }
 
 HWND DesktopScreenWin::GetHWNDFromNativeView(gfx::NativeView window) const {
-  aura::WindowTreeHost* host = window->GetHost();
+  aura::WindowTreeHost* host = window ? window->GetHost() : NULL;
   return host ? host->GetAcceleratedWidget() : NULL;
 }
 
diff --git a/ui/views/widget/desktop_aura/desktop_screen_x11.cc b/ui/views/widget/desktop_aura/desktop_screen_x11.cc
index 3432afe40d033984d5c898b9b395a38253eff235..bbc8064b1232ff911b7855db64fbec6107f7a273 100644
--- a/ui/views/widget/desktop_aura/desktop_screen_x11.cc
+++ b/ui/views/widget/desktop_aura/desktop_screen_x11.cc
@@ -183,6 +183,7 @@ display::Display DesktopScreenX11::GetDisplayNearestWindow(
   // created before we create the aura::WindowEventDispatcher. So we ask what
   // the DRWHX11 believes the window bounds are instead of going through the
   // aura::Window's screen bounds.
+  if (window) {
   aura::WindowTreeHost* host = window->GetHost();
   if (host) {
     DesktopWindowTreeHostX11* rwh = DesktopWindowTreeHostX11::GetHostForXID(
@@ -190,6 +191,7 @@ display::Display DesktopScreenX11::GetDisplayNearestWindow(
     if (rwh)
       return GetDisplayMatching(rwh->GetX11RootWindowBounds());
   }
+  }
 
   return GetPrimaryDisplay();
 }
diff --git a/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc b/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
index 448baf898079c28dab37da1087c36257225f5e8c..81be0d3be7a9f3b8414f77a45758ef9453a4d195 100644
--- a/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
+++ b/ui/views/widget/desktop_aura/desktop_window_tree_host_x11.cc
@@ -64,6 +64,8 @@
 #include "ui/wm/core/compound_event_filter.h"
 #include "ui/wm/core/window_util.h"
 
+#include "ui/views/widget/widget_delegate.h"
+
 DECLARE_UI_CLASS_PROPERTY_TYPE(views::DesktopWindowTreeHostX11*);
 
 namespace views {
@@ -1356,6 +1358,15 @@ void DesktopWindowTreeHostX11::InitX11Window(
   swa.background_pixmap = None;
   swa.bit_gravity = NorthWestGravity;
 
+  // NOTE(espen@vivaldi.com): Set override_redirect for the window we use to
+  // generate thumbnails. It will be opened offscreen. Override_redirect
+  // will disable window managers from forcing the window to open in a visible
+  // area.
+  if (params.thumbnail_window) {
+    swa.override_redirect = True;
+    attribute_mask |= CWOverrideRedirect;
+  }
+
   ::Atom window_type;
   switch (params.type) {
     case Widget::InitParams::TYPE_MENU:
@@ -1786,6 +1797,22 @@ void DesktopWindowTreeHostX11::DispatchMouseEvent(ui::MouseEvent* event) {
   if (event->IsAnyButton() || event->IsMouseWheelEvent())
     FlashFrame(false);
 
+  // Support for page history navigation with mouse. Only when capture to avoid
+  // navigating when a context menu etc is open (vivaldi/espen).
+  if (event->type() == ui::ET_MOUSE_RELEASED && HasCapture()) {
+    bool back_button_pressed =
+      (event->changed_button_flags() == ui::EF_BACK_MOUSE_BUTTON);
+    bool forward_button_pressed =
+      (event->changed_button_flags() == ui::EF_FORWARD_MOUSE_BUTTON);
+    if (back_button_pressed || forward_button_pressed) {
+      Widget* widget = native_widget_delegate_->AsWidget();
+      if (widget && widget->widget_delegate()) {
+        widget->widget_delegate()->HandleKeyboardCode(back_button_pressed ?
+            ui::VKEY_BROWSER_BACK : ui::VKEY_BROWSER_FORWARD);
+      }
+    }
+  }
+
   if (!g_current_capture || g_current_capture == this) {
     SendEventToProcessor(event);
   } else {
diff --git a/ui/views/widget/native_widget_aura.cc b/ui/views/widget/native_widget_aura.cc
index 9c2c1b0f8a93f313059fefc14ad323b641308f51..fbb37203fb9e6e18c019ad3e9e037c14a9f1858f 100644
--- a/ui/views/widget/native_widget_aura.cc
+++ b/ui/views/widget/native_widget_aura.cc
@@ -1178,6 +1178,8 @@ void NativeWidgetPrivate::ReparentNativeView(gfx::NativeView native_view,
     // matches our previous behaviour; the global stacking client would almost
     // always reattach the window to the same RootWindow.
     aura::Window* root_window = native_view->GetRootWindow();
+    if (!root_window)
+      return;
     aura::client::ParentWindowWithContext(
         native_view, root_window, root_window->GetBoundsInScreen());
   }
diff --git a/ui/views/widget/widget.cc b/ui/views/widget/widget.cc
index 5ff65fdbd3aeda262b71e96eee4e627472e019f1..8588240bdf51637944def78e817b83bbe4073809 100644
--- a/ui/views/widget/widget.cc
+++ b/ui/views/widget/widget.cc
@@ -136,6 +136,7 @@ Widget::InitParams::InitParams(Type type)
       layer_type(ui::LAYER_TEXTURED),
       context(nullptr),
       force_show_in_taskbar(false),
+      thumbnail_window(false),
       force_software_compositing(false) {}
 
 Widget::InitParams::InitParams(const InitParams& other) = default;
diff --git a/ui/views/widget/widget.h b/ui/views/widget/widget.h
index a7affa69a5dda14fa1270c324f994de108217488..ebc7a547429fdb151ab249f4bf90d525bdfb9f96 100644
--- a/ui/views/widget/widget.h
+++ b/ui/views/widget/widget.h
@@ -283,6 +283,9 @@ class VIEWS_EXPORT Widget : public internal::NativeWidgetDelegate,
     // If true, forces the window to be shown in the taskbar, even for window
     // types that do not appear in the taskbar by default (popup and bubble).
     bool force_show_in_taskbar;
+    // If true, this is a window used for generating off-screen thumbnails and
+    // should not appear in the taskbar.
+    bool thumbnail_window;
     // Only used by X11, for root level windows. Specifies the res_name and
     // res_class fields, respectively, of the WM_CLASS window property. Controls
     // window grouping and desktop file matching in Linux window managers.
@@ -657,6 +660,8 @@ class VIEWS_EXPORT Widget : public internal::NativeWidgetDelegate,
   void set_frame_type(FrameType frame_type) { frame_type_ = frame_type; }
   FrameType frame_type() const { return frame_type_; }
 
+  void SetSavedShowState(ui::WindowShowState saved_show_state) { saved_show_state_ = saved_show_state; }
+
   // Creates an appropriate NonClientFrameView for this widget. The
   // WidgetDelegate is given the first opportunity to create one, followed by
   // the NativeWidget implementation. If both return NULL, a default one is
diff --git a/ui/views/widget/widget_delegate.cc b/ui/views/widget/widget_delegate.cc
index a7696103df3117b7dd32cf0d8786ea80cacd7f37..9a3e5bc85adcd0b28b6eef9639a7fc49ad0ab655 100644
--- a/ui/views/widget/widget_delegate.cc
+++ b/ui/views/widget/widget_delegate.cc
@@ -120,6 +120,9 @@ bool WidgetDelegate::ExecuteWindowsCommand(int command_id) {
   return false;
 }
 
+void WidgetDelegate::HandleKeyboardCode(ui::KeyboardCode code) {
+}
+
 std::string WidgetDelegate::GetWindowName() const {
   return std::string();
 }
diff --git a/ui/views/widget/widget_delegate.h b/ui/views/widget/widget_delegate.h
index bd4d8a7a1de290efe33f61084d1640e85e1badd2..b2a9249b38d29dc54ea0a69fd4271aba2f6972b1 100644
--- a/ui/views/widget/widget_delegate.h
+++ b/ui/views/widget/widget_delegate.h
@@ -104,6 +104,9 @@ class VIEWS_EXPORT WidgetDelegate {
   // was handled, false if it was not.
   virtual bool ExecuteWindowsCommand(int command_id);
 
+  // Execute a command in the window's controller. (vivaldi/espen).
+  virtual void HandleKeyboardCode(ui::KeyboardCode code);
+
   // Returns the window's name identifier. Used to identify this window for
   // state restoration.
   virtual std::string GetWindowName() const;
diff --git a/ui/views/widget/widget_hwnd_utils.cc b/ui/views/widget/widget_hwnd_utils.cc
index b843416dac5fe668527a0032f7df6d5fba63c6fd..5fcb415a02be594afb8d9e76600d5c610f43f0e4 100644
--- a/ui/views/widget/widget_hwnd_utils.cc
+++ b/ui/views/widget/widget_hwnd_utils.cc
@@ -113,7 +113,7 @@ void CalculateWindowStylesFromInitParams(
 
       // See layered window comment above.
       if (*ex_style & WS_EX_COMPOSITED)
-        *style &= ~(WS_THICKFRAME | WS_CAPTION);
+        *style &= ~(/*WS_THICKFRAME |*/ WS_CAPTION);
       break;
     }
     case Widget::InitParams::TYPE_CONTROL:
diff --git a/ui/views/win/hwnd_message_handler.cc b/ui/views/win/hwnd_message_handler.cc
index ac7645344f2e60f7ce4fe9484df2342fe752e60c..a58cb5a804ccadc78dde0a6154bb5fe4aae3c4de 100644
--- a/ui/views/win/hwnd_message_handler.cc
+++ b/ui/views/win/hwnd_message_handler.cc
@@ -1151,8 +1151,12 @@ bool HWNDMessageHandler::GetClientAreaInsets(gfx::Insets* insets) const {
     // Windows automatically adds a standard width border to all sides when a
     // window is maximized.
     int border_thickness = GetSystemMetrics(SM_CXSIZEFRAME);
+// NOTE(jarle@vivaldi.com): The adjustment of the border by 1 pixel causes
+// VB-17434, VB-22957 (and many, many duplicates). Ref. Chromium bug 157728.
+#if !defined(VIVALDI_BUILD)
     if (!delegate_->HasFrame())
       border_thickness -= 1;
+#endif
     *insets = gfx::Insets(
         border_thickness, border_thickness, border_thickness, border_thickness);
     return true;
diff --git a/ui/webui/resources/js/util.js b/ui/webui/resources/js/util.js
index 28a60d2ce1378e6c9609170c4aad62c3d547c856..6b990c35625f8457bf282b6efcccb28023500436 100644
--- a/ui/webui/resources/js/util.js
+++ b/ui/webui/resources/js/util.js
@@ -210,6 +210,13 @@ function queryRequiredElement(selectors, opt_context) {
       element, HTMLElement, 'Missing required element: ' + selectors);
 }
 
+// We don't have access to window.parent, so we'll receive the window
+// message like this.
+var vivaldiWindow = null;
+window.addEventListener("message", function(event) {
+  vivaldiWindow = event.source;
+});
+
 // Handle click on a link. If the link points to a chrome: or file: url, then
 // call into the browser to do the navigation.
 ['click', 'auxclick'].forEach(function(eventName) {
@@ -245,6 +252,13 @@ function queryRequiredElement(selectors, opt_context) {
       return;
 
     anchor = /** @type {!HTMLAnchorElement} */ (anchor);
+
+    // Send this URL to vivaldi.
+    if (vivaldiWindow) {
+      e.preventDefault();
+      vivaldiWindow.postMessage(anchor.href, "*");
+    }
+
     if ((anchor.protocol == 'file:' || anchor.protocol == 'about:') &&
         (e.button == 0 || e.button == 1)) {
       chrome.send('navigateToUrl', [
