diff --git a/net/base/url_util.h b/net/base/url_util.h
index 239a93cd4d742f0011dba1241064a4803b1bb11b..7d081e8be28fa18c54f037a2b8e8190e516e1a25 100644
--- a/net/base/url_util.h
+++ b/net/base/url_util.h
@@ -11,6 +11,7 @@
 #include "base/strings/string16.h"
 #include "base/strings/string_piece.h"
 #include "net/base/net_export.h"
+#include "url/gurl.h"
 #include "url/third_party/mozilla/url_parse.h"
 
 class GURL;
@@ -62,7 +63,7 @@ class NET_EXPORT QueryIterator {
   void Advance();
 
  private:
-  const GURL& url_;
+  GURL url_;
   url::Component query_;
   bool at_end_;
   url::Component key_;
diff --git a/net/ssl/ssl_config.cc b/net/ssl/ssl_config.cc
index e3ced671925e5e0f978fde5f8cb569c964eb094d..616eea70a7f59a8e8f75f5503e310630fdfd13a6 100644
--- a/net/ssl/ssl_config.cc
+++ b/net/ssl/ssl_config.cc
@@ -19,7 +19,7 @@ SSLConfig::CertAndStatus::CertAndStatus() : cert_status(0) {}
 SSLConfig::CertAndStatus::~CertAndStatus() {}
 
 SSLConfig::SSLConfig()
-    : rev_checking_enabled(false),
+    : rev_checking_enabled(true),
       rev_checking_required_local_anchors(false),
       version_min(kDefaultSSLVersionMin),
       version_max(kDefaultSSLVersionMax),
diff --git a/net/test/embedded_test_server/embedded_test_server.cc b/net/test/embedded_test_server/embedded_test_server.cc
index 8f10e1ca301bf55318704d7bb147bc80db2690d6..dc91dbb6d2552f64c37718f8bfedc0a96257037d 100644
--- a/net/test/embedded_test_server/embedded_test_server.cc
+++ b/net/test/embedded_test_server/embedded_test_server.cc
@@ -236,7 +236,8 @@ scoped_refptr<X509Certificate> EmbeddedTestServer::GetCertificate() const {
 
 void EmbeddedTestServer::ServeFilesFromDirectory(
     const base::FilePath& directory) {
-  RegisterRequestHandler(base::Bind(&HandleFileRequest, directory));
+  base::FilePath abs_dir = base::MakeAbsoluteFilePath(directory);
+  RegisterRequestHandler(base::Bind(&HandleFileRequest, abs_dir));
 }
 
 void EmbeddedTestServer::ServeFilesFromSourceDirectory(
