diff --git a/net/http/transport_security_state.cc b/net/http/transport_security_state.cc
index 4ffe9e8796f8676689b0c01517b8a8ae08867f5e..79a86657d9d7b4181c9a0ba43ee78e754f144617 100644
--- a/net/http/transport_security_state.cc
+++ b/net/http/transport_security_state.cc
@@ -757,7 +757,7 @@ TransportSecurityState::TransportSecurityState()
       sent_expect_ct_reports_cache_(kMaxReportCacheEntries) {
 // Static pinning is only enabled for official builds to make sure that
 // others don't end up with pins that cannot be easily updated.
-#if !defined(GOOGLE_CHROME_BUILD) || defined(OS_ANDROID) || defined(OS_IOS)
+#if (!defined(GOOGLE_CHROME_BUILD) && !defined (VIVALDI_BUILD))|| defined(OS_ANDROID) || defined(OS_IOS)
   enable_static_pins_ = false;
   enable_static_expect_ct_ = false;
 #endif
diff --git a/net/ssl/ssl_config.cc b/net/ssl/ssl_config.cc
index 89f81335f0582ea626292a1a7e822bb0175ccd13..efcfc3d51db1e6cf61afaf90f44f8f228a837261 100644
--- a/net/ssl/ssl_config.cc
+++ b/net/ssl/ssl_config.cc
@@ -6,6 +6,8 @@
 
 #include "net/cert/cert_verifier.h"
 
+#include "app/vivaldi_apptools.h"
+
 namespace net {
 
 const uint16_t kDefaultSSLVersionMin = SSL_PROTOCOL_VERSION_TLS1;
@@ -23,7 +25,7 @@ SSLConfig::CertAndStatus::CertAndStatus(const CertAndStatus& other)
 SSLConfig::CertAndStatus::~CertAndStatus() = default;
 
 SSLConfig::SSLConfig()
-    : rev_checking_enabled(false),
+    : rev_checking_enabled(vivaldi::IsVivaldiRunning()), // false when testing
       rev_checking_required_local_anchors(false),
       sha1_local_anchors_enabled(true),
       common_name_fallback_local_anchors_enabled(true),
diff --git a/net/test/embedded_test_server/embedded_test_server.cc b/net/test/embedded_test_server/embedded_test_server.cc
index 83112a2baee65a2d12c2ac58d856cf5e09fa620f..8eead700b4ac5ba2afa409cc7970c0bdfbe5385a 100644
--- a/net/test/embedded_test_server/embedded_test_server.cc
+++ b/net/test/embedded_test_server/embedded_test_server.cc
@@ -291,7 +291,8 @@ scoped_refptr<X509Certificate> EmbeddedTestServer::GetCertificate() const {
 
 void EmbeddedTestServer::ServeFilesFromDirectory(
     const base::FilePath& directory) {
-  RegisterRequestHandler(base::Bind(&HandleFileRequest, directory));
+  base::FilePath abs_dir = base::MakeAbsoluteFilePath(directory);
+  RegisterRequestHandler(base::Bind(&HandleFileRequest, abs_dir));
 }
 
 void EmbeddedTestServer::ServeFilesFromSourceDirectory(
diff --git a/services/service_manager/embedder/main.cc b/services/service_manager/embedder/main.cc
index 7acf0e44fea4f87517236c4adc735c1634532035..069c6626cdca6bc9c930a8acb6361ed41870bc97 100644
--- a/services/service_manager/embedder/main.cc
+++ b/services/service_manager/embedder/main.cc
@@ -76,7 +76,7 @@ namespace {
 
 // Maximum message size allowed to be read from a Mojo message pipe in any
 // service manager embedder process.
-constexpr size_t kMaximumMojoMessageSize = 128 * 1024 * 1024;
+constexpr size_t kMaximumMojoMessageSize = 256 * 1024 * 1024;
 
 class ServiceProcessLauncherDelegateImpl
     : public service_manager::ServiceProcessLauncherDelegate {
diff --git a/services/ui/gpu/gpu_service.cc b/services/ui/gpu/gpu_service.cc
index ef4877ec1bbfa24af83bf76438f68fbf37f0e02d..f808c24d3db90b396015c2677377ed71f712b484 100644
--- a/services/ui/gpu/gpu_service.cc
+++ b/services/ui/gpu/gpu_service.cc
@@ -45,6 +45,10 @@
 #include "media/gpu/content_video_view_overlay_allocator.h"
 #endif
 
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+#include "platform_media/gpu/pipeline/propmedia_gpu_channel_manager.h"
+#endif
+
 namespace ui {
 
 namespace {
@@ -116,6 +120,11 @@ GpuService::~GpuService() {
     wait.Wait();
   }
   media_gpu_channel_manager_.reset();
+
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+  proprietary_media_gpu_channel_manager_.reset();
+#endif
+
   gpu_channel_manager_.reset();
   owned_sync_point_manager_.reset();
 
@@ -192,6 +201,12 @@ void GpuService::InitializeWithHost(mojom::GpuHostPtr gpu_host,
 
   media_gpu_channel_manager_.reset(
       new media::MediaGpuChannelManager(gpu_channel_manager_.get()));
+
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+  proprietary_media_gpu_channel_manager_.reset(
+      new gpu::ProprietaryMediaGpuChannelManager(gpu_channel_manager_.get()));
+#endif
+
   if (watchdog_thread())
     watchdog_thread()->AddPowerObserver();
 }
@@ -350,6 +365,11 @@ void GpuService::DidCreateOffscreenContext(const GURL& active_url) {
 void GpuService::DidDestroyChannel(int client_id) {
   DCHECK(main_runner_->BelongsToCurrentThread());
   media_gpu_channel_manager_->RemoveChannel(client_id);
+
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+  proprietary_media_gpu_channel_manager_->RemoveChannel(client_id);
+#endif
+
   (*gpu_host_)->DidDestroyChannel(client_id);
 }
 
@@ -417,6 +437,10 @@ void GpuService::EstablishGpuChannel(
 
   media_gpu_channel_manager_->AddChannel(client_id);
 
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+  proprietary_media_gpu_channel_manager_->AddChannel(client_id);
+#endif
+
   callback.Run(std::move(pipe.handle1));
 }
 
diff --git a/services/ui/gpu/gpu_service.h b/services/ui/gpu/gpu_service.h
index e92a25a722b0b4e2e8bae2c0322f5b6725026068..0bee644fefae665abeb99e9ecba9e341464eff44 100644
--- a/services/ui/gpu/gpu_service.h
+++ b/services/ui/gpu/gpu_service.h
@@ -31,6 +31,9 @@ class GpuMemoryBufferFactory;
 class GpuWatchdogThread;
 class Scheduler;
 class SyncPointManager;
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+class ProprietaryMediaGpuChannelManager;
+#endif
 }
 
 namespace media {
@@ -186,6 +189,11 @@ class GpuService : public gpu::GpuChannelManagerDelegate,
   std::unique_ptr<gpu::GpuChannelManager> gpu_channel_manager_;
   std::unique_ptr<media::MediaGpuChannelManager> media_gpu_channel_manager_;
 
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+  std::unique_ptr<gpu::ProprietaryMediaGpuChannelManager>
+      proprietary_media_gpu_channel_manager_;
+#endif
+
   // On some platforms (e.g. android webview), the SyncPointManager comes from
   // external sources.
   std::unique_ptr<gpu::SyncPointManager> owned_sync_point_manager_;
diff --git a/services/ui/public/cpp/gpu/gpu.cc b/services/ui/public/cpp/gpu/gpu.cc
index 936dc5d0f89175ff46e486dfe42ea42d35f0cc2a..97478f1cea35f22380f8223fbf7a0dd675d17999 100644
--- a/services/ui/public/cpp/gpu/gpu.cc
+++ b/services/ui/public/cpp/gpu/gpu.cc
@@ -147,6 +147,10 @@ scoped_refptr<gpu::GpuChannelHost> Gpu::EstablishGpuChannelSync() {
   return gpu_channel_;
 }
 
+void Gpu::SetForceAllowAccessToGpu(bool enable) {
+  gpu_->SetForceAllowAccessToGpu(enable);
+}
+
 gpu::GpuMemoryBufferManager* Gpu::GetGpuMemoryBufferManager() {
   return gpu_memory_buffer_manager_.get();
 }
diff --git a/services/ui/public/cpp/gpu/gpu.h b/services/ui/public/cpp/gpu/gpu.h
index c64eeaeb71d9cc178ec9f8711a39975e717b822d..04aa88c4b32b9b3d5b9d109685df8bdd0b2fc25d 100644
--- a/services/ui/public/cpp/gpu/gpu.h
+++ b/services/ui/public/cpp/gpu/gpu.h
@@ -53,6 +53,7 @@ class Gpu : public gpu::GpuChannelHostFactory,
   void EstablishGpuChannel(
       const gpu::GpuChannelEstablishedCallback& callback) override;
   scoped_refptr<gpu::GpuChannelHost> EstablishGpuChannelSync() override;
+  void SetForceAllowAccessToGpu(bool enable) override;
   gpu::GpuMemoryBufferManager* GetGpuMemoryBufferManager() override;
 
  private:
diff --git a/services/ui/public/interfaces/gpu.mojom b/services/ui/public/interfaces/gpu.mojom
index 6bbd3b380658bf1b038a154db0743c3f268125dd..c53cc0af1159600c84e93f43d9cf396313ba026b 100644
--- a/services/ui/public/interfaces/gpu.mojom
+++ b/services/ui/public/interfaces/gpu.mojom
@@ -20,6 +20,8 @@ interface Gpu {
                             handle<message_pipe>? channel_handle,
                             gpu.mojom.GpuInfo gpu_info);
 
+  SetForceAllowAccessToGpu(bool enable);
+
   // Creates a new JpegDecodeAccelerator and binds it to |jda|.
   CreateJpegDecodeAccelerator(media.mojom.GpuJpegDecodeAccelerator& jda);
 
diff --git a/services/ui/ws/gpu_client.h b/services/ui/ws/gpu_client.h
index ea3a106e78250de05f7d7283d88404a43b026dd7..742b8e9a08ccb5c78132906e864f07a2d665f44a 100644
--- a/services/ui/ws/gpu_client.h
+++ b/services/ui/ws/gpu_client.h
@@ -44,6 +44,7 @@ class GpuClient : public mojom::Gpu {
   // mojom::Gpu overrides:
   void EstablishGpuChannel(
       const EstablishGpuChannelCallback& callback) override;
+  void SetForceAllowAccessToGpu(bool enable) override {}
   void CreateJpegDecodeAccelerator(
       media::mojom::GpuJpegDecodeAcceleratorRequest jda_request) override;
   void CreateVideoEncodeAccelerator(
