diff --git a/net/base/url_util.h b/net/base/url_util.h
index 75f7b620be0686210434d30c4452d76efb40997f..6bcc5fd606315dba2d995a45e6b49099a03d8e63 100644
--- a/net/base/url_util.h
+++ b/net/base/url_util.h
@@ -16,6 +16,7 @@
 #include "base/strings/string16.h"
 #include "base/strings/string_piece.h"
 #include "net/base/net_export.h"
+#include "url/gurl.h"
 #include "url/third_party/mozilla/url_parse.h"
 
 class GURL;
@@ -71,7 +72,7 @@ class NET_EXPORT QueryIterator {
   void Advance();
 
  private:
-  const GURL& url_;
+  GURL url_;
   url::Component query_;
   bool at_end_;
   url::Component key_;
diff --git a/net/quic/stream_sequencer_buffer_test.cc b/net/quic/stream_sequencer_buffer_test.cc
index 348205ffd2c7e04fd95d25e9f47880d2bc811576..0300519acf1e76285eaf483affbdd279b0f9dc65 100644
--- a/net/quic/stream_sequencer_buffer_test.cc
+++ b/net/quic/stream_sequencer_buffer_test.cc
@@ -283,39 +283,6 @@ TEST_F(StreamSequencerBufferTest, OnStreamDataWithoutOverlap) {
   EXPECT_TRUE(helper_->CheckBufferInvariants());
 }
 
-TEST_F(StreamSequencerBufferTest, OnStreamDataInLongStreamWithOverlap) {
-  // Assume a stream has already buffered almost 4GB.
-  uint64_t total_bytes_read = pow(2, 32) - 1;
-  helper_->set_total_bytes_read(total_bytes_read);
-  helper_->set_gaps(std::list<Gap>(
-      1, Gap(total_bytes_read, std::numeric_limits<QuicStreamOffset>::max())));
-
-  // Three new out of order frames arrive.
-  const size_t kBytesToWrite = 100;
-  string source(kBytesToWrite, 'a');
-  size_t written;
-  // Frame [2^32 + 500, 2^32 + 600).
-  QuicStreamOffset offset = pow(2, 32) + 500;
-  EXPECT_EQ(
-      QUIC_NO_ERROR,
-      buffer_->OnStreamData(offset, source, clock_.ApproximateNow(), &written));
-  EXPECT_EQ(2, helper_->GapSize());
-
-  // Frame [2^32 + 700, 2^32 + 800).
-  offset = pow(2, 32) + 700;
-  EXPECT_EQ(
-      QUIC_NO_ERROR,
-      buffer_->OnStreamData(offset, source, clock_.ApproximateNow(), &written));
-  EXPECT_EQ(3, helper_->GapSize());
-
-  // Another frame [2^32 + 300, 2^32 + 400).
-  offset = pow(2, 32) + 300;
-  EXPECT_EQ(
-      QUIC_NO_ERROR,
-      buffer_->OnStreamData(offset, source, clock_.ApproximateNow(), &written));
-  EXPECT_EQ(4, helper_->GapSize());
-}
-
 TEST_F(StreamSequencerBufferTest, OnStreamDataTillEnd) {
   // Write 50 bytes to the end.
   const size_t kBytesToWrite = 50;
diff --git a/net/ssl/ssl_config.cc b/net/ssl/ssl_config.cc
index 4a71b5948cbebe5b94d71d0f274817fd425b77da..371dcf0c0565cf5aaf59e673ea6bf8dc38fc8cc0 100644
--- a/net/ssl/ssl_config.cc
+++ b/net/ssl/ssl_config.cc
@@ -19,7 +19,7 @@ SSLConfig::CertAndStatus::CertAndStatus() : cert_status(0) {}
 SSLConfig::CertAndStatus::~CertAndStatus() {}
 
 SSLConfig::SSLConfig()
-    : rev_checking_enabled(false),
+    : rev_checking_enabled(true),
       rev_checking_required_local_anchors(false),
       version_min(kDefaultSSLVersionMin),
       version_max(kDefaultSSLVersionMax),
diff --git a/net/test/embedded_test_server/embedded_test_server.cc b/net/test/embedded_test_server/embedded_test_server.cc
index 8f10e1ca301bf55318704d7bb147bc80db2690d6..dc91dbb6d2552f64c37718f8bfedc0a96257037d 100644
--- a/net/test/embedded_test_server/embedded_test_server.cc
+++ b/net/test/embedded_test_server/embedded_test_server.cc
@@ -236,7 +236,8 @@ scoped_refptr<X509Certificate> EmbeddedTestServer::GetCertificate() const {
 
 void EmbeddedTestServer::ServeFilesFromDirectory(
     const base::FilePath& directory) {
-  RegisterRequestHandler(base::Bind(&HandleFileRequest, directory));
+  base::FilePath abs_dir = base::MakeAbsoluteFilePath(directory);
+  RegisterRequestHandler(base::Bind(&HandleFileRequest, abs_dir));
 }
 
 void EmbeddedTestServer::ServeFilesFromSourceDirectory(
