diff --git a/net/base/url_util.h b/net/base/url_util.h
index 75f7b620be0686210434d30c4452d76efb40997f..6bcc5fd606315dba2d995a45e6b49099a03d8e63 100644
--- a/net/base/url_util.h
+++ b/net/base/url_util.h
@@ -16,6 +16,7 @@
 #include "base/strings/string16.h"
 #include "base/strings/string_piece.h"
 #include "net/base/net_export.h"
+#include "url/gurl.h"
 #include "url/third_party/mozilla/url_parse.h"
 
 class GURL;
@@ -71,7 +72,7 @@ class NET_EXPORT QueryIterator {
   void Advance();
 
  private:
-  const GURL& url_;
+  GURL url_;
   url::Component query_;
   bool at_end_;
   url::Component key_;
diff --git a/net/ssl/ssl_config.cc b/net/ssl/ssl_config.cc
index 4a71b5948cbebe5b94d71d0f274817fd425b77da..371dcf0c0565cf5aaf59e673ea6bf8dc38fc8cc0 100644
--- a/net/ssl/ssl_config.cc
+++ b/net/ssl/ssl_config.cc
@@ -19,7 +19,7 @@ SSLConfig::CertAndStatus::CertAndStatus() : cert_status(0) {}
 SSLConfig::CertAndStatus::~CertAndStatus() {}
 
 SSLConfig::SSLConfig()
-    : rev_checking_enabled(false),
+    : rev_checking_enabled(true),
       rev_checking_required_local_anchors(false),
       version_min(kDefaultSSLVersionMin),
       version_max(kDefaultSSLVersionMax),
diff --git a/net/test/embedded_test_server/embedded_test_server.cc b/net/test/embedded_test_server/embedded_test_server.cc
index eb09b96786793f83dd7b1da3b18fc03f0276e028..979cea8afcd6232501be1d6360b87b4302b4bcab 100644
--- a/net/test/embedded_test_server/embedded_test_server.cc
+++ b/net/test/embedded_test_server/embedded_test_server.cc
@@ -259,7 +259,8 @@ scoped_refptr<X509Certificate> EmbeddedTestServer::GetCertificate() const {
 
 void EmbeddedTestServer::ServeFilesFromDirectory(
     const base::FilePath& directory) {
-  RegisterRequestHandler(base::Bind(&HandleFileRequest, directory));
+  base::FilePath abs_dir = base::MakeAbsoluteFilePath(directory);
+  RegisterRequestHandler(base::Bind(&HandleFileRequest, abs_dir));
 }
 
 void EmbeddedTestServer::ServeFilesFromSourceDirectory(
