diff --git a/net/cert/cert_verify_proc_unittest.cc b/net/cert/cert_verify_proc_unittest.cc
index 93437474b3373e42aa14165d6cfa9dd86d0036e0..0d880ce372313ae1a70ec9c60b57d81eb2bf6a06 100644
--- a/net/cert/cert_verify_proc_unittest.cc
+++ b/net/cert/cert_verify_proc_unittest.cc
@@ -700,6 +700,13 @@ class CertVerifyProcInspectSignatureAlgorithmsTest : public ::testing::Test {
     LOG(INFO) << "Skipping test on iOS because certs with mismatched "
                  "algorithms cannot be imported";
     return false;
+#elif defined(OS_MACOSX)
+    if (base::mac::IsAtLeastOS10_12()) {
+      LOG(INFO) << "Skipping test on macOS >= 10.12 because certs with "
+                   "mismatched algorithms cannot be imported";
+      return false;
+    }
+    return true;
 #else
     return true;
 #endif
diff --git a/net/ssl/ssl_config.cc b/net/ssl/ssl_config.cc
index 726ce48f1e9b53eb14f1bf0c9762a2a3b356a9b2..3357da6323501e3ec36d09156053f67e52d6d8af 100644
--- a/net/ssl/ssl_config.cc
+++ b/net/ssl/ssl_config.cc
@@ -6,6 +6,8 @@
 
 #include "net/cert/cert_verifier.h"
 
+#include "app/vivaldi_apptools.h"
+
 namespace net {
 
 const uint16_t kDefaultSSLVersionMin = SSL_PROTOCOL_VERSION_TLS1;
@@ -21,7 +23,7 @@ SSLConfig::CertAndStatus::CertAndStatus(const CertAndStatus& other)
 SSLConfig::CertAndStatus::~CertAndStatus() = default;
 
 SSLConfig::SSLConfig()
-    : rev_checking_enabled(false),
+    : rev_checking_enabled(vivaldi::IsVivaldiRunning()), // false when testing
       rev_checking_required_local_anchors(false),
       sha1_local_anchors_enabled(true),
       common_name_fallback_local_anchors_enabled(true),
diff --git a/net/test/embedded_test_server/embedded_test_server.cc b/net/test/embedded_test_server/embedded_test_server.cc
index 0f89bd9c9e4ad8aec00628a3a8ea2bdc849abddc..d528a6f070a09d8dcc4ad59abde6cda75640e79d 100644
--- a/net/test/embedded_test_server/embedded_test_server.cc
+++ b/net/test/embedded_test_server/embedded_test_server.cc
@@ -291,7 +291,8 @@ scoped_refptr<X509Certificate> EmbeddedTestServer::GetCertificate() const {
 
 void EmbeddedTestServer::ServeFilesFromDirectory(
     const base::FilePath& directory) {
-  RegisterRequestHandler(base::Bind(&HandleFileRequest, directory));
+  base::FilePath abs_dir = base::MakeAbsoluteFilePath(directory);
+  RegisterRequestHandler(base::Bind(&HandleFileRequest, abs_dir));
 }
 
 void EmbeddedTestServer::ServeFilesFromSourceDirectory(
diff --git a/services/ui/gpu/gpu_service.cc b/services/ui/gpu/gpu_service.cc
index f5e5b1f179bcdf34da4450fb22d8905a4a2a21e7..42115f3187a19f17cf2757cfab930057a8a7a883 100644
--- a/services/ui/gpu/gpu_service.cc
+++ b/services/ui/gpu/gpu_service.cc
@@ -35,6 +35,10 @@
 #include "ui/gl/init/gl_factory.h"
 #include "url/gurl.h"
 
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+#include "platform_media/gpu/pipeline/propmedia_gpu_channel_manager.h"
+#endif
+
 namespace ui {
 
 GpuService::GpuService(const gpu::GPUInfo& gpu_info,
@@ -90,7 +94,11 @@ void GpuService::InitializeWithHost(mojom::GpuHostPtr gpu_host,
   // Defer creation of the render thread. This is to prevent it from handling
   // IPC messages before the sandbox has been enabled and all other necessary
   // initialization has succeeded.
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+  gpu_channel_manager_.reset(new gpu::ProprietaryMediaGpuChannelManager(
+#else
   gpu_channel_manager_.reset(new gpu::GpuChannelManager(
+#endif
       gpu_preferences_, this, watchdog_thread_.get(),
       base::ThreadTaskRunnerHandle::Get().get(), io_runner_.get(),
       shutdown_event ? shutdown_event : &shutdown_event_, sync_point_manager_,
diff --git a/services/ui/public/cpp/gpu/gpu.cc b/services/ui/public/cpp/gpu/gpu.cc
index 45ba2595fcd7a5c9a20961d317c748552c20f130..4a7a2ed8379e7eef4d7bd1d2a76ff6404939fff0 100644
--- a/services/ui/public/cpp/gpu/gpu.cc
+++ b/services/ui/public/cpp/gpu/gpu.cc
@@ -102,11 +102,11 @@ void Gpu::EstablishGpuChannel(
     connector_->BindInterface(ui::mojom::kServiceName, &gpu_);
   else
     interface_provider_->GetInterface(&gpu_);
-  gpu_->EstablishGpuChannel(
+  gpu_->EstablishGpuChannel(false,
       base::Bind(&Gpu::OnEstablishedGpuChannel, base::Unretained(this)));
 }
 
-scoped_refptr<gpu::GpuChannelHost> Gpu::EstablishGpuChannelSync() {
+scoped_refptr<gpu::GpuChannelHost> Gpu::EstablishGpuChannelSync(bool force_access_to_gpu) {
   DCHECK(IsMainThread());
   if (GetGpuChannel())
     return gpu_channel_;
@@ -120,7 +120,7 @@ scoped_refptr<gpu::GpuChannelHost> Gpu::EstablishGpuChannelSync() {
     interface_provider_->GetInterface(&gpu_);
 
   mojo::SyncCallRestrictions::ScopedAllowSyncCall allow_sync_call;
-  if (!gpu_->EstablishGpuChannel(&client_id, &channel_handle, &gpu_info)) {
+  if (!gpu_->EstablishGpuChannel(force_access_to_gpu, &client_id, &channel_handle, &gpu_info)) {
     DLOG(WARNING)
         << "Channel encountered error while establishing gpu channel.";
     return nullptr;
diff --git a/services/ui/public/cpp/gpu/gpu.h b/services/ui/public/cpp/gpu/gpu.h
index f99250a65d4376a258ead4a999646095ac263d7b..03864c7d5efdcc3c00021f3fdb2a31272aa1e592 100644
--- a/services/ui/public/cpp/gpu/gpu.h
+++ b/services/ui/public/cpp/gpu/gpu.h
@@ -50,7 +50,7 @@ class Gpu : public gpu::GpuChannelHostFactory,
   // gpu::GpuChannelEstablishFactory:
   void EstablishGpuChannel(
       const gpu::GpuChannelEstablishedCallback& callback) override;
-  scoped_refptr<gpu::GpuChannelHost> EstablishGpuChannelSync() override;
+  scoped_refptr<gpu::GpuChannelHost> EstablishGpuChannelSync(bool force_access_to_gpu = false) override;
   gpu::GpuMemoryBufferManager* GetGpuMemoryBufferManager() override;
 
  private:
diff --git a/services/ui/public/interfaces/gpu.mojom b/services/ui/public/interfaces/gpu.mojom
index 552b9ebc9dc68071fc4ebd7eebbddda1e7d66c45..7df1925123cce3d400c481831d7f3078b4ef9cd5 100644
--- a/services/ui/public/interfaces/gpu.mojom
+++ b/services/ui/public/interfaces/gpu.mojom
@@ -14,7 +14,7 @@ interface Gpu {
   // client. The GPU service responds with client ID, IPC handle and
   // GPUInfo.
   [Sync]
-  EstablishGpuChannel() => (int32 client_id,
+  EstablishGpuChannel(bool force_access_to_gpu) => (int32 client_id,
                             handle<message_pipe>? channel_handle,
                             gpu.mojom.GpuInfo gpu_info);
 
diff --git a/services/ui/ws/gpu_host.cc b/services/ui/ws/gpu_host.cc
index bfa771d2719096c64545be4756218e0c0cb422b4..a0ba66cd1fbf9e70577788ccdef22644867fdf1b 100644
--- a/services/ui/ws/gpu_host.cc
+++ b/services/ui/ws/gpu_host.cc
@@ -57,7 +57,7 @@ class GpuClient : public mojom::Gpu {
   }
 
   // mojom::Gpu overrides:
-  void EstablishGpuChannel(
+  void EstablishGpuChannel(bool force_access_to_gpu,
       const EstablishGpuChannelCallback& callback) override {
     // TODO(sad): crbug.com/617415 figure out how to generate a meaningful
     // tracing id.
