diff --git a/net/http/transport_security_state.cc b/net/http/transport_security_state.cc
index a433afd0217830b49ddb3381ad55afeae2381840..1ba82c52519ff1f913333c8b4b09a180fb472216 100644
--- a/net/http/transport_security_state.cc
+++ b/net/http/transport_security_state.cc
@@ -758,7 +758,7 @@ TransportSecurityState::TransportSecurityState()
       sent_expect_ct_reports_cache_(kMaxReportCacheEntries) {
 // Static pinning is only enabled for official builds to make sure that
 // others don't end up with pins that cannot be easily updated.
-#if !defined(GOOGLE_CHROME_BUILD) || defined(OS_ANDROID) || defined(OS_IOS)
+#if (!defined(GOOGLE_CHROME_BUILD) && !defined (VIVALDI_BUILD))|| defined(OS_ANDROID) || defined(OS_IOS)
   enable_static_pins_ = false;
   enable_static_expect_ct_ = false;
 #endif
diff --git a/net/ssl/ssl_config.cc b/net/ssl/ssl_config.cc
index a460ebe038f6d512cb8da3c575141a0039851d68..3857278850ab423f0c136a5ae07c388d56daaabc 100644
--- a/net/ssl/ssl_config.cc
+++ b/net/ssl/ssl_config.cc
@@ -6,6 +6,8 @@
 
 #include "net/cert/cert_verifier.h"
 
+#include "app/vivaldi_apptools.h"
+
 namespace net {
 
 const uint16_t kDefaultSSLVersionMin = SSL_PROTOCOL_VERSION_TLS1;
@@ -22,7 +24,7 @@ SSLConfig::CertAndStatus::CertAndStatus(const CertAndStatus& other) = default;
 SSLConfig::CertAndStatus::~CertAndStatus() = default;
 
 SSLConfig::SSLConfig()
-    : rev_checking_enabled(false),
+    : rev_checking_enabled(vivaldi::IsVivaldiRunning()), // false when testing
       rev_checking_required_local_anchors(false),
       sha1_local_anchors_enabled(true),
       common_name_fallback_local_anchors_enabled(true),
diff --git a/net/test/embedded_test_server/embedded_test_server.cc b/net/test/embedded_test_server/embedded_test_server.cc
index de416f22aed61545158089a30b9c54af6cc643cd..ac08ee4b0bacf42467179b1ced2ee62d8b053539 100644
--- a/net/test/embedded_test_server/embedded_test_server.cc
+++ b/net/test/embedded_test_server/embedded_test_server.cc
@@ -296,7 +296,8 @@ scoped_refptr<X509Certificate> EmbeddedTestServer::GetCertificate() const {
 
 void EmbeddedTestServer::ServeFilesFromDirectory(
     const base::FilePath& directory) {
-  RegisterRequestHandler(base::Bind(&HandleFileRequest, directory));
+  base::FilePath abs_dir = base::MakeAbsoluteFilePath(directory);
+  RegisterRequestHandler(base::Bind(&HandleFileRequest, abs_dir));
 }
 
 void EmbeddedTestServer::ServeFilesFromSourceDirectory(
diff --git a/services/service_manager/embedder/main.cc b/services/service_manager/embedder/main.cc
index 08ea94a11772017ff760ec88ba4835eea0231df2..9b84c25893b2dd4b9206bac6a696f2633c645433 100644
--- a/services/service_manager/embedder/main.cc
+++ b/services/service_manager/embedder/main.cc
@@ -76,7 +76,7 @@ namespace {
 
 // Maximum message size allowed to be read from a Mojo message pipe in any
 // service manager embedder process.
-constexpr size_t kMaximumMojoMessageSize = 128 * 1024 * 1024;
+constexpr size_t kMaximumMojoMessageSize = 256 * 1024 * 1024;
 
 class ServiceProcessLauncherDelegateImpl
     : public service_manager::ServiceProcessLauncherDelegate {
diff --git a/services/ui/public/cpp/gpu/gpu.cc b/services/ui/public/cpp/gpu/gpu.cc
index 35e067f35162e85a0fc82512b929094e0384d242..f0f75bc4e27a20bab288ca2de7aba828aeddcfe0 100644
--- a/services/ui/public/cpp/gpu/gpu.cc
+++ b/services/ui/public/cpp/gpu/gpu.cc
@@ -259,6 +259,10 @@ scoped_refptr<gpu::GpuChannelHost> Gpu::EstablishGpuChannelSync(
   return gpu_channel_;
 }
 
+void Gpu::SetForceAllowAccessToGpu(bool enable) {
+  (*gpu_)->SetForceAllowAccessToGpu(enable);
+}
+
 gpu::GpuMemoryBufferManager* Gpu::GetGpuMemoryBufferManager() {
   return gpu_memory_buffer_manager_.get();
 }
diff --git a/services/ui/public/cpp/gpu/gpu.h b/services/ui/public/cpp/gpu/gpu.h
index 933d0dc67bbeb0e8db177c130a565501265472e0..4e2341a5f4c5fba52fb7b9f13fcaa3821345112c 100644
--- a/services/ui/public/cpp/gpu/gpu.h
+++ b/services/ui/public/cpp/gpu/gpu.h
@@ -50,6 +50,7 @@ class Gpu : public gpu::GpuChannelEstablishFactory {
       const gpu::GpuChannelEstablishedCallback& callback) override;
   scoped_refptr<gpu::GpuChannelHost> EstablishGpuChannelSync(
       bool* connection_error) override;
+  void SetForceAllowAccessToGpu(bool enable) override;
   gpu::GpuMemoryBufferManager* GetGpuMemoryBufferManager() override;
 
   void LoseChannel();
diff --git a/services/ui/public/interfaces/gpu.mojom b/services/ui/public/interfaces/gpu.mojom
index f0806bbaf2515ec58261d802adbdd200b1b72b71..0e218d327d7012d17bb2dcfb5f6228e5fe83a0b2 100644
--- a/services/ui/public/interfaces/gpu.mojom
+++ b/services/ui/public/interfaces/gpu.mojom
@@ -21,6 +21,8 @@ interface Gpu {
                             gpu.mojom.GpuInfo gpu_info,
                             gpu.mojom.GpuFeatureInfo gpu_feature_info);
 
+  SetForceAllowAccessToGpu(bool enable);
+
   // Creates a new JpegDecodeAccelerator and binds it to |jda|.
   CreateJpegDecodeAccelerator(media.mojom.JpegDecodeAccelerator& jda);
 
diff --git a/services/ui/ws/gpu_client.h b/services/ui/ws/gpu_client.h
index 9887a5f1b1329ad63250f7b19e3bb93ef00c0b85..089262f10188dada21ee74be3acd8ef45064c256 100644
--- a/services/ui/ws/gpu_client.h
+++ b/services/ui/ws/gpu_client.h
@@ -46,6 +46,7 @@ class GpuClient : public mojom::Gpu {
   // mojom::Gpu overrides:
   void EstablishGpuChannel(
       const EstablishGpuChannelCallback& callback) override;
+  void SetForceAllowAccessToGpu(bool enable) override {}
   void CreateJpegDecodeAccelerator(
       media::mojom::JpegDecodeAcceleratorRequest jda_request) override;
   void CreateVideoEncodeAcceleratorProvider(
