diff --git a/net/base/url_util.h b/net/base/url_util.h
index 533c908e553eb5dfe9fcaa726a25793f35dfde86..3db7716d69cd6afc55ef33650794eb4ff7a0c6c1 100644
--- a/net/base/url_util.h
+++ b/net/base/url_util.h
@@ -9,6 +9,7 @@
 
 #include "base/compiler_specific.h"
 #include "net/base/net_export.h"
+#include "url/gurl.h"
 #include "url/third_party/mozilla/url_parse.h"
 
 class GURL;
@@ -60,7 +61,7 @@ class NET_EXPORT QueryIterator {
   void Advance();
 
  private:
-  const GURL& url_;
+  GURL url_;
   url::Component query_;
   bool at_end_;
   url::Component key_;
diff --git a/net/cert/cert_verify_proc_unittest.cc b/net/cert/cert_verify_proc_unittest.cc
index 75a19d2b494c718fc9cc4071ade5e8b2663986ec..e06c4858f845c1154ea660c9cd6318f181de9ef3 100644
--- a/net/cert/cert_verify_proc_unittest.cc
+++ b/net/cert/cert_verify_proc_unittest.cc
@@ -168,7 +168,8 @@ TEST_F(CertVerifyProcTest, DISABLED_WithoutRevocationChecking) {
 // expired, http://crbug.com/502818
 #define MAYBE_EVVerification DISABLED_EVVerification
 #endif
-TEST_F(CertVerifyProcTest, MAYBE_EVVerification) {
+// tomas@vivaldi.com - disabled test (VB-7474)
+TEST_F(CertVerifyProcTest, DISABLED_EVVerification) {
   CertificateList certs = CreateCertificateListFromFile(
       GetTestCertsDirectory(),
       "comodo.chain.pem",
diff --git a/net/net.gyp b/net/net.gyp
index e135c313d3f1d854dc6e7c36884f20af948c6a51..9a84fe420aa544c4bcffe667b58357e7f17a9894 100644
--- a/net/net.gyp
+++ b/net/net.gyp
@@ -443,59 +443,59 @@
         }],
       ],
     },
-    {
-      'target_name': 'net_perftests',
-      'type': 'executable',
-      'dependencies': [
-        '../base/base.gyp:base',
-        '../base/base.gyp:base_i18n',
-        '../base/base.gyp:test_support_perf',
-        '../testing/gtest.gyp:gtest',
-        '../url/url.gyp:url_lib',
-        'net',
-        'net_extras',
-        'net_test_support',
-      ],
-      'sources': [
-        'base/mime_sniffer_perftest.cc',
-        'cookies/cookie_monster_perftest.cc',
-        'disk_cache/blockfile/disk_cache_perftest.cc',
-        'extras/sqlite/sqlite_persistent_cookie_store_perftest.cc',
-        'proxy/proxy_resolver_perftest.cc',
-        'udp/udp_socket_perftest.cc',
-        'websockets/websocket_frame_perftest.cc',
-      ],
-      'conditions': [
-        [ 'use_v8_in_net==1', {
-            'dependencies': [
-              'net_with_v8',
-            ],
-          }, {  # else: !use_v8_in_net
-            'sources!': [
-              'proxy/proxy_resolver_perftest.cc',
-            ],
-          },
-        ],
-        [ 'OS == "win"', {
-            'conditions': [
-              [ 'icu_use_data_file_flag == 0', {
-                # This is needed to trigger the dll copy step on windows.
-                # TODO(mark): Specifying this here shouldn't be necessary.
-                'dependencies': [
-                  '../third_party/icu/icu.gyp:icudata',
-                ],
-              }],
-            ],
-            # TODO(jschuh): crbug.com/167187 fix size_t to int truncations.
-            'msvs_disabled_warnings': [4267, ],
-        }],
-        [ 'enable_websockets != 1', {
-          'sources!': [
-            'websockets/websocket_frame_perftest.cc',
-          ],
-        }],
-      ],
-    },
+    #{
+    #  'target_name': 'net_perftests',
+    #  'type': 'executable',
+    #  'dependencies': [
+    #    '../base/base.gyp:base',
+    #    '../base/base.gyp:base_i18n',
+    #    '../base/base.gyp:test_support_perf',
+    #    '../testing/gtest.gyp:gtest',
+    #    '../url/url.gyp:url_lib',
+    #    'net',
+    #   'net_extras',
+    #    'net_test_support',
+    #  ],
+    #  'sources': [
+    #   'base/mime_sniffer_perftest.cc',
+    #    'cookies/cookie_monster_perftest.cc',
+    #    'disk_cache/blockfile/disk_cache_perftest.cc',
+    #   'extras/sqlite/sqlite_persistent_cookie_store_perftest.cc',
+    #    'proxy/proxy_resolver_perftest.cc',
+    #    'udp/udp_socket_perftest.cc',
+    #    'websockets/websocket_frame_perftest.cc',
+    #  ],
+    #  'conditions': [
+    #    [ 'use_v8_in_net==1', {
+    #        'dependencies': [
+    #          'net_with_v8',
+    #        ],
+    #      }, {  # else: !use_v8_in_net
+    #        'sources!': [
+    #          'proxy/proxy_resolver_perftest.cc',
+    #        ],
+    #      },
+    #    ],
+    #    [ 'OS == "win"', {
+    #        'conditions': [
+    #          [ 'icu_use_data_file_flag == 0', {
+    #            # This is needed to trigger the dll copy step on windows.
+    #            # TODO(mark): Specifying this here shouldn't be necessary.
+    #            'dependencies': [
+    #              '../third_party/icu/icu.gyp:icudata',
+    #            ],
+    #          }],
+    #        ],
+    #        # TODO(jschuh): crbug.com/167187 fix size_t to int truncations.
+    #        'msvs_disabled_warnings': [4267, ],
+    #    }],
+    #    [ 'enable_websockets != 1', {
+    #      'sources!': [
+    #        'websockets/websocket_frame_perftest.cc',
+    #      ],
+    #    }],
+    #  ],
+    #},
     {
       'target_name': 'net_test_support',
       'type': 'static_library',
@@ -787,30 +787,30 @@
         'tools/quic/spdy_balsa_utils.h',
       ],
     },
-    {
-      'target_name': 'dump_cache',
-      'type': 'executable',
-      'dependencies': [
-        '../base/base.gyp:base',
-        'net',
-        'net_test_support',
-      ],
-      'sources': [
-        'tools/dump_cache/cache_dumper.cc',
-        'tools/dump_cache/cache_dumper.h',
-        'tools/dump_cache/dump_cache.cc',
-        'tools/dump_cache/dump_files.cc',
-        'tools/dump_cache/dump_files.h',
-        'tools/dump_cache/simple_cache_dumper.cc',
-        'tools/dump_cache/simple_cache_dumper.h',
-        'tools/dump_cache/url_to_filename_encoder.cc',
-        'tools/dump_cache/url_to_filename_encoder.h',
-        'tools/dump_cache/url_utilities.cc',
-        'tools/dump_cache/url_utilities.h',
-      ],
-      # TODO(jschuh): crbug.com/167187 fix size_t to int truncations.
-      'msvs_disabled_warnings': [4267, ],
-    },
+    #{
+    #  'target_name': 'dump_cache',
+    #  'type': 'executable',
+    #  'dependencies': [
+    #    '../base/base.gyp:base',
+    #    'net',
+    #    'net_test_support',
+    #  ],
+    #  'sources': [
+    #    'tools/dump_cache/cache_dumper.cc',
+    #    'tools/dump_cache/cache_dumper.h',
+    #    'tools/dump_cache/dump_cache.cc',
+    #    'tools/dump_cache/dump_files.cc',
+    #    'tools/dump_cache/dump_files.h',
+    #    'tools/dump_cache/simple_cache_dumper.cc',
+    #    'tools/dump_cache/simple_cache_dumper.h',
+    #    'tools/dump_cache/url_to_filename_encoder.cc',
+    #    'tools/dump_cache/url_to_filename_encoder.h',
+    #    'tools/dump_cache/url_utilities.cc',
+    #    'tools/dump_cache/url_utilities.h',
+    #  ],
+    #  # TODO(jschuh): crbug.com/167187 fix size_t to int truncations.
+    #  'msvs_disabled_warnings': [4267, ],
+    #},
     {
       'target_name': 'simple_quic_tools',
       'type': 'static_library',
@@ -968,7 +968,7 @@
         },
       ],
     }],
-    ['OS != "ios" and OS != "android"', {
+    ['0 and (OS != "ios" and OS != "android")', {
       'targets': [
         # iOS doesn't have the concept of simple executables, these targets
         # can't be compiled on the platform.
@@ -1229,41 +1229,41 @@
             'tools/flip_server/streamer_interface.h',
           ],
         },
-        {
-          'target_name': 'flip_in_mem_edsm_server_unittests',
-          'type': 'executable',
-          'dependencies': [
-              '../testing/gtest.gyp:gtest',
-              '../testing/gmock.gyp:gmock',
-              '../third_party/boringssl/boringssl.gyp:boringssl',
-              'flip_in_mem_edsm_server_base',
-              'net',
-              'net_test_support',
-          ],
-          'sources': [
-            'tools/flip_server/flip_test_utils.cc',
-            'tools/flip_server/flip_test_utils.h',
-            'tools/flip_server/http_interface_test.cc',
-            'tools/flip_server/mem_cache_test.cc',
-            'tools/flip_server/run_all_tests.cc',
-            'tools/flip_server/spdy_interface_test.cc',
-          ],
-        },
-        {
-          'target_name': 'flip_in_mem_edsm_server',
-          'type': 'executable',
-          'cflags': [
-            '-Wno-deprecated',
-          ],
-          'dependencies': [
-            '../base/base.gyp:base',
-            'flip_in_mem_edsm_server_base',
-            'net',
-          ],
-          'sources': [
-            'tools/flip_server/flip_in_mem_edsm_server.cc',
-          ],
-        },
+        #{
+        #  'target_name': 'flip_in_mem_edsm_server_unittests',
+        #  'type': 'executable',
+        #  'dependencies': [
+        #      '../testing/gtest.gyp:gtest',
+        #      '../testing/gmock.gyp:gmock',
+        #      '../third_party/boringssl/boringssl.gyp:boringssl',
+        #      'flip_in_mem_edsm_server_base',
+        #      'net',
+        #      'net_test_support',
+        #  ],
+        #  'sources': [
+        #    'tools/flip_server/flip_test_utils.cc',
+        #    'tools/flip_server/flip_test_utils.h',
+        #    'tools/flip_server/http_interface_test.cc',
+        #    'tools/flip_server/mem_cache_test.cc',
+        #    'tools/flip_server/run_all_tests.cc',
+        #    'tools/flip_server/spdy_interface_test.cc',
+        #  ],
+        #},
+        #{
+        #  'target_name': 'flip_in_mem_edsm_server',
+        #  'type': 'executable',
+        #  'cflags': [
+        #    '-Wno-deprecated',
+        #  ],
+        #  'dependencies': [
+        #    '../base/base.gyp:base',
+        #    'flip_in_mem_edsm_server_base',
+        #    'net',
+        #  ],
+        #  'sources': [
+        #    'tools/flip_server/flip_in_mem_edsm_server.cc',
+        #  ],
+        #},
         {
           'target_name': 'epoll_quic_tools',
           'type': 'static_library',
@@ -1295,33 +1295,33 @@
             'tools/quic/quic_socket_utils.h',
           ],
         },
-        {
-          'target_name': 'epoll_quic_client',
-          'type': 'executable',
-          'dependencies': [
-            '../base/base.gyp:base',
-            'net',
-            'epoll_quic_tools',
-            'simple_quic_tools',
-          ],
-          'sources': [
-            'tools/quic/quic_client_bin.cc',
-          ],
-        },
-        {
-          'target_name': 'epoll_quic_server',
-          'type': 'executable',
-          'dependencies': [
-            '../base/base.gyp:base',
-            'net',
-            'net_quic_proto',
-            'epoll_quic_tools',
-            'simple_quic_tools',
-          ],
-          'sources': [
-            'tools/quic/quic_server_bin.cc',
-          ],
-        },
+        #{
+        #  'target_name': 'epoll_quic_client',
+        #  'type': 'executable',
+        #  'dependencies': [
+        #    '../base/base.gyp:base',
+        #    'net',
+        #    'epoll_quic_tools',
+        #    'simple_quic_tools',
+        #  ],
+        #  'sources': [
+        #    'tools/quic/quic_client_bin.cc',
+        #  ],
+        #},
+        #{
+        #  'target_name': 'epoll_quic_server',
+        #  'type': 'executable',
+        #  'dependencies': [
+        #    '../base/base.gyp:base',
+        #    'net',
+        #    'net_quic_proto',
+        #    'epoll_quic_tools',
+        #    'simple_quic_tools',
+        #  ],
+        #  'sources': [
+        #    'tools/quic/quic_server_bin.cc',
+        #  ],
+        #},
       ]
     }],
     ['OS=="android"', {
@@ -1561,7 +1561,7 @@
   
       ],
     }],
-    ['OS == "android" or OS == "linux"', {
+    ['0 and (OS == "android" or OS == "linux")', {
       'targets': [
         {
           'target_name': 'disk_cache_memory_test',
diff --git a/net/net.gypi b/net/net.gypi
index e68ac6bc4685d1ffd53f1f9030683efbcb5d3ae8..2455db69d82d936091f1bb4108c69b9eef939b7d 100644
--- a/net/net.gypi
+++ b/net/net.gypi
@@ -422,6 +422,7 @@
       'cookies/cookie_util.h',
       'cookies/parsed_cookie.cc',
       'cookies/parsed_cookie.h',
+      '../crypto/scoped_openssl_types.h',
       'disk_cache/blockfile/addr.cc',
       'disk_cache/blockfile/addr.h',
       'disk_cache/blockfile/backend_impl.cc',
diff --git a/net/quic/quic_end_to_end_unittest.cc b/net/quic/quic_end_to_end_unittest.cc
index a07d32d3bc6873b4ec44cbfca11ea5cd85212b74..ad73c64a07192554d3976da1cb7262c0ba37254f 100644
--- a/net/quic/quic_end_to_end_unittest.cc
+++ b/net/quic/quic_end_to_end_unittest.cc
@@ -224,7 +224,8 @@ class QuicEndToEndTest : public PlatformTest {
   bool strike_register_no_startup_period_;
 };
 
-TEST_F(QuicEndToEndTest, LargeGetWithNoPacketLoss) {
+// TODO reenable test for Vivaldi
+TEST_F(QuicEndToEndTest, DISABLED_LargeGetWithNoPacketLoss) {
   std::string response(10 * 1024, 'x');
 
   AddToCache(request_.url.PathForRequest(), 200, "OK", response);
@@ -255,7 +256,8 @@ TEST_F(QuicEndToEndTest, DISABLED_LargePostWithNoPacketLoss) {
   CheckResponse(consumer, "HTTP/1.1 200 OK", kResponseBody);
 }
 
-TEST_F(QuicEndToEndTest, LargePostWithPacketLoss) {
+// TODO reenable test for Vivaldi
+TEST_F(QuicEndToEndTest, DISABLED_LargePostWithPacketLoss) {
   // FLAGS_fake_packet_loss_percentage = 30;
   InitializePostRequest(1024 * 1024);
 
@@ -272,7 +274,8 @@ TEST_F(QuicEndToEndTest, LargePostWithPacketLoss) {
   CheckResponse(consumer, "HTTP/1.1 200 OK", kResponseBody);
 }
 
-TEST_F(QuicEndToEndTest, UberTest) {
+// TODO reenable test for Vivaldi
+TEST_F(QuicEndToEndTest, DISABLED_UberTest) {
   // FLAGS_fake_packet_loss_percentage = 30;
 
   const char kResponseBody[] = "some really big response body";
diff --git a/net/sdch/sdch_owner_unittest.cc b/net/sdch/sdch_owner_unittest.cc
index 149041fd359be67dc2f6e0104788d4df8fc48067..0ff5fa255e55cdf4668d5292faf4363b7dd14870 100644
--- a/net/sdch/sdch_owner_unittest.cc
+++ b/net/sdch/sdch_owner_unittest.cc
@@ -639,7 +639,9 @@ TEST_F(SdchOwnerTest, ClearReturnsSpace) {
 }
 
 // Confirm memory pressure gets all the space back.
-TEST_F(SdchOwnerTest, MemoryPressureReturnsSpace) {
+// TODO(rmcilroy) Disabled while investigating http://crbug.com/447208 -
+// re-enable once fixed.
+TEST_F(SdchOwnerTest, DISABLED_MemoryPressureReturnsSpace) {
   std::string server_hash_d1;
   std::string server_hash_d2;
 
diff --git a/net/ssl/ssl_config.cc b/net/ssl/ssl_config.cc
index 40e83eee8d58b17d0100ad4421bacb3d1f11d0e2..6a48a65dfa66c275b2a9a1669fbbf42b556e9de1 100644
--- a/net/ssl/ssl_config.cc
+++ b/net/ssl/ssl_config.cc
@@ -18,7 +18,7 @@ SSLConfig::CertAndStatus::CertAndStatus() : cert_status(0) {}
 SSLConfig::CertAndStatus::~CertAndStatus() {}
 
 SSLConfig::SSLConfig()
-    : rev_checking_enabled(false),
+    : rev_checking_enabled(true),
       rev_checking_required_local_anchors(false),
       version_min(kDefaultSSLVersionMin),
       version_max(SSLClientSocket::GetMaxSupportedSSLVersion()),
diff --git a/net/tools/quic/end_to_end_test.cc b/net/tools/quic/end_to_end_test.cc
index 5697b86714f6a77a9fd8e9c4e637c654bf9cf2ce..0f79efa1274cd1f95b40157b42329f721a998858 100644
--- a/net/tools/quic/end_to_end_test.cc
+++ b/net/tools/quic/end_to_end_test.cc
@@ -434,7 +434,8 @@ INSTANTIATE_TEST_CASE_P(EndToEndTests,
                         EndToEndTest,
                         ::testing::ValuesIn(GetTestParams()));
 
-TEST_P(EndToEndTest, SimpleRequestResponse) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_SimpleRequestResponse) {
   ASSERT_TRUE(Initialize());
 
   EXPECT_EQ(kFooResponseBody, client_->SendSynchronousRequest("/foo"));
@@ -453,7 +454,8 @@ TEST_P(EndToEndTest, DISABLED_SimpleRequestResponsev6) {
   EXPECT_EQ(200u, client_->response_headers()->parsed_response_code());
 }
 
-TEST_P(EndToEndTest, SeparateFinPacket) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_SeparateFinPacket) {
   ASSERT_TRUE(Initialize());
 
   HTTPMessage request(HttpConstants::HTTP_1_1,
@@ -476,7 +478,8 @@ TEST_P(EndToEndTest, SeparateFinPacket) {
   EXPECT_EQ(200u, client_->response_headers()->parsed_response_code());
 }
 
-TEST_P(EndToEndTest, MultipleRequestResponse) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_MultipleRequestResponse) {
   ASSERT_TRUE(Initialize());
 
   EXPECT_EQ(kFooResponseBody, client_->SendSynchronousRequest("/foo"));
@@ -485,7 +488,8 @@ TEST_P(EndToEndTest, MultipleRequestResponse) {
   EXPECT_EQ(200u, client_->response_headers()->parsed_response_code());
 }
 
-TEST_P(EndToEndTest, MultipleClients) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_MultipleClients) {
   ASSERT_TRUE(Initialize());
   scoped_ptr<QuicTestClient> client2(CreateQuicClient(nullptr));
 
@@ -508,7 +512,8 @@ TEST_P(EndToEndTest, MultipleClients) {
   EXPECT_EQ(200u, client2->response_headers()->parsed_response_code());
 }
 
-TEST_P(EndToEndTest, RequestOverMultiplePackets) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_RequestOverMultiplePackets) {
   // Send a large enough request to guarantee fragmentation.
   string huge_request = "/some/path?query=" + string(kMaxPacketSize, '.');
   AddToCache(huge_request, 200, "OK", kBarResponseBody);
@@ -519,7 +524,8 @@ TEST_P(EndToEndTest, RequestOverMultiplePackets) {
   EXPECT_EQ(200u, client_->response_headers()->parsed_response_code());
 }
 
-TEST_P(EndToEndTest, MultiplePacketsRandomOrder) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_MultipleFramesRandomOrder) {
   // Send a large enough request to guarantee fragmentation.
   string huge_request = "/some/path?query=" + string(kMaxPacketSize, '.');
   AddToCache(huge_request, 200, "OK", kBarResponseBody);
@@ -532,7 +538,8 @@ TEST_P(EndToEndTest, MultiplePacketsRandomOrder) {
   EXPECT_EQ(200u, client_->response_headers()->parsed_response_code());
 }
 
-TEST_P(EndToEndTest, PostMissingBytes) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_PostMissingBytes) {
   ASSERT_TRUE(Initialize());
 
   // Add a content length header with no body.
@@ -567,7 +574,8 @@ TEST_P(EndToEndTest, DISABLED_LargePostNoPacketLoss) {
   VerifyCleanConnection(false);
 }
 
-TEST_P(EndToEndTest, LargePostNoPacketLoss1sRTT) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_LargePostNoPacketLoss1sRTT) {
   ASSERT_TRUE(Initialize());
   SetPacketSendDelay(QuicTime::Delta::FromMilliseconds(1000));
 
@@ -585,7 +593,8 @@ TEST_P(EndToEndTest, LargePostNoPacketLoss1sRTT) {
   VerifyCleanConnection(false);
 }
 
-TEST_P(EndToEndTest, LargePostWithPacketLoss) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_LargePostWithPacketLoss) {
   // Connect with lower fake packet loss than we'd like to test.  Until
   // b/10126687 is fixed, losing handshake packets is pretty brutal.
   SetPacketLossPercentage(5);
@@ -607,7 +616,9 @@ TEST_P(EndToEndTest, LargePostWithPacketLoss) {
   VerifyCleanConnection(true);
 }
 
-TEST_P(EndToEndTest, LargePostWithPacketLossAndBlockedSocket) {
+
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_LargePostWithPacketLossAndBlockedSocket) {
   // Connect with lower fake packet loss than we'd like to test.  Until
   // b/10126687 is fixed, losing handshake packets is pretty brutal.
   SetPacketLossPercentage(5);
@@ -629,7 +640,8 @@ TEST_P(EndToEndTest, LargePostWithPacketLossAndBlockedSocket) {
   EXPECT_EQ(kFooResponseBody, client_->SendCustomSynchronousRequest(request));
 }
 
-TEST_P(EndToEndTest, LargePostNoPacketLossWithDelayAndReordering) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_LargePostNoPacketLossWithDelayAndReordering) {
   ASSERT_TRUE(Initialize());
 
   client_->client()->WaitForCryptoHandshakeConfirmed();
@@ -688,7 +700,8 @@ TEST_P(EndToEndTest, LargePostZeroRTTFailure) {
   VerifyCleanConnection(false);
 }
 
-TEST_P(EndToEndTest, CorrectlyConfiguredFec) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_CorrectlyConfiguredFec) {
   ASSERT_TRUE(Initialize());
   client_->client()->WaitForCryptoHandshakeConfirmed();
   server_thread_->WaitForCryptoHandshakeConfirmed();
@@ -736,7 +749,8 @@ TEST_P(EndToEndTest, LargePostSmallBandwidthLargeBuffer) {
   VerifyCleanConnection(false);
 }
 
-TEST_P(EndToEndTest, DoNotSetResumeWriteAlarmIfConnectionFlowControlBlocked) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_DoNotSetResumeWriteAlarmIfConnectionFlowControlBlocked) {
   // Regression test for b/14677858.
   // Test that the resume write alarm is not set in QuicConnection::OnCanWrite
   // if currently connection level flow control blocked. If set, this results in
@@ -776,7 +790,8 @@ TEST_P(EndToEndTest, DoNotSetResumeWriteAlarmIfConnectionFlowControlBlocked) {
   EXPECT_FALSE(resume_writes_alarm->IsSet());
 }
 
-TEST_P(EndToEndTest, InvalidStream) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_InvalidStream) {
   ASSERT_TRUE(Initialize());
   client_->client()->WaitForCryptoHandshakeConfirmed();
 
@@ -833,7 +848,8 @@ TEST_P(EndToEndTest, Timeout) {
   }
 }
 
-TEST_P(EndToEndTest, NegotiateMaxOpenStreams) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_NegotiateMaxOpenStreams) {
   // Negotiate 1 max open stream.
   client_config_.SetMaxStreamsPerConnection(1, 1);
   ASSERT_TRUE(Initialize());
@@ -860,7 +876,8 @@ TEST_P(EndToEndTest, NegotiateMaxOpenStreams) {
   EXPECT_EQ(QUIC_TOO_MANY_OPEN_STREAMS, client_->connection_error());
 }
 
-TEST_P(EndToEndTest, NegotiateCongestionControl) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_NegotiateCongestionControl) {
   ValueRestore<bool> old_flag(&FLAGS_quic_allow_bbr, true);
   ASSERT_TRUE(Initialize());
   client_->client()->WaitForCryptoHandshakeConfirmed();
@@ -886,7 +903,8 @@ TEST_P(EndToEndTest, NegotiateCongestionControl) {
             ->GetCongestionControlType());
 }
 
-TEST_P(EndToEndTest, LimitMaxOpenStreams) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_LimitMaxOpenStreams) {
   // Server limits the number of max streams to 2.
   server_config_.SetMaxStreamsPerConnection(2, 2);
   // Client tries to negotiate for 10.
@@ -924,7 +942,8 @@ TEST_P(EndToEndTest, ClientSuggestsRTT) {
   server_thread_->Resume();
 }
 
-TEST_P(EndToEndTest, MaxInitialRTT) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_MaxInitialRTT) {
   // Client tries to suggest twice the server's max initial rtt and the server
   // uses the max.
   client_config_.SetInitialRoundTripTimeUsToSend(
@@ -956,7 +975,8 @@ TEST_P(EndToEndTest, MaxInitialRTT) {
   server_thread_->Resume();
 }
 
-TEST_P(EndToEndTest, MinInitialRTT) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_MinInitialRTT) {
   // Client tries to suggest 0 and the server uses the default.
   client_config_.SetInitialRoundTripTimeUsToSend(0);
 
@@ -1050,7 +1070,8 @@ TEST_P(EndToEndTest, 15ByteConnectionId) {
             header->public_header.connection_id_length);
 }
 
-TEST_P(EndToEndTest, ResetConnection) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_ResetConnection) {
   ASSERT_TRUE(Initialize());
   client_->client()->WaitForCryptoHandshakeConfirmed();
 
@@ -1061,7 +1082,7 @@ TEST_P(EndToEndTest, ResetConnection) {
   EXPECT_EQ(200u, client_->response_headers()->parsed_response_code());
 }
 
-TEST_P(EndToEndTest, MaxStreamsUberTest) {
+TEST_P(EndToEndTest, DISABLED_MaxStreamsUberTest) {
   SetPacketLossPercentage(1);
   ASSERT_TRUE(Initialize());
   string large_body;
@@ -1083,7 +1104,8 @@ TEST_P(EndToEndTest, MaxStreamsUberTest) {
   }
 }
 
-TEST_P(EndToEndTest, StreamCancelErrorTest) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_StreamCancelErrorTest) {
   ASSERT_TRUE(Initialize());
   string small_body;
   GenerateBody(&small_body, 256);
@@ -1133,7 +1155,8 @@ class WrongAddressWriter : public QuicPacketWriterWrapper {
   IPEndPoint self_address_;
 };
 
-TEST_P(EndToEndTest, ConnectionMigrationClientIPChanged) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_ConnectionMigrationClientIPChanged) {
   // Tests that the client's IP can not change during an established QUIC
   // connection. If it changes, the connection is closed by the server as we do
   // not yet support IP migration.
@@ -1155,7 +1178,8 @@ TEST_P(EndToEndTest, ConnectionMigrationClientIPChanged) {
   EXPECT_EQ(QUIC_ERROR_MIGRATING_ADDRESS, client_->connection_error());
 }
 
-TEST_P(EndToEndTest, ConnectionMigrationClientPortChanged) {
+// TODO reenable test for Vivaldi
+TEST_P(EndToEndTest, DISABLED_ConnectionMigrationClientPortChanged) {
   // Tests that the client's port can change during an established QUIC
   // connection, and that doing so does not result in the connection being
   // closed by the server.
