diff --git a/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.h b/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.h
index 7f75480fc51a583d70dfe6d8d0d6cd7d05a3c49d..acb01cf4daa60ce2d099a2e63d09ed52253d2e78 100644
--- a/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.h
+++ b/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.h
@@ -27,7 +27,8 @@ class IOSChromeTabRestoreServiceClient
       const std::string& app_name,
       const gfx::Rect& bounds,
       ui::WindowShowState show_state,
-      const std::string& workspace) override;
+      const std::string& workspace,
+      const std::string& ext_data) override;
   sessions::LiveTabContext* FindLiveTabContextForTab(
       const sessions::LiveTab* tab) override;
   sessions::LiveTabContext* FindLiveTabContextWithID(
diff --git a/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.mm b/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.mm
index 3976e4753cbfd4893cd24c09263b46a1e3c511ae..6e254c841327ebeeb6729c2699ea54f00ce20918 100644
--- a/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.mm
+++ b/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.mm
@@ -74,7 +74,8 @@ IOSChromeTabRestoreServiceClient::CreateLiveTabContext(
     const std::string& /* app_name */,
     const gfx::Rect& /* bounds */,
     ui::WindowShowState /* show_state */,
-    const std::string& /* workspace */) {
+    const std::string& /* workspace */,
+    const std::string& ext_data) {
   return TabRestoreServiceDelegateImplIOSFactory::GetForBrowserState(
       browser_state_);
 }
