diff --git a/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.h b/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.h
index 7f75480fc51a583d70dfe6d8d0d6cd7d05a3c49d..acb01cf4daa60ce2d099a2e63d09ed52253d2e78 100644
--- a/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.h
+++ b/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.h
@@ -27,7 +27,8 @@ class IOSChromeTabRestoreServiceClient
       const std::string& app_name,
       const gfx::Rect& bounds,
       ui::WindowShowState show_state,
-      const std::string& workspace) override;
+      const std::string& workspace,
+      const std::string& ext_data) override;
   sessions::LiveTabContext* FindLiveTabContextForTab(
       const sessions::LiveTab* tab) override;
   sessions::LiveTabContext* FindLiveTabContextWithID(
diff --git a/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.mm b/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.mm
index a84e9d9402049da9925b5ba448e0bf4904dd9ee8..a3310c62dbe01d2bdc89e2dbcf585cd7c04c9c3f 100644
--- a/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.mm
+++ b/ios/chrome/browser/sessions/ios_chrome_tab_restore_service_client.mm
@@ -73,7 +73,8 @@ IOSChromeTabRestoreServiceClient::CreateLiveTabContext(
     const std::string& /* app_name */,
     const gfx::Rect& /* bounds */,
     ui::WindowShowState /* show_state */,
-    const std::string& /* workspace */) {
+    const std::string& /* workspace */,
+    const std::string& ext_data) {
   return TabRestoreServiceDelegateImplIOSFactory::GetForBrowserState(
       browser_state_);
 }
diff --git a/ios/third_party/material_components_ios/BUILD.gn b/ios/third_party/material_components_ios/BUILD.gn
index 37035edc00fb34cf65a4859c48299e8b5e5b5efd..86cdf065b98dbf7ff4f4b61fbff7b8aad8a07ce1 100644
--- a/ios/third_party/material_components_ios/BUILD.gn
+++ b/ios/third_party/material_components_ios/BUILD.gn
@@ -298,7 +298,7 @@ foreach(locale, _locales) {
       "src/components/Collections/src/MaterialCollections.bundle/Resources/$locale.lproj/MaterialCollections.strings",
     ]
     outputs = [
-      "{{bundle_resources_dir}}/MaterialCollections.bundle/Resources/$locale.lproj/{{source_file_part}}",
+      "{{bundle_root_dir}}/MaterialCollections.bundle/Resources/$locale.lproj/{{source_file_part}}",
     ]
   }
 
