diff --git a/third_party/catapult/tracing/.gitignore b/third_party/catapult/tracing/.gitignore
new file mode 100644
index 0000000000000000000000000000000000000000..c17a21f63f6bb0e8e75fecdb74a64cb959a13c6e
--- /dev/null
+++ b/third_party/catapult/tracing/.gitignore
@@ -0,0 +1,9 @@
+*.pyc
+*.swp
+.idea
+.DS_Store
+*~
+.project
+.settings
+bin/trace_viewer*.html
+test_data/measurmt-traces
diff --git a/third_party/catapult/tracing/.npmignore b/third_party/catapult/tracing/.npmignore
new file mode 100644
index 0000000000000000000000000000000000000000..70d51d239f102e9bb346b88227609a7ce29b0e33
--- /dev/null
+++ b/third_party/catapult/tracing/.npmignore
@@ -0,0 +1,17 @@
+test_data
+skp_data
+third_party
+
+images
+tracing_build
+build
+
+app.yaml
+*_test.html
+
+*.gyp
+*.gypi
+*.gn
+
+*.py
+*.pyc
diff --git a/third_party/catapult/tracing/BUILD.gn b/third_party/catapult/tracing/BUILD.gn
new file mode 100644
index 0000000000000000000000000000000000000000..f81a2d5ff40415b44e25d454de7b11428d7712fa
--- /dev/null
+++ b/third_party/catapult/tracing/BUILD.gn
@@ -0,0 +1,27 @@
+# Copyright 2014 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+gypi_values = exec_script("//build/gypi_to_gn.py",
+                          [ rebase_path("trace_viewer.gypi") ],
+                          "scope",
+                          [ "trace_viewer.gypi" ])
+
+# TODO: ideally this would go into the target_gen_dir, but this requires some
+# changes to the scripts that process them.
+output_resource_dir = "$root_gen_dir/content/browser/tracing"
+
+action("generate_about_tracing") {
+  script = "bin/generate_about_tracing_contents"
+
+  inputs = gypi_values.tracing_css_files + gypi_values.tracing_js_html_files +
+      gypi_values.tracing_img_files
+  outputs = [
+    "$output_resource_dir/about_tracing.js",
+    "$output_resource_dir/about_tracing.html",
+  ]
+
+  args = [
+    "--outdir", rebase_path(output_resource_dir, root_build_dir),
+  ]
+}
diff --git a/third_party/catapult/tracing/LICENSE b/third_party/catapult/tracing/LICENSE
new file mode 100644
index 0000000000000000000000000000000000000000..e6c0d72a5122171deb4c458991d1c7547f31a2f0
--- /dev/null
+++ b/third_party/catapult/tracing/LICENSE
@@ -0,0 +1,27 @@
+// Copyright (c) 2012 The Chromium Authors. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
diff --git a/third_party/catapult/tracing/OWNERS b/third_party/catapult/tracing/OWNERS
new file mode 100644
index 0000000000000000000000000000000000000000..cddffe86de5aa842391ae6a2f0a63b3157a57120
--- /dev/null
+++ b/third_party/catapult/tracing/OWNERS
@@ -0,0 +1,19 @@
+# Owners: These folks have a very good understanding of the tracing
+# system and how the UI bits fit together. Good for reviews of large
+# architectual changes, or areas of the code where you may need more
+# information.
+nduca@chromium.org
+petrcermak@chromium.org
+dsinclair@chromium.org
+fmeawad@chromium.org
+charliea@chromium.org
+eakuefner@chromium.org
+benjhayden@chromium.org
+
+# For changes related to system_health/loading_metric.html and its tests.
+kouhei@chromium.org
+
+# Tracing, in general, doesn't require an owners stamp to commit. If you
+# don't think the complexity requires the review of the above you can use
+# any tracing committer.
+*
diff --git a/third_party/catapult/tracing/PRESUBMIT.py b/third_party/catapult/tracing/PRESUBMIT.py
new file mode 100644
index 0000000000000000000000000000000000000000..32f06c0831b4c26d43c760f1e21991d464f272a2
--- /dev/null
+++ b/third_party/catapult/tracing/PRESUBMIT.py
@@ -0,0 +1,64 @@
+# Copyright 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import sys
+
+def _RunArgs(args, input_api):
+  p = input_api.subprocess.Popen(args, stdout=input_api.subprocess.PIPE,
+                                 stderr=input_api.subprocess.STDOUT)
+  out, _ = p.communicate()
+  return (out, p.returncode)
+
+
+def _CheckRegisteredMetrics(input_api, output_api):
+  """ Check that all tracing metrics are imported in all_metrics.html """
+  results = []
+  tracing_dir = input_api.PresubmitLocalPath()
+  out, return_code = _RunArgs(
+      [input_api.python_executable,
+       input_api.os_path.join(tracing_dir, 'bin', 'validate_all_metrics')],
+      input_api)
+  if return_code:
+    results.append(output_api.PresubmitError(
+        'Failed validate_all_metrics: ', long_text=out))
+  return results
+
+
+def CheckChangeOnUpload(input_api, output_api):
+  return _CheckChange(input_api, output_api)
+
+
+def CheckChangeOnCommit(input_api, output_api):
+  return _CheckChange(input_api, output_api)
+
+
+def _CheckChange(input_api, output_api):
+  results = []
+
+  original_sys_path = sys.path
+  try:
+    sys.path += [input_api.PresubmitLocalPath()]
+    from tracing_build import check_gypi
+    error = check_gypi.GypiCheck()
+    if error:
+      results.append(output_api.PresubmitError(error))
+  finally:
+    sys.path = original_sys_path
+
+  results += input_api.RunTests(input_api.canned_checks.GetPylint(
+      input_api, output_api, extra_paths_list=_GetPathsToPrepend(input_api),
+      pylintrc='../pylintrc'))
+
+  results += _CheckRegisteredMetrics(input_api, output_api)
+
+  return results
+
+
+def _GetPathsToPrepend(input_api):
+  project_dir = input_api.PresubmitLocalPath()
+  catapult_dir = input_api.os_path.join(project_dir, '..')
+  return [
+      project_dir,
+      input_api.os_path.join(catapult_dir, 'third_party', 'mock'),
+  ]
diff --git a/third_party/catapult/tracing/README.md b/third_party/catapult/tracing/README.md
new file mode 100644
index 0000000000000000000000000000000000000000..46dbe820b0bd6b47d9ce924d13cef7806ec8a2d3
--- /dev/null
+++ b/third_party/catapult/tracing/README.md
@@ -0,0 +1,51 @@
+
+<!-- Copyright 2015 The Chromium Authors. All rights reserved.
+     Use of this source code is governed by a BSD-style license that can be
+     found in the LICENSE file.
+-->
+![Trace Viewer Logo](https://raw.githubusercontent.com/catapult-project/catapult/master/tracing/images/trace-viewer-circle-blue.png)
+
+Trace-Viewer is the javascript frontend for Chrome [about:tracing](http://dev.chromium.org/developers/how-tos/trace-event-profiling-tool) and [Android
+systrace](http://developer.android.com/tools/help/systrace.html).
+
+It provides rich analysis and visualization capabilities for many types of trace
+files. Its particularly good at viewing linux kernel traces (aka [ftrace](https://www.kernel.org/doc/Documentation/trace/ftrace.txt)) and Chrome's
+[trace_event format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview). Trace viewer can be [embedded](https://github.com/catapult-project/catapult/wiki/Embedding-Trace-Viewer) as a component in your own code, or used from a plain checkout to turn trace files into standalone, emailable HTML files from the commandline:
+
+```
+$CATAPULT/tracing/bin/trace2html my_trace.json --output=my_trace.html && open my_trace.html
+```
+
+Its easy to [extend trace viewer](https://github.com/catapult-project/catapult/wiki/Extending-and-Customizing-Trace-Viewer) to support your favorite trace format, or add domain specific visualizations to the UI to simplify drilling down into complex data.
+
+Contributing, quick version
+===
+We welcome contributions! To hack on this code.
+
+There are two type of tests.
+
+### In the browser
+
+Run http server `$CATAPULT/bin/run_dev_server`. In any browser, navigate to `http://localhost:8003/`
+
+**Unit tests**| **Descripton**
+--- | ---
+All tests | http://localhost:8003/tests.html
+All tests with short format | http://localhost:8003/tracing/tests.html?shortFormat
+An individual test suite(such as ui/foo_test.js) | http://localhost:8003/tests.html?testSuiteName=ui.foo
+Tests named foo| http://localhost:8003/tests.html?testFilterString=foo
+
+### On command
+
+**Unit tests**| **Description**
+--- | ---
+All python tests | `$CATAPULT/tracing/bin/run_py_tests`
+All tracing tests in d8 environment | `$CATAPULT/tracing/bin/run_vinn_tests`
+All tests | `$CATAPULT/tracing/bin/run_tests`
+
+Make sure tests pass before sending us changelist. **We use rietveld for codereview**. For more details, esp on rietveld, [read our contributing guide](https://github.com/catapult-project/catapult/blob/master/CONTRIBUTING.md) or check out the [trace viewer wiki](https://github.com/catapult-project/catapult/wiki/Trace-Viewer-Getting-Started).
+
+Contact Us
+===
+Join our Google Group:
+* [tracing@chromium.org](https://groups.google.com/a/chromium.org/forum/#!forum/tracing)
diff --git a/third_party/catapult/tracing/app.yaml b/third_party/catapult/tracing/app.yaml
new file mode 100644
index 0000000000000000000000000000000000000000..40839ddfb07929136c584381bd8fff6fa50d8762
--- /dev/null
+++ b/third_party/catapult/tracing/app.yaml
@@ -0,0 +1,58 @@
+# Copyright 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+runtime: python27
+threadsafe: true
+api_version: 1
+
+handlers:
+
+- url: /base
+  static_dir: tracing/tracing/base
+  secure: always
+
+- url: /core
+  static_dir: tracing/tracing/core
+  secure: always
+
+- url: /extras
+  static_dir: tracing/tracing/extras
+  secure: always
+
+- url: /ui
+  static_dir: tracing/tracing/ui
+  secure: always
+
+- url: /components
+  static_dir: tracing/third_party/components
+  secure: always
+
+- url: /trace_viewer
+  static_dir: tracing/tracing
+  secure: always
+
+- url: /gl-matrix-min.js
+  static_files: tracing/third_party/gl-matrix/dist/gl-matrix-min.js
+  upload: tracing/third_party/gl-matrix/dist/gl-matrix-min.js
+  secure: always
+
+- url: /mannwhitneyu.js
+  static_files: tracing/third_party/mannwhitneyu/mannwhitneyu.js
+  upload: tracing/third_party/mannwhitneyu/mannwhitneyu.js
+  secure: always
+
+- url: /jszip.min.js
+  static_files: tracing/third_party/jszip/jszip.min.js
+  upload: tracing/third_party/jszip/jszip.min.js
+  secure: always
+
+- url: /d3.min.js
+  static_files: tracing/third_party/d3/d3.min.js
+  upload: tracing/third_party/d3/d3.min.js
+  secure: always
+
+- url: /.*
+  static_files: tracing/tracing/ui/extras/drive/index.html
+  upload: tracing/tracing/ui/extras/drive/index.html
+  secure: always
diff --git a/third_party/catapult/tracing/bin/PRESUBMIT.py b/third_party/catapult/tracing/bin/PRESUBMIT.py
new file mode 100644
index 0000000000000000000000000000000000000000..799215d729da79358e32f41905256623b926e829
--- /dev/null
+++ b/third_party/catapult/tracing/bin/PRESUBMIT.py
@@ -0,0 +1,19 @@
+# Copyright 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+import os
+
+def CheckChange(input_api, output_api):
+  init_py_path = os.path.join(input_api.PresubmitLocalPath(), '__init__.py')
+  res = []
+  if os.path.exists(init_py_path):
+    res += [output_api.PresubmitError(
+        '__init__.py is not allowed to exist in bin/')]
+  return res
+
+def CheckChangeOnUpload(input_api, output_api):
+  return CheckChange(input_api, output_api)
+
+
+def CheckChangeOnCommit(input_api, output_api):
+  return CheckChange(input_api, output_api)
diff --git a/third_party/catapult/tracing/bin/compare_samples b/third_party/catapult/tracing/bin/compare_samples
new file mode 100755
index 0000000000000000000000000000000000000000..ff56ccc69faa370daedf828494eaaf6a50daacaf
--- /dev/null
+++ b/third_party/catapult/tracing/bin/compare_samples
@@ -0,0 +1,54 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import argparse
+import os
+import sys
+
+tracing_path = os.path.abspath(os.path.join(os.path.dirname(__file__),
+                               '..'))
+sys.path.append(tracing_path)
+from tracing.metrics import compare_samples
+
+def Main(argv):
+  parser = argparse.ArgumentParser(
+      description='Compare samples.')
+  parser.add_argument('sample_a', type=str,
+                      help='comma-separated list of paths to valuesets from '
+                           'sample a')
+  parser.add_argument('sample_b', type=str,
+                      help='comma-separated list of paths to valuesets from '
+                           'sample b')
+  parser.add_argument('metric', type=str,
+                      help='name of the metric to compare')
+  parser.add_argument('--chartjson', dest='format', action='store_const',
+                      const='chartjson',
+                      help='assume chartjson format for the input data')
+  parser.add_argument('--buildbot', dest='format', action='store_const',
+                      const='buildbot',
+                      help='assume buildbot result line format for the data')
+  parser.add_argument('--valueset', dest='format', action='store_const',
+                      const='valueset',
+                      help='assume valueset format for the input data')
+  args = parser.parse_args(argv[1:])
+
+  if not args.format:
+    filename = os.path.basename(sample_a.split(',')[0])
+    if filename == 'results-valueset.json':
+      args.format = 'valueset'
+    else:
+      args.format = 'chartjson'
+
+  vinn_result = compare_samples.CompareSamples(
+      args.sample_a,
+      args.sample_b,
+      args.metric,
+      args.format
+  )
+  print vinn_result.stdout
+  return vinn_result.returncode
+
+if __name__ == '__main__':
+  sys.exit(Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/generate_about_tracing_contents b/third_party/catapult/tracing/bin/generate_about_tracing_contents
new file mode 100755
index 0000000000000000000000000000000000000000..feee8828976e83e0dfcbebcee69634594a651a5c
--- /dev/null
+++ b/third_party/catapult/tracing/bin/generate_about_tracing_contents
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+# Copyright (c) 2014 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..'))
+  sys.path.append(tracing_path)
+  from tracing_build import generate_about_tracing_contents
+  sys.exit(generate_about_tracing_contents.Main(sys.argv[1:]))
diff --git a/third_party/catapult/tracing/bin/html2trace b/third_party/catapult/tracing/bin/html2trace
new file mode 100755
index 0000000000000000000000000000000000000000..82161fb1cfd64f440f078cb904706bc19a6347ca
--- /dev/null
+++ b/third_party/catapult/tracing/bin/html2trace
@@ -0,0 +1,53 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import argparse
+import codecs
+import sys
+import os
+
+tracing_path = os.path.abspath(os.path.join(
+  os.path.dirname(os.path.realpath(__file__)), '..'))
+sys.path.append(tracing_path)
+from tracing_build import html2trace
+
+
+def main():
+  parser = argparse.ArgumentParser(description='Extract trace data from an '
+                                   'HTML trace.', add_help=False)
+  parser.add_argument('html_path', metavar='HTML_PATH',
+                      help='HTML file path (input).')
+  parser.add_argument('trace_path', metavar='TRACE_PATH',
+                      help='Trace file path (output). If the HTML file '
+                      'contains more than one trace data block, the first '
+                      'block will be extracted into %(metavar)s and the rest '
+                      'will be extracted into separate files %(metavar)s.1, '
+                      '%(metavar)s.2, etc.')
+  parser.add_argument('--gzipped_output', choices=['true', 'false', 'auto'],
+                      default='auto', help='Flag whether the output trace '
+                      'file should be gzipped.')
+  parser.add_argument('-q', '--quiet', action='store_true',
+                      help='Don\'t print the saved file name(s).')
+  parser.add_argument('-h', '--help', action='help',
+                      help='Show this help message and exit.')
+  args = parser.parse_args()
+
+  if args.gzipped_output == 'true':
+    gzipped_output = True
+  elif args.gzipped_output == 'false':
+    gzipped_output = False
+  else:
+    gzipped_output = args.trace_path.endswith('.gz')
+
+  with codecs.open(args.html_path, mode='r', encoding='utf-8') as html_file:
+    saved_paths = html2trace.CopyTraceDataFromHTMLFilePath(
+        html_file, args.trace_path, gzipped_output)
+
+  if not args.quiet:
+    print '\n'.join(saved_paths)
+
+
+if __name__ == '__main__':
+  sys.exit(main())
diff --git a/third_party/catapult/tracing/bin/index.html b/third_party/catapult/tracing/bin/index.html
new file mode 100644
index 0000000000000000000000000000000000000000..d48ccc92a4a95696dbc4f8c988886ed386e89226
--- /dev/null
+++ b/third_party/catapult/tracing/bin/index.html
@@ -0,0 +1,115 @@
+<!DOCTYPE html>
+<!--
+Copyright (c) 2014 The Chromium Authors. All rights reserved.
+Use of this source code is governed by a BSD-style license that can be
+found in the LICENSE file.
+-->
+<head>
+<script>
+'use strict';
+
+function onTraceViewerImportFail() {
+  document.addEventListener('DOMContentLoaded', function() {
+    document.body.textContent =
+        'tracing/bin/trace_viewer_full.html is missing. ' +
+        'Run vulcanize_trace_viewer from $TRACE_VIEWER and reload.';
+  });
+}
+</script>
+<link rel="import" href="trace_viewer_full.html"
+      onerror="onTraceViewerImportFail(event)">
+
+<style>
+  html, body {
+    box-sizing: border-box;
+    overflow: hidden;
+    margin: 0px;
+    padding: 0;
+    width: 100%;
+    height: 100%;
+  }
+  #trace-viewer {
+    width: 100%;
+    height: 100%;
+  }
+  #trace-viewer:focus {
+    outline: none;
+  }
+</style>
+<script>
+'use strict';
+
+(function() {
+  var viewer;
+  var url;
+  var model;
+
+  function load() {
+    var req = new XMLHttpRequest();
+    var is_binary = /[.]gz$/.test(url) || /[.]zip$/.test(url);
+    req.overrideMimeType('text/plain; charset=x-user-defined');
+    req.open('GET', url, true);
+    if (is_binary)
+      req.responseType = 'arraybuffer';
+
+    req.onreadystatechange = function(event) {
+      if (req.readyState !== 4)
+        return;
+
+      window.setTimeout(function() {
+        if (req.status === 200)
+          onResult(is_binary ? req.response : req.responseText);
+        else
+          onResultFail(req.status);
+      }, 0);
+    };
+    req.send(null);
+  }
+
+  function onResultFail(err) {
+    var overlay = new tr.ui.b.Overlay();
+    overlay.textContent = err + ': ' + url + ' could not be loaded';
+    overlay.title = 'Failed to fetch data';
+    overlay.visible = true;
+  }
+
+  function onResult(result) {
+    model = new tr.Model();
+    var i = new tr.importer.Import(model);
+    var p = i.importTracesWithProgressDialog([result]);
+    p.then(onModelLoaded, onImportFail);
+  }
+
+  function onModelLoaded() {
+    viewer.model = model;
+    viewer.viewTitle = url;
+  }
+
+  function onImportFail() {
+    var overlay = new tr.ui.b.Overlay();
+    overlay.textContent = tr.b.normalizeException(err).message;
+    overlay.title = 'Import error';
+    overlay.visible = true;
+  }
+
+  document.addEventListener('DOMContentLoaded', function() {
+    var container = document.createElement('track-view-container');
+    container.id = 'track_view_container';
+
+    viewer = document.createElement('tr-ui-timeline-view');
+    viewer.track_view_container = container;
+    Polymer.dom(viewer).appendChild(container);
+
+    viewer.id = 'trace-viewer';
+    viewer.globalMode = true;
+    Polymer.dom(document.body).appendChild(viewer);
+
+    url = '../test_data/big_trace.json';
+    load();
+  });
+}());
+</script>
+</head>
+<body>
+</body>
+</html>
diff --git a/third_party/catapult/tracing/bin/map_traces b/third_party/catapult/tracing/bin/map_traces
new file mode 100755
index 0000000000000000000000000000000000000000..af58491504df566af15f282d56f819f8e4963e2e
--- /dev/null
+++ b/third_party/catapult/tracing/bin/map_traces
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+# Copyright (c) 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(os.path.dirname(__file__),
+                                              '..'))
+  sys.path.append(tracing_path)
+  from tracing.mre import map_traces
+  sys.exit(map_traces.Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/merge_histograms b/third_party/catapult/tracing/bin/merge_histograms
new file mode 100755
index 0000000000000000000000000000000000000000..c4df3ce9d581067ef8758468aa6b43c01c70c61c
--- /dev/null
+++ b/third_party/catapult/tracing/bin/merge_histograms
@@ -0,0 +1,29 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import argparse
+import json
+import os
+import sys
+
+tracing_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+sys.path.append(tracing_path)
+from tracing.value import merge_histograms
+
+def Main(argv):
+  parser = argparse.ArgumentParser(
+      description='Merge Histograms.')
+  parser.add_argument('input', type=str,
+                      help='Path to a HistogramSet JSON file. (input)')
+  parser.add_argument('output', type=str,
+                      help='Path to a HistogramSet JSON file. (output)')
+  args = parser.parse_args(argv[1:])
+
+  merged = merge_histograms.MergeHistograms(args.input)
+  json.dump(merged, file(args.output, 'w'))
+  return 0
+
+if __name__ == '__main__':
+  sys.exit(Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/merge_traces b/third_party/catapult/tracing/bin/merge_traces
new file mode 100755
index 0000000000000000000000000000000000000000..4b610dd0699b3421efe41187150aae58b0ebc561
--- /dev/null
+++ b/third_party/catapult/tracing/bin/merge_traces
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..'))
+  sys.path.append(tracing_path)
+  from tracing_build import merge_traces
+  sys.exit(merge_traces.Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/results2json b/third_party/catapult/tracing/bin/results2json
new file mode 100755
index 0000000000000000000000000000000000000000..7b3b2b24dad43666de968f3b872755dcb671cbd0
--- /dev/null
+++ b/third_party/catapult/tracing/bin/results2json
@@ -0,0 +1,35 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import argparse
+import json
+import sys
+import os
+
+tracing_path = os.path.abspath(os.path.join(
+  os.path.dirname(os.path.realpath(__file__)), '..'))
+sys.path.append(tracing_path)
+from tracing import results_renderer
+
+
+def main():
+  parser = argparse.ArgumentParser(
+      description='Extract HistogramSet JSON from results2.html.',
+      add_help=False)
+  parser.add_argument('html_path', metavar='HTML_PATH',
+                      help='HTML file path (input).')
+  parser.add_argument('json_path', metavar='JSON_PATH',
+                      help='JSON file path (input/output).')
+  parser.add_argument('-h', '--help', action='help',
+                      help='Show this help message and exit.')
+  args = parser.parse_args()
+
+  histograms = results_renderer.ReadExistingResults(open(args.html_path, 'r'))
+  if os.path.exists(args.json_path):
+    histograms.extend(json.load(open(args.json_path, 'r')))
+  json.dump(histograms, open(args.json_path, 'w'))
+
+if __name__ == '__main__':
+  sys.exit(main())
diff --git a/third_party/catapult/tracing/bin/run_dev_server_tests b/third_party/catapult/tracing/bin/run_dev_server_tests
new file mode 100755
index 0000000000000000000000000000000000000000..bd09476721e433f991b8193a6d0de767594cc92d
--- /dev/null
+++ b/third_party/catapult/tracing/bin/run_dev_server_tests
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+# Copyright (c) 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..', '..'))
+  sys.path.append(tracing_path)
+  from catapult_build import run_dev_server_tests
+  sys.exit(run_dev_server_tests.Main(sys.argv + ['--tests=tracing']))
diff --git a/third_party/catapult/tracing/bin/run_metric b/third_party/catapult/tracing/bin/run_metric
new file mode 100755
index 0000000000000000000000000000000000000000..c1d913c5658a495a6076ea80366165e7a824c065
--- /dev/null
+++ b/third_party/catapult/tracing/bin/run_metric
@@ -0,0 +1,75 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import argparse
+import codecs
+import json
+import os
+import sys
+
+sys.path.insert(1, os.path.join(os.path.dirname(__file__), '..'))
+from tracing import results_renderer
+from tracing.metrics import metric_runner
+from tracing.metrics import discover
+
+def Main(argv):
+  all_metrics = discover.DiscoverMetrics(
+      ['/tracing/metrics/all_metrics.html'])
+
+  parser = argparse.ArgumentParser(
+      description='Runs metrics on local traces')
+  parser.add_argument('trace_file_or_dir',
+                      help='A trace file, or a dir containing trace files')
+  parser.add_argument('metrics', nargs='+',
+                      help=('Function names of registered metrics '
+                            '(not filenames.) '
+                            'Available metrics are: %s' %
+                            ', '.join(all_metrics)),
+                      choices=all_metrics, metavar='metricName')
+  parser.add_argument('--filename', default='results2', type=str,
+                      help='Output file name (no extension)')
+  parser.add_argument('--reset', action='store_true',
+                      help=('Whether to ignore existing results in HTML file '
+                            '(if it exists'))
+  parser.add_argument('--also-output-json', action='store_true',
+                      help=('Also output json file containing values. Note that'
+                            'this only contains the results of current run'))
+
+  args = parser.parse_args(argv[1:])
+  trace_file_or_dir = os.path.abspath(args.trace_file_or_dir)
+
+  if os.path.isdir(trace_file_or_dir):
+    trace_dir = trace_file_or_dir
+    traces = [os.path.join(trace_dir, trace) for trace in os.listdir(trace_dir)]
+  else:
+    traces = [trace_file_or_dir]
+
+  failures = []
+  histograms = []
+  for trace_url, mre_result in metric_runner.RunMetricOnTraces(
+      traces, args.metrics).iteritems():
+    failures.extend(mre_result.failures)
+    histograms.extend(mre_result.pairs.get('histograms', []))
+
+  if failures:
+    print 'Running metric failed:'
+    for failure in failures:
+      print failure['stack']
+
+  output_file = args.filename + '.html'
+  open(output_file, 'a').close() # Create file if it doesn't exist.
+  with codecs.open(output_file, mode='r+', encoding='utf-8') as output_stream:
+    results_renderer.RenderHTMLView(histograms, output_stream, args.reset)
+    print 'HTML result created in file://' + os.path.abspath(output_file)
+
+  if args.also_output_json:
+    output_file = args.filename + '.json'
+    with open(output_file, 'w') as f:
+      json.dump(histograms, f, indent=2, sort_keys=True, separators=(',', ': '))
+    print 'JSON result created in file://' + os.path.abspath(output_file)
+
+
+if __name__ == '__main__':
+  sys.exit(Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/run_node_tests b/third_party/catapult/tracing/bin/run_node_tests
new file mode 100755
index 0000000000000000000000000000000000000000..d044e0f1c2276f97735490d73199c677cb19a682
--- /dev/null
+++ b/third_party/catapult/tracing/bin/run_node_tests
@@ -0,0 +1,23 @@
+#!/usr/bin/env node
+// Copyright (c) 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+var fs = require('fs');
+var path = require('path');
+
+var catapultPath = fs.realpathSync(path.join(__dirname, '..', '..'));
+var catapultBuildPath = path.join(catapultPath, 'catapult_build');
+
+var node_bootstrap = require(path.join(catapultBuildPath, 'node_bootstrap.js'));
+
+HTMLImportsLoader.addArrayToSourcePath(
+    node_bootstrap.getSourcePathsForProject('tracing'));
+
+// Go!
+var headless_test_module_filenames =
+    node_bootstrap.getHeadlessTestModuleFilenamesForProject('tracing');
+
+HTMLImportsLoader.loadHTML('/tracing/base/headless_tests.html');
+tr.b.unittest.loadAndRunTests(headless_test_module_filenames);
\ No newline at end of file
diff --git a/third_party/catapult/tracing/bin/run_py_tests b/third_party/catapult/tracing/bin/run_py_tests
new file mode 100755
index 0000000000000000000000000000000000000000..23a6ceaa31f22d5098bef1471898b0111e9d0826
--- /dev/null
+++ b/third_party/catapult/tracing/bin/run_py_tests
@@ -0,0 +1,41 @@
+#!/usr/bin/env python
+# Copyright (c) 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import platform
+import sys
+
+_CATAPULT_PATH = os.path.abspath(
+    os.path.join(
+        os.path.dirname(os.path.realpath(__file__)),
+        os.path.pardir,
+        os.path.pardir))
+_TRACING_PATH = os.path.join(_CATAPULT_PATH, 'tracing')
+
+
+def _RunTestsOrDie(top_level_dir):
+  exit_code = run_with_typ.Run(top_level_dir, path=[_TRACING_PATH])
+  if exit_code:
+    sys.exit(exit_code)
+
+
+def _AddToPathIfNeeded(path):
+  if path not in sys.path:
+    sys.path.insert(0, path)
+
+
+if __name__ == '__main__':
+  _AddToPathIfNeeded(_CATAPULT_PATH)
+
+  from hooks import install
+  if '--no-install-hooks' in sys.argv:
+    sys.argv.remove('--no-install-hooks')
+  else:
+    install.InstallHooks()
+
+  from catapult_build import run_with_typ
+  _RunTestsOrDie(os.path.join(_TRACING_PATH, 'tracing'))
+  _RunTestsOrDie(os.path.join(_TRACING_PATH, 'tracing_build'))
+  sys.exit(0)
diff --git a/third_party/catapult/tracing/bin/run_tests b/third_party/catapult/tracing/bin/run_tests
new file mode 100755
index 0000000000000000000000000000000000000000..df95b1d6c4e3251b158391f369061f46dd12f3ce
--- /dev/null
+++ b/third_party/catapult/tracing/bin/run_tests
@@ -0,0 +1,25 @@
+#!/usr/bin/env python
+# Copyright (c) 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+
+_THIS_PATH = os.path.dirname(os.path.realpath(__file__))
+_TESTS = [
+    {'path': os.path.join(_THIS_PATH, 'run_py_tests')},
+    {'path': os.path.join(_THIS_PATH, 'run_vinn_tests'),
+     'disabled': {'win32'}},
+    {'path': os.path.join(_THIS_PATH, 'run_dev_server_tests'),
+     'chrome_path_arg': True}
+]
+
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..'))
+  sys.path.append(tracing_path)
+  from catapult_build import test_runner
+  sys.exit(test_runner.Main('tracing', _TESTS, sys.argv))
diff --git a/third_party/catapult/tracing/bin/run_vinn_tests b/third_party/catapult/tracing/bin/run_vinn_tests
new file mode 100755
index 0000000000000000000000000000000000000000..42ec180f119bcc8836c0d6678146a5cbfd2af392
--- /dev/null
+++ b/third_party/catapult/tracing/bin/run_vinn_tests
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+# Copyright (c) 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..'))
+  sys.path.append(tracing_path)
+  from tracing_build import run_vinn_tests
+  sys.exit(run_vinn_tests.Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/slim_trace b/third_party/catapult/tracing/bin/slim_trace
new file mode 100755
index 0000000000000000000000000000000000000000..2e170777840855ede68c7d119a21f7f41fd47c64
--- /dev/null
+++ b/third_party/catapult/tracing/bin/slim_trace
@@ -0,0 +1,16 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import sys
+import os
+
+tracing_path = os.path.abspath(os.path.join(
+  os.path.dirname(os.path.realpath(__file__)), '..'))
+sys.path.append(tracing_path)
+from tracing_build import slim_trace
+
+
+if __name__ == '__main__':
+  sys.exit(slim_trace.Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/strip_memory_infra_trace b/third_party/catapult/tracing/bin/strip_memory_infra_trace
new file mode 100755
index 0000000000000000000000000000000000000000..8bdc8375d75847ebb62ec39e9ec21721684ecdb6
--- /dev/null
+++ b/third_party/catapult/tracing/bin/strip_memory_infra_trace
@@ -0,0 +1,13 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+  sys.path.append(tracing_path)
+  from tracing_build import strip_memory_infra_trace
+  sys.exit(strip_memory_infra_trace.Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/symbolize_trace b/third_party/catapult/tracing/bin/symbolize_trace
new file mode 100755
index 0000000000000000000000000000000000000000..ecfbb673bf0c46bbbcb4dce0f414a68388f237fe
--- /dev/null
+++ b/third_party/catapult/tracing/bin/symbolize_trace
@@ -0,0 +1,409 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import argparse
+import bisect
+import collections
+import gzip
+import json
+import os
+import re
+import subprocess
+import sys
+
+_SYMBOLS_PATH = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)),
+    '..',
+    'third_party',
+    'symbols'))
+sys.path.append(_SYMBOLS_PATH)
+# pylint: disable=import-error
+import symbols.elf_symbolizer as elf_symbolizer
+
+
+# Relevant trace event phases from Chromium's
+# src/base/trace_event/common/trace_event_common.h.
+TRACE_EVENT_PHASE_METADATA = 'M'
+TRACE_EVENT_PHASE_MEMORY_DUMP = 'v'
+
+
+# Matches Android library paths, supports both K (/data/app-lib/<>/lib.so)
+# as well as L+ (/data/app/<>/lib/<>/lib.so). Library name is available
+# via 'name' group.
+ANDROID_PATH_MATCHER = re.compile(
+    r'^/data/(?:'
+      r'app/[^/]+/lib/[^/]+/|'
+      r'app-lib/[^/]+/|'
+      r'data/[^/]+/incremental-install-files/lib/'
+    r')(?P<name>.*\.so)')
+
+# Subpath of output path where unstripped libraries are stored.
+ANDROID_UNSTRIPPED_SUBPATH = 'lib.unstripped'
+
+
+def FindInSystemPath(binary_name):
+  paths = os.environ['PATH'].split(os.pathsep)
+  for path in paths:
+    binary_path = os.path.join(path, binary_name)
+    if os.path.isfile(binary_path):
+      return binary_path
+  return None
+
+
+def IsSymbolizableFile(file_path):
+  result = subprocess.check_output(['file', '-0', file_path])
+  type_string = result[result.find('\0') + 1:]
+  return bool(re.match(r'\: (ELF|Mach-O) (32|64)-bit\b', type_string))
+
+
+class ProcessMemoryMaps(object):
+  """Represents 'process_mmaps' trace file entry."""
+
+  class Region(object):
+    def __init__(self, start_address, size, file_path):
+      self._start_address = start_address
+      self._size = size
+      self._file_path = file_path
+
+    @property
+    def start_address(self):
+      return self._start_address
+
+    @property
+    def end_address(self):
+      return self._start_address + self._size
+
+    @property
+    def size(self):
+      return self._size
+
+    @property
+    def file_path(self):
+      return self._file_path
+
+    def __cmp__(self, other):
+      if isinstance(other, type(self)):
+        return long(self._start_address).__cmp__(long(other._start_address))
+      elif isinstance(other, (long, int)):
+        return long(self._start_address).__cmp__(long(other))
+      else:
+        raise Exception('Cannot compare with %s' % type(other))
+
+    def __repr__(self):
+      return 'Region(0x{:X} - 0x{:X}, {})'.format(
+          self.start_address, self.end_address, self.file_path)
+
+  def __init__(self, process_mmaps):
+    """Parses 'process_mmaps' dictionary."""
+
+    regions = []
+    for region_value in process_mmaps['vm_regions']:
+      regions.append(self.Region(
+          long(region_value['sa'], 16),
+          long(region_value['sz'], 16),
+          region_value['mf']))
+    regions.sort()
+
+    # Copy regions without duplicates and check for overlaps.
+    self._regions = []
+    previous_region = None
+    for region in regions:
+      if previous_region is not None:
+        if region == previous_region:
+          continue
+        assert region.start_address >= previous_region.end_address, \
+            'Regions {} and {} overlap.'.format(previous_region, region)
+      previous_region = region
+      self._regions.append(region)
+
+  @property
+  def regions(self):
+    return self._regions
+
+  def FindRegion(self, address):
+    """Finds region containing |address|. Returns None if none found."""
+
+    region_index = bisect.bisect_right(self._regions, address) - 1
+    if region_index >= 0:
+      region = self._regions[region_index]
+      if address >= region.start_address and address < region.end_address:
+        return region
+    return None
+
+
+class StackFrames(object):
+  """Represents 'stackFrames' trace file entry."""
+
+  class PCFrame(object):
+    def __init__(self, pc, frame):
+      self._modified = False
+      self._pc = pc
+      self._frame = frame
+
+    @property
+    def modified(self):
+      return self._modified
+
+    @property
+    def pc(self):
+      return self._pc
+
+    @property
+    def name(self):
+      return self._frame['name']
+
+    @name.setter
+    def name(self, value):
+      self._modified = True
+      self._frame['name'] = value
+
+  def __init__(self, stack_frames):
+    """Constructs object using 'stackFrames' dictionary."""
+    self._pc_frames = []
+    for frame in stack_frames.itervalues():
+      pc_frame = self._ParsePCFrame(frame)
+      if pc_frame:
+        self._pc_frames.append(pc_frame)
+
+  @property
+  def pc_frames(self):
+    return self._pc_frames
+
+  @property
+  def modified(self):
+    return any(f.modified for f in self._pc_frames)
+
+  _PC_TAG = 'pc:'
+
+  @classmethod
+  def _ParsePCFrame(self, frame):
+    name = frame['name']
+    if not name.startswith(self._PC_TAG):
+      return None
+    pc = long(name[len(self._PC_TAG):], 16)
+    return self.PCFrame(pc, frame)
+
+
+class Process(object):
+  """Holds various bits of information about a process in a trace file."""
+
+  def __init__(self, pid):
+    self.pid = pid
+    self.name = None
+    self.mmaps = None
+    self.stack_frames = None
+
+
+def CollectProcesses(trace):
+  """Parses trace dictionary and returns pid->Process map of all processes
+     suitable for symbolization (which have both mmaps and stack_frames).
+  """
+
+  process_map = {}
+
+  # Android traces produced via 'chrome://inspect/?tracing#devices' are
+  # just list of events.
+  events = trace if isinstance(trace, list) else trace['traceEvents']
+  for event in events:
+    name = event.get('name')
+    if not name:
+      continue
+
+    pid = event['pid']
+    process = process_map.get(pid)
+    if process is None:
+      process = Process(pid)
+      process_map[pid] = process
+
+    phase = event['ph']
+    if phase == TRACE_EVENT_PHASE_METADATA:
+      if name == 'process_name':
+        process.name = event['args']['name']
+      elif name == 'stackFrames':
+        process.stack_frames = StackFrames(event['args']['stackFrames'])
+    elif phase == TRACE_EVENT_PHASE_MEMORY_DUMP:
+      process_mmaps = event['args']['dumps'].get('process_mmaps')
+      if process_mmaps:
+        # TODO(dskiba): this parses all process_mmaps, but retains only the
+        #               last one. We need to parse only once (lazy parsing?).
+        process.mmaps = ProcessMemoryMaps(process_mmaps)
+
+  return [p for p in process_map.itervalues() if p.mmaps and p.stack_frames]
+
+
+class SymbolizableFile(object):
+  """Holds file path, addresses to symbolize and stack frames to update.
+
+  This class is a link between ELFSymbolizer and a trace file: it specifies
+  what to symbolize (addresses) and what to update with the symbolization
+  result (frames).
+  """
+  def __init__(self, file_path):
+    self.path = file_path
+    self.symbolizable_path = file_path # path to use for symbolization
+    self.frames_by_address = collections.defaultdict(list)
+
+
+def ResolveSymbolizableFiles(processes):
+  """Resolves and groups PCs into list of SymbolizableFiles.
+
+  As part of the grouping process, this function resolves PC from each stack
+  frame to the corresponding mmap region. Stack frames that failed to resolve
+  are symbolized with '<unresolved>'.
+  """
+  symfile_by_path = {}
+  for process in processes:
+    for frame in process.stack_frames.pc_frames:
+      region = process.mmaps.FindRegion(frame.pc)
+      if region is None:
+        frame.name = '<unresolved>'
+        continue
+
+      symfile = symfile_by_path.get(region.file_path)
+      if symfile is None:
+        symfile = SymbolizableFile(region.file_path)
+        symfile_by_path[symfile.path] = symfile
+
+      relative_pc = frame.pc - region.start_address
+      symfile.frames_by_address[relative_pc].append(frame)
+  return symfile_by_path.values()
+
+
+def SymbolizeFiles(symfiles, addr2line_path):
+  """Symbolizes each file in the given list of SymbolizableFiles
+     and updates stack frames with symbolization results."""
+  print 'Symbolizing...'
+
+  def _SubPrintf(message, *args):
+    print ('  ' + message).format(*args)
+
+  symbolized = False
+  for symfile in symfiles:
+    unsymbolized_name = '<{}>'.format(
+        symfile.path if symfile.path else 'unnamed')
+
+    problem = None
+    if not os.path.isabs(symfile.symbolizable_path):
+      problem = 'not a file'
+    elif not os.path.isfile(symfile.symbolizable_path):
+      problem = "file doesn't exist"
+    elif not IsSymbolizableFile(symfile.symbolizable_path):
+      problem = 'file is not symbolizable'
+    if problem:
+      _SubPrintf("Won't symbolize {} PCs for '{}': {}.",
+                 len(symfile.frames_by_address),
+                 symfile.symbolizable_path,
+                 problem)
+      for frames in symfile.frames_by_address.itervalues():
+        for frame in frames:
+          frame.name = unsymbolized_name
+      continue
+
+    def _SymbolizerCallback(sym_info, frames):
+      # Unwind inline chain to the top.
+      while sym_info.inlined_by:
+        sym_info = sym_info.inlined_by
+
+      symbolized_name = sym_info.name if sym_info.name else unsymbolized_name
+      for frame in frames:
+        frame.name = symbolized_name
+
+    symbolizer = elf_symbolizer.ELFSymbolizer(symfile.symbolizable_path,
+                                              addr2line_path,
+                                              _SymbolizerCallback,
+                                              inlines=True)
+
+    _SubPrintf('Symbolizing {} PCs from {}...',
+               len(symfile.frames_by_address),
+               symfile.path)
+
+    for address, frames in symfile.frames_by_address.iteritems():
+      # SymbolizeAsync() asserts that the type of address is int. We operate
+      # on longs (since they are raw pointers possibly from 64-bit processes).
+      # It's OK to cast here because we're passing relative PC, which should
+      # always fit into int.
+      symbolizer.SymbolizeAsync(int(address), frames)
+
+    symbolizer.Join()
+    symbolized = True
+
+  return symbolized
+
+
+def HaveFilesFromAndroid(symfiles):
+  return any(ANDROID_PATH_MATCHER.match(f.path) for f in symfiles)
+
+
+def RemapAndroidFiles(symfiles, output_path):
+  for symfile in symfiles:
+    match = ANDROID_PATH_MATCHER.match(symfile.path)
+    if match:
+      name = match.group('name')
+      symfile.symbolizable_path = os.path.join(
+          output_path, ANDROID_UNSTRIPPED_SUBPATH, name)
+    else:
+      # Clobber file path to trigger "not a file" problem in SymbolizeFiles().
+      # Without this, files won't be symbolized with "file not found" problem,
+      # which is not accurate.
+      symfile.symbolizable_path = 'android://{}'.format(symfile.path)
+
+
+# Suffix used for backup files.
+BACKUP_FILE_TAG = '.BACKUP'
+
+def main():
+  parser = argparse.ArgumentParser()
+  parser.add_argument('file',
+                      help='Trace file to symbolize (.json or .json.gz)')
+  parser.add_argument('--no-backup',
+                      dest='backup', default='true', action='store_false',
+                      help="Don't create {} files".format(BACKUP_FILE_TAG))
+  parser.add_argument('--output-directory',
+                      help='The path to the build output directory, such ' +
+                           'as out/Debug. Only needed for Android.')
+  options = parser.parse_args()
+
+  trace_file_path = options.file
+  def _OpenTraceFile(mode):
+    if trace_file_path.endswith('.gz'):
+      return gzip.open(trace_file_path, mode + 'b')
+    else:
+      return open(trace_file_path, mode + 't')
+
+  addr2line_path = FindInSystemPath('addr2line')
+  if addr2line_path is None:
+    sys.exit("Can't symbolize - no addr2line in PATH.")
+
+  print 'Reading trace file...'
+  with _OpenTraceFile('r') as trace_file:
+    trace = json.load(trace_file)
+
+  processes = CollectProcesses(trace)
+  symfiles = ResolveSymbolizableFiles(processes)
+
+  # Android trace files don't have any indication they are from Android.
+  # So we're checking for Android-specific paths.
+  if HaveFilesFromAndroid(symfiles):
+    if not options.output_directory:
+      parser.error('The trace file appears to be from Android. Please '
+                   "specify output directory (e.g. 'out/Debug') to properly "
+                   'symbolize it.')
+    RemapAndroidFiles(symfiles, os.path.abspath(options.output_directory))
+
+  if SymbolizeFiles(symfiles, addr2line_path):
+    if options.backup:
+      backup_file_path = trace_file_path + BACKUP_FILE_TAG
+      print 'Backing up trace file to {}...'.format(backup_file_path)
+      os.rename(trace_file_path, backup_file_path)
+
+    print 'Updating trace file...'
+    with _OpenTraceFile('w') as trace_file:
+      json.dump(trace, trace_file)
+  else:
+    print 'No PCs symbolized - not updating trace file.'
+
+
+if __name__ == '__main__':
+  main()
diff --git a/third_party/catapult/tracing/bin/trace2html b/third_party/catapult/tracing/bin/trace2html
new file mode 100755
index 0000000000000000000000000000000000000000..0457d39a66b2c6357a8ca0060f4ca657cf34a049
--- /dev/null
+++ b/third_party/catapult/tracing/bin/trace2html
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+# Copyright (c) 2014 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import sys
+import os
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..'))
+  sys.path.append(tracing_path)
+  from tracing_build import trace2html
+  sys.exit(trace2html.Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/update_gypi b/third_party/catapult/tracing/bin/update_gypi
new file mode 100755
index 0000000000000000000000000000000000000000..a4e54fdf3485595a8d55644aefa2dd0f1ecf29a0
--- /dev/null
+++ b/third_party/catapult/tracing/bin/update_gypi
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+# Copyright (c) 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..'))
+  sys.path.append(tracing_path)
+  from tracing_build import update_gypi
+  sys.exit(update_gypi.Update())
diff --git a/third_party/catapult/tracing/bin/validate_all_metrics b/third_party/catapult/tracing/bin/validate_all_metrics
new file mode 100755
index 0000000000000000000000000000000000000000..b4b26307bc172f8b158c90cc1cc864547bda7708
--- /dev/null
+++ b/third_party/catapult/tracing/bin/validate_all_metrics
@@ -0,0 +1,43 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import argparse
+import json
+import os
+import string
+import sys
+
+sys.path.insert(
+    1,
+    os.path.join(os.path.dirname(os.path.realpath(__file__)), '..'))
+from tracing.metrics import discover
+import tracing_project
+
+
+def Main():
+  all_registered_metrics = set(discover.DiscoverMetrics(
+      ['/tracing/metrics/all_metrics.html']))
+  all_modules = list(
+      '/' + rel_path for rel_path in
+      tracing_project.TracingProject().FindAllMetricsModuleRelPaths())
+  all_possible_metrics = set(discover.DiscoverMetrics(all_modules))
+  unregistered_metrics = all_possible_metrics - all_registered_metrics
+  if unregistered_metrics:
+    print ('These metrics are unregistered: %s. Please import their modules in '
+           'tracing/tracing/metrics/all_metrics.html' %
+           ', '.join(unregistered_metrics))
+    return 1
+  uppercased_metrics = []
+  for m in all_possible_metrics:
+    if str.isupper(m[0]):
+      uppercased_metrics.append(m)
+  if uppercased_metrics:
+    print ('These metrics must be renamed to start with a lower-case: %s' %
+           uppercased_metrics)
+    return 1
+  return 0
+
+if __name__ == '__main__':
+  sys.exit(Main())
diff --git a/third_party/catapult/tracing/bin/valueset2html b/third_party/catapult/tracing/bin/valueset2html
new file mode 100755
index 0000000000000000000000000000000000000000..2e950ba392c742ec5bb3044f5da9808f01ac75d5
--- /dev/null
+++ b/third_party/catapult/tracing/bin/valueset2html
@@ -0,0 +1,55 @@
+#!/usr/bin/env python
+# Copyright 2016 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import argparse
+import codecs
+import json
+import sys
+import os
+
+tracing_path = os.path.abspath(os.path.join(
+  os.path.dirname(os.path.realpath(__file__)), '..'))
+sys.path.append(tracing_path)
+from tracing import results_renderer
+
+
+def main():
+  parser = argparse.ArgumentParser(
+      description='Upgrade a results2 instance or add a new HistogramSet.',
+      add_help=False)
+  parser.add_argument('html_path', metavar='HTML_PATH',
+                      help='HTML file path (output).')
+  parser.add_argument('-h', '--help', action='help',
+                      help='Show this help message and exit.')
+  parser.add_argument('--html', nargs='+', default=[],
+                      help='Zero or more HTML file paths (input).')
+  parser.add_argument('--json', nargs='+', default=[],
+                      help='Zero or more HistogramSet JSON file paths (input).')
+  parser.add_argument('--mapresults', nargs='+', default=[],
+                      help='Zero or more map results JSON file paths (input).')
+  args = parser.parse_args()
+
+  histograms = []
+
+  for html_path in args.html:
+    histograms.extend(results_renderer.ReadExistingResults(
+        open(html_path, 'r')))
+
+  for json_path in args.json:
+    histograms.extend(json.load(open(json_path, 'r')))
+
+  for json_path in args.mapresults:
+    for filename, results in json.load(open(json_path, 'r')).iteritems():
+      for histogram in results['pairs']['histograms']:
+        histogram['diagnostics']['iteration']['storyUrl'] = filename
+        histograms.append(histogram)
+
+  open(args.html_path, 'a').close() # Create file if it doesn't exist.
+  with codecs.open(args.html_path,
+                   mode='r+', encoding='utf-8') as output_stream:
+    results_renderer.RenderHTMLView(histograms, output_stream)
+
+if __name__ == '__main__':
+  sys.exit(main())
diff --git a/third_party/catapult/tracing/bin/vulcanize_trace_viewer b/third_party/catapult/tracing/bin/vulcanize_trace_viewer
new file mode 100755
index 0000000000000000000000000000000000000000..c3602d40447536b69c319b0a926e8932ffcafb37
--- /dev/null
+++ b/third_party/catapult/tracing/bin/vulcanize_trace_viewer
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+# Copyright (c) 2014 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import os
+import sys
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..'))
+  sys.path.append(tracing_path)
+  from tracing_build import vulcanize_trace_viewer
+  sys.exit(vulcanize_trace_viewer.Main(sys.argv))
diff --git a/third_party/catapult/tracing/bin/why_imported b/third_party/catapult/tracing/bin/why_imported
new file mode 100755
index 0000000000000000000000000000000000000000..9b5a59a6f49058f61592e5c71cee7da2fa63a9bc
--- /dev/null
+++ b/third_party/catapult/tracing/bin/why_imported
@@ -0,0 +1,48 @@
+#!/usr/bin/env python
+# Copyright 2015 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+"""Produces a dot file showing dependency relationships between modules.
+
+The dot file contains a text-based representation of a directed graph that
+explains why given module names were included in a trace_viewer config.
+
+Example usage:
+$ ./why_imported tracing.ui.analysis.analysis_view > ~/analysis_view.dot
+
+This can then be converted to a graphical representation with the dot tool:
+$ dot -Grankdir=LR -Tpng ~/analysis_view.dot -o ~/analysis_view.png
+"""
+
+import os
+import sys
+import argparse
+
+
+def Main():
+  project = tracing_project.TracingProject()
+
+  parser = argparse.ArgumentParser(
+      usage='%(prog)s <options> moduleNames', epilog=__doc__)
+  parser.add_argument('--config', choices=project.GetConfigNames())
+  parser.add_argument('module_names', nargs='+')
+  args = parser.parse_args()
+
+  if args.config:
+    names = [project.GetModuleNameForConfigName(options.config)]
+    vulcanizer = project.CreateVulcanizer()
+    load_sequence = vulcanizer.CalcLoadSequenceForModuleNames(names)
+  else:
+    parser.error('No config specified.')
+  print vulcanizer.GetDominatorGraphForModulesNamed(
+      args.module_names, load_sequence)
+
+
+if __name__ == '__main__':
+  tracing_path = os.path.abspath(os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), '..'))
+  sys.path.append(tracing_path)
+  import tracing_project
+  tracing_project.UpdateSysPathIfNeeded()
+  sys.exit(Main())
