diff --git a/third_party/cld_2/src/public/compact_lang_det.h b/third_party/cld_2/src/public/compact_lang_det.h
new file mode 100644
index 0000000000000000000000000000000000000000..4a6f36663b418553d30f11b3da6d1bb3da22abc3
--- /dev/null
+++ b/third_party/cld_2/src/public/compact_lang_det.h
@@ -0,0 +1,429 @@
+// Copyright 2013 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// Author: dsites@google.com (Dick Sites)
+//
+
+// NOTE:
+// Baybayin (ancient script of the Philippines) is detected as TAGALOG.
+// Chu Nom (Vietnamese ancient Han characters) is detected as VIETNAMESE.
+// HAITIAN_CREOLE is detected as such.
+// NORWEGIAN and NORWEGIAN_N are detected separately (but not robustly)
+// PORTUGUESE, PORTUGUESE_P, and PORTUGUESE_B are all detected as PORTUGUESE.
+// ROMANIAN-Latin is detected as ROMANIAN; ROMANIAN-Cyrillic as ROMANIAN.
+// BOSNIAN is not detected as such, but likely scores as Croatian or Serbian.
+// MONTENEGRIN is not detected as such, but likely scores as Serbian.
+// CROATIAN is detected in the Latin script
+// SERBIAN is detected in the Cyrililc and Latin scripts
+// Zhuang is detected in the Latin script only.
+//
+// The languages X_PIG_LATIN and X_KLINGON are detected in the
+//  extended calls ExtDetectLanguageSummary().
+//
+// UNKNOWN_LANGUAGE is returned if no language's internal reliablity measure
+//  is high enough. This happens with non-text input such as the bytes of a
+//  JPEG, and also with text in languages outside training set.
+//
+// The following languages are to be detected in multiple scripts:
+//  AZERBAIJANI (Latin, Cyrillic*, Arabic*)
+//  BURMESE (Latin, Myanmar)
+//  HAUSA (Latin, Arabic)
+//  KASHMIRI (Arabic, Devanagari)
+//  KAZAKH (Latin, Cyrillic, Arabic)
+//  KURDISH (Latin, Arabic)
+//  KYRGYZ (Cyrillic, Arabic)
+//  LIMBU (Devanagari, Limbu)
+//  MONGOLIAN (Cyrillic, Mongolian)
+//  SANSKRIT (Latin, Devanagari)
+//  SINDHI (Arabic, Devanagari)
+//  TAGALOG (Latin, Tagalog)
+//  TAJIK (Cyrillic, Arabic*)
+//  TATAR (Latin, Cyrillic, Arabic)
+//  TURKMEN (Latin, Cyrillic, Arabic)
+//  UIGHUR (Latin, Cyrillic, Arabic)
+//  UZBEK (Latin, Cyrillic, Arabic)
+//
+// * Due to a shortage of training text, AZERBAIJANI is not currently detected
+//   in Arabic or Cyrillic scripts, nor TAJIK in Arabic script.
+//
+
+#ifndef I18N_ENCODINGS_CLD2_PUBLIC_COMPACT_LANG_DET_H_
+#define I18N_ENCODINGS_CLD2_PUBLIC_COMPACT_LANG_DET_H_
+
+#include <stdint.h>
+#include <vector>
+#include "../internal/integral_types.h"   // For uint8 etc.
+#include "../internal/lang_script.h"      // For Language
+
+namespace CLD2 {
+
+// NOTE: If you cannot prove the the input text is valid UTF-8 by design because
+// it went through a known-good conversion program, call one of the *CheckUTF8
+// routines. For example, never trust raw user-supplied bytes. It is especially
+// important to do a UTF8-to-UTF8 conversion on raw bytes that claim to be
+// UTF-8, using a converter that guarantees to produce valid UTF-8, turning
+// other byte sequences into the Unicode replacement character U+FFFD (deleting
+// or turning into space or question-mark can create security holes).
+
+  // Scan interchange-valid UTF-8 bytes and detect most likely language,
+  // or set of languages.
+  //
+  // Design goals:
+  //   Skip over big stretches of HTML tags
+  //   Able to return ranges of different languages
+  //   Relatively small tables and relatively fast processing
+  //   Thread safe
+  //
+  // For HTML documents, tags are skipped, along with <script> ... </script>
+  // and <style> ... </style> sequences, and entities are expanded.
+  //
+  // We distinguish between bytes of the raw input buffer and bytes of non-tag
+  // text letters. Since tags can be over 50% of the bytes of an HTML Page,
+  // and are nearly all seven-bit ASCII English, we prefer to distinguish
+  // language mixture fractions based on just the non-tag text.
+  //
+  // Inputs: text and text_length
+  //  Code skips HTML tags and expands HTML entities, unless
+  //  is_plain_text is true
+  // Outputs:
+  //  language3 is an array of the top 3 languages or UNKNOWN_LANGUAGE
+  //  percent3 is an array of the text percentages 0..100 of the top 3 languages
+  //  text_bytes is the amount of non-tag/letters-only text found
+  //  is_reliable set true if the returned Language is some amount more
+  //   probable then the second-best Language. Calculation is a complex function
+  //   of the length of the text and the different-script runs of text.
+  // Return value: the most likely Language for the majority of the input text
+  //  Length 0 input returns UNKNOWN_LANGUAGE. Very short indeterminate text
+  //  defaults to ENGLISH.
+  //
+  // The first two versions return ENGLISH instead of UNKNOWN_LANGUAGE, for
+  // backwards compatibility with a different detector.
+  //
+  // The third version may return UNKNOWN_LANGUAGE, and also returns extended
+  // language codes from lang_script.h
+  //
+
+
+  // Instead of individual arguments, pass in hints as an initialized struct
+  // Init to {NULL, NULL, UNKNOWN_ENCODING, UNKNOWN_LANGUAGE} if not known.
+  //
+  // Pass in hints whenever possible; doing so improves detection accuracy. The
+  // set of passed-in hints are all information that is external to the text
+  // itself.
+  //
+  // The content_language_hint is intended to come from an HTTP header
+  // Content-Language: field, the tld_hint from the hostname of a URL, the
+  // encoding-hint from an encoding detector applied to the input
+  // document, and the language hint from any other context you might have.
+  // The lang= tags inside an HTML document will be picked up as hints
+  // by code within the compact language detector.
+
+  typedef struct {
+    const char* content_language_hint;      // "mi,en" boosts Maori and English
+    const char* tld_hint;                   // "id" boosts Indonesian
+    int encoding_hint;                      // SJS boosts Japanese
+    Language language_hint;                 // ITALIAN boosts it
+  } CLDHints;
+
+  static const int32 kMaxResultChunkBytes = 0x7fffffff;
+
+  // Note: this was initially over-optimized to fit into 8 bytes,
+  // causing too much work to deal with with greater than 16-bit byte lengths.
+  // For returning a vector of per-language pieces of the input buffer
+  // Unreliable and too-short are mapped to UNKNOWN_LANGUAGE
+  typedef struct {
+    int offset;                 // Starting byte offset in original buffer
+    int32 bytes;                // Number of bytes in chunk
+    uint16 lang1;               // Top lang, as full Language. Apply
+                                //  static_cast<Language>() to this short value.
+    uint16 pad;                 // Make multiple of 4 bytes
+  } ResultChunk;
+  typedef std::vector<ResultChunk> ResultChunkVector;
+
+
+  // These initial simple versions all cascade through the full-blown last
+  // version which it would be better for you to use directly because you will
+  // get better results passing in any available hints.
+
+  // Scan interchange-valid UTF-8 bytes and detect most likely language
+  // If the input is in fact not valid UTF-8, this returns immediately with
+  // the result value UNKNOWN_LANGUAGE and is_reliable set to false.
+  //
+  // In all cases, valid_prefix_bytes will be set to the number of leading
+  // bytes that are valid UTF-8. If this is < buffer_length, there is invalid
+  // input starting at the following byte.
+  Language DetectLanguageCheckUTF8(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          bool* is_reliable,
+                          int* valid_prefix_bytes);
+
+  // Use this one ONLY if you can prove the the input text is valid UTF-8 by
+  // design because it went through a known-good conversion program.
+  // Scan interchange-valid UTF-8 bytes and detect most likely language
+  Language DetectLanguage(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          bool* is_reliable);
+
+  // Use this one ONLY if you can prove the the input text is valid UTF-8 by
+  // design because it went through a known-good conversion program.
+  // Scan interchange-valid UTF-8 bytes and detect list of top 3 languages.
+  // language3[0] is usually also the return value
+  Language DetectLanguageSummary(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          Language* language3,
+                          int* percent3,
+                          int* text_bytes,
+                          bool* is_reliable);
+
+  // Use this one ONLY if you can prove the the input text is valid UTF-8 by
+  // design because it went through a known-good conversion program.
+  // Same as above, with hints supplied
+  // Scan interchange-valid UTF-8 bytes and detect list of top 3 languages.
+  // language3[0] is usually also the return value
+  Language DetectLanguageSummary(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          const char* tld_hint,       // "id" boosts Indonesian
+                          int encoding_hint,          // SJS boosts Japanese
+                          Language language_hint,     // ITALIAN boosts it
+                          Language* language3,
+                          int* percent3,
+                          int* text_bytes,
+                          bool* is_reliable);
+
+  // Use this one ONLY if you can prove the the input text is valid UTF-8 by
+  // design because it went through a known-good conversion program.
+  // Scan interchange-valid UTF-8 bytes and detect list of top 3 extended
+  // languages.
+  //
+  // Extended languages are additional interface languages and Unicode
+  // single-language scripts, from lang_script.h
+  //
+  // language3[0] is usually also the return value
+  Language ExtDetectLanguageSummary(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          Language* language3,
+                          int* percent3,
+                          int* text_bytes,
+                          bool* is_reliable);
+
+  // Use this one ONLY if you can prove the the input text is valid UTF-8 by
+  // design because it went through a known-good conversion program.
+  // Same as above, with hints supplied
+  // Scan interchange-valid UTF-8 bytes and detect list of top 3 extended
+  // languages.
+  //
+  // Extended languages are additional Google interface languages and Unicode
+  // single-language scripts, from lang_script.h
+  //
+  // language3[0] is usually also the return value
+  Language ExtDetectLanguageSummary(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          const char* tld_hint,       // "id" boosts Indonesian
+                          int encoding_hint,          // SJS boosts Japanese
+                          Language language_hint,     // ITALIAN boosts it
+                          Language* language3,
+                          int* percent3,
+                          int* text_bytes,
+                          bool* is_reliable);
+
+  // Use this one ONLY if you can prove the the input text is valid UTF-8 by
+  // design because it went through a known-good conversion program.
+  // Same as above, and also returns 3 internal language scores as a ratio to
+  // normal score for real text in that language. Scores close to 1.0 indicate
+  // normal text, while scores far away from 1.0 indicate badly-skewed text or
+  // gibberish
+  //
+  Language ExtDetectLanguageSummary(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          const char* tld_hint,       // "id" boosts Indonesian
+                          int encoding_hint,          // SJS boosts Japanese
+                          Language language_hint,     // ITALIAN boosts it
+                          Language* language3,
+                          int* percent3,
+                          double* normalized_score3,
+                          int* text_bytes,
+                          bool* is_reliable);
+
+
+  // Use this one.
+  //
+  // Hints are collected into a struct.
+  // Flags are passed in (normally zero).
+  //
+  // Also returns 3 internal language scores as a ratio to
+  // normal score for real text in that language. Scores close to 1.0 indicate
+  // normal text, while scores far away from 1.0 indicate badly-skewed text or
+  // gibberish
+  //
+  // Returns a vector of chunks in different languages, so that caller may
+  // spell-check, translate, or otherwise process different parts of the input
+  // buffer in language-dependant ways.
+  //
+  // If the input is in fact not valid UTF-8, this returns immediately with
+  // the result value UNKNOWN_LANGUAGE and is_reliable set to false.
+  //
+  // In all cases, valid_prefix_bytes will be set to the number of leading
+  // bytes that are valid UTF-8. If this is < buffer_length, there is invalid
+  // input starting at the following byte.
+  Language ExtDetectLanguageSummaryCheckUTF8(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          const CLDHints* cld_hints,
+                          int flags,
+                          Language* language3,
+                          int* percent3,
+                          double* normalized_score3,
+                          ResultChunkVector* resultchunkvector,
+                          int* text_bytes,
+                          bool* is_reliable,
+                          int* valid_prefix_bytes);
+
+  // Use this one ONLY if you can prove the the input text is valid UTF-8 by
+  // design because it went through a known-good conversion program.
+  //
+  // Hints are collected into a struct.
+  // Flags are passed in (normally zero).
+  //
+  // Also returns 3 internal language scores as a ratio to
+  // normal score for real text in that language. Scores close to 1.0 indicate
+  // normal text, while scores far away from 1.0 indicate badly-skewed text or
+  // gibberish
+  //
+  // Returns a vector of chunks in different languages, so that caller may
+  // spell-check, translate, or otherwaise process different parts of the input
+  // buffer in language-dependant ways.
+  //
+  Language ExtDetectLanguageSummary(
+                          const char* buffer,
+                          int buffer_length,
+                          bool is_plain_text,
+                          const CLDHints* cld_hints,
+                          int flags,
+                          Language* language3,
+                          int* percent3,
+                          double* normalized_score3,
+                          ResultChunkVector* resultchunkvector,
+                          int* text_bytes,
+                          bool* is_reliable);
+
+  // Return version text string
+  // String is "code_version - data_build_date"
+  const char* DetectLanguageVersion();
+
+
+  // Public use flags, debug output controls
+  static const int kCLDFlagScoreAsQuads = 0x0100;  // Force Greek, etc. => quads
+  static const int kCLDFlagHtml =         0x0200;  // Debug HTML => stderr
+  static const int kCLDFlagCr =           0x0400;  // <cr> per chunk if HTML
+  static const int kCLDFlagVerbose =      0x0800;  // More debug HTML => stderr
+  static const int kCLDFlagQuiet =        0x1000;  // Less debug HTML => stderr
+  static const int kCLDFlagEcho =         0x2000;  // Echo input => stderr
+  static const int kCLDFlagBestEffort =   0x4000;  // Give best-effort answer,
+                                                   // even on short text
+
+
+/***
+
+Flag meanings:
+ kCLDFlagScoreAsQuads
+   Normally, several languages are detected solely by their Unicode script.
+   Combined with appropritate lookup tables, this flag forces them instead
+   to be detected via quadgrams. This can be a useful refinement when looking
+   for meaningful text in these languages, instead of just character sets.
+   The default tables do not support this use.
+ kCLDFlagHtml
+   For each detection call, write an HTML file to stderr, showing the text
+   chunks and their detected languages.
+ kCLDFlagCr
+   In that HTML file, force a new line for each chunk.
+ kCLDFlagVerbose
+   In that HTML file, show every lookup entry.
+ kCLDFlagQuiet
+   In that HTML file, suppress most of the output detail.
+ kCLDFlagEcho
+  Echo every input buffer to stderr.
+ kCLDFlagBestEffort
+  Give best-effort answer, instead of UNKNOWN_LANGUAGE. May be useful for
+  short text if the caller prefers an approximate answer over none.
+
+***/
+
+// Debug output: Print the resultchunkvector to file f
+void DumpResultChunkVector(FILE* f, const char* src,
+                           ResultChunkVector* resultchunkvector);
+
+// If compiled with dynamic mode, load data from the specified file location.
+// If other data has already been loaded, it is discarded and the data is read
+// in from the specified file location again (even if the file has not changed).
+// If data needs to be loaded in a context where direct access to the file
+// system is either undesireable or impossible, use loadDataFromRawAddress
+// instead to read the data from an arbitrary region in memory (such as a
+// mmap-ed file).
+// WARNING: Before calling one of the provided "loadData" methods, language
+// detection will always fail and will always return the unknown language.
+// If not compiled with dynamic mode, this method does nothing.
+void loadDataFromFile(const char* fileName);
+
+// If compiled with dynamic mode, load data from the specified location in
+// memory.
+// This method is provided as an alternative to loadDataFromFile() for use cases
+// where the loading process may not have direct access to the file system,
+// e.g., where the direct process knows the pointer to an mmap region in system
+// memory where the data file's contents have been loaded.
+// If other data has already been loaded, it is discarded and the data is read
+// in from the specified location again (even if it has not changed).
+// WARNING: Before calling one of the provided "loadData" methods, language
+// detection will always fail and will always return the unknown language.
+// If not compiled with dynamic mode, this method does nothing.
+void loadDataFromRawAddress(const void* rawAddress, const uint32_t length);
+
+// If compiled with dynamic mode, unload the data that was previously loaded
+// via loadDataFromFile() or loadDataFromRawAddress().
+// WARNING: After calling this method, language detection will no longer work
+// and will always return the unknown language.
+// If not compiled with dynamic mode, this method does nothing.
+void unloadData();
+
+// Returns true if and only if data has been loaded via a call to
+// loadDataFromFile(...) or loadDataFromRawAddress(...) and has not been
+// subsequently unladed via a call to unloadData().
+// If not compiled with dynamic mode, this method always returns true (because
+// data has been statically linked).
+bool isDataLoaded();
+
+// Returns true if and only if compiled with dynamic mode, otherwise returns
+// false. Callers can use this to make runtime checks for whether or not CLD2
+// data needs to be dynamically initialized or not, instead of relying on the
+// CLD2_DYNAMIC_MODE define.
+bool isDataDynamic();
+
+};      // End namespace CLD2
+
+#endif  // I18N_ENCODINGS_CLD2_PUBLIC_COMPACT_LANG_DET_H_
diff --git a/third_party/cld_2/src/public/encodings.h b/third_party/cld_2/src/public/encodings.h
new file mode 100644
index 0000000000000000000000000000000000000000..1eb8f0a1518eff7a7045b6db264971781cd659e5
--- /dev/null
+++ b/third_party/cld_2/src/public/encodings.h
@@ -0,0 +1,169 @@
+// Copyright 2013 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// Author: dsites@google.com (Dick Sites)
+//
+
+#ifndef I18N_ENCODINGS_CLD2_PUBLIC_ENCODINGS_H__
+#define I18N_ENCODINGS_CLD2_PUBLIC_ENCODINGS_H__
+
+namespace CLD2 {
+
+enum Encoding {
+  ISO_8859_1           =  0,  //   ASCII
+  ISO_8859_2           =  1,  //   Latin2
+  ISO_8859_3           =  2,  //
+  ISO_8859_4           =  3,  //   Latin4
+  ISO_8859_5           =  4,  //   ISO-8859-5
+  ISO_8859_6           =  5,  //   Arabic
+  ISO_8859_7           =  6,  //   Greek
+  ISO_8859_8           =  7,  //   Hebrew
+  ISO_8859_9           =  8,  //
+  ISO_8859_10          =  9,  //
+  JAPANESE_EUC_JP      = 10,  //   EUC_JP
+  JAPANESE_SHIFT_JIS   = 11,  //   SJS
+  JAPANESE_JIS         = 12,  //   JIS
+  CHINESE_BIG5         = 13,  //   BIG5
+  CHINESE_GB           = 14,  //   GB
+  CHINESE_EUC_CN       = 15,  // Misnamed. Should be EUC_TW. Was Basis Tech
+                              // CNS11643EUC, before that   EUC-CN(!)
+  KOREAN_EUC_KR        = 16,  //   KSC
+  UNICODE_UNUSED       = 17,  //   Unicode
+  CHINESE_EUC_DEC      = 18,  // Misnamed. Should be EUC_TW. Was
+                              // CNS11643EUC, before that   EUC.
+  CHINESE_CNS          = 19,  // Misnamed. Should be EUC_TW. Was
+                              // CNS11643EUC, before that   CNS.
+  CHINESE_BIG5_CP950   = 20,  //   BIG5_CP950
+  JAPANESE_CP932       = 21,  //   CP932
+  UTF8                 = 22,
+  UNKNOWN_ENCODING     = 23,
+  ASCII_7BIT           = 24,  // ISO_8859_1 with all characters <= 127.
+  RUSSIAN_KOI8_R       = 25,  //   KOI8R
+  RUSSIAN_CP1251       = 26,  //   CP1251
+
+  //----------------------------------------------------------
+  MSFT_CP1252          = 27,  // 27: CP1252 aka MSFT euro ascii
+  RUSSIAN_KOI8_RU      = 28,  // CP21866 aka KOI8-U, used for Ukrainian.
+                              // Misnamed, this is _not_ KOI8-RU but KOI8-U.
+                              // KOI8-U is used much more often than KOI8-RU.
+  MSFT_CP1250          = 29,  // CP1250 aka MSFT eastern european
+  ISO_8859_15          = 30,  // aka ISO_8859_0 aka ISO_8859_1 euroized
+  //----------------------------------------------------------
+
+  //----------------------------------------------------------
+  MSFT_CP1254          = 31,  // used for Turkish
+  MSFT_CP1257          = 32,  // used in Baltic countries
+  //----------------------------------------------------------
+
+  //----------------------------------------------------------
+  //----------------------------------------------------------
+  ISO_8859_11          = 33,  // aka TIS-620, used for Thai
+  MSFT_CP874           = 34,  // used for Thai
+  MSFT_CP1256          = 35,  // used for Arabic
+
+  //----------------------------------------------------------
+  MSFT_CP1255          = 36,  // Logical Hebrew Microsoft
+  ISO_8859_8_I         = 37,  // Iso Hebrew Logical
+  HEBREW_VISUAL        = 38,  // Iso Hebrew Visual
+  //----------------------------------------------------------
+
+  //----------------------------------------------------------
+  CZECH_CP852          = 39,
+  CZECH_CSN_369103     = 40,  // aka ISO_IR_139 aka KOI8_CS
+  MSFT_CP1253          = 41,  // used for Greek
+  RUSSIAN_CP866        = 42,
+  //----------------------------------------------------------
+
+  //----------------------------------------------------------
+  // Handled by iconv in glibc
+  ISO_8859_13          = 43,
+  ISO_2022_KR          = 44,
+  GBK                  = 45,
+  GB18030              = 46,
+  BIG5_HKSCS           = 47,
+  ISO_2022_CN          = 48,
+
+  //-----------------------------------------------------------
+  // Following 4 encodings are deprecated (font encodings)
+  TSCII                = 49,
+  TAMIL_MONO           = 50,
+  TAMIL_BI             = 51,
+  JAGRAN               = 52,
+
+
+  MACINTOSH_ROMAN      = 53,
+  UTF7                 = 54,
+
+  //-----------------------------------------------------------
+  // Following 2 encodings are deprecated (font encodings)
+  BHASKAR              = 55,  // Indic encoding - Devanagari
+  HTCHANAKYA           = 56,  // 56 Indic encoding - Devanagari
+
+  //-----------------------------------------------------------
+  UTF16BE              = 57,  // big-endian UTF-16
+  UTF16LE              = 58,  // little-endian UTF-16
+  UTF32BE              = 59,  // big-endian UTF-32
+  UTF32LE              = 60,  // little-endian UTF-32
+  //-----------------------------------------------------------
+
+  //-----------------------------------------------------------
+  // An encoding that means "This is not text, but it may have some
+  // simple ASCII text embedded". Intended input conversion
+  // is to keep strings of >=4 seven-bit ASCII characters
+  BINARYENC            = 61,
+  //-----------------------------------------------------------
+
+  //-----------------------------------------------------------
+  // Some Web pages allow a mixture of HZ-GB and GB-2312 by using
+  // ~{ ... ~} for 2-byte pairs, and the browsers support this.
+  HZ_GB_2312           = 62,
+  //-----------------------------------------------------------
+
+  //-----------------------------------------------------------
+  // Some external vendors make the common input error of
+  // converting MSFT_CP1252 to UTF8 *twice*.
+  UTF8UTF8             = 63,
+  //-----------------------------------------------------------
+
+  //-----------------------------------------------------------
+  // Following 6 encodings are deprecated (font encodings)
+  TAM_ELANGO           = 64,  // Elango - Tamil
+  TAM_LTTMBARANI       = 65,  // Barani - Tamil
+  TAM_SHREE            = 66,  // Shree - Tamil
+  TAM_TBOOMIS          = 67,  // TBoomis - Tamil
+  TAM_TMNEWS           = 68,  // TMNews - Tamil
+  TAM_WEBTAMIL         = 69,  // Webtamil - Tamil
+  //-----------------------------------------------------------
+
+  //-----------------------------------------------------------
+  // Shift_JIS variants used by Japanese cell phone carriers.
+  KDDI_SHIFT_JIS       = 70,
+  DOCOMO_SHIFT_JIS     = 71,
+  SOFTBANK_SHIFT_JIS   = 72,
+  // ISO-2022-JP variants used by KDDI and SoftBank.
+  KDDI_ISO_2022_JP     = 73,
+  SOFTBANK_ISO_2022_JP = 74,
+  //-----------------------------------------------------------
+
+  NUM_ENCODINGS        = 75,  // Always keep this at the end. It is not a
+                              // valid Encoding enum, it is only used to
+                              // indicate the total number of Encodings.
+};
+
+}       // End namespace CLD2
+
+#endif  // I18N_ENCODINGS_CLD2_PUBLIC_ENCODINGS_H__
+
+
diff --git a/third_party/colorama/src/.gitignore b/third_party/colorama/src/.gitignore
new file mode 100644
index 0000000000000000000000000000000000000000..888499b429c93590ff3dd598c7bf16d38c22e269
--- /dev/null
+++ b/third_party/colorama/src/.gitignore
@@ -0,0 +1,7 @@
+*.pyc
+*.egg-info
+MANIFEST
+dist
+tags
+virtualenv
+
diff --git a/third_party/colorama/src/CHANGELOG.rst b/third_party/colorama/src/CHANGELOG.rst
new file mode 100644
index 0000000000000000000000000000000000000000..bbd59348f08506bef3401938d5896f9414c1805f
--- /dev/null
+++ b/third_party/colorama/src/CHANGELOG.rst
@@ -0,0 +1,103 @@
+0.3.1
+  * Fixed crash on exit with closed stdout, with thanks to Marc Abramowitz.
+  * Now uses setuptools if available, and falls back to distutils if not.
+  * setup.py no longer imports anything from colorama source.
+0.3.0
+  * Move repository to Git, https://github.com/tartley/colorama. (My Mercurial
+    repo seemed to be corrupted, I couldn't commit nor view patches of old
+    commits, even on fresh checkouts.)
+  * Fix always-crash on non-Windows platforms, reported by Matt McCormick.
+  * Fix issue #47, incompatible with pyreadline.
+0.2.7
+    Fix problem under 64-bit windows due to ctypes HANDLE size.
+    Submitted by the rather magnificent Ben Hoyt.
+    This fixes issue 43 (http://code.google.com/p/colorama/issues/detail?id=43)
+0.2.6
+    Add copyright & licensing info to every file, as requested by a large
+    downstream project which has problems making sure that all 3rd party
+    contributions have appropriate license.
+0.2.5
+    Severeral documentation & demo fixes.
+0.2.4
+    Fix to work on Windows 7.
+    Python 3 compatibility in docs and demos.
+    Add handling for 'cursor up' and 'get position' ANSI codes.
+0.2.3
+	Split changelog out into separate file.
+0.2.2
+    Fix bug which caused init() to raise, introduced in 0.2.1.
+    Remove asserts which cause problems in various circumstances. At least
+    some users saw asserts fail on 'success' returned from win32 functions,
+    even though the win32 functions appear to have worked correctly.
+0.2.1
+    Completely broken: I added a bug which caused init() to raise.
+    Added some documentation for cursor positioning and clear screen to README.
+    Add 'reinit' and 'deinit' functions, as suggested by Charles FOL and
+    Romanov DA.
+0.2
+    Merge in changes from Daniel Griffith: Add ANSI cursor positioning &
+    partial support for clear screen. Patch submitted by Oscar Lester, don't
+    send RESET_ALL to non-tty. Demos split into separate files and moved into
+    their own directory. Tweak sys.path in demos so they run against local
+    source, not installed version of Colorama.
+0.1.18
+    Fix README (no such attr as Fore.DEFAULT, etc), kindly reported by nodakai.
+0.1.17
+    Prevent printing of garbage ANSI codes upon installing with pip
+0.1.16
+    Re-upload to fix previous error. Make clean now removes old MANIFEST.
+0.1.15
+    Completely broken. Distribution was empty due to leftover invalid MANIFEST
+    file from building on a different platform.
+    Fix python3 incompatibility kindly reported by G |uumlaut| nter Kolousek
+0.1.14
+    Fix hard-coded reset to white-on-black colors. Fore.RESET, Back.RESET
+    and Style.RESET_ALL now revert to the colors as they were when init()
+    was called. Some lessons hopefully learned about testing prior to release.
+0.1.13
+    Completely broken: barfed when installed using pip.
+0.1.12
+    Completely broken: contained no source code. double oops.
+0.1.11
+    Completely broken: fatal import errors on Ubuntu. oops.
+0.1.10
+    Stop emulating 'bright' text with bright backgrounds.
+    Display 'normal' text using win32 normal foreground instead of bright.
+    Drop support for 'dim' text.
+0.1.9
+    Fix incompatibility with Python 2.5 and earlier.
+    Remove setup.py dependency on setuptools, now uses stdlib distutils.
+0.1.8
+    Fix ghastly errors all over the place on Ubuntu.
+    Add init kwargs 'convert' and 'strip', which supercede the old 'wrap'.
+0.1.7
+    Python 3 compatible.
+    Fix: Now strips ansi on windows without necessarily converting it to
+    win32 calls (eg. if output is not a tty.)
+    Fix: Flaky interaction of interleaved ansi sent to stdout and stderr.
+    Improved demo.sh (hg checkout only.)
+0.1.6
+    Fix ansi sequences with no params now default to parmlist of [0].
+    Fix flaky behaviour of autoreset and reset_all atexit.
+    Fix stacking of repeated atexit calls - now just called once.
+    Fix ghastly import problems while running tests.
+    'demo.py' (hg checkout only) now demonstrates autoreset and reset atexit.
+    Provide colorama.VERSION, used by setup.py.
+    Tests defanged so they no longer actually change terminal color when run.
+0.1.5
+    Now works on Ubuntu.
+0.1.4
+    Implemented RESET_ALL on application exit
+0.1.3
+    Implemented init(wrap=False)
+0.1.2
+    Implemented init(autoreset=True)
+0.1.1
+    Minor tidy
+0.1
+    Works on Windows for foreground color, background color, bright or dim
+
+
+.. |uumlaut| unicode:: U+00FC .. u with umlaut
+   :trim:
+
diff --git a/third_party/colorama/src/LICENSE.txt b/third_party/colorama/src/LICENSE.txt
new file mode 100644
index 0000000000000000000000000000000000000000..5f567799f395f33acd05abbc5d409dd126ce0dc8
--- /dev/null
+++ b/third_party/colorama/src/LICENSE.txt
@@ -0,0 +1,28 @@
+Copyright (c) 2010 Jonathan Hartley
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+
+* Redistributions of source code must retain the above copyright notice, this
+  list of conditions and the following disclaimer.
+
+* Redistributions in binary form must reproduce the above copyright notice,
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
+
+* Neither the name of the copyright holders, nor those of its contributors
+  may be used to endorse or promote products derived from this software without
+  specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
diff --git a/third_party/colorama/src/MANIFEST.in b/third_party/colorama/src/MANIFEST.in
new file mode 100644
index 0000000000000000000000000000000000000000..b7f36646e3d1d2678bea3f332c5e640bef4e203a
--- /dev/null
+++ b/third_party/colorama/src/MANIFEST.in
@@ -0,0 +1,3 @@
+include LICENSE.txt CHANGELOG.rst
+recursive-include demos *.py *.bat *.sh
+
diff --git a/third_party/colorama/src/Makefile b/third_party/colorama/src/Makefile
new file mode 100644
index 0000000000000000000000000000000000000000..20c5602af96036ce3df187c18a4291c5b561aff6
--- /dev/null
+++ b/third_party/colorama/src/Makefile
@@ -0,0 +1,33 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+
+# This makefile is just a cheatsheet to remind me of some commonly used
+# commands. I generally am executing these commands on Ubuntu, or on WindowsXP
+# with Cygwin binaries at the start of the PATH.
+
+NAME=colorama
+
+clean:
+	-rm -rf build dist MANIFEST colorama.egg-info
+	-find . -name '*.py[oc]' -exec rm {} \;
+.PHONY: clean
+
+sdist: clean
+	python setup.py sdist --formats=zip,gztar
+.PHONY: sdist
+
+register: clean
+	python setup.py sdist --formats=zip,gztar register 
+.PHONY: release
+
+upload: clean
+	python setup.py sdist --formats=zip,gztar register upload
+.PHONY: release
+
+test:
+	python -m unittest discover -p *_test.py
+.PHONY: test
+
+tags:
+	ctags -R ${NAME}
+.PHONY: tags
+
diff --git a/third_party/colorama/src/README.txt b/third_party/colorama/src/README.txt
new file mode 100644
index 0000000000000000000000000000000000000000..a87e16a789143f298ed85238867a51bea0beea48
--- /dev/null
+++ b/third_party/colorama/src/README.txt
@@ -0,0 +1,310 @@
+Download and docs:
+    http://pypi.python.org/pypi/colorama
+Development:
+    https://github.com/tartley/colorama
+Discussion group:
+     https://groups.google.com/forum/#!forum/python-colorama
+
+Description
+===========
+
+Makes ANSI escape character sequences for producing colored terminal text and
+cursor positioning work under MS Windows.
+
+ANSI escape character sequences have long been used to produce colored terminal
+text and cursor positioning on Unix and Macs. Colorama makes this work on
+Windows, too, by wrapping stdout, stripping ANSI sequences it finds (which
+otherwise show up as gobbledygook in your output), and converting them into the
+appropriate win32 calls to modify the state of the terminal. On other platforms,
+Colorama does nothing.
+
+Colorama also provides some shortcuts to help generate ANSI sequences
+but works fine in conjunction with any other ANSI sequence generation library,
+such as Termcolor (http://pypi.python.org/pypi/termcolor.)
+
+This has the upshot of providing a simple cross-platform API for printing
+colored terminal text from Python, and has the happy side-effect that existing
+applications or libraries which use ANSI sequences to produce colored output on
+Linux or Macs can now also work on Windows, simply by calling
+``colorama.init()``.
+
+An alternative approach is to install 'ansi.sys' on Windows machines, which
+provides the same behaviour for all applications running in terminals. Colorama
+is intended for situations where that isn't easy (e.g. maybe your app doesn't
+have an installer.)
+
+Demo scripts in the source code repository prints some colored text using
+ANSI sequences. Compare their output under Gnome-terminal's built in ANSI
+handling, versus on Windows Command-Prompt using Colorama:
+
+.. image:: https://github.com/tartley/colorama/raw/master/screenshots/ubuntu-demo.png
+    :width: 661
+    :height: 357
+    :alt: ANSI sequences on Ubuntu under gnome-terminal.
+
+.. image:: https://github.com/tartley/colorama/raw/master/screenshots/windows-demo.png
+    :width: 668
+    :height: 325
+    :alt: Same ANSI sequences on Windows, using Colorama.
+
+These screengrabs show that Colorama on Windows does not support ANSI 'dim
+text': it looks the same as 'normal text'.
+
+
+License
+=======
+
+Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+
+
+Dependencies
+============
+
+None, other than Python. Tested on Python 2.5.5, 2.6.5, 2.7, 3.1.2, and 3.2
+
+Usage
+=====
+
+Initialisation
+--------------
+
+Applications should initialise Colorama using::
+
+    from colorama import init
+    init()
+
+If you are on Windows, the call to ``init()`` will start filtering ANSI escape
+sequences out of any text sent to stdout or stderr, and will replace them with
+equivalent Win32 calls.
+
+Calling ``init()`` has no effect on other platforms (unless you request other
+optional functionality, see keyword args below.) The intention is that
+applications can call ``init()`` unconditionally on all platforms, after which
+ANSI output should just work.
+
+To stop using colorama before your program exits, simply call ``deinit()``.
+This will restore stdout and stderr to their original values, so that Colorama
+is disabled. To start using Colorama again, call ``reinit()``, which wraps
+stdout and stderr again, but is cheaper to call than doing ``init()`` all over
+again.
+
+
+Colored Output
+--------------
+
+Cross-platform printing of colored text can then be done using Colorama's
+constant shorthand for ANSI escape sequences::
+
+    from colorama import Fore, Back, Style
+    print(Fore.RED + 'some red text')
+    print(Back.GREEN + 'and with a green background')
+    print(Style.DIM + 'and in dim text')
+    print(Fore.RESET + Back.RESET + Style.RESET_ALL)
+    print('back to normal now')
+
+or simply by manually printing ANSI sequences from your own code::
+
+    print('/033[31m' + 'some red text')
+    print('/033[30m' # and reset to default color)
+
+or Colorama can be used happily in conjunction with existing ANSI libraries
+such as Termcolor::
+
+    from colorama import init
+    from termcolor import colored
+
+    # use Colorama to make Termcolor work on Windows too
+    init()
+
+    # then use Termcolor for all colored text output
+    print(colored('Hello, World!', 'green', 'on_red'))
+
+Available formatting constants are::
+
+    Fore: BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, WHITE, RESET.
+    Back: BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, WHITE, RESET.
+    Style: DIM, NORMAL, BRIGHT, RESET_ALL
+
+Style.RESET_ALL resets foreground, background and brightness. Colorama will
+perform this reset automatically on program exit.
+
+
+Cursor Positioning
+------------------
+
+ANSI codes to reposition the cursor are supported. See demos/demo06.py for
+an example of how to generate them.
+
+
+Init Keyword Args
+-----------------
+
+``init()`` accepts some kwargs to override default behaviour.
+
+init(autoreset=False):
+    If you find yourself repeatedly sending reset sequences to turn off color
+    changes at the end of every print, then ``init(autoreset=True)`` will
+    automate that::
+
+        from colorama import init
+        init(autoreset=True)
+        print(Fore.RED + 'some red text')
+        print('automatically back to default color again')
+
+init(strip=None):
+    Pass ``True`` or ``False`` to override whether ansi codes should be
+    stripped from the output. The default behaviour is to strip if on Windows.
+
+init(convert=None):
+    Pass ``True`` or ``False`` to override whether to convert ansi codes in the
+    output into win32 calls. The default behaviour is to convert if on Windows
+    and output is to a tty (terminal).
+
+init(wrap=True):
+    On Windows, colorama works by replacing ``sys.stdout`` and ``sys.stderr``
+    with proxy objects, which override the .write() method to do their work. If
+    this wrapping causes you problems, then this can be disabled by passing
+    ``init(wrap=False)``. The default behaviour is to wrap if autoreset or
+    strip or convert are True.
+
+    When wrapping is disabled, colored printing on non-Windows platforms will
+    continue to work as normal. To do cross-platform colored output, you can
+    use Colorama's ``AnsiToWin32`` proxy directly::
+
+        import sys
+        from colorama import init, AnsiToWin32
+        init(wrap=False)
+        stream = AnsiToWin32(sys.stderr).stream
+
+        # Python 2
+        print >>stream, Fore.BLUE + 'blue text on stderr'
+
+        # Python 3
+        print(Fore.BLUE + 'blue text on stderr', file=stream)
+
+
+Status & Known Problems
+=======================
+
+I've personally only tested it on WinXP (CMD, Console2), Ubuntu
+(gnome-terminal, xterm), and OSX.
+
+Some presumably valid ANSI sequences aren't recognised (see details below)
+but to my knowledge nobody has yet complained about this. Puzzling.
+
+See outstanding issues and wishlist, still on Google code for the time being:
+http://code.google.com/p/colorama/issues/list
+
+If anything doesn't work for you, or doesn't do what you expected or hoped for,
+I'd love to hear about it on that issues list, would be delighted by patches,
+and would be happy to grant commit access to anyone who submits a working patch
+or two.
+
+
+Recognised ANSI Sequences
+=========================
+
+ANSI sequences generally take the form:
+
+    ESC [ <param> ; <param> ... <command>
+
+Where <param> is an integer, and <command> is a single letter. Zero or more
+params are passed to a <command>. If no params are passed, it is generally
+synonymous with passing a single zero. No spaces exist in the sequence, they
+have just been inserted here to make it easy to read.
+
+The only ANSI sequences that colorama converts into win32 calls are::
+
+    ESC [ 0 m       # reset all (colors and brightness)
+    ESC [ 1 m       # bright
+    ESC [ 2 m       # dim (looks same as normal brightness)
+    ESC [ 22 m      # normal brightness
+
+    # FOREGROUND:
+    ESC [ 30 m      # black
+    ESC [ 31 m      # red
+    ESC [ 32 m      # green
+    ESC [ 33 m      # yellow
+    ESC [ 34 m      # blue
+    ESC [ 35 m      # magenta
+    ESC [ 36 m      # cyan
+    ESC [ 37 m      # white
+    ESC [ 39 m      # reset
+
+    # BACKGROUND
+    ESC [ 40 m      # black
+    ESC [ 41 m      # red
+    ESC [ 42 m      # green
+    ESC [ 43 m      # yellow
+    ESC [ 44 m      # blue
+    ESC [ 45 m      # magenta
+    ESC [ 46 m      # cyan
+    ESC [ 47 m      # white
+    ESC [ 49 m      # reset
+
+    # cursor positioning
+    ESC [ y;x H     # position cursor at x across, y down
+
+    # clear the screen
+    ESC [ mode J    # clear the screen. Only mode 2 (clear entire screen)
+                    # is supported. It should be easy to add other modes,
+                    # let me know if that would be useful.
+
+Multiple numeric params to the 'm' command can be combined into a single
+sequence, eg::
+
+    ESC [ 36 ; 45 ; 1 m     # bright cyan text on magenta background
+
+All other ANSI sequences of the form ``ESC [ <param> ; <param> ... <command>``
+are silently stripped from the output on Windows.
+
+Any other form of ANSI sequence, such as single-character codes or alternative
+initial characters, are not recognised nor stripped. It would be cool to add
+them though. Let me know if it would be useful for you, via the issues on
+google code.
+
+
+Development
+===========
+
+Help and fixes welcome! Ask Jonathan for commit rights, you'll get them.
+
+Running tests requires:
+
+- Michael Foord's 'mock' module to be installed.
+- Tests are written using the 2010 era updates to 'unittest', and require to
+  be run either using Python2.7 or greater, or else to have Michael Foord's
+  'unittest2' module installed.
+
+To run tests::
+
+   python -m unittest discover -p *_test.py 
+
+This, like a few other handy commands, is captured in a Makefile.
+
+If using nose to run the tests, pass the -s flag required because 'nosetests'
+otherwise applies a proxy of its own to stdout, which confuses the unit tests.
+
+
+Contact
+=======
+
+Created by Jonathan Hartley, tartley@tartley.com
+
+
+Thanks
+======
+| Marc Abramowitz, for reporting and fixing a crash on exit with closed stdout.
+|   and for providing a solution to issue #7's setuptools/distutils debate.
+| User 'eryksun', for guidance on correctly instantiating ctypes.windll.
+| Matthew McCormick for politely pointing out a longstanding crash on non-Win.
+| Ben Hoyt, for a magnificent fix under 64-bit Windows.
+| Jesse@EmptySquare for submitting a fix for examples in the README.
+| User 'jamessp', an observant documentation fix for cursor positioning.
+| User 'vaal1239', Dave Mckee & Lackner Kristof for a tiny but much-needed Win7 fix.
+| Julien Stuyck, for wisely suggesting Python3 compatible updates to README.
+| Daniel Griffith for multiple fabulous patches.
+| Oscar Lesta for valuable fix to stop ANSI chars being sent to non-tty output.
+| Roger Binns, for many suggestions, valuable feedback, & bug reports.
+| Tim Golden for thought and much appreciated feedback on the initial idea.
+
diff --git a/third_party/colorama/src/colorama/__init__.py b/third_party/colorama/src/colorama/__init__.py
new file mode 100644
index 0000000000000000000000000000000000000000..0856986fab5d503a83fdff6d4a1302a629026b31
--- /dev/null
+++ b/third_party/colorama/src/colorama/__init__.py
@@ -0,0 +1,7 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+from .initialise import init, deinit, reinit
+from .ansi import Fore, Back, Style
+from .ansitowin32 import AnsiToWin32
+
+__version__ = '0.3.1'
+
diff --git a/third_party/colorama/src/colorama/ansi.py b/third_party/colorama/src/colorama/ansi.py
new file mode 100644
index 0000000000000000000000000000000000000000..5dfe374ceb58fbf0a2d4da9160bef06a27d470ac
--- /dev/null
+++ b/third_party/colorama/src/colorama/ansi.py
@@ -0,0 +1,50 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+'''
+This module generates ANSI character codes to printing colors to terminals.
+See: http://en.wikipedia.org/wiki/ANSI_escape_code
+'''
+
+CSI = '\033['
+
+def code_to_chars(code):
+    return CSI + str(code) + 'm'
+
+class AnsiCodes(object):
+    def __init__(self, codes):
+        for name in dir(codes):
+            if not name.startswith('_'):
+                value = getattr(codes, name)
+                setattr(self, name, code_to_chars(value))
+
+class AnsiFore:
+    BLACK   = 30
+    RED     = 31
+    GREEN   = 32
+    YELLOW  = 33
+    BLUE    = 34
+    MAGENTA = 35
+    CYAN    = 36
+    WHITE   = 37
+    RESET   = 39
+
+class AnsiBack:
+    BLACK   = 40
+    RED     = 41
+    GREEN   = 42
+    YELLOW  = 43
+    BLUE    = 44
+    MAGENTA = 45
+    CYAN    = 46
+    WHITE   = 47
+    RESET   = 49
+
+class AnsiStyle:
+    BRIGHT    = 1
+    DIM       = 2
+    NORMAL    = 22
+    RESET_ALL = 0
+
+Fore = AnsiCodes( AnsiFore )
+Back = AnsiCodes( AnsiBack )
+Style = AnsiCodes( AnsiStyle )
+
diff --git a/third_party/colorama/src/colorama/ansitowin32.py b/third_party/colorama/src/colorama/ansitowin32.py
new file mode 100644
index 0000000000000000000000000000000000000000..0053a7758227697de70be55fbca69068247edc17
--- /dev/null
+++ b/third_party/colorama/src/colorama/ansitowin32.py
@@ -0,0 +1,190 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+import re
+import sys
+
+from .ansi import AnsiFore, AnsiBack, AnsiStyle, Style
+from .winterm import WinTerm, WinColor, WinStyle
+from .win32 import windll
+
+
+winterm = None
+if windll is not None:
+    winterm = WinTerm()
+
+
+def is_a_tty(stream):
+    return hasattr(stream, 'isatty') and stream.isatty()
+
+
+class StreamWrapper(object):
+    '''
+    Wraps a stream (such as stdout), acting as a transparent proxy for all
+    attribute access apart from method 'write()', which is delegated to our
+    Converter instance.
+    '''
+    def __init__(self, wrapped, converter):
+        # double-underscore everything to prevent clashes with names of
+        # attributes on the wrapped stream object.
+        self.__wrapped = wrapped
+        self.__convertor = converter
+
+    def __getattr__(self, name):
+        return getattr(self.__wrapped, name)
+
+    def write(self, text):
+        self.__convertor.write(text)
+
+
+class AnsiToWin32(object):
+    '''
+    Implements a 'write()' method which, on Windows, will strip ANSI character
+    sequences from the text, and if outputting to a tty, will convert them into
+    win32 function calls.
+    '''
+    ANSI_RE = re.compile('\033\[((?:\d|;)*)([a-zA-Z])')
+
+    def __init__(self, wrapped, convert=None, strip=None, autoreset=False):
+        # The wrapped stream (normally sys.stdout or sys.stderr)
+        self.wrapped = wrapped
+
+        # should we reset colors to defaults after every .write()
+        self.autoreset = autoreset
+
+        # create the proxy wrapping our output stream
+        self.stream = StreamWrapper(wrapped, self)
+
+        on_windows = sys.platform.startswith('win')
+
+        # should we strip ANSI sequences from our output?
+        if strip is None:
+            strip = on_windows
+        self.strip = strip
+
+        # should we should convert ANSI sequences into win32 calls?
+        if convert is None:
+            convert = on_windows and is_a_tty(wrapped)
+        self.convert = convert
+
+        # dict of ansi codes to win32 functions and parameters
+        self.win32_calls = self.get_win32_calls()
+
+        # are we wrapping stderr?
+        self.on_stderr = self.wrapped is sys.stderr
+
+
+    def should_wrap(self):
+        '''
+        True if this class is actually needed. If false, then the output
+        stream will not be affected, nor will win32 calls be issued, so
+        wrapping stdout is not actually required. This will generally be
+        False on non-Windows platforms, unless optional functionality like
+        autoreset has been requested using kwargs to init()
+        '''
+        return self.convert or self.strip or self.autoreset
+
+
+    def get_win32_calls(self):
+        if self.convert and winterm:
+            return {
+                AnsiStyle.RESET_ALL: (winterm.reset_all, ),
+                AnsiStyle.BRIGHT: (winterm.style, WinStyle.BRIGHT),
+                AnsiStyle.DIM: (winterm.style, WinStyle.NORMAL),
+                AnsiStyle.NORMAL: (winterm.style, WinStyle.NORMAL),
+                AnsiFore.BLACK: (winterm.fore, WinColor.BLACK),
+                AnsiFore.RED: (winterm.fore, WinColor.RED),
+                AnsiFore.GREEN: (winterm.fore, WinColor.GREEN),
+                AnsiFore.YELLOW: (winterm.fore, WinColor.YELLOW),
+                AnsiFore.BLUE: (winterm.fore, WinColor.BLUE),
+                AnsiFore.MAGENTA: (winterm.fore, WinColor.MAGENTA),
+                AnsiFore.CYAN: (winterm.fore, WinColor.CYAN),
+                AnsiFore.WHITE: (winterm.fore, WinColor.GREY),
+                AnsiFore.RESET: (winterm.fore, ),
+                AnsiBack.BLACK: (winterm.back, WinColor.BLACK),
+                AnsiBack.RED: (winterm.back, WinColor.RED),
+                AnsiBack.GREEN: (winterm.back, WinColor.GREEN),
+                AnsiBack.YELLOW: (winterm.back, WinColor.YELLOW),
+                AnsiBack.BLUE: (winterm.back, WinColor.BLUE),
+                AnsiBack.MAGENTA: (winterm.back, WinColor.MAGENTA),
+                AnsiBack.CYAN: (winterm.back, WinColor.CYAN),
+                AnsiBack.WHITE: (winterm.back, WinColor.GREY),
+                AnsiBack.RESET: (winterm.back, ),
+            }
+
+
+    def write(self, text):
+        if self.strip or self.convert:
+            self.write_and_convert(text)
+        else:
+            self.wrapped.write(text)
+            self.wrapped.flush()
+        if self.autoreset:
+            self.reset_all()
+
+
+    def reset_all(self):
+        if self.convert:
+            self.call_win32('m', (0,))
+        elif not self.wrapped.closed and is_a_tty(self.wrapped):
+            self.wrapped.write(Style.RESET_ALL)
+
+
+    def write_and_convert(self, text):
+        '''
+        Write the given text to our wrapped stream, stripping any ANSI
+        sequences from the text, and optionally converting them into win32
+        calls.
+        '''
+        cursor = 0
+        for match in self.ANSI_RE.finditer(text):
+            start, end = match.span()
+            self.write_plain_text(text, cursor, start)
+            self.convert_ansi(*match.groups())
+            cursor = end
+        self.write_plain_text(text, cursor, len(text))
+
+
+    def write_plain_text(self, text, start, end):
+        if start < end:
+            self.wrapped.write(text[start:end])
+            self.wrapped.flush()
+
+
+    def convert_ansi(self, paramstring, command):
+        if self.convert:
+            params = self.extract_params(paramstring)
+            self.call_win32(command, params)
+
+
+    def extract_params(self, paramstring):
+        def split(paramstring):
+            for p in paramstring.split(';'):
+                if p != '':
+                    yield int(p)
+        return tuple(split(paramstring))
+
+
+    def call_win32(self, command, params):
+        if params == []:
+            params = [0]
+        if command == 'm':
+            for param in params:
+                if param in self.win32_calls:
+                    func_args = self.win32_calls[param]
+                    func = func_args[0]
+                    args = func_args[1:]
+                    kwargs = dict(on_stderr=self.on_stderr)
+                    func(*args, **kwargs)
+        elif command in ('H', 'f'): # set cursor position
+            func = winterm.set_cursor_position
+            func(params, on_stderr=self.on_stderr)
+        elif command in ('J'):
+            func = winterm.erase_data
+            func(params, on_stderr=self.on_stderr)
+        elif command == 'A':
+            if params == () or params == None:
+                num_rows = 1
+            else:
+                num_rows = params[0]
+            func = winterm.cursor_up
+            func(num_rows, on_stderr=self.on_stderr)
+
diff --git a/third_party/colorama/src/colorama/initialise.py b/third_party/colorama/src/colorama/initialise.py
new file mode 100644
index 0000000000000000000000000000000000000000..cba3676dd9c26b7db3cc7c5186ae4feffebbda88
--- /dev/null
+++ b/third_party/colorama/src/colorama/initialise.py
@@ -0,0 +1,56 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+import atexit
+import sys
+
+from .ansitowin32 import AnsiToWin32
+
+
+orig_stdout = sys.stdout
+orig_stderr = sys.stderr
+
+wrapped_stdout = sys.stdout
+wrapped_stderr = sys.stderr
+
+atexit_done = False
+
+
+def reset_all():
+    AnsiToWin32(orig_stdout).reset_all()
+
+
+def init(autoreset=False, convert=None, strip=None, wrap=True):
+
+    if not wrap and any([autoreset, convert, strip]):
+        raise ValueError('wrap=False conflicts with any other arg=True')
+
+    global wrapped_stdout, wrapped_stderr
+    sys.stdout = wrapped_stdout = \
+        wrap_stream(orig_stdout, convert, strip, autoreset, wrap)
+    sys.stderr = wrapped_stderr = \
+        wrap_stream(orig_stderr, convert, strip, autoreset, wrap)
+
+    global atexit_done
+    if not atexit_done:
+        atexit.register(reset_all)
+        atexit_done = True
+
+
+def deinit():
+    sys.stdout = orig_stdout
+    sys.stderr = orig_stderr
+
+
+def reinit():
+    sys.stdout = wrapped_stdout
+    sys.stderr = wrapped_stdout
+
+
+def wrap_stream(stream, convert, strip, autoreset, wrap):
+    if wrap:
+        wrapper = AnsiToWin32(stream,
+            convert=convert, strip=strip, autoreset=autoreset)
+        if wrapper.should_wrap():
+            stream = wrapper.stream
+    return stream
+
+
diff --git a/third_party/colorama/src/colorama/tests/__init__.py b/third_party/colorama/src/colorama/tests/__init__.py
new file mode 100644
index 0000000000000000000000000000000000000000..8c5661e93a205bf4fb22404d4fc50f902cc31369
--- /dev/null
+++ b/third_party/colorama/src/colorama/tests/__init__.py
@@ -0,0 +1 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
diff --git a/third_party/colorama/src/colorama/tests/ansi_test.py b/third_party/colorama/src/colorama/tests/ansi_test.py
new file mode 100644
index 0000000000000000000000000000000000000000..f4c1f74fee08ca1908da0ec2e9cdb472cbeecd2e
--- /dev/null
+++ b/third_party/colorama/src/colorama/tests/ansi_test.py
@@ -0,0 +1,61 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+import sys
+try:
+    from unittest2 import TestCase, main
+except ImportError:
+    from unittest import TestCase, main
+
+from ..ansi import Fore, Back, Style
+from ..ansitowin32 import AnsiToWin32
+
+
+stdout_orig = sys.stdout
+stderr_orig = sys.stderr
+
+
+class AnsiTest(TestCase):
+    
+    def setUp(self):
+        # sanity check: stdout should be a file or StringIO object.
+        # It will only be AnsiToWin32 if init() has previously wrapped it
+        self.assertNotEqual(type(sys.stdout), AnsiToWin32)
+        self.assertNotEqual(type(sys.stderr), AnsiToWin32)
+
+    def tearDown(self):
+        sys.stdout = stdout_orig
+        sys.stderr = stderr_orig
+
+
+    def testForeAttributes(self):
+        self.assertEqual(Fore.BLACK, '\033[30m')
+        self.assertEqual(Fore.RED, '\033[31m')
+        self.assertEqual(Fore.GREEN, '\033[32m')
+        self.assertEqual(Fore.YELLOW, '\033[33m')
+        self.assertEqual(Fore.BLUE, '\033[34m')
+        self.assertEqual(Fore.MAGENTA, '\033[35m')
+        self.assertEqual(Fore.CYAN, '\033[36m')
+        self.assertEqual(Fore.WHITE, '\033[37m')
+        self.assertEqual(Fore.RESET, '\033[39m')
+
+
+    def testBackAttributes(self):
+        self.assertEqual(Back.BLACK, '\033[40m')
+        self.assertEqual(Back.RED, '\033[41m')
+        self.assertEqual(Back.GREEN, '\033[42m')
+        self.assertEqual(Back.YELLOW, '\033[43m')
+        self.assertEqual(Back.BLUE, '\033[44m')
+        self.assertEqual(Back.MAGENTA, '\033[45m')
+        self.assertEqual(Back.CYAN, '\033[46m')
+        self.assertEqual(Back.WHITE, '\033[47m')
+        self.assertEqual(Back.RESET, '\033[49m')
+
+
+    def testStyleAttributes(self):
+        self.assertEqual(Style.DIM, '\033[2m')
+        self.assertEqual(Style.NORMAL, '\033[22m')
+        self.assertEqual(Style.BRIGHT, '\033[1m')
+
+
+if __name__ == '__main__':
+    main()
+
diff --git a/third_party/colorama/src/colorama/tests/ansitowin32_test.py b/third_party/colorama/src/colorama/tests/ansitowin32_test.py
new file mode 100644
index 0000000000000000000000000000000000000000..bb2f73a40f47a6b1e45bae4449eacd96eadb887c
--- /dev/null
+++ b/third_party/colorama/src/colorama/tests/ansitowin32_test.py
@@ -0,0 +1,206 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+try:
+    # python3
+    from io import StringIO
+except ImportError:
+    # python2
+    import StringIO
+
+try:
+    # with unittest2 installed, presumably is Python <= 2.6
+    from unittest2 import TestCase, main
+except ImportError:
+    # without unittest2 installed, hopefully is Python > 2.6
+    from unittest import TestCase, main
+
+from mock import Mock, patch
+
+from .utils import platform
+
+from ..ansi import Style
+from ..ansitowin32 import AnsiToWin32, StreamWrapper
+
+
+
+class StreamWrapperTest(TestCase):
+
+    def testIsAProxy(self):
+        mockStream = Mock()
+        wrapper = StreamWrapper(mockStream, None)
+        self.assertTrue( wrapper.random_attr is mockStream.random_attr )
+
+    def testDelegatesWrite(self):
+        mockStream = Mock()
+        mockConverter = Mock()
+        wrapper = StreamWrapper(mockStream, mockConverter)
+        wrapper.write('hello')
+        self.assertTrue(mockConverter.write.call_args, (('hello',), {}))
+
+
+class AnsiToWin32Test(TestCase):
+
+    def testInit(self):
+        mockStdout = object()
+        auto = object()
+        stream = AnsiToWin32(mockStdout, autoreset=auto)
+        self.assertEqual(stream.wrapped, mockStdout)
+        self.assertEqual(stream.autoreset, auto)
+
+    @patch('colorama.ansitowin32.winterm', None)
+    def testStripIsTrueOnWindows(self):
+        with platform('windows'):
+            stream = AnsiToWin32(None)
+            self.assertTrue(stream.strip)
+
+    def testStripIsFalseOffWindows(self):
+        with platform('darwin'):
+            stream = AnsiToWin32(None)
+            self.assertFalse(stream.strip)
+
+
+    def testWriteStripsAnsi(self):
+        mockStdout = Mock()
+        stream = AnsiToWin32(mockStdout)
+        stream.wrapped = Mock()
+        stream.write_and_convert = Mock()
+        stream.strip = True
+
+        stream.write('abc')
+
+        self.assertFalse(stream.wrapped.write.called)
+        self.assertEqual(stream.write_and_convert.call_args, (('abc',), {}))
+
+
+    def testWriteDoesNotStripAnsi(self):
+        mockStdout = Mock()
+        stream = AnsiToWin32(mockStdout)
+        stream.wrapped = Mock()
+        stream.write_and_convert = Mock()
+        stream.strip = False
+        stream.convert = False
+
+        stream.write('abc')
+        
+        self.assertFalse(stream.write_and_convert.called)
+        self.assertEqual(stream.wrapped.write.call_args, (('abc',), {}))
+
+
+    def assert_autoresets(self, convert, autoreset=True):
+        stream = AnsiToWin32(Mock())
+        stream.convert = convert
+        stream.reset_all = Mock()
+        stream.autoreset = autoreset
+        stream.winterm = Mock()
+
+        stream.write('abc')
+        
+        self.assertEqual(stream.reset_all.called, autoreset)
+
+    def testWriteAutoresets(self):
+        self.assert_autoresets(convert=True)
+        self.assert_autoresets(convert=False)
+        self.assert_autoresets(convert=True, autoreset=False)
+        self.assert_autoresets(convert=False, autoreset=False)
+
+    def testWriteAndConvertWritesPlainText(self):
+        stream = AnsiToWin32(Mock())
+        stream.write_and_convert( 'abc' )
+        self.assertEqual( stream.wrapped.write.call_args, (('abc',), {}) )
+
+    def testWriteAndConvertStripsAllValidAnsi(self):
+        stream = AnsiToWin32(Mock())
+        stream.call_win32 = Mock()
+        data = [
+            'abc\033[mdef',
+            'abc\033[0mdef',
+            'abc\033[2mdef',
+            'abc\033[02mdef',
+            'abc\033[002mdef',
+            'abc\033[40mdef',
+            'abc\033[040mdef',
+            'abc\033[0;1mdef',
+            'abc\033[40;50mdef',
+            'abc\033[50;30;40mdef',
+            'abc\033[Adef',
+            'abc\033[0Gdef',
+            'abc\033[1;20;128Hdef',
+        ]
+        for datum in data:
+            stream.wrapped.write.reset_mock()
+            stream.write_and_convert( datum )
+            self.assertEqual(
+               [args[0] for args in stream.wrapped.write.call_args_list],
+               [ ('abc',), ('def',) ]
+            )
+
+    def testWriteAndConvertSkipsEmptySnippets(self):
+        stream = AnsiToWin32(Mock())
+        stream.call_win32 = Mock()
+        stream.write_and_convert( '\033[40m\033[41m' )
+        self.assertFalse( stream.wrapped.write.called )
+
+    def testWriteAndConvertCallsWin32WithParamsAndCommand(self):
+        stream = AnsiToWin32(Mock())
+        stream.convert = True
+        stream.call_win32 = Mock()
+        stream.extract_params = Mock(return_value='params')
+        data = {
+            'abc\033[adef':         ('a', 'params'),
+            'abc\033[;;bdef':       ('b', 'params'),
+            'abc\033[0cdef':        ('c', 'params'),
+            'abc\033[;;0;;Gdef':    ('G', 'params'),
+            'abc\033[1;20;128Hdef': ('H', 'params'),
+        }
+        for datum, expected in data.items():
+            stream.call_win32.reset_mock()
+            stream.write_and_convert( datum )
+            self.assertEqual( stream.call_win32.call_args[0], expected )
+
+    def test_reset_all_shouldnt_raise_on_closed_orig_stdout(self):
+        stream = StringIO()
+        converter = AnsiToWin32(stream)
+        stream.close()
+
+        converter.reset_all()
+
+
+    def testExtractParams(self):
+        stream = AnsiToWin32(Mock())
+        data = {
+            '':               (),
+            ';;':             (),
+            '2':              (2,),
+            ';;002;;':        (2,),
+            '0;1':            (0, 1),
+            ';;003;;456;;':   (3, 456),
+            '11;22;33;44;55': (11, 22, 33, 44, 55),
+        }
+        for datum, expected in data.items():
+            self.assertEqual(stream.extract_params(datum), expected)
+
+    def testCallWin32UsesLookup(self):
+        listener = Mock()
+        stream = AnsiToWin32(listener)
+        stream.win32_calls = {
+            1: (lambda *_, **__: listener(11),),
+            2: (lambda *_, **__: listener(22),),
+            3: (lambda *_, **__: listener(33),),
+        }
+        stream.call_win32('m', (3, 1, 99, 2))
+        self.assertEqual(
+            [a[0][0] for a in listener.call_args_list],
+            [33, 11, 22] )
+
+    def testCallWin32DefaultsToParams0(self):
+        mockStdout = Mock()
+        stream = AnsiToWin32(mockStdout)
+        stream.win32_calls = {0: (mockStdout.reset,)}
+        
+        stream.call_win32('m', [])
+
+        self.assertTrue(mockStdout.reset.called)
+
+
+if __name__ == '__main__':
+    main()
+
diff --git a/third_party/colorama/src/colorama/tests/initialise_test.py b/third_party/colorama/src/colorama/tests/initialise_test.py
new file mode 100644
index 0000000000000000000000000000000000000000..3074f3338f60fabc55d237e9e9b31e740eb85b18
--- /dev/null
+++ b/third_party/colorama/src/colorama/tests/initialise_test.py
@@ -0,0 +1,117 @@
+# Copyright Jonathan Hartley 2013. BSD 3-Clause license, see LICENSE file.
+import sys
+try:
+    from unittest2 import TestCase, main
+except ImportError:
+    from unittest import TestCase, main
+
+from mock import patch
+
+from .utils import platform, redirected_output
+
+from ..initialise import init
+from ..ansitowin32 import StreamWrapper
+
+
+orig_stdout = sys.stdout
+orig_stderr = sys.stderr
+
+
+class InitTest(TestCase):
+
+    def setUp(self):
+        # sanity check
+        self.assertNotWrapped()
+
+    def tearDown(self):
+        sys.stdout = orig_stdout
+        sys.stderr = orig_stderr
+
+    def assertWrapped(self):
+        self.assertIsNot(sys.stdout, orig_stdout, 'stdout should be wrapped')
+        self.assertIsNot(sys.stderr, orig_stderr, 'stderr should be wrapped')
+        self.assertTrue(isinstance(sys.stdout, StreamWrapper),
+            'bad stdout wrapper')
+        self.assertTrue(isinstance(sys.stderr, StreamWrapper),
+            'bad stderr wrapper')
+
+    def assertNotWrapped(self):
+        self.assertIs(sys.stdout, orig_stdout, 'stdout should not be wrapped')
+        self.assertIs(sys.stderr, orig_stderr, 'stderr should not be wrapped')
+
+    @patch('colorama.initialise.reset_all')
+    def testInitWrapsOnWindows(self, _):
+        with platform('windows'):
+            init()
+            self.assertWrapped()
+
+    def testInitDoesntWrapOnNonWindows(self):
+        with platform('darwin'):
+            init()
+            self.assertNotWrapped()
+
+    def testInitAutoresetOnWrapsOnAllPlatforms(self):
+        with platform('darwin'):
+            init(autoreset=True)
+            self.assertWrapped()
+
+    def testInitWrapOffDoesntWrapOnWindows(self):
+        with platform('windows'):
+            init(wrap=False)
+            self.assertNotWrapped()
+
+    def testInitWrapOffWillUnwrapIfRequired(self):
+        with platform('windows'):
+            init()
+            init(wrap=False)
+            self.assertNotWrapped()
+
+    def testInitWrapOffIncompatibleWithAutoresetOn(self):
+        self.assertRaises(ValueError, lambda: init(autoreset=True, wrap=False))
+
+    @patch('colorama.ansitowin32.winterm', None)
+    def testInitOnlyWrapsOnce(self):
+        with platform('windows'):
+            init()
+            init()
+            self.assertWrapped()
+
+    @patch('colorama.win32.SetConsoleTextAttribute')
+    @patch('colorama.initialise.AnsiToWin32')
+    def testAutoResetPassedOn(self, mockATW32, _):
+        with platform('windows'):
+            init(autoreset=True)
+            self.assertEqual(len(mockATW32.call_args_list), 2)
+            self.assertEqual(mockATW32.call_args_list[1][1]['autoreset'], True)
+            self.assertEqual(mockATW32.call_args_list[0][1]['autoreset'], True)
+
+    @patch('colorama.initialise.AnsiToWin32')
+    def testAutoResetChangeable(self, mockATW32):
+        with platform('windows'):
+            init()
+
+            init(autoreset=True)
+            self.assertEqual(len(mockATW32.call_args_list), 4)
+            self.assertEqual(mockATW32.call_args_list[2][1]['autoreset'], True)
+            self.assertEqual(mockATW32.call_args_list[3][1]['autoreset'], True)
+
+            init()
+            self.assertEqual(len(mockATW32.call_args_list), 6)
+            self.assertEqual(
+                mockATW32.call_args_list[4][1]['autoreset'], False)
+            self.assertEqual(
+                mockATW32.call_args_list[5][1]['autoreset'], False)
+
+
+    @patch('colorama.initialise.atexit.register')
+    def testAtexitRegisteredOnlyOnce(self, mockRegister):
+        init()
+        self.assertTrue(mockRegister.called)
+        mockRegister.reset_mock()
+        init()
+        self.assertFalse(mockRegister.called)
+
+
+if __name__ == '__main__':
+    main()
+
