diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/tests_innerHTML_1.dat b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/tests_innerHTML_1.dat
new file mode 100644
index 0000000000000000000000000000000000000000..6c78661e0cb83ebf19ff8db1fe5c7695d3489df6
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/tests_innerHTML_1.dat
@@ -0,0 +1,741 @@
+#data
+<body><span>
+#errors
+#document-fragment
+body
+#document
+| <span>
+
+#data
+<span><body>
+#errors
+#document-fragment
+body
+#document
+| <span>
+
+#data
+<span><body>
+#errors
+#document-fragment
+div
+#document
+| <span>
+
+#data
+<body><span>
+#errors
+#document-fragment
+html
+#document
+| <head>
+| <body>
+|   <span>
+
+#data
+<frameset><span>
+#errors
+#document-fragment
+body
+#document
+| <span>
+
+#data
+<span><frameset>
+#errors
+#document-fragment
+body
+#document
+| <span>
+
+#data
+<span><frameset>
+#errors
+#document-fragment
+div
+#document
+| <span>
+
+#data
+<frameset><span>
+#errors
+#document-fragment
+html
+#document
+| <head>
+| <frameset>
+
+#data
+<table><tr>
+#errors
+#document-fragment
+table
+#document
+| <tbody>
+|   <tr>
+
+#data
+</table><tr>
+#errors
+#document-fragment
+table
+#document
+| <tbody>
+|   <tr>
+
+#data
+<a>
+#errors
+#document-fragment
+table
+#document
+| <a>
+
+#data
+<a>
+#errors
+#document-fragment
+table
+#document
+| <a>
+
+#data
+<a><caption>a
+#errors
+#document-fragment
+table
+#document
+| <a>
+| <caption>
+|   "a"
+
+#data
+<a><colgroup><col>
+#errors
+#document-fragment
+table
+#document
+| <a>
+| <colgroup>
+|   <col>
+
+#data
+<a><tbody><tr>
+#errors
+#document-fragment
+table
+#document
+| <a>
+| <tbody>
+|   <tr>
+
+#data
+<a><tfoot><tr>
+#errors
+#document-fragment
+table
+#document
+| <a>
+| <tfoot>
+|   <tr>
+
+#data
+<a><thead><tr>
+#errors
+#document-fragment
+table
+#document
+| <a>
+| <thead>
+|   <tr>
+
+#data
+<a><tr>
+#errors
+#document-fragment
+table
+#document
+| <a>
+| <tbody>
+|   <tr>
+
+#data
+<a><th>
+#errors
+#document-fragment
+table
+#document
+| <a>
+| <tbody>
+|   <tr>
+|     <th>
+
+#data
+<a><td>
+#errors
+#document-fragment
+table
+#document
+| <a>
+| <tbody>
+|   <tr>
+|     <td>
+
+#data
+<table></table><tbody>
+#errors
+#document-fragment
+caption
+#document
+| <table>
+
+#data
+</table><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+
+#data
+<span></table>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+
+#data
+</caption><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+
+#data
+<span></caption><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><caption><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><col><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><colgroup><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><html><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><tbody><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><td><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><tfoot><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><thead><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><th><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span><tr><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+<span></table><span>
+#errors
+#document-fragment
+caption
+#document
+| <span>
+|   <span>
+
+#data
+</colgroup><col>
+#errors
+#document-fragment
+colgroup
+#document
+| <col>
+
+#data
+<a><col>
+#errors
+#document-fragment
+colgroup
+#document
+| <col>
+
+#data
+<caption><a>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+
+#data
+<col><a>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+
+#data
+<colgroup><a>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+
+#data
+<tbody><a>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+
+#data
+<tfoot><a>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+
+#data
+<thead><a>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+
+#data
+</table><a>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+
+#data
+<a><tr>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+| <tr>
+
+#data
+<a><td>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+| <tr>
+|   <td>
+
+#data
+<a><td>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+| <tr>
+|   <td>
+
+#data
+<a><td>
+#errors
+#document-fragment
+tbody
+#document
+| <a>
+| <tr>
+|   <td>
+
+#data
+<td><table><tbody><a><tr>
+#errors
+#document-fragment
+tbody
+#document
+| <tr>
+|   <td>
+|     <a>
+|     <table>
+|       <tbody>
+|         <tr>
+
+#data
+</tr><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+<td><table><a><tr></tr><tr>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+|   <a>
+|   <table>
+|     <tbody>
+|       <tr>
+|       <tr>
+
+#data
+<caption><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+<col><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+<colgroup><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+<tbody><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+<tfoot><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+<thead><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+<tr><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+</table><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+
+#data
+<td><table></table><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+|   <table>
+| <td>
+
+#data
+<td><table></table><td>
+#errors
+#document-fragment
+tr
+#document
+| <td>
+|   <table>
+| <td>
+
+#data
+<caption><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+<col><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+<colgroup><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+<tbody><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+<tfoot><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+<th><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+<thead><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+<tr><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+</table><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+</tbody><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+</td><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+</tfoot><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+</thead><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+</th><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+</tr><a>
+#errors
+#document-fragment
+td
+#document
+| <a>
+
+#data
+<table><td><td>
+#errors
+#document-fragment
+td
+#document
+| <table>
+|   <tbody>
+|     <tr>
+|       <td>
+|       <td>
+
+#data
+</select><option>
+#errors
+#document-fragment
+select
+#document
+| <option>
+
+#data
+<input><option>
+#errors
+#document-fragment
+select
+#document
+| <option>
+
+#data
+<keygen><option>
+#errors
+#document-fragment
+select
+#document
+| <option>
+
+#data
+<textarea><option>
+#errors
+#document-fragment
+select
+#document
+| <option>
+
+#data
+</html><!--abc-->
+#errors
+#document-fragment
+html
+#document
+| <head>
+| <body>
+| <!-- abc -->
+
+#data
+</frameset><frame>
+#errors
+#document-fragment
+frameset
+#document
+| <frame>
+
+#data
+#errors
+#document-fragment
+html
+#document
+| <head>
+| <body>
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/tricky01.dat b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/tricky01.dat
new file mode 100644
index 0000000000000000000000000000000000000000..0841992448623b2730c58be384ba1112f15f0bf5
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/tricky01.dat
@@ -0,0 +1,261 @@
+#data
+<b><p>Bold </b> Not bold</p>
+Also not bold.
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <b>
+|     <p>
+|       <b>
+|         "Bold "
+|       " Not bold"
+|     "
+Also not bold."
+
+#data
+<html>
+<font color=red><i>Italic and Red<p>Italic and Red </font> Just italic.</p> Italic only.</i> Plain
+<p>I should not be red. <font color=red>Red. <i>Italic and red.</p>
+<p>Italic and red. </i> Red.</font> I should not be red.</p>
+<b>Bold <i>Bold and italic</b> Only Italic </i> Plain
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <font>
+|       color="red"
+|       <i>
+|         "Italic and Red"
+|     <i>
+|       <p>
+|         <font>
+|           color="red"
+|           "Italic and Red "
+|         " Just italic."
+|       " Italic only."
+|     " Plain
+"
+|     <p>
+|       "I should not be red. "
+|       <font>
+|         color="red"
+|         "Red. "
+|         <i>
+|           "Italic and red."
+|     <font>
+|       color="red"
+|       <i>
+|         "
+"
+|     <p>
+|       <font>
+|         color="red"
+|         <i>
+|           "Italic and red. "
+|         " Red."
+|       " I should not be red."
+|     "
+"
+|     <b>
+|       "Bold "
+|       <i>
+|         "Bold and italic"
+|     <i>
+|       " Only Italic "
+|     " Plain"
+
+#data
+<html><body>
+<p><font size="7">First paragraph.</p>
+<p>Second paragraph.</p></font>
+<b><p><i>Bold and Italic</b> Italic</p>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "
+"
+|     <p>
+|       <font>
+|         size="7"
+|         "First paragraph."
+|     <font>
+|       size="7"
+|       "
+"
+|       <p>
+|         "Second paragraph."
+|     "
+"
+|     <b>
+|     <p>
+|       <b>
+|         <i>
+|           "Bold and Italic"
+|       <i>
+|         " Italic"
+
+#data
+<html>
+<dl>
+<dt><b>Boo
+<dd>Goo?
+</dl>
+</html>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <dl>
+|       "
+"
+|       <dt>
+|         <b>
+|           "Boo
+"
+|       <dd>
+|         <b>
+|           "Goo?
+"
+|     <b>
+|       "
+"
+
+#data
+<html><body>
+<label><a><div>Hello<div>World</div></a></label>  
+</body></html>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "
+"
+|     <label>
+|       <a>
+|       <div>
+|         <a>
+|           "Hello"
+|           <div>
+|             "World"
+|         "  
+"
+
+#data
+<table><center> <font>a</center> <img> <tr><td> </td> </tr> </table>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <center>
+|       " "
+|       <font>
+|         "a"
+|     <font>
+|       <img>
+|       " "
+|     <table>
+|       " "
+|       <tbody>
+|         <tr>
+|           <td>
+|             " "
+|           " "
+|         " "
+
+#data
+<table><tr><p><a><p>You should see this text.
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <p>
+|       <a>
+|     <p>
+|       <a>
+|         "You should see this text."
+|     <table>
+|       <tbody>
+|         <tr>
+
+#data
+<TABLE>
+<TR>
+<CENTER><CENTER><TD></TD></TR><TR>
+<FONT>
+<TABLE><tr></tr></TABLE>
+</P>
+<a></font><font></a>
+This page contains an insanely badly-nested tag sequence.
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <center>
+|       <center>
+|     <font>
+|       "
+"
+|     <table>
+|       "
+"
+|       <tbody>
+|         <tr>
+|           "
+"
+|           <td>
+|         <tr>
+|           "
+"
+|     <table>
+|       <tbody>
+|         <tr>
+|     <font>
+|       "
+"
+|       <p>
+|       "
+"
+|       <a>
+|     <a>
+|       <font>
+|     <font>
+|       "
+This page contains an insanely badly-nested tag sequence."
+
+#data
+<html>
+<body>
+<b><nobr><div>This text is in a div inside a nobr</nobr>More text that should not be in the nobr, i.e., the
+nobr should have closed the div inside it implicitly. </b><pre>A pre tag outside everything else.</pre>
+</body>
+</html>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "
+"
+|     <b>
+|       <nobr>
+|     <div>
+|       <b>
+|         <nobr>
+|           "This text is in a div inside a nobr"
+|         "More text that should not be in the nobr, i.e., the
+nobr should have closed the div inside it implicitly. "
+|       <pre>
+|         "A pre tag outside everything else."
+|       "
+
+"
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/webkit01.dat b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/webkit01.dat
new file mode 100644
index 0000000000000000000000000000000000000000..06bc436b6164ca4061cfa2bad22576c32f3c796f
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/webkit01.dat
@@ -0,0 +1,594 @@
+#data
+Test
+#errors
+Line: 1 Col: 4 Unexpected non-space characters. Expected DOCTYPE.
+#document
+| <html>
+|   <head>
+|   <body>
+|     "Test"
+
+#data
+<div></div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <div>
+
+#data
+<div>Test</div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <div>
+|       "Test"
+
+#data
+<di
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+
+#data
+<div>Hello</div>
+<script>
+console.log("PASS");
+</script>
+<div>Bye</div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <div>
+|       "Hello"
+|     "
+"
+|     <script>
+|       "
+console.log("PASS");
+"
+|     "
+"
+|     <div>
+|       "Bye"
+
+#data
+<div foo="bar">Hello</div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <div>
+|       foo="bar"
+|       "Hello"
+
+#data
+<div>Hello</div>
+<script>
+console.log("FOO<span>BAR</span>BAZ");
+</script>
+<div>Bye</div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <div>
+|       "Hello"
+|     "
+"
+|     <script>
+|       "
+console.log("FOO<span>BAR</span>BAZ");
+"
+|     "
+"
+|     <div>
+|       "Bye"
+
+#data
+<foo bar="baz"></foo><potato quack="duck"></potato>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <foo>
+|       bar="baz"
+|     <potato>
+|       quack="duck"
+
+#data
+<foo bar="baz"><potato quack="duck"></potato></foo>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <foo>
+|       bar="baz"
+|       <potato>
+|         quack="duck"
+
+#data
+<foo></foo bar="baz"><potato></potato quack="duck">
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <foo>
+|     <potato>
+
+#data
+</ tttt>
+#errors
+#document
+| <!--  tttt -->
+| <html>
+|   <head>
+|   <body>
+
+#data
+<div FOO ><img><img></div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <div>
+|       foo=""
+|       <img>
+|       <img>
+
+#data
+<p>Test</p<p>Test2</p>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <p>
+|       "TestTest2"
+
+#data
+<rdar://problem/6869687>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <rdar:>
+|       6869687=""
+|       problem=""
+
+#data
+<A>test< /A>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <a>
+|       "test< /A>"
+
+#data
+&lt;
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "<"
+
+#data
+<body foo='bar'><body foo='baz' yo='mama'>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     foo="bar"
+|     yo="mama"
+
+#data
+<body></br foo="bar"></body>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <br>
+
+#data
+<bdy><br foo="bar"></body>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <bdy>
+|       <br>
+|         foo="bar"
+
+#data
+<body></body></br foo="bar">
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <br>
+
+#data
+<bdy></body><br foo="bar">
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <bdy>
+|       <br>
+|         foo="bar"
+
+#data
+<html><body></body></html><!-- Hi there -->
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+| <!--  Hi there  -->
+
+#data
+<html><body></body></html>x<!-- Hi there -->
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "x"
+|     <!--  Hi there  -->
+
+#data
+<html><body></body></html>x<!-- Hi there --></html><!-- Again -->
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "x"
+|     <!--  Hi there  -->
+| <!--  Again  -->
+
+#data
+<html><body></body></html>x<!-- Hi there --></body></html><!-- Again -->
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "x"
+|     <!--  Hi there  -->
+| <!--  Again  -->
+
+#data
+<html><body><ruby><div><rp>xx</rp></div></ruby></body></html>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <ruby>
+|       <div>
+|         <rp>
+|           "xx"
+
+#data
+<html><body><ruby><div><rt>xx</rt></div></ruby></body></html>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <ruby>
+|       <div>
+|         <rt>
+|           "xx"
+
+#data
+<html><frameset><!--1--><noframes>A</noframes><!--2--></frameset><!--3--><noframes>B</noframes><!--4--></html><!--5--><noframes>C</noframes><!--6-->
+#errors
+#document
+| <html>
+|   <head>
+|   <frameset>
+|     <!-- 1 -->
+|     <noframes>
+|       "A"
+|     <!-- 2 -->
+|   <!-- 3 -->
+|   <noframes>
+|     "B"
+|   <!-- 4 -->
+|   <noframes>
+|     "C"
+| <!-- 5 -->
+| <!-- 6 -->
+
+#data
+<select><option>A<select><option>B<select><option>C<select><option>D<select><option>E<select><option>F<select><option>G<select>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <select>
+|       <option>
+|         "A"
+|     <option>
+|       "B"
+|       <select>
+|         <option>
+|           "C"
+|     <option>
+|       "D"
+|       <select>
+|         <option>
+|           "E"
+|     <option>
+|       "F"
+|       <select>
+|         <option>
+|           "G"
+
+#data
+<dd><dd><dt><dt><dd><li><li>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <dd>
+|     <dd>
+|     <dt>
+|     <dt>
+|     <dd>
+|       <li>
+|       <li>
+
+#data
+<div><b></div><div><nobr>a<nobr>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <div>
+|       <b>
+|     <div>
+|       <b>
+|         <nobr>
+|           "a"
+|         <nobr>
+
+#data
+<head></head>
+<body></body>
+#errors
+#document
+| <html>
+|   <head>
+|   "
+"
+|   <body>
+
+#data
+<head></head> <style></style>ddd
+#errors
+#document
+| <html>
+|   <head>
+|     <style>
+|   " "
+|   <body>
+|     "ddd"
+
+#data
+<kbd><table></kbd><col><select><tr>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <kbd>
+|       <select>
+|       <table>
+|         <colgroup>
+|           <col>
+|         <tbody>
+|           <tr>
+
+#data
+<kbd><table></kbd><col><select><tr></table><div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <kbd>
+|       <select>
+|       <table>
+|         <colgroup>
+|           <col>
+|         <tbody>
+|           <tr>
+|       <div>
+
+#data
+<a><li><style></style><title></title></a>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <a>
+|     <li>
+|       <a>
+|         <style>
+|         <title>
+
+#data
+<font></p><p><meta><title></title></font>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <font>
+|       <p>
+|     <p>
+|       <font>
+|         <meta>
+|         <title>
+
+#data
+<a><center><title></title><a>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <a>
+|     <center>
+|       <a>
+|         <title>
+|       <a>
+
+#data
+<svg><title><div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <svg svg>
+|       <svg title>
+|         <div>
+
+#data
+<svg><title><rect><div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <svg svg>
+|       <svg title>
+|         <rect>
+|           <div>
+
+#data
+<svg><title><svg><div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <svg svg>
+|       <svg title>
+|         <svg svg>
+|         <div>
+
+#data
+<img <="" FAIL>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <img>
+|       <=""
+|       fail=""
+
+#data
+<ul><li><div id='foo'/>A</li><li>B<div>C</div></li></ul>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <ul>
+|       <li>
+|         <div>
+|           id="foo"
+|           "A"
+|       <li>
+|         "B"
+|         <div>
+|           "C"
+
+#data
+<svg><em><desc></em>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <svg svg>
+|     <em>
+|       <desc>
+
+#data
+<svg><tfoot></mi><td>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <svg svg>
+|       <svg tfoot>
+|         <svg td>
+
+#data
+<math><mrow><mrow><mn>1</mn></mrow><mi>a</mi></mrow></math>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <math math>
+|       <math mrow>
+|         <math mrow>
+|           <math mn>
+|             "1"
+|         <math mi>
+|           "a"
+
+#data
+<!doctype html><input type="hidden"><frameset>
+#errors
+#document
+| <!DOCTYPE html>
+| <html>
+|   <head>
+|   <frameset>
+
+#data
+<!doctype html><input type="button"><frameset>
+#errors
+#document
+| <!DOCTYPE html>
+| <html>
+|   <head>
+|   <body>
+|     <input>
+|       type="button"
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/webkit02.dat b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/webkit02.dat
new file mode 100644
index 0000000000000000000000000000000000000000..468879b3c5348ff11ef0808be066e8fce380d9c7
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction/webkit02.dat
@@ -0,0 +1,94 @@
+#data
+<foo bar=qux/>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <foo>
+|       bar="qux/"
+
+#data
+<p id="status"><noscript><strong>A</strong></noscript><span>B</span></p>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <p>
+|       id="status"
+|       <noscript>
+|         "<strong>A</strong>"
+|       <span>
+|         "B"
+
+#data
+<div><sarcasm><div></div></sarcasm></div>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <div>
+|       <sarcasm>
+|         <div>
+
+#data
+<html><body><img src="" border="0" alt="><div>A</div></body></html>
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+
+#data
+<table><td></tbody>A
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "A"
+|     <table>
+|       <tbody>
+|         <tr>
+|           <td>
+
+#data
+<table><td></thead>A
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <table>
+|       <tbody>
+|         <tr>
+|           <td>
+|             "A"
+
+#data
+<table><td></tfoot>A
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <table>
+|       <tbody>
+|         <tr>
+|           <td>
+|             "A"
+
+#data
+<table><thead><td></tbody>A
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <table>
+|       <thead>
+|         <tr>
+|           <td>
+|             "A"
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_jsdom/document_write.dat b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_jsdom/document_write.dat
new file mode 100644
index 0000000000000000000000000000000000000000..d1ca2fc014335d1d8c9e6e69fb0ec125b82e570d
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_jsdom/document_write.dat
@@ -0,0 +1,39 @@
+#data
+1<script>document.write("2")</script>3
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "1"
+|     <script>
+|       "document.write("2")"
+|     "23"
+
+#data
+1<script>document.write("<script>document.write('2')</scr"+ "ipt><script>document.write('3')</scr" + "ipt>")</script>4
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "1"
+|     <script>
+|       "document.write("<script>document.write('2')</scr"+ "ipt><script>document.write('3')</scr" + "ipt>")"
+|     <script>
+|       "document.write('2')"
+|     "2"
+|     <script>
+|       "document.write('3')"
+|     "34"
+
+#data
+<script>document.write('<!--');</script>I should be commented-->
+#errors
+#document
+| <html>
+|   <head>
+|     <script>
+|       "document.write('<!--');"
+|     <!-- I should be commented -->
+|   <body>
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_options/decode_html_entities.dat b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_options/decode_html_entities.dat
new file mode 100644
index 0000000000000000000000000000000000000000..4c5e95b8126b77dcf044f4eda545e7b8d51e18cb
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_options/decode_html_entities.dat
@@ -0,0 +1,22 @@
+#data
+<html><head></head><body>&nbsp;</body></html>
+#disable-html-entities-decoding
+true
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     "&nbsp;"
+
+#data
+<html><head></head><body><a href="&amp;test&copy;"></a></body></html>
+#disable-html-entities-decoding
+true
+#errors
+#document
+| <html>
+|   <head>
+|   <body>
+|     <a>
+|       href="&amp;test&copy;"
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_regression/gh40_form_in_template.dat b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_regression/gh40_form_in_template.dat
new file mode 100644
index 0000000000000000000000000000000000000000..b73aa19ab2c1365474bd3bb97a57366b5c5f0ba0
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/data/tree_construction_regression/gh40_form_in_template.dat
@@ -0,0 +1,13 @@
+#data
+<template><form><input name="q"></form><div>second</div></template>
+#document-fragment
+template
+#errors
+#document
+| <template>
+|   content
+|     <form>
+|       <input>
+|         name="q"
+|     <div>
+|       "second"
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/formatting_element_list_test.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/formatting_element_list_test.js
new file mode 100644
index 0000000000000000000000000000000000000000..87da7b1f4439fd4d3d600bd122818ea8c35b97e8
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/formatting_element_list_test.js
@@ -0,0 +1,193 @@
+var assert = require('assert'),
+    html = require('../../lib/common/html'),
+    FormattingElementList = require('../../lib/tree_construction/formatting_element_list'),
+    TestUtils = require('../test_utils');
+
+//Aliases
+var $ = html.TAG_NAMES,
+    NS = html.NAMESPACES;
+
+TestUtils.generateTestsForEachTreeAdapter(module.exports, function (_test, treeAdapter) {
+    _test['Insert marker'] = function () {
+        var list = new FormattingElementList(treeAdapter);
+
+        list.insertMarker();
+        assert.strictEqual(list.length, 1);
+        assert.strictEqual(list.entries[0].type, FormattingElementList.MARKER_ENTRY);
+
+        list.insertMarker();
+        assert.strictEqual(list.length, 2);
+        assert.strictEqual(list.entries[1].type, FormattingElementList.MARKER_ENTRY);
+    };
+
+    _test['Push element'] = function () {
+        var list = new FormattingElementList(treeAdapter),
+            element1Token = 'token1',
+            element2Token = 'token2',
+            element1 = treeAdapter.createElement($.DIV, NS.HTML, []),
+            element2 = treeAdapter.createElement($.P, NS.HTML, []);
+
+        list.pushElement(element1, element1Token);
+        assert.strictEqual(list.length, 1);
+        assert.strictEqual(list.entries[0].type, FormattingElementList.ELEMENT_ENTRY);
+        assert.strictEqual(list.entries[0].element, element1);
+        assert.strictEqual(list.entries[0].token, element1Token);
+
+        list.pushElement(element2, element2Token);
+        assert.strictEqual(list.length, 2);
+        assert.strictEqual(list.entries[1].type, FormattingElementList.ELEMENT_ENTRY);
+        assert.strictEqual(list.entries[1].element, element2);
+        assert.strictEqual(list.entries[1].token, element2Token);
+    };
+
+    _test['Insert element after bookmark'] = function () {
+        var list = new FormattingElementList(treeAdapter),
+            token = 'token1',
+            element1 = treeAdapter.createElement($.DIV, NS.HTML, []),
+            element2 = treeAdapter.createElement($.P, NS.HTML, []),
+            element3 = treeAdapter.createElement($.SPAN, NS.HTML, []),
+            element4 = treeAdapter.createElement($.TITLE, NS.HTML, []);
+
+        list.pushElement(element1, token);
+        list.bookmark = list.entries[0];
+
+        list.pushElement(element2, token);
+        list.pushElement(element3, token);
+
+        list.insertElementAfterBookmark(element4, token);
+
+        assert.strictEqual(list.length, 4);
+        assert.strictEqual(list.entries[1].element, element4);
+    };
+
+    _test['Push element - Noah Ark condition'] = function () {
+        var list = new FormattingElementList(treeAdapter),
+            token1 = 'token1',
+            token2 = 'token2',
+            token3 = 'token3',
+            token4 = 'token4',
+            token5 = 'token5',
+            token6 = 'token6',
+            element1 = treeAdapter.createElement($.DIV, NS.HTML, [
+                {name: 'attr1', value: 'val1'},
+                {name: 'attr2', value: 'val2'}
+            ]),
+            element2 = treeAdapter.createElement($.DIV, NS.HTML, [
+                {name: 'attr1', value: 'val1'},
+                {name: 'attr2', value: 'someOtherValue'}
+            ]);
+
+        list.pushElement(element1, token1);
+        list.pushElement(element1, token2);
+        list.pushElement(element2, token3);
+        list.pushElement(element1, token4);
+
+        assert.strictEqual(list.length, 4);
+        assert.strictEqual(list.entries[0].token, token1);
+        assert.strictEqual(list.entries[1].token, token2);
+        assert.strictEqual(list.entries[2].token, token3);
+        assert.strictEqual(list.entries[3].token, token4);
+
+        list.pushElement(element1, token5);
+
+        assert.strictEqual(list.length, 4);
+        assert.strictEqual(list.entries[0].token, token2);
+        assert.strictEqual(list.entries[1].token, token3);
+        assert.strictEqual(list.entries[2].token, token4);
+        assert.strictEqual(list.entries[3].token, token5);
+
+        list.insertMarker();
+        list.pushElement(element1, token6);
+
+        assert.strictEqual(list.length, 6);
+        assert.strictEqual(list.entries[0].token, token2);
+        assert.strictEqual(list.entries[1].token, token3);
+        assert.strictEqual(list.entries[2].token, token4);
+        assert.strictEqual(list.entries[3].token, token5);
+        assert.strictEqual(list.entries[4].type, FormattingElementList.MARKER_ENTRY);
+        assert.strictEqual(list.entries[5].token, token6);
+    };
+
+    _test['Clear to the last marker'] = function () {
+        var list = new FormattingElementList(treeAdapter),
+            token = 'token',
+            element1 = treeAdapter.createElement($.DIV, NS.HTML, [
+                {name: 'attr1', value: 'val1'},
+                {name: 'attr2', value: 'val2'}
+            ]),
+            element2 = treeAdapter.createElement($.DIV, NS.HTML, [
+                {name: 'attr1', value: 'val1'},
+                {name: 'attr2', value: 'someOtherValue'}
+            ]);
+
+        list.pushElement(element1, token);
+        list.pushElement(element2, token);
+        list.insertMarker();
+        list.pushElement(element1, token);
+        list.pushElement(element1, token);
+        list.pushElement(element2, token);
+
+        list.clearToLastMarker();
+
+        assert.strictEqual(list.length, 2);
+    };
+
+    _test['Remove entry'] = function () {
+        var list = new FormattingElementList(treeAdapter),
+            token = 'token',
+            element1 = treeAdapter.createElement($.DIV, NS.HTML, [
+                {name: 'attr1', value: 'val1'},
+                {name: 'attr2', value: 'val2'}
+            ]),
+            element2 = treeAdapter.createElement($.DIV, NS.HTML, [
+                {name: 'attr1', value: 'val1'},
+                {name: 'attr2', value: 'someOtherValue'}
+            ]);
+
+        list.pushElement(element1, token);
+        list.pushElement(element2, token);
+        list.pushElement(element2, token);
+
+        list.removeEntry(list.entries[0]);
+
+        assert.strictEqual(list.length, 2);
+
+        for (var i = list.length - 1; i >= 0; i--)
+            assert.notStrictEqual(list.entries[i].element, element1);
+    };
+
+    _test['Get entry in scope with given tag name'] = function () {
+        var list = new FormattingElementList(treeAdapter),
+            token = 'token',
+            element = treeAdapter.createElement($.DIV, NS.HTML, []);
+
+        assert.ok(!list.getElementEntryInScopeWithTagName($.DIV));
+
+        list.pushElement(element, token);
+        list.pushElement(element, token);
+        assert.strictEqual(list.getElementEntryInScopeWithTagName($.DIV), list.entries[1]);
+
+        list.insertMarker();
+        assert.ok(!list.getElementEntryInScopeWithTagName($.DIV));
+
+        list.pushElement(element, token);
+        assert.strictEqual(list.getElementEntryInScopeWithTagName($.DIV), list.entries[3]);
+    };
+
+    _test['Get element entry'] = function () {
+        var list = new FormattingElementList(treeAdapter),
+            token = 'token',
+            element1 = treeAdapter.createElement($.DIV, NS.HTML, []),
+            element2 = treeAdapter.createElement($.A, NS.HTML, []);
+
+        list.pushElement(element2, token);
+        list.pushElement(element1, token);
+        list.pushElement(element2, token);
+        list.insertMarker();
+
+        var entry = list.getElementEntry(element1);
+        assert.strictEqual(entry.type, FormattingElementList.ELEMENT_ENTRY);
+        assert.strictEqual(entry.token, token);
+        assert.strictEqual(entry.element, element1);
+    };
+});
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/jsdom_parser_test.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/jsdom_parser_test.js
new file mode 100644
index 0000000000000000000000000000000000000000..bed6585cef518bdb02b2a2ae84a7d78c32c1a612
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/jsdom_parser_test.js
@@ -0,0 +1,148 @@
+var assert = require('assert'),
+    path = require('path'),
+    HTML = require('../../lib/common/html'),
+    JsDomParser = require('../../index').JsDomParser,
+    Serializer = require('../../index').TreeSerializer,
+    TestUtils = require('../test_utils');
+
+exports['State guard'] = function () {
+    var docHtml = '<script>Yoyo</script>',
+        getParsingUnit = function () {
+            var parsing = JsDomParser.parseDocument(docHtml);
+
+            parsing.done(function noop() {
+                //NOTE: do nothing =)
+            });
+
+            return parsing;
+        };
+
+    assert.throws(function () {
+        var parsing = getParsingUnit();
+
+        parsing.suspend();
+        parsing.suspend();
+    });
+
+    assert.throws(function () {
+        var parsing = getParsingUnit();
+
+        parsing.resume();
+    });
+};
+
+exports['Reentrancy'] = function (done) {
+    var serializer = new Serializer(),
+        asyncAssertionCount = 0,
+        docHtml1 = '<!DOCTYPE html><html><head><script>Yoyo</script></head><body></body></html>',
+        docHtml2 = '<!DOCTYPE html><html><head></head><body>Beep boop</body></html>',
+        fragments = [
+            '<div>Hey ya!</div>',
+            '<p><a href="#"></a></p>',
+            '<template><td>yo</td></template>',
+            '<select><template><option></option></template></select>'
+        ];
+
+    var parsing1 = JsDomParser.parseDocument(docHtml1)
+
+        .done(function (document1) {
+            var actual = serializer.serialize(document1);
+
+            assert.strictEqual(asyncAssertionCount, 5);
+            assert.ok(actual === docHtml1, TestUtils.getStringDiffMsg(actual, docHtml1));
+            done();
+        })
+
+        .handleScripts(function () {
+            parsing1.suspend();
+
+            setTimeout(function () {
+                fragments.forEach(function (fragment) {
+                    var actual = serializer.serialize(JsDomParser.parseInnerHtml(fragment));
+                    assert.ok(actual === fragment, TestUtils.getStringDiffMsg(actual, fragment));
+                    asyncAssertionCount++;
+                });
+
+                var parsing2 = JsDomParser.parseDocument(docHtml2);
+
+                parsing2.done(function (document2) {
+                    var actual = serializer.serialize(document2);
+
+                    assert.ok(actual === docHtml2, TestUtils.getStringDiffMsg(actual, docHtml2));
+                    asyncAssertionCount++;
+                    parsing1.resume();
+                });
+            });
+        });
+};
+
+
+TestUtils.generateTestsForEachTreeAdapter(module.exports, function (_test, treeAdapter) {
+    function getFullTestName(test) {
+        return ['JsDomParser - ', test.idx, '.', test.setName, ' - ', test.input].join('');
+    }
+
+    function parseDocument(input, callback) {
+        var parsing = JsDomParser.parseDocument(input, treeAdapter)
+
+            .done(callback)
+
+            .handleScripts(function (document, scriptElement) {
+                parsing.suspend();
+
+                //NOTE: test parser suspension in different modes (sync and async).
+                //If we have script then execute it and resume parser asynchronously.
+                //Otherwise - resume synchronously.
+                var scriptTextNode = treeAdapter.getChildNodes(scriptElement)[0],
+                    script = scriptTextNode && treeAdapter.getTextNodeContent(scriptTextNode);
+
+                //NOTE: don't pollute test runner output by console.log() calls from test scripts
+                if (script && script.trim() && script.indexOf('console.log') === -1) {
+                    setTimeout(function () {
+                        //NOTE: mock document for script evaluation
+                        document.write = function (html) {
+                            parsing.documentWrite(html);
+                        };
+
+                        try {
+                            eval(script);
+                        } catch (err) {
+                            //NOTE: ignore broken scripts from test data
+                        }
+
+                        parsing.resume();
+                    });
+                }
+
+                else
+                    parsing.resume();
+            });
+    }
+
+    //Here we go..
+    TestUtils.loadTreeConstructionTestData([
+        path.join(__dirname, '../data/tree_construction'),
+        path.join(__dirname, '../data/tree_construction_jsdom')
+    ], treeAdapter).forEach(function (test) {
+        _test[getFullTestName(test)] = function (done) {
+            function assertResult(result) {
+                var actual = TestUtils.serializeToTestDataFormat(result, treeAdapter),
+                    msg = TestUtils.prettyPrintParserAssertionArgs(actual, test.expected);
+
+                assert.strictEqual(actual, test.expected, msg);
+
+                done();
+            }
+
+            if (test.fragmentContext) {
+                var result = JsDomParser.parseInnerHtml(test.input, test.fragmentContext, treeAdapter);
+                assertResult(result);
+            }
+
+            else
+                parseDocument(test.input, assertResult);
+        };
+    });
+});
+
+
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/open_element_stack_test.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/open_element_stack_test.js
new file mode 100644
index 0000000000000000000000000000000000000000..5298579eb17ee0667f36e24a3049dc31ff7fc266
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/open_element_stack_test.js
@@ -0,0 +1,471 @@
+var assert = require('assert'),
+    HTML = require('../../lib/common/html'),
+    OpenElementStack = require('../../lib/tree_construction/open_element_stack'),
+    TestUtils = require('../test_utils');
+
+//Aliases
+var $ = HTML.TAG_NAMES,
+    NS = HTML.NAMESPACES;
+
+TestUtils.generateTestsForEachTreeAdapter(module.exports, function (_test, treeAdapter) {
+    _test['Push element'] = function () {
+        var document = treeAdapter.createDocument(),
+            element1 = treeAdapter.createElement('#element1', 'namespace1', []),
+            element2 = treeAdapter.createElement('#element2', 'namespace2', []),
+            stack = new OpenElementStack(document, treeAdapter);
+
+        assert.strictEqual(stack.current, document);
+        assert.strictEqual(stack.stackTop, -1);
+
+        stack.push(element1);
+        assert.strictEqual(stack.current, element1);
+        assert.strictEqual(stack.currentTagName, treeAdapter.getTagName(element1));
+        assert.strictEqual(stack.stackTop, 0);
+
+        stack.push(element2);
+        assert.strictEqual(stack.current, element2);
+        assert.strictEqual(stack.currentTagName, treeAdapter.getTagName(element2));
+        assert.strictEqual(stack.stackTop, 1);
+    };
+
+    _test['Pop element'] = function () {
+        var element = treeAdapter.createElement('#element', 'namespace1', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(element);
+        stack.push('#element2');
+        stack.pop();
+        assert.strictEqual(stack.current, element);
+        assert.strictEqual(stack.currentTagName, treeAdapter.getTagName(element));
+        assert.strictEqual(stack.stackTop, 0);
+
+        stack.pop();
+        assert.ok(!stack.current);
+        assert.ok(!stack.currentTagName);
+        assert.strictEqual(stack.stackTop, -1);
+    };
+
+    _test['Replace element'] = function () {
+        var element = treeAdapter.createElement('#element', 'namespace', []),
+            newElement = treeAdapter.createElement('#newElement', 'newElementNamespace', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push('#element2');
+        stack.push(element);
+        stack.replace(element, newElement);
+        assert.strictEqual(stack.current, newElement);
+        assert.strictEqual(stack.currentTagName, treeAdapter.getTagName(newElement));
+        assert.strictEqual(stack.stackTop, 1);
+    };
+
+    _test['Insert element after element'] = function () {
+        var element1 = treeAdapter.createElement('#element1', 'namespace1', []),
+            element2 = treeAdapter.createElement('#element2', 'namespace2', []),
+            element3 = treeAdapter.createElement('#element3', 'namespace3', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(element1);
+        stack.push(element2);
+        stack.insertAfter(element1, element3);
+        assert.strictEqual(stack.stackTop, 2);
+        assert.strictEqual(stack.items[1], element3);
+
+        stack.insertAfter(element2, element1);
+        assert.strictEqual(stack.stackTop, 3);
+        assert.strictEqual(stack.current, element1);
+        assert.strictEqual(stack.currentTagName, treeAdapter.getTagName(element1));
+    };
+
+    _test['Pop elements until popped with given tagName'] = function () {
+        var element1 = treeAdapter.createElement('#element1', '', []),
+            element2 = treeAdapter.createElement('#element2', '', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(element2);
+        stack.push(element2);
+        stack.push(element2);
+        stack.push(element2);
+        stack.popUntilTagNamePopped(treeAdapter.getTagName(element1));
+        assert.ok(!stack.current);
+        assert.strictEqual(stack.stackTop, -1);
+
+        stack.push(element2);
+        stack.push(element1);
+        stack.push(element2);
+        stack.popUntilTagNamePopped(treeAdapter.getTagName(element1));
+        assert.strictEqual(stack.current, element2);
+        assert.strictEqual(stack.stackTop, 0);
+    };
+
+    _test['Pop elements until given element popped'] = function () {
+        var element1 = treeAdapter.createElement('#element1', '', []),
+            element2 = treeAdapter.createElement('#element2', '', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(element2);
+        stack.push(element2);
+        stack.push(element2);
+        stack.push(element2);
+        stack.popUntilElementPopped(element1);
+        assert.ok(!stack.current);
+        assert.strictEqual(stack.stackTop, -1);
+
+        stack.push(element2);
+        stack.push(element1);
+        stack.push(element2);
+        stack.popUntilElementPopped(element1);
+        assert.strictEqual(stack.current, element2);
+        assert.strictEqual(stack.stackTop, 0);
+    };
+
+    _test['Pop elements until numbered header popped'] = function () {
+        var element1 = treeAdapter.createElement($.H3, '', []),
+            element2 = treeAdapter.createElement('#element2', '', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(element2);
+        stack.push(element2);
+        stack.push(element2);
+        stack.push(element2);
+        stack.popUntilNumberedHeaderPopped();
+        assert.ok(!stack.current);
+        assert.strictEqual(stack.stackTop, -1);
+
+        stack.push(element2);
+        stack.push(element1);
+        stack.push(element2);
+        stack.popUntilNumberedHeaderPopped();
+        assert.strictEqual(stack.current, element2);
+        assert.strictEqual(stack.stackTop, 0);
+    };
+
+    _test['Pop all up to <html> element'] = function () {
+        var htmlElement = treeAdapter.createElement($.HTML, NS.HTML, []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(htmlElement);
+        stack.push('#element1');
+        stack.push('#element2');
+
+        stack.popAllUpToHtmlElement();
+        assert.strictEqual(stack.current, htmlElement);
+    };
+
+    _test['Clear back to a table context'] = function () {
+        var htmlElement = treeAdapter.createElement($.HTML, '', []),
+            tableElement = treeAdapter.createElement($.TABLE, '', []),
+            divElement = treeAdapter.createElement($.DIV, '', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(htmlElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.clearBackToTableContext();
+        assert.strictEqual(stack.current, htmlElement);
+        assert.strictEqual(stack.stackTop, 0);
+
+        stack.push(divElement);
+        stack.push(tableElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.clearBackToTableContext();
+        assert.strictEqual(stack.current, tableElement);
+        assert.strictEqual(stack.stackTop, 2);
+    };
+
+    _test['Clear back to a table body context'] = function () {
+        var htmlElement = treeAdapter.createElement($.HTML, '', []),
+            theadElement = treeAdapter.createElement($.THEAD, '', []),
+            divElement = treeAdapter.createElement($.DIV, '', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(htmlElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.clearBackToTableBodyContext();
+        assert.strictEqual(stack.current, htmlElement);
+        assert.strictEqual(stack.stackTop, 0);
+
+        stack.push(divElement);
+        stack.push(theadElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.clearBackToTableBodyContext();
+        assert.strictEqual(stack.current, theadElement);
+        assert.strictEqual(stack.stackTop, 2);
+    };
+
+    _test['Clear back to a table row context'] = function () {
+        var htmlElement = treeAdapter.createElement($.HTML, '', []),
+            trElement = treeAdapter.createElement($.TR, '', []),
+            divElement = treeAdapter.createElement($.DIV, '', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(htmlElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.clearBackToTableRowContext();
+        assert.strictEqual(stack.current, htmlElement);
+        assert.strictEqual(stack.stackTop, 0);
+
+        stack.push(divElement);
+        stack.push(trElement);
+        stack.push(divElement);
+        stack.push(divElement);
+        stack.clearBackToTableRowContext();
+        assert.strictEqual(stack.current, trElement);
+        assert.strictEqual(stack.stackTop, 2);
+    };
+
+    _test['Remove element'] = function () {
+        var element = treeAdapter.createElement('#element', '', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(element);
+        stack.push(treeAdapter.createElement('element1', '', []));
+        stack.push(treeAdapter.createElement('element2', '', []));
+
+        stack.remove(element);
+
+        assert.strictEqual(stack.stackTop, 1);
+
+        for (var i = stack.stackTop; i >= 0; i--)
+            assert.notStrictEqual(stack.items[i], element);
+    };
+
+    _test['Try peek properly nested <body> element'] = function () {
+        var bodyElement = treeAdapter.createElement($.BODY, '', []),
+            stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, '', []));
+        stack.push(bodyElement);
+        stack.push(treeAdapter.createElement($.DIV, '', []));
+        assert.strictEqual(stack.tryPeekProperlyNestedBodyElement(), bodyElement);
+
+        stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+        stack.push(treeAdapter.createElement($.HTML, '', []));
+        assert.ok(!stack.tryPeekProperlyNestedBodyElement());
+    };
+
+    _test['Is root <html> element current'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, '', []));
+        assert.ok(stack.isRootHtmlElementCurrent());
+
+        stack.push(treeAdapter.createElement($.DIV, '', []));
+        assert.ok(!stack.isRootHtmlElementCurrent());
+    };
+
+    _test['Get common ancestor'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter),
+            element = treeAdapter.createElement('#element', '', []),
+            ancestor = treeAdapter.createElement('#ancestor', '', []);
+
+        stack.push(treeAdapter.createElement('#someElement', '', []));
+        assert.ok(!stack.getCommonAncestor(element));
+
+        stack.pop();
+        assert.ok(!stack.getCommonAncestor(element));
+
+        stack.push(element);
+        assert.ok(!stack.getCommonAncestor(element));
+
+        stack.push(treeAdapter.createElement('#someElement', '', []));
+        stack.push(ancestor);
+        stack.push(element);
+        assert.strictEqual(stack.getCommonAncestor(element), ancestor);
+    };
+
+    _test['Contains element'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter),
+            element = treeAdapter.createElement('#element', '', []);
+
+        stack.push(treeAdapter.createElement('#someElement', '', []));
+        assert.ok(!stack.contains(element));
+
+        stack.push(element);
+        assert.ok(stack.contains(element));
+    };
+
+    _test['Has element in scope'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        assert.ok(!stack.hasInScope($.P));
+
+        stack.push(treeAdapter.createElement($.P, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.UL, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.BUTTON, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        assert.ok(stack.hasInScope($.P));
+
+        stack.push(treeAdapter.createElement($.TITLE, NS.SVG, []));
+        assert.ok(!stack.hasInScope($.P));
+    };
+
+    _test['Has numbered header in scope'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        assert.ok(!stack.hasNumberedHeaderInScope());
+
+        stack.push(treeAdapter.createElement($.P, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.UL, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.H3, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        assert.ok(stack.hasNumberedHeaderInScope());
+
+        stack.push(treeAdapter.createElement($.TITLE, NS.SVG, []));
+        assert.ok(!stack.hasNumberedHeaderInScope());
+
+        stack.push(treeAdapter.createElement($.H6, NS.HTML, []));
+        assert.ok(stack.hasNumberedHeaderInScope());
+    };
+
+    _test['Has element in list item scope'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        assert.ok(!stack.hasInListItemScope($.P));
+
+        stack.push(treeAdapter.createElement($.P, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.BUTTON, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        assert.ok(stack.hasInListItemScope($.P));
+
+        stack.push(treeAdapter.createElement($.UL, NS.HTML, []));
+        assert.ok(!stack.hasInListItemScope($.P));
+    };
+
+    _test['Has element in button scope'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        assert.ok(!stack.hasInButtonScope($.P));
+
+        stack.push(treeAdapter.createElement($.P, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.UL, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        assert.ok(stack.hasInButtonScope($.P));
+
+        stack.push(treeAdapter.createElement($.BUTTON, NS.HTML, []));
+        assert.ok(!stack.hasInButtonScope($.P));
+    };
+
+
+    _test['Has element in table scope'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        assert.ok(!stack.hasInTableScope($.P));
+
+        stack.push(treeAdapter.createElement($.P, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.UL, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.TD, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        assert.ok(stack.hasInTableScope($.P));
+
+        stack.push(treeAdapter.createElement($.TABLE, NS.HTML, []));
+        assert.ok(!stack.hasInTableScope($.P));
+    };
+
+    _test['Has table body context in table scope'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        assert.ok(!stack.hasTableBodyContextInTableScope());
+
+        stack.push(treeAdapter.createElement($.TABLE, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.UL, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.TBODY, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        assert.ok(stack.hasTableBodyContextInTableScope());
+
+        stack.push(treeAdapter.createElement($.TABLE, NS.HTML, []));
+        assert.ok(!stack.hasTableBodyContextInTableScope());
+
+        stack.push(treeAdapter.createElement($.TFOOT, NS.HTML, []));
+        assert.ok(stack.hasTableBodyContextInTableScope());
+    };
+
+    _test['Has element in select scope'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        assert.ok(!stack.hasInSelectScope($.P));
+
+        stack.push(treeAdapter.createElement($.P, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        assert.ok(stack.hasInSelectScope($.P));
+
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        assert.ok(!stack.hasInSelectScope($.P));
+    };
+
+    _test['Generate implied end tags'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.LI, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.LI, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.P, NS.HTML, []));
+
+        stack.generateImpliedEndTags();
+
+        assert.strictEqual(stack.stackTop, 2);
+        assert.strictEqual(stack.currentTagName, $.DIV);
+    };
+
+    _test['Generate implied end tags with exclusion'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.LI, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.DIV, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.LI, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.P, NS.HTML, []));
+
+        stack.generateImpliedEndTagsWithExclusion($.LI);
+
+        assert.strictEqual(stack.stackTop, 3);
+        assert.strictEqual(stack.currentTagName, $.LI);
+    };
+
+    _test['Template count'] = function () {
+        var stack = new OpenElementStack(treeAdapter.createDocument(), treeAdapter);
+
+        stack.push(treeAdapter.createElement($.HTML, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.TEMPLATE, NS.MATHML, []));
+        assert.strictEqual(stack.tmplCount, 0);
+
+        stack.push(treeAdapter.createElement($.TEMPLATE, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.LI, NS.HTML, []));
+        assert.strictEqual(stack.tmplCount, 1);
+
+        stack.push(treeAdapter.createElement($.OPTION, NS.HTML, []));
+        stack.push(treeAdapter.createElement($.TEMPLATE, NS.HTML, []));
+        assert.strictEqual(stack.tmplCount, 2);
+
+        stack.pop();
+        assert.strictEqual(stack.tmplCount, 1);
+
+        stack.pop();
+        stack.pop();
+        stack.pop();
+        assert.strictEqual(stack.tmplCount, 0);
+    };
+});
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/parser_test.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/parser_test.js
new file mode 100644
index 0000000000000000000000000000000000000000..61b8a7542c4bdc48de5a64162717d57f724ba147
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/parser_test.js
@@ -0,0 +1,136 @@
+var assert = require('assert'),
+    path = require('path'),
+    HTML = require('../../lib/common/html'),
+    parse5 = require('../../index'),
+    Parser = parse5.Parser,
+    Serializer = parse5.Serializer,
+    TestUtils = require('../test_utils');
+
+
+TestUtils.generateTestsForEachTreeAdapter(module.exports, function (_test, treeAdapter) {
+    function getFullTestName(test) {
+        return ['Parser(', test.dirName, ') - ', test.idx, '.', test.setName, ' - ', test.input].join('');
+    }
+
+    function getLocationFullTestName(test) {
+        return ['Parser(Location info) - ', test.name].join('');
+    }
+
+    function walkTree(document, handler) {
+        for (var stack = treeAdapter.getChildNodes(document).slice(); stack.length;) {
+            var node = stack.shift(),
+                children = treeAdapter.getChildNodes(node);
+
+            handler(node);
+
+            if (children && children.length)
+                stack = children.concat(stack);
+        }
+    }
+
+    //Here we go..
+    TestUtils.loadTreeConstructionTestData([
+        path.join(__dirname, '../data/tree_construction'),
+        path.join(__dirname, '../data/tree_construction_regression'),
+        path.join(__dirname, '../data/tree_construction_options')
+    ], treeAdapter).forEach(function (test) {
+        _test[getFullTestName(test)] = function () {
+            var parser = new Parser(treeAdapter, {
+                    decodeHtmlEntities: !test.disableEntitiesDecoding
+                }),
+                result = test.fragmentContext ?
+                         parser.parseFragment(test.input, test.fragmentContext) :
+                         parser.parse(test.input),
+                actual = TestUtils.serializeToTestDataFormat(result, treeAdapter),
+                msg = TestUtils.prettyPrintParserAssertionArgs(actual, test.expected);
+
+            assert.strictEqual(actual, test.expected, msg);
+        };
+    });
+
+
+    //Location info tests
+    TestUtils.loadSerializationTestData(path.join(__dirname, '../data/serialization')).forEach(function (test) {
+        //NOTE: the idea of this test is the following: we parse document with the location info.
+        //Then for each node in the tree we run serializer and compare results with the substring
+        //obtained via location info from the expected serialization results.
+        _test[getLocationFullTestName(test)] = function () {
+            var parser = new Parser(treeAdapter, {
+                    locationInfo: true,
+                    decodeHtmlEntities: false
+                }),
+                serializer = new Serializer(treeAdapter, {
+                    encodeHtmlEntities: false
+                }),
+                html = test.expected,
+                document = parser.parse(html);
+
+            walkTree(document, function (node) {
+                if (node.__location !== null) {
+                    var fragment = treeAdapter.createDocumentFragment();
+
+                    treeAdapter.appendChild(fragment, node);
+
+                    var expected = serializer.serialize(fragment),
+                        actual = html.substring(node.__location.start, node.__location.end);
+
+                    expected = TestUtils.removeNewLines(expected);
+                    actual = TestUtils.removeNewLines(actual);
+
+                    //NOTE: use ok assertion, so output will not be polluted by the whole content of the strings
+                    assert.ok(actual === expected, TestUtils.getStringDiffMsg(actual, expected));
+
+                    if (node.__location.startTag) {
+                        //NOTE: Based on the idea that the serialized fragment starts with the startTag
+                        var length = node.__location.startTag.end - node.__location.startTag.start,
+                            expectedStartTag = serializer.serialize(fragment).substring(0, length),
+                            actualStartTag = html.substring(node.__location.startTag.start, node.__location.startTag.end);
+
+                        expectedStartTag = TestUtils.removeNewLines(expectedStartTag);
+                        actualStartTag = TestUtils.removeNewLines(actualStartTag);
+
+                        assert.ok(expectedStartTag === actualStartTag, TestUtils.getStringDiffMsg(actualStartTag, expectedStartTag));
+                    }
+
+                    if (node.__location.endTag) {
+                        //NOTE: Based on the idea that the serialized fragment ends with the endTag
+                        var length = node.__location.endTag.end - node.__location.endTag.start,
+                            expectedEndTag = serializer.serialize(fragment).slice(-length),
+                            actualEndTag = html.substring(node.__location.endTag.start, node.__location.endTag.end);
+
+                        expectedEndTag = TestUtils.removeNewLines(expectedEndTag);
+                        actualEndTag = TestUtils.removeNewLines(actualEndTag);
+
+                        assert.ok(expectedEndTag === actualEndTag, TestUtils.getStringDiffMsg(actualEndTag, expectedEndTag));
+                    }
+                }
+            });
+        };
+    });
+
+    exports['Regression - location info for the implicitly generated <body>, <html> and <head> (GH-44)'] = function () {
+        var html = '</head><div class="test"></div></body></html>',
+            parser = new Parser(treeAdapter, {
+                locationInfo: true,
+                decodeHtmlEntities: false
+            }),
+            document = parser.parse(html);
+
+        //NOTE: location info for all implicitly generated elements should be null
+        walkTree(document, function (node) {
+            if (treeAdapter.getTagName(node) !== HTML.TAG_NAMES.DIV)
+                assert.strictEqual(node.__location, null);
+        });
+    };
+});
+
+
+exports['Regression - HTML5 Legacy Doctype Misparsed with htmlparser2 tree adapter (GH-45)'] = function () {
+    var html = '<!DOCTYPE html SYSTEM "about:legacy-compat"><html><head></head><body>Hi there!</body></html>',
+        parser = new Parser(parse5.TreeAdapters.htmlparser2),
+        document = parser.parse(html);
+
+    assert.strictEqual(document.childNodes[0].data, '!DOCTYPE html SYSTEM "about:legacy-compat"');
+};
+
+
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/serializer_test.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/serializer_test.js
new file mode 100644
index 0000000000000000000000000000000000000000..c3c4369e717c6cf288e63539625e82f4052849b8
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/serializer_test.js
@@ -0,0 +1,151 @@
+var assert = require('assert'),
+    path = require('path'),
+    parse5 = require('../../index'),
+    Parser = parse5.Parser,
+    Serializer = parse5.Serializer,
+    TestUtils = require('../test_utils');
+
+
+exports['Backward compatibility - parse5.TreeSerializer'] = function () {
+    assert.strictEqual(Serializer, parse5.TreeSerializer);
+};
+
+exports['Regression - Get text node\'s parent tagName only if it\'s an Element node (GH-38)'] = {
+    test: function () {
+        var parser = new Parser(),
+            serializer = new Serializer(),
+            document = parser.parse('<template>yo<div></div>42</template>'),
+            originalGetTagName = this.originalGetTagName = parse5.TreeAdapters.default.getTagName;
+
+        parse5.TreeAdapters.default.getTagName = function (element) {
+            assert.ok(element.tagName);
+
+            return originalGetTagName(element);
+        };
+
+        serializer.serialize(document);
+    },
+
+    after: function () {
+        parse5.TreeAdapters.default.getTagName = this.originalGetTagName;
+    }
+};
+
+exports['Regression - SYSTEM-only doctype serialization'] = function () {
+    var html = '<!DOCTYPE html SYSTEM "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">' +
+               '<html><head></head><body></body></html>',
+        parser = new Parser(),
+        serializer = new Serializer(),
+        document = parser.parse(html),
+        serializedResult = serializer.serialize(document);
+
+    assert.strictEqual(serializedResult, html);
+};
+
+exports['Regression - Escaping of doctypes with quotes in them'] = function () {
+    var htmlStrs = [
+            '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" ' +
+            '"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">' +
+            '<html><head></head><body></body></html>',
+
+            '<!DOCTYPE html PUBLIC \'-//W3C//"DTD" XHTML 1.0 Transitional//EN\' ' +
+            '"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">' +
+            '<html><head></head><body></body></html>',
+
+            '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" ' +
+            '\'http://www.w3.org/TR/xhtml1/DTD/"xhtml1-transitional.dtd"\'>' +
+            '<html><head></head><body></body></html>'
+        ],
+        parser = new Parser(),
+        serializer = new Serializer();
+
+    htmlStrs.forEach(function (html) {
+        var document = parser.parse(html),
+            serializedResult = serializer.serialize(document);
+
+        assert.strictEqual(serializedResult, html);
+    });
+};
+
+exports['Regression - new line in <pre> tag'] = function () {
+    var htmlStrs = [
+            {
+                src: '<!DOCTYPE html><html><head></head><body><pre>\ntest</pre></body></html>',
+                expected: '<!DOCTYPE html><html><head></head><body><pre>test</pre></body></html>'
+            },
+
+            {
+                src: '<!DOCTYPE html><html><head></head><body><pre>\n\ntest</pre></body></html>',
+                expected: '<!DOCTYPE html><html><head></head><body><pre>\n\ntest</pre></body></html>'
+            }
+        ],
+        parser = new Parser(),
+        serializer = new Serializer();
+
+    htmlStrs.forEach(function (htmlStr) {
+        var document = parser.parse(htmlStr.src),
+            serializedResult = serializer.serialize(document);
+
+        assert.strictEqual(serializedResult, htmlStr.expected);
+    });
+};
+
+exports['Options - encodeHtmlEntities'] = function () {
+    var testHtmlCases = [
+            {
+                options: {encodeHtmlEntities: true},
+                src: '<!DOCTYPE html><html><head></head><body>&</body></html>',
+                expected: '<!DOCTYPE html><html><head></head><body>&amp;</body></html>'
+            },
+
+            {
+                options: {encodeHtmlEntities: false},
+                src: '<!DOCTYPE html><html><head></head><body>&</body></html>',
+                expected: '<!DOCTYPE html><html><head></head><body>&</body></html>'
+            },
+            {
+                options: {encodeHtmlEntities: true},
+                src: '<!DOCTYPE html><html><head></head><body><a href="http://example.com?hello=1&world=2"></a></body></html>',
+                expected: '<!DOCTYPE html><html><head></head><body><a href="http://example.com?hello=1&amp;world=2"></a></body></html>'
+            },
+
+            {
+                options: {encodeHtmlEntities: false},
+                src: '<!DOCTYPE html><html><head></head><body><a href="http://example.com?hello=1&world=2"></a></body></html>',
+                expected: '<!DOCTYPE html><html><head></head><body><a href="http://example.com?hello=1&world=2"></a></body></html>'
+            }
+        ],
+        parser = new Parser();
+
+    testHtmlCases.forEach(function (testCase) {
+        var serializer = new Serializer(null, testCase.options);
+        var document = parser.parse(testCase.src),
+            serializedResult = serializer.serialize(document);
+
+        assert.strictEqual(serializedResult, testCase.expected);
+    });
+};
+
+
+TestUtils.generateTestsForEachTreeAdapter(module.exports, function (_test, treeAdapter) {
+    function getFullTestName(test) {
+        return ['Serializer - ', test.idx, '.', test.name].join('');
+    }
+
+    var testDataDir = path.join(__dirname, '../data/serialization');
+
+    //Here we go..
+    TestUtils.loadSerializationTestData(testDataDir).forEach(function (test) {
+        _test[getFullTestName(test)] = function () {
+            var parser = new Parser(treeAdapter),
+                serializer = new Serializer(treeAdapter),
+                document = parser.parse(test.src),
+                serializedResult = TestUtils.removeNewLines(serializer.serialize(document)),
+                expected = TestUtils.removeNewLines(test.expected);
+
+            //NOTE: use ok assertion, so output will not be polluted by the whole content of the strings
+            assert.ok(serializedResult === expected, TestUtils.getStringDiffMsg(serializedResult, expected));
+        };
+    });
+
+});
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/simple_api_parser_test.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/simple_api_parser_test.js
new file mode 100644
index 0000000000000000000000000000000000000000..264f7eeb86b5d81fbf69d2b1c63bb97682c12e5f
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/simple_api_parser_test.js
@@ -0,0 +1,117 @@
+var assert = require('assert'),
+    path = require('path'),
+    SimpleApiParser = require('../../index').SimpleApiParser,
+    TestUtils = require('../test_utils');
+
+function getFullTestName(test, idx) {
+    return ['SimpleApiParser - ', idx, '.', test.name].join('');
+}
+
+function sanitizeForComparison(str) {
+    return TestUtils.removeNewLines(str)
+        .replace(/\s/g, '')
+        .replace(/'/g, '"')
+        .toLowerCase();
+}
+
+
+function createTest(html, expected, options) {
+    return function () {
+        //NOTE: the idea of the test is to serialize back given HTML using SimpleApiParser handlers
+        var actual = '',
+            parser = new SimpleApiParser({
+                doctype: function (name, publicId, systemId) {
+                    actual += '<!DOCTYPE ' + name;
+
+                    if (publicId !== null)
+                        actual += ' PUBLIC "' + publicId + '"';
+
+                    else if (systemId !== null)
+                        actual += ' SYSTEM';
+
+                    if (systemId !== null)
+                        actual += ' "' + systemId + '"';
+
+
+                    actual += '>';
+                },
+
+                startTag: function (tagName, attrs, selfClosing) {
+                    actual += '<' + tagName;
+
+                    if (attrs.length) {
+                        for (var i = 0; i < attrs.length; i++)
+                            actual += ' ' + attrs[i].name + '="' + attrs[i].value + '"';
+                    }
+
+                    actual += selfClosing ? '/>' : '>';
+                },
+
+                endTag: function (tagName) {
+                    actual += '</' + tagName + '>';
+                },
+
+                text: function (text) {
+                    actual += text;
+                },
+
+                comment: function (text) {
+                    actual += '<!--' + text + '-->';
+                }
+            }, options);
+
+        parser.parse(html);
+
+        expected = sanitizeForComparison(expected);
+        actual = sanitizeForComparison(actual);
+
+        //NOTE: use ok assertion, so output will not be polluted by the whole content of the strings
+        assert.ok(actual === expected, TestUtils.getStringDiffMsg(actual, expected));
+    };
+}
+
+TestUtils.loadSerializationTestData(path.join(__dirname, '../data/simple_api_parsing'))
+    .concat([
+        {
+            name: 'Options - decodeHtmlEntities (text)',
+            src: '<div>&amp;&copy;</div>',
+            expected: '<div>&amp;&copy;</div>',
+            options: {
+                decodeHtmlEntities: false
+            }
+        },
+
+        {
+            name: 'Options - decodeHtmlEntities (attributes)',
+            src: '<a href = "&amp;test&lt;" &copy;>Yo</a>',
+            expected: '<a href = "&amp;test&lt;" &copy;="">Yo</a>',
+            options: {
+                decodeHtmlEntities: false
+            }
+        }
+    ])
+    .forEach(function (test, idx) {
+        var testName = getFullTestName(test, idx);
+
+        exports[testName] = createTest(test.src, test.expected, test.options);
+
+        exports['Options - locationInfo - ' + testName] = function () {
+            //NOTE: we've already tested the correctness of the location info with the Tokenizer tests.
+            //So here we just check that SimpleApiParser provides this info in the handlers.
+            var handlers = ['doctype', 'startTag', 'endTag', 'text', 'comment'].reduce(function (handlers, key) {
+                    handlers[key] = function () {
+                        var locationInfo = arguments[arguments.length - 1];
+
+                        assert.strictEqual(typeof locationInfo.start, 'number');
+                        assert.strictEqual(typeof locationInfo.end, 'number');
+                    };
+                    return handlers;
+                }, {}),
+                parser = new SimpleApiParser(handlers, {locationInfo: true});
+
+            parser.parse(test.src);
+        };
+    });
+
+
+
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/tokenizer_test.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/tokenizer_test.js
new file mode 100644
index 0000000000000000000000000000000000000000..45b1801763d9d43880e3aeadf65a2ba877398a36
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/fixtures/tokenizer_test.js
@@ -0,0 +1,251 @@
+var assert = require('assert'),
+    fs = require('fs'),
+    path = require('path'),
+    Tokenizer = require('../../lib/tokenization/tokenizer');
+
+function tokenize(html, initialState, lastStartTag) {
+    var tokenizer = new Tokenizer(html),
+        nextToken = null,
+        out = [];
+
+    tokenizer.state = initialState;
+
+    if (lastStartTag)
+        tokenizer.lastStartTagName = lastStartTag;
+
+    do {
+        nextToken = tokenizer.getNextToken();
+
+        //NOTE: append current token to the output sequence in html5lib test suite compatible format
+        switch (nextToken.type) {
+            case Tokenizer.CHARACTER_TOKEN:
+            case Tokenizer.NULL_CHARACTER_TOKEN:
+            case Tokenizer.WHITESPACE_CHARACTER_TOKEN:
+                out.push(['Character', nextToken.chars]);
+                break;
+
+            case Tokenizer.START_TAG_TOKEN:
+                var reformatedAttrs = {};
+
+                nextToken.attrs.forEach(function (attr) {
+                    reformatedAttrs[attr.name] = attr.value;
+                });
+
+                var startTagEntry = [
+                    'StartTag',
+                    nextToken.tagName,
+                    reformatedAttrs
+                ];
+
+                if (nextToken.selfClosing)
+                    startTagEntry.push(true);
+
+                out.push(startTagEntry);
+                break;
+
+            case Tokenizer.END_TAG_TOKEN:
+                out.push(['EndTag', nextToken.tagName]);
+                break;
+
+            case Tokenizer.COMMENT_TOKEN:
+                out.push(['Comment', nextToken.data]);
+                break;
+
+            case Tokenizer.DOCTYPE_TOKEN:
+                out.push([
+                    'DOCTYPE',
+                    nextToken.name,
+                    nextToken.publicId,
+                    nextToken.systemId,
+                    !nextToken.forceQuirks
+                ]);
+                break;
+        }
+    } while (nextToken.type !== Tokenizer.EOF_TOKEN);
+
+    return concatCharacterTokens(out)
+}
+
+function unicodeUnescape(str) {
+    return str.replace(/\\u([\d\w]{4})/gi, function (match, chCodeStr) {
+        return String.fromCharCode(parseInt(chCodeStr, 16));
+    });
+}
+
+function unescapeDescrIO(testDescr) {
+    testDescr.input = unicodeUnescape(testDescr.input);
+
+    testDescr.output.forEach(function (tokenEntry) {
+        if (tokenEntry === 'ParseError')
+            return;
+
+        //NOTE: unescape token tagName (for StartTag and EndTag tokens), comment data (for Comment token),
+        //character token data (for Character token).
+        tokenEntry[1] = unicodeUnescape(tokenEntry[1]);
+
+        //NOTE: unescape token attributes(if we have them).
+        if (tokenEntry.length > 2) {
+            Object.keys(tokenEntry).forEach(function (attrName) {
+                var attrVal = tokenEntry[attrName];
+
+                delete tokenEntry[attrName];
+                tokenEntry[unicodeUnescape(attrName)] = unicodeUnescape(attrVal);
+            });
+        }
+    });
+}
+
+function concatCharacterTokens(tokenEntries) {
+    var result = [];
+
+    tokenEntries.forEach(function (tokenEntry) {
+        if (tokenEntry[0] === 'Character') {
+            var lastEntry = result[result.length - 1];
+
+            if (lastEntry && lastEntry[0] === 'Character') {
+                lastEntry[1] += tokenEntry[1];
+                return;
+            }
+        }
+
+        result.push(tokenEntry);
+    });
+
+    return result;
+}
+
+function getTokenizerSuitableStateName(testDataStateName) {
+    return testDataStateName.toUpperCase().replace(/\s/g, '_');
+}
+
+function loadTests() {
+    var dataDirPath = path.join(__dirname, '../data/tokenization'),
+        testSetFileNames = fs.readdirSync(dataDirPath),
+        testIdx = 0,
+        tests = [];
+
+    testSetFileNames.forEach(function (fileName) {
+        var filePath = path.join(dataDirPath, fileName),
+            testSetJson = fs.readFileSync(filePath).toString(),
+            testSet = JSON.parse(testSetJson),
+            testDescrs = testSet.tests,
+            setName = fileName.replace('.test', '');
+
+        testDescrs.forEach(function (descr) {
+            if (!descr.initialStates)
+                descr.initialStates = ['Data state'];
+
+            if (descr.doubleEscaped)
+                unescapeDescrIO(descr);
+
+            var expected = [];
+
+            descr.output.forEach(function (tokenEntry) {
+                if (tokenEntry !== 'ParseError')
+                    expected.push(tokenEntry);
+            });
+
+            descr.initialStates.forEach(function (initialState) {
+                tests.push({
+                    idx: ++testIdx,
+                    setName: setName,
+                    name: descr.description,
+                    input: descr.input,
+                    expected: concatCharacterTokens(expected),
+                    initialState: getTokenizerSuitableStateName(initialState),
+                    lastStartTag: descr.lastStartTag
+                });
+            });
+        });
+    });
+
+    return tests;
+}
+
+function getFullTestName(test) {
+    return ['Tokenizer - ' +
+            test.idx, '.', test.setName, ' - ', test.name, ' - Initial state: ', test.initialState].join('');
+}
+
+//Here we go..
+loadTests().forEach(function (test) {
+    exports[getFullTestName(test)] = function () {
+        var out = tokenize(test.input, test.initialState, test.lastStartTag);
+
+        assert.deepEqual(out, test.expected);
+    };
+});
+
+
+exports['Options - locationInfo'] = function () {
+    var testCases = [
+        {
+            initialMode: Tokenizer.MODE.DATA,
+            lastStartTagName: '',
+            htmlChunks: [
+                '\r\n', '<!DOCTYPE html>', '\n',
+                '<!-- Test -->', '\n',
+                '<head>',
+                '\n   ', '<meta charset="utf-8">', '<title>', '   ', 'node.js', '\u0000', '</title>', '\n',
+                '</head>', '\n',
+                '<body id="front">', '\n',
+                '<div id="intro">',
+                '\n   ', '<p>',
+                '\n       ', 'Node.js', ' ', 'is', ' ', 'a',
+                '\n       ', 'platform', ' ', 'built', ' ', 'on',
+                '\n       ', '<a href="http://code.google.com/p/v8/">',
+                '\n       ', 'Chrome\'s', ' ', 'JavaScript', ' ', 'runtime',
+                '\n       ', '</a>', '\n',
+                '</div>',
+                '<body>'
+            ]
+        },
+        {
+            initialMode: Tokenizer.MODE.RCDATA,
+            lastStartTagName: 'title',
+            htmlChunks: [
+                '<div>Test',
+                ' \n   ', 'hey', ' ', 'ya!', '</title>', '<!--Yo-->'
+            ]
+        },
+        {
+            initialMode: Tokenizer.MODE.RAWTEXT,
+            lastStartTagName: 'style',
+            htmlChunks: [
+                '.header{', ' \n   ', 'color:red;', '\n', '}', '</style>', 'Some', ' ', 'text'
+            ]
+        },
+        {
+            initialMode: Tokenizer.MODE.SCRIPT_DATA,
+            lastStartTagName: 'script',
+            htmlChunks: [
+                'var', ' ', 'a=c', ' ', '-', ' ', 'd;', '\n', 'a<--d;', '</script>', '<div>'
+            ]
+        },
+        {
+            initialMode: Tokenizer.MODE.PLAINTEXT,
+            lastStartTagName: 'plaintext',
+            htmlChunks: [
+                'Text', ' \n', 'Test</plaintext><div>'
+            ]
+        }
+
+    ];
+
+    testCases.forEach(function (testCase) {
+        var html = testCase.htmlChunks.join(''),
+            tokenizer = new Tokenizer(html, {locationInfo: true});
+
+        tokenizer.state = testCase.initialMode;
+        tokenizer.lastStartTagName = testCase.lastStartTagName;
+
+        for (var token = tokenizer.getNextToken(), i = 0; token.type !== Tokenizer.EOF_TOKEN;) {
+            var chunk = html.substring(token.location.start, token.location.end);
+
+            assert.strictEqual(chunk, testCase.htmlChunks[i]);
+
+            token = tokenizer.getNextToken();
+            i++;
+        }
+    });
+};
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/run_tests.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/run_tests.js
new file mode 100644
index 0000000000000000000000000000000000000000..89fa89449e331d9b0c842697905dcbffdb3d9935
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/run_tests.js
@@ -0,0 +1,18 @@
+var fs = require('fs'),
+    path = require('path'),
+    Mocha = require('mocha');
+
+var fixturesDir = path.join(__dirname, './fixtures'),
+    mocha = new Mocha()
+        .ui('exports')
+        .reporter('progress');
+
+fs.readdirSync(fixturesDir).forEach(function (file) {
+    mocha.addFile(path.join(fixturesDir, file));
+});
+
+mocha.run(function (failed) {
+    process.on('exit', function () {
+        process.exit(failed);
+    });
+});
\ No newline at end of file
diff --git a/third_party/catapult/third_party/vinn/third_party/parse5/test/test_utils.js b/third_party/catapult/third_party/vinn/third_party/parse5/test/test_utils.js
new file mode 100644
index 0000000000000000000000000000000000000000..6bdc06c694fe436739486c35af8a34ba3930e918
--- /dev/null
+++ b/third_party/catapult/third_party/vinn/third_party/parse5/test/test_utils.js
@@ -0,0 +1,241 @@
+var fs = require('fs'),
+    path = require('path'),
+    parse5 = require('../index'),
+    HTML = require('../lib/common/html');
+
+function addSlashes(str) {
+    return str
+        .replace(/\t/g, '\\t')
+        .replace(/\n/g, '\\n')
+        .replace(/\f/g, '\\f')
+        .replace(/\r/g, '\\r');
+}
+
+function createDiffMarker(markerPosition) {
+    var marker = '';
+
+    for (var i = 0; i < markerPosition - 1; i++)
+        marker += ' ';
+
+    return marker + '^\n';
+}
+
+//NOTE: creates test suites for each available tree adapter.
+exports.generateTestsForEachTreeAdapter = function (moduleExports, ctor) {
+    Object.keys(parse5.TreeAdapters).forEach(function (adapterName) {
+        var tests = {},
+            adapter = parse5.TreeAdapters[adapterName];
+
+        ctor(tests, adapter);
+
+        Object.keys(tests).forEach(function (testName) {
+            moduleExports['Tree adapter: ' + adapterName + ' - ' + testName] = tests[testName];
+        });
+    });
+};
+
+exports.getStringDiffMsg = function (actual, expected) {
+    for (var i = 0; i < expected.length; i++) {
+        if (actual[i] !== expected[i]) {
+            var diffMsg = '\nString differ at index ' + i + '\n';
+
+            var expectedStr = 'Expected: ' + addSlashes(expected.substring(i - 100, i + 1)),
+                expectedDiffMarker = createDiffMarker(expectedStr.length);
+
+            diffMsg += expectedStr + addSlashes(expected.substring(i + 1, i + 20)) + '\n' + expectedDiffMarker;
+
+            var actualStr = 'Actual:   ' + addSlashes(actual.substring(i - 100, i + 1)),
+                actualDiffMarker = createDiffMarker(actualStr.length);
+
+            diffMsg += actualStr + addSlashes(actual.substring(i + 1, i + 20)) + '\n' + actualDiffMarker;
+
+            return diffMsg;
+        }
+    }
+
+    return '';
+};
+
+exports.removeNewLines = function (str) {
+    return str
+        .replace(/\r/g, '')
+        .replace(/\n/g, '');
+};
+
+function normalizeNewLine(str) {
+    return str.replace(/\r\n/g, '\n');
+}
+
+exports.loadSerializationTestData = function (dataDirPath) {
+    var testSetFileDirs = fs.readdirSync(dataDirPath),
+        tests = [];
+
+    testSetFileDirs.forEach(function (dirName) {
+        var srcFilePath = path.join(dataDirPath, dirName, 'src.html'),
+            expectedFilePath = path.join(dataDirPath, dirName, 'expected.html'),
+            src = fs.readFileSync(srcFilePath).toString(),
+            expected = fs.readFileSync(expectedFilePath).toString();
+
+        tests.push({
+            name: dirName,
+            src: normalizeNewLine(src),
+            expected: normalizeNewLine(expected)
+        });
+    });
+
+    return tests;
+};
+
+exports.loadTreeConstructionTestData = function (dataDirs, treeAdapter) {
+    var testIdx = 0,
+        tests = [];
+
+    dataDirs.forEach(function (dataDirPath) {
+        var testSetFileNames = fs.readdirSync(dataDirPath),
+            dirName = path.basename(dataDirPath);
+
+        testSetFileNames.forEach(function (fileName) {
+            var filePath = path.join(dataDirPath, fileName),
+                testSet = fs.readFileSync(filePath).toString(),
+                setName = fileName.replace('.dat', ''),
+                testDescrs = [],
+                curDirective = '',
+                curDescr = null;
+
+            testSet.split(/\r?\n/).forEach(function (line) {
+                if (line === '#data') {
+                    curDescr = {};
+                    testDescrs.push(curDescr);
+                }
+
+                if (line[0] === '#') {
+                    curDirective = line;
+                    curDescr[curDirective] = [];
+                }
+
+                else
+                    curDescr[curDirective].push(line);
+            });
+
+            testDescrs.forEach(function (descr) {
+                var fragmentContextTagName = descr['#document-fragment'] && descr['#document-fragment'].join('');
+
+                tests.push({
+                    idx: ++testIdx,
+                    setName: setName,
+                    dirName: dirName,
+                    input: descr['#data'].join('\r\n'),
+                    expected: descr['#document'].join('\n'),
+                    expectedErrors: descr['#errors'],
+                    disableEntitiesDecoding: !!descr['#disable-html-entities-decoding'],
+                    fragmentContext: fragmentContextTagName &&
+                                     treeAdapter.createElement(fragmentContextTagName, HTML.NAMESPACES.HTML, [])
+                });
+            });
+        });
+    });
+
+    return tests;
+};
+
+exports.serializeToTestDataFormat = function (rootNode, treeAdapter) {
+    function getSerializedTreeIndent(indent) {
+        var str = '|';
+
+        for (var i = 0; i < indent + 1; i++)
+            str += ' ';
+
+        return str;
+    }
+
+    function getElementSerializedNamespaceURI(element) {
+        switch (treeAdapter.getNamespaceURI(element)) {
+            case HTML.NAMESPACES.SVG:
+                return 'svg ';
+            case HTML.NAMESPACES.MATHML:
+                return 'math ';
+            default :
+                return '';
+        }
+    }
+
+    function serializeNodeList(nodes, indent) {
+        var str = '';
+
+        nodes.forEach(function (node) {
+            str += getSerializedTreeIndent(indent);
+
+            if (treeAdapter.isCommentNode(node))
+                str += '<!-- ' + treeAdapter.getCommentNodeContent(node) + ' -->\n';
+
+            else if (treeAdapter.isTextNode(node))
+                str += '"' + treeAdapter.getTextNodeContent(node) + '"\n';
+
+            else if (treeAdapter.isDocumentTypeNode(node)) {
+                var parts = [],
+                    publicId = treeAdapter.getDocumentTypeNodePublicId(node),
+                    systemId = treeAdapter.getDocumentTypeNodeSystemId(node);
+
+                str += '<!DOCTYPE';
+
+                parts.push(treeAdapter.getDocumentTypeNodeName(node) || '');
+
+                if (publicId !== null || systemId !== null) {
+                    parts.push('"' + (publicId || '') + '"');
+                    parts.push('"' + (systemId || '') + '"');
+                }
+
+                parts.forEach(function (part) {
+                    str += ' ' + part;
+                });
+
+                str += '>\n';
+            }
+
+            else {
+                var tn = treeAdapter.getTagName(node);
+
+                str += '<' + getElementSerializedNamespaceURI(node) + tn + '>\n';
+
+                var childrenIndent = indent + 2,
+                    serializedAttrs = [];
+
+                treeAdapter.getAttrList(node).forEach(function (attr) {
+                    var attrStr = getSerializedTreeIndent(childrenIndent);
+
+                    if (attr.prefix)
+                        attrStr += attr.prefix + ' ';
+
+                    attrStr += attr.name + '="' + attr.value + '"\n';
+
+                    serializedAttrs.push(attrStr);
+                });
+
+                str += serializedAttrs.sort().join('');
+
+                if (tn === HTML.TAG_NAMES.TEMPLATE && treeAdapter.getNamespaceURI(node) === HTML.NAMESPACES.HTML) {
+                    str += getSerializedTreeIndent(childrenIndent) + 'content\n';
+                    childrenIndent += 2;
+                    node = treeAdapter.getChildNodes(node)[0];
+                }
+
+                str += serializeNodeList(treeAdapter.getChildNodes(node), childrenIndent);
+            }
+        });
+
+        return str;
+    }
+
+    return serializeNodeList(treeAdapter.getChildNodes(rootNode), 0);
+};
+
+exports.prettyPrintParserAssertionArgs = function (actual, expected) {
+    var msg = '\nExpected:\n';
+    msg += '-----------------\n';
+    msg += expected + '\n';
+    msg += '\nActual:\n';
+    msg += '-----------------\n';
+    msg += actual + '\n';
+
+    return msg;
+};
