diff --git a/.gitattributes b/.gitattributes
index dfb4ad17849f64e8f69fd55c8bfb15e91a928ac2..89a1820a711ae536bc0157ddf9a164deeb4c2e2d 100644
--- a/.gitattributes
+++ b/.gitattributes
@@ -3,3 +3,12 @@
 
 # Force LF checkout of the pins files to avoid transport_security_state_generator errors.
 /net/http/*.pins  text eol=lf
+
+*       -text
+*.cpp text eol=lf
+*.cc text eol=lf
+*.h text eol=lf
+*.mm text eol=lf
+*.gn text eol=lf
+*.gni text eol=lf
+*.sh text eol=lf
diff --git a/.gitmodules b/.gitmodules
new file mode 100644
index 0000000000000000000000000000000000000000..98e32ff975ded7ec5707b83b713c8bf6be7be4f6
--- /dev/null
+++ b/.gitmodules
@@ -0,0 +1,255 @@
+[submodule "buildtools"]
+	path = buildtools
+	url = ../buildtools.git
+[submodule "chrome/installer/mac/third_party/xz/xz"]
+	path = chrome/installer/mac/third_party/xz/xz
+	url = ../deps/xz.git
+[submodule "chrome/test/data/perf/canvas_bench"]
+	path = chrome/test/data/perf/canvas_bench
+	url = ../canvas_bench.git
+[submodule "chrome/test/data/perf/frame_rate/content"]
+	path = chrome/test/data/perf/frame_rate/content
+	url = ../frame_rate/content.git
+[submodule "chrome/test/data/vr/webvr_info"]
+	path = chrome/test/data/vr/webvr_info
+	url = ../../external/github.com/toji/webvr.info.git
+[submodule "media/cdm/api"]
+	path = media/cdm/api
+	url = ../cdm.git
+[submodule "third_party/SPIRV-Tools/src"]
+	path = third_party/SPIRV-Tools/src
+	url = ../../external/github.com/KhronosGroup/SPIRV-Tools.git
+[submodule "third_party/angle"]
+	path = third_party/angle
+	url = ../../angle/angle.git
+[submodule "third_party/bidichecker"]
+	path = third_party/bidichecker
+	url = ../../external/bidichecker/lib.git
+[submodule "third_party/boringssl/src"]
+	path = third_party/boringssl/src
+	url = ../../boringssl/boringssl.git
+[submodule "third_party/breakpad/breakpad"]
+	path = third_party/breakpad/breakpad
+	url = ../../breakpad/breakpad.git
+[submodule "third_party/catapult"]
+	path = third_party/catapult
+	url = ../../catapult.git
+[submodule "third_party/ced/src"]
+	path = third_party/ced/src
+	url = ../../external/github.com/google/compact_enc_det.git
+[submodule "third_party/cld_3/src"]
+	path = third_party/cld_3/src
+	url = ../../external/github.com/google/cld_3.git
+[submodule "third_party/colorama/src"]
+	path = third_party/colorama/src
+	url = ../../external/colorama.git
+[submodule "third_party/crc32c/src"]
+	path = third_party/crc32c/src
+	url = ../../external/github.com/google/crc32c.git
+[submodule "third_party/cros_system_api"]
+	path = third_party/cros_system_api
+	url = ../../chromiumos/platform/system_api.git
+[submodule "third_party/depot_tools"]
+	path = third_party/depot_tools
+	url = ../tools/depot_tools.git
+[submodule "third_party/deqp/src"]
+	path = third_party/deqp/src
+	url = ../../android/platform/external/deqp.git
+[submodule "third_party/dom_distiller_js/dist"]
+	path = third_party/dom_distiller_js/dist
+	url = ../dom-distiller/dist.git
+[submodule "third_party/ffmpeg"]
+	path = third_party/ffmpeg
+	url = ../third_party/ffmpeg.git
+[submodule "third_party/flac"]
+	path = third_party/flac
+	url = ../deps/flac.git
+[submodule "third_party/flatbuffers/src"]
+	path = third_party/flatbuffers/src
+	url = ../../external/github.com/google/flatbuffers.git
+[submodule "third_party/fontconfig/src"]
+	path = third_party/fontconfig/src
+	url = ../../external/fontconfig.git
+[submodule "third_party/freetype/src"]
+	path = third_party/freetype/src
+	url = ../src/third_party/freetype2.git
+[submodule "third_party/gestures/gestures"]
+	path = third_party/gestures/gestures
+	url = ../../chromiumos/platform/gestures.git
+[submodule "third_party/glslang-angle/src"]
+	path = third_party/glslang-angle/src
+	url = ../../android/platform/external/shaderc/glslang.git
+[submodule "third_party/glslang/src"]
+	path = third_party/glslang/src
+	url = ../../external/github.com/google/glslang.git
+[submodule "third_party/google_toolbox_for_mac/src"]
+	path = third_party/google_toolbox_for_mac/src
+	url = ../../external/github.com/google/google-toolbox-for-mac.git
+[submodule "third_party/googletest/src"]
+	path = third_party/googletest/src
+	url = ../../external/github.com/google/googletest.git
+[submodule "third_party/hunspell_dictionaries"]
+	path = third_party/hunspell_dictionaries
+	url = ../deps/hunspell_dictionaries.git
+[submodule "third_party/icu"]
+	path = third_party/icu
+	url = ../deps/icu.git
+[submodule "third_party/jsoncpp/source"]
+	path = third_party/jsoncpp/source
+	url = ../../external/github.com/open-source-parsers/jsoncpp.git
+[submodule "third_party/leveldatabase/src"]
+	path = third_party/leveldatabase/src
+	url = ../../external/leveldb.git
+[submodule "third_party/libFuzzer/src"]
+	path = third_party/libFuzzer/src
+	url = ../llvm-project/compiler-rt/lib/fuzzer.git
+[submodule "third_party/libaddressinput/src"]
+	path = third_party/libaddressinput/src
+	url = ../../external/libaddressinput.git
+[submodule "third_party/libdrm/src"]
+	path = third_party/libdrm/src
+	url = ../../chromiumos/third_party/libdrm.git
+[submodule "third_party/libevdev/src"]
+	path = third_party/libevdev/src
+	url = ../../chromiumos/platform/libevdev.git
+[submodule "third_party/libjpeg_turbo"]
+	path = third_party/libjpeg_turbo
+	url = ../deps/libjpeg_turbo.git
+[submodule "third_party/liblouis/src"]
+	path = third_party/liblouis/src
+	url = ../../external/liblouis-github.git
+[submodule "third_party/libphonenumber/dist"]
+	path = third_party/libphonenumber/dist
+	url = ../../external/libphonenumber.git
+[submodule "third_party/libprotobuf-mutator/src"]
+	path = third_party/libprotobuf-mutator/src
+	url = ../../external/github.com/google/libprotobuf-mutator.git
+[submodule "third_party/libsrtp"]
+	path = third_party/libsrtp
+	url = ../deps/libsrtp.git
+[submodule "third_party/libsync/src"]
+	path = third_party/libsync/src
+	url = ../../aosp/platform/system/core/libsync.git
+[submodule "third_party/libvpx/source/libvpx"]
+	path = third_party/libvpx/source/libvpx
+	url = ../../webm/libvpx.git
+[submodule "third_party/libwebm/source"]
+	path = third_party/libwebm/source
+	url = ../../webm/libwebm.git
+[submodule "third_party/libyuv"]
+	path = third_party/libyuv
+	url = ../../libyuv/libyuv.git
+[submodule "third_party/lighttpd"]
+	path = third_party/lighttpd
+	url = ../deps/lighttpd.git
+[submodule "third_party/lss"]
+	path = third_party/lss
+	url = ../../linux-syscall-support.git
+[submodule "third_party/mesa/src"]
+	path = third_party/mesa/src
+	url = ../deps/mesa.git
+[submodule "third_party/minigbm/src"]
+	path = third_party/minigbm/src
+	url = ../../chromiumos/platform/minigbm.git
+[submodule "third_party/minizip/src"]
+	path = third_party/minizip/src
+	url = ../../external/github.com/nmoinvaz/minizip.git
+[submodule "third_party/openh264/src"]
+	path = third_party/openh264/src
+	url = ../../external/github.com/cisco/openh264.git
+[submodule "third_party/openmax_dl"]
+	path = third_party/openmax_dl
+	url = ../../webrtc/deps/third_party/openmax.git
+[submodule "third_party/pdfium"]
+	path = third_party/pdfium
+	url = ../../pdfium/pdfium.git
+[submodule "third_party/perfetto"]
+	path = third_party/perfetto
+	url = ../../android/platform/external/perfetto.git
+[submodule "third_party/pyftpdlib/src"]
+	path = third_party/pyftpdlib/src
+	url = ../../external/pyftpdlib.git
+[submodule "third_party/pywebsocket/src"]
+	path = third_party/pywebsocket/src
+	url = ../../external/github.com/google/pywebsocket.git
+[submodule "third_party/re2/src"]
+	path = third_party/re2/src
+	url = ../../external/github.com/google/re2.git
+[submodule "third_party/sfntly/src"]
+	path = third_party/sfntly/src
+	url = ../../external/github.com/googlei18n/sfntly.git
+[submodule "third_party/shaderc/src"]
+	path = third_party/shaderc/src
+	url = ../../external/github.com/google/shaderc.git
+[submodule "third_party/skia"]
+	path = third_party/skia
+	url = ../../skia/skia.git
+[submodule "third_party/smhasher/src"]
+	path = third_party/smhasher/src
+	url = ../../external/smhasher.git
+[submodule "third_party/snappy/src"]
+	path = third_party/snappy/src
+	url = ../../external/github.com/google/snappy.git
+[submodule "third_party/spirv-headers/src"]
+	path = third_party/spirv-headers/src
+	url = ../../android/platform/external/shaderc/spirv-headers.git
+[submodule "third_party/spirv-tools-angle/src"]
+	path = third_party/spirv-tools-angle/src
+	url = ../../android/platform/external/shaderc/spirv-tools.git
+[submodule "third_party/swiftshader"]
+	path = third_party/swiftshader
+	url = ../../swiftshader/SwiftShader.git
+[submodule "third_party/usrsctp/usrsctplib"]
+	path = third_party/usrsctp/usrsctplib
+	url = ../../external/github.com/sctplab/usrsctp.git
+[submodule "third_party/visualmetrics/src"]
+	path = third_party/visualmetrics/src
+	url = ../../external/github.com/WPO-Foundation/visualmetrics.git
+[submodule "third_party/vulkan-validation-layers/src"]
+	path = third_party/vulkan-validation-layers/src
+	url = ../../android/platform/external/vulkan-validation-layers.git
+[submodule "third_party/wayland-protocols/src"]
+	path = third_party/wayland-protocols/src
+	url = ../../external/anongit.freedesktop.org/git/wayland/wayland-protocols.git
+[submodule "third_party/wayland/src"]
+	path = third_party/wayland/src
+	url = ../../external/anongit.freedesktop.org/git/wayland/wayland.git
+[submodule "third_party/wds/src"]
+	path = third_party/wds/src
+	url = ../../external/github.com/01org/wds.git
+[submodule "third_party/webdriver/pylib"]
+	path = third_party/webdriver/pylib
+	url = ../../external/selenium/py.git
+[submodule "third_party/webgl/src"]
+	path = third_party/webgl/src
+	url = ../../external/khronosgroup/webgl.git
+[submodule "third_party/webrtc"]
+	path = third_party/webrtc
+	url = ../../webrtc/src.git
+[submodule "third_party/xdg-utils"]
+	path = third_party/xdg-utils
+	url = ../deps/xdg-utils.git
+[submodule "third_party/yasm/source/patched-yasm"]
+	path = third_party/yasm/source/patched-yasm
+	url = ../deps/yasm/patched-yasm.git
+[submodule "tools/gyp"]
+	path = tools/gyp
+	url = ../../external/gyp.git
+[submodule "tools/page_cycler/acid3"]
+	path = tools/page_cycler/acid3
+	url = ../deps/acid3.git
+[submodule "tools/swarming_client"]
+	path = tools/swarming_client
+	url = ../../infra/luci/client-py.git
+[submodule "v8"]
+	path = v8
+	url = ../../v8/v8.git
+[submodule "third_party/pefile"]
+	path = third_party/pefile
+	url = ../../external/pefile.git
+[submodule "third_party/bison"]
+	path = third_party/bison
+	url = ../deps/bison.git
+[submodule "third_party/gperf"]
+	path = third_party/gperf
+	url = ../deps/gperf.git
diff --git a/android_webview/BUILD.gn b/android_webview/BUILD.gn
index 4e12f9a38c02668373fef145262ff52422159f17..c9dfff5e933578859da926d12f0318076ca96050 100644
--- a/android_webview/BUILD.gn
+++ b/android_webview/BUILD.gn
@@ -172,7 +172,7 @@ repack("repack_100_percent") {
     "//third_party/WebKit/public:scaled_resources_100_percent",
     "//ui/resources",
   ]
-  output = "$target_gen_dir/chrome_100_percent.pak"
+  output = "$target_gen_dir/vivaldi_100_percent.pak"
 
   if (enable_resource_whitelist_generation) {
     repack_whitelist = system_webview_pak_whitelist
@@ -182,7 +182,7 @@ repack("repack_100_percent") {
 
 android_assets("pak_file_assets") {
   sources = [
-    "$target_gen_dir/chrome_100_percent.pak",
+    "$target_gen_dir/vivaldi_100_percent.pak",
     "$target_gen_dir/resources.pak",
   ]
   deps = [
diff --git a/apps/launcher.cc b/apps/launcher.cc
index 6623473fb4ab6e9c777f35b0a0fd7f058cfb0d90..fe823b9935806f9215f62d879a72e6bc34459b09 100644
--- a/apps/launcher.cc
+++ b/apps/launcher.cc
@@ -44,6 +44,9 @@
 #include "net/base/filename_util.h"
 #include "url/gurl.h"
 
+#include "app/vivaldi_apptools.h"
+#include "components/url_formatter/url_fixer.h"
+
 #if defined(OS_CHROMEOS)
 #include "components/user_manager/user_manager.h"
 #endif
@@ -393,6 +396,27 @@ void LaunchPlatformAppWithCommandLineAndLaunchId(
     }
   }
 
+  if (vivaldi::IsVivaldiRunning()) {
+    if (command_line.GetArgs().size() > 0) {
+      // To simplistic fixing of the supplied URL.  For more complete
+      // fixing see: StartupBrowserCreator::GetURLsFromCommandLine()
+      GURL url;
+      {
+        base::ThreadRestrictions::ScopedAllowIO allow_io;
+        url = url_formatter::FixupRelativeFile(
+            current_directory, base::FilePath(command_line.GetArgs()[0]));
+      }
+      //GURL url = GURL(command_line.GetArgs()[0]);
+      AppRuntimeEventRouter::DispatchOnLaunchedEventWithUrl(
+        context, app,
+        "None", //const std::string& handler_id,
+        url, //const GURL& url,
+        GURL() //const GURL& referrer_url);
+        );
+      return;
+    }
+  }
+
 #if defined(OS_WIN)
   base::CommandLine::StringType about_blank_url(
       base::ASCIIToUTF16(url::kAboutBlankURL));
diff --git a/base/base_paths_mac.mm b/base/base_paths_mac.mm
index 9576237110c83599f0f4f4b30cf6b5ff07ff0c09..16d61b7f4d5604687b35020edb14673f5c193ed1 100644
--- a/base/base_paths_mac.mm
+++ b/base/base_paths_mac.mm
@@ -99,6 +99,9 @@ bool PathProviderMac(int key, base::FilePath* result) {
         // src/xcodebuild/{Debug|Release}/base_unittests
         *result = result->DirName().DirName();
       }
+#endif
+#if defined(VIVALDI_BUILD)
+      *result = result->Append(FILE_PATH_LITERAL("chromium"));
 #endif
       return true;
     case base::DIR_USER_DESKTOP:
diff --git a/base/base_paths_posix.cc b/base/base_paths_posix.cc
index 00a15696cacda65ab7e22cf1571455048f0fb6e7..238195aa5d5c994cd17e724431014ed2c8faf7cb 100644
--- a/base/base_paths_posix.cc
+++ b/base/base_paths_posix.cc
@@ -94,7 +94,11 @@ bool PathProviderPosix(int key, FilePath* result) {
       // On POSIX, unit tests execute two levels deep from the source root.
       // For example:  out/{Debug|Release}/net_unittest
       if (PathService::Get(DIR_EXE, &path)) {
-        *result = path.DirName().DirName();
+        *result = path.DirName().DirName()
+#if defined(VIVALDI_BUILD)
+            .Append(FILE_PATH_LITERAL("chromium"))
+#endif
+          ;
         return true;
       }
 
diff --git a/base/base_paths_win.cc b/base/base_paths_win.cc
index 4029a49f96ccafc4efed0253fcb8b4efabb4ed98..a92baa6933e93f8d365fd899a07dbb527c53d56b 100644
--- a/base/base_paths_win.cc
+++ b/base/base_paths_win.cc
@@ -129,7 +129,11 @@ bool PathProviderWin(int key, FilePath* result) {
       // On Windows, unit tests execute two levels deep from the source root.
       // For example:  chrome/{Debug|Release}/ui_tests.exe
       PathService::Get(base::DIR_EXE, &executableDir);
-      cur = executableDir.DirName().DirName();
+      cur = executableDir.DirName().DirName()
+#if defined(VIVALDI_BUILD)
+            .Append(FILE_PATH_LITERAL("chromium"))
+#endif
+            ;
       break;
     }
     case base::DIR_APP_SHORTCUTS: {
diff --git a/base/files/file_util_posix.cc b/base/files/file_util_posix.cc
index d2eb4531d8ea82fdb18729cd907f532ec2a69c10..6e58cd705e1474c6cb21f5e55477c9a27daee860 100644
--- a/base/files/file_util_posix.cc
+++ b/base/files/file_util_posix.cc
@@ -135,7 +135,9 @@ std::string TempFileName() {
   return StringPrintf(".%s.XXXXXX", base::mac::BaseBundleID());
 #endif
 
-#if defined(GOOGLE_CHROME_BUILD)
+#if defined(VIVALDI_BUILD)
+  return std::string(".com.vivaldi.Vivaldi.XXXXXX");
+#elif defined(GOOGLE_CHROME_BUILD)
   return std::string(".com.google.Chrome.XXXXXX");
 #else
   return std::string(".org.chromium.Chromium.XXXXXX");
diff --git a/base/mac/foundation_util.mm b/base/mac/foundation_util.mm
index eb8284ee58fef43df064d3365dd1b0e8a65c9f49..73573df4029b9f51baf96ddfdcff5db050a4a212 100644
--- a/base/mac/foundation_util.mm
+++ b/base/mac/foundation_util.mm
@@ -247,7 +247,9 @@ const char* BaseBundleID() {
     return base_bundle_id;
   }
 
-#if defined(GOOGLE_CHROME_BUILD)
+#if defined(VIVALDI_BUILD)
+  return "com.vivaldi.Vivaldi";
+#elif defined(GOOGLE_CHROME_BUILD)
   return "com.google.Chrome";
 #else
   return "org.chromium.Chromium";
diff --git a/base/memory/shared_memory_posix.cc b/base/memory/shared_memory_posix.cc
index d3163e5a6406027b84e4893b3791db571c812b84..b378e40abb9514b41b06f27e27e1363669b2530b 100644
--- a/base/memory/shared_memory_posix.cc
+++ b/base/memory/shared_memory_posix.cc
@@ -367,7 +367,9 @@ bool SharedMemory::FilePathForMemoryName(const std::string& mem_name,
   if (!GetShmemTempDir(false, &temp_dir))
     return false;
 
-#if defined(GOOGLE_CHROME_BUILD)
+#if defined(VIVALDI_BUILD)
+  static const char kShmem[] = "com.vivaldi.Vivaldi";
+#elif defined(GOOGLE_CHROME_BUILD)
   static const char kShmem[] = "com.google.Chrome.shmem.";
 #else
   static const char kShmem[] = "org.chromium.Chromium.shmem.";
diff --git a/base/path_service.cc b/base/path_service.cc
index 6ac501eafeeb464d4592e2815b2c4e5a34b4df82..a40990b49f1e5dcedde24460c748ca5f7a97feb7 100644
--- a/base/path_service.cc
+++ b/base/path_service.cc
@@ -18,6 +18,8 @@
 #include "base/synchronization/lock.h"
 #include "build/build_config.h"
 
+#include "base/threading/thread_restrictions.h"
+
 namespace base {
 
 bool PathProvider(int key, FilePath* result);
@@ -214,8 +216,13 @@ bool PathService::Get(int key, FilePath* result) {
     return false;
 
   if (path.ReferencesParent()) {
+    // Lift the thread's disallowance of blocking IO, if necessary,
+    // just for the absolute path operation
+    bool last_ioallow = base::ThreadRestrictions::SetIOAllowed(true);
     // Make sure path service never returns a path with ".." in it.
     path = MakeAbsoluteFilePath(path);
+    // Turn it back on, if it was on.
+    base::ThreadRestrictions::SetIOAllowed(last_ioallow);
     if (path.empty())
       return false;
   }
diff --git a/base/process/process_win.cc b/base/process/process_win.cc
index 10a67f3026d054f017be558899141d7b6ef58651..1f28be2bafa0567d8006ddaefb4de4c014e27d04 100644
--- a/base/process/process_win.cc
+++ b/base/process/process_win.cc
@@ -68,7 +68,8 @@ Process Process::OpenWithAccess(ProcessId pid, DWORD desired_access) {
 
 // static
 Process Process::DeprecatedGetProcessFromHandle(ProcessHandle handle) {
-  DCHECK_NE(handle, ::GetCurrentProcess());
+  // andre@vivaldi.com ; this was hitting in --single-process mode
+  //  DCHECK_NE(handle, ::GetCurrentProcess());
   ProcessHandle out_handle;
   if (!::DuplicateHandle(GetCurrentProcess(), handle,
                          GetCurrentProcess(), &out_handle,
diff --git a/base/test/launcher/unit_test_launcher.cc b/base/test/launcher/unit_test_launcher.cc
index 17be5293aea6f4b5c64a42b34d59205aed3d10d1..ed0b4e21120f0ca581ecf289cf09f8ef1090352e 100644
--- a/base/test/launcher/unit_test_launcher.cc
+++ b/base/test/launcher/unit_test_launcher.cc
@@ -40,6 +40,9 @@
 #include "base/files/file_descriptor_watcher_posix.h"
 #endif
 
+#include "app/vivaldi_apptools.h"
+#include "base/vivaldi_switches.h"
+
 namespace base {
 
 namespace {
@@ -191,6 +194,9 @@ int LaunchUnitTestsInternal(const RunTestSuiteCallback& run_test_suite,
                             int default_batch_limit,
                             bool use_job_objects,
                             const Closure& gtest_init) {
+  vivaldi::CommandLineAppendSwitchNoDup(CommandLine::ForCurrentProcess(),
+                                        switches::kDisableVivaldi);
+
 #if defined(OS_ANDROID)
   // We can't easily fork on Android, just run the test suite directly.
   return run_test_suite.Run();
diff --git a/base/test/test_suite.cc b/base/test/test_suite.cc
index 15399e6bb7755ed613bb2aaeed1c86302f35c34c..5ca13b2282d899824c6395dc66cfe7fc4aa0fef9 100644
--- a/base/test/test_suite.cc
+++ b/base/test/test_suite.cc
@@ -53,6 +53,8 @@
 #include "base/strings/string_util.h"
 #include "third_party/icu/source/common/unicode/uloc.h"
 #endif
+#else
+#include "base/i18n/rtl.h"
 #endif
 
 #if defined(OS_ANDROID)
@@ -163,6 +165,7 @@ TestSuite::~TestSuite() {
 
 void TestSuite::InitializeFromCommandLine(int argc, char** argv) {
   initialized_command_line_ = CommandLine::Init(argc, argv);
+  CommandLine::ForCurrentProcess()->AppendSwitch("--disable-vivaldi");;
   testing::InitGoogleTest(&argc, argv);
   testing::InitGoogleMock(&argc, argv);
 
@@ -175,6 +178,7 @@ void TestSuite::InitializeFromCommandLine(int argc, char** argv) {
 void TestSuite::InitializeFromCommandLine(int argc, wchar_t** argv) {
   // Windows CommandLine::Init ignores argv anyway.
   initialized_command_line_ = CommandLine::Init(argc, NULL);
+  CommandLine::ForCurrentProcess()->AppendSwitch("--disable-vivaldi");;
   testing::InitGoogleTest(&argc, argv);
   testing::InitGoogleMock(&argc, argv);
 }
@@ -455,6 +459,8 @@ void TestSuite::Initialize() {
   if (EndsWith(default_locale, "POSIX", CompareCase::INSENSITIVE_ASCII))
     i18n::SetICUDefaultLocale("en_US");
 #endif
+#else
+  i18n::SetICUDefaultLocale("en_US");
 #endif
 
   // Enable SequencedWorkerPool in tests.
diff --git a/base/threading/thread_restrictions.h b/base/threading/thread_restrictions.h
index abfde3aaab1f1bc6cb8f925576b8e049e00cc798..05094b9f74698f416a47ffeb5f3dad1d90759bd3 100644
--- a/base/threading/thread_restrictions.h
+++ b/base/threading/thread_restrictions.h
@@ -73,6 +73,14 @@ class BlockingUrlProtocol;
 namespace midi {
 class TaskService;  // https://crbug.com/796830
 }
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+// IPCAudioDecoder needs to use base::ThreadRestrictions::ScopedAllowWait
+// to mimic a synchronous API for AudioFileReader
+namespace media {
+  class IPCAudioDecoder;
+  class IPCFactory;
+}
+#endif // USE_SYSTEM_PROPRIETARY_CODECS
 namespace mojo {
 class SyncCallRestrictions;
 namespace edk {
@@ -418,6 +426,13 @@ class BASE_EXPORT ThreadRestrictions {
   friend class ui::CommandBufferLocal;
   friend class ui::GpuState;
 
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+  // IPCAudioDecoder needs to use base::ThreadRestrictions::ScopedAllowWait
+  // to mimic a synchronous API for AudioFileReader
+  friend class media::IPCAudioDecoder;
+  friend class media::IPCFactory;
+#endif // USE_SYSTEM_PROPRIETARY_CODECS
+
   // END ALLOWED USAGE.
   // BEGIN USAGE THAT NEEDS TO BE FIXED.
   friend class ::chromeos::BlockingMethodCaller;  // http://crbug.com/125360
diff --git a/cc/tiles/software_image_decode_cache.cc b/cc/tiles/software_image_decode_cache.cc
index 765d94a68348769abd864bce01535ce063aab127..44b74da7427d23f5d2da7ced6f44949acab18b36 100644
--- a/cc/tiles/software_image_decode_cache.cc
+++ b/cc/tiles/software_image_decode_cache.cc
@@ -28,7 +28,7 @@ namespace {
 // if more items are locked. That is, locked items ignore this limit.
 // Depending on the memory state of the system, we limit the amount of items
 // differently.
-const size_t kNormalMaxItemsInCacheForSoftware = 1000;
+const size_t kNormalMaxItemsInCacheForSoftware = 200;
 const size_t kThrottledMaxItemsInCacheForSoftware = 100;
 const size_t kSuspendedMaxItemsInCacheForSoftware = 0;
 
diff --git a/chrome_elf/chrome_elf.ver b/chrome_elf/chrome_elf.ver
index 78f890e7ece7dc50b0ce3bb7784b88bc53bd7792..4c3c51d2be91a116f2df34ac9778c65af9a0b025 100644
--- a/chrome_elf/chrome_elf.ver
+++ b/chrome_elf/chrome_elf.ver
@@ -1,2 +1,2 @@
 INTERNAL_NAME=chrome_elf_dll
-ORIGINAL_FILENAME=chrome_elf.dll
+ORIGINAL_FILENAME=vivaldi_elf.dll
diff --git a/chrome_elf/chrome_elf_x64.def b/chrome_elf/chrome_elf_x64.def
index 690cfae2d01e9b46dbc175cfa2b74deb0cf32456..8bf38d9a89b08fe9df90954df9aa57eb5358f18b 100644
--- a/chrome_elf/chrome_elf_x64.def
+++ b/chrome_elf/chrome_elf_x64.def
@@ -1,7 +1,7 @@
 ; Copyright 2013 The Chromium Authors. All rights reserved.
 ; Use of this source code is governed by a BSD-style license that can be
 ; found in the LICENSE file.
-LIBRARY  "chrome_elf.dll"
+LIBRARY  "vivaldi_elf.dll"
 
 EXPORTS
   ; When functions are added to this file, they must also be added to
diff --git a/chrome_elf/chrome_elf_x86.def b/chrome_elf/chrome_elf_x86.def
index 7a59f7e0146c6fb4ca4ed46d5ea6fe97b5bcaf65..29b4d76d70daf8c2912a7d95d8b7de02ae8c1add 100644
--- a/chrome_elf/chrome_elf_x86.def
+++ b/chrome_elf/chrome_elf_x86.def
@@ -1,7 +1,7 @@
 ; Copyright 2013 The Chromium Authors. All rights reserved.
 ; Use of this source code is governed by a BSD-style license that can be
 ; found in the LICENSE file.
-LIBRARY  "chrome_elf.dll"
+LIBRARY  "vivaldi_elf.dll"
 
 EXPORTS
   ; When functions are added to this file, they must also be added to
diff --git a/chrome_elf/elf_imports_unittest.cc b/chrome_elf/elf_imports_unittest.cc
index dfcb9b9fddda68da516aa874fc5373df8908490e..422598827e50626bec82a935c2946c3ff1d173a8 100644
--- a/chrome_elf/elf_imports_unittest.cc
+++ b/chrome_elf/elf_imports_unittest.cc
@@ -74,7 +74,7 @@ class ELFImportsTest : public testing::Test {
 TEST_F(ELFImportsTest, ChromeElfSanityCheck) {
   base::FilePath dll;
   ASSERT_TRUE(PathService::Get(base::DIR_EXE, &dll));
-  dll = dll.Append(L"chrome_elf.dll");
+  dll = dll.Append(L"vivaldi_elf.dll");
 
   std::vector<std::string> elf_imports;
   GetImports(dll, &elf_imports);
@@ -168,7 +168,7 @@ TEST_F(ELFImportsTest, ChromeExeSanityCheck) {
 
   base::FilePath exe;
   ASSERT_TRUE(PathService::Get(base::DIR_EXE, &exe));
-  exe = exe.Append(L"chrome.exe");
+  exe = exe.Append(L"vivaldi.exe");
   GetImports(exe, &exe_imports);
 
   // Check that chrome.exe has imports.
@@ -177,7 +177,7 @@ TEST_F(ELFImportsTest, ChromeExeSanityCheck) {
          "target was built, instead of chrome_elf_import_unittests.exe";
 
   // Chrome.exe's first import must be ELF.
-  EXPECT_EQ("chrome_elf.dll", exe_imports[0])
+  EXPECT_EQ("vivaldi_elf.dll", exe_imports[0]) <<
       << "Illegal import order in chrome.exe (ensure the "
          "chrome_elf_import_unittest "
          "target was built, instead of just chrome_elf_import_unittests.exe)";
diff --git a/device/geolocation/network_location_request.cc b/device/geolocation/network_location_request.cc
index a9a739e4f9a707faaf195f0747b560197b97a1be..0d716d1f0792740c806b6b5db89fed87c5e09ba1 100644
--- a/device/geolocation/network_location_request.cc
+++ b/device/geolocation/network_location_request.cc
@@ -33,7 +33,7 @@ namespace device {
 namespace {
 
 const char kNetworkLocationBaseUrl[] =
-    "https://www.googleapis.com/geolocation/v1/geolocate";
+    "https://location.services.mozilla.com/v1/geolocate";
 
 const char kLocationString[] = "location";
 const char kLatitudeString[] = "lat";
@@ -112,7 +112,10 @@ NetworkLocationRequest::NetworkLocationRequest(
     const std::string& api_key,
     LocationResponseCallback callback)
     : url_context_(std::move(context)),
-      api_key_(api_key),
+      api_key_(
+          // Use Mozilla Location Services API key.
+          "c18ef960-6b92-4abc-91b8-fc7e954062ee"
+        ),
       location_response_callback_(callback) {}
 
 NetworkLocationRequest::~NetworkLocationRequest() = default;
diff --git a/google_apis/gaia/gaia_auth_util.cc b/google_apis/gaia/gaia_auth_util.cc
index 537016deb66538ba10b75cf1caec9c01b5eb9faf..1651c761f214c88be0e1816abd07f73004680483 100644
--- a/google_apis/gaia/gaia_auth_util.cc
+++ b/google_apis/gaia/gaia_auth_util.cc
@@ -21,6 +21,8 @@
 #include "net/url_request/url_request.h"
 #include "url/gurl.h"
 
+#include "app/vivaldi_apptools.h"
+
 namespace gaia {
 
 namespace {
@@ -32,6 +34,8 @@ const void* const kURLRequestUserDataKey = &kURLRequestUserDataKey;
 
 std::string CanonicalizeEmailImpl(const std::string& email_address,
                                   bool change_googlemail_to_gmail) {
+  if(vivaldi::IsVivaldiRunning())
+    return email_address;
   std::vector<std::string> parts = base::SplitString(
       email_address, "@", base::TRIM_WHITESPACE, base::SPLIT_WANT_ALL);
   if (parts.size() != 2U) {
diff --git a/google_apis/gaia/oauth2_token_service.h b/google_apis/gaia/oauth2_token_service.h
index f1e0f88107c3ccf742bd23f5baa9200201ceeac6..b00b3b3db7d1ed67e75aec7cd9f7858bcfad32df 100644
--- a/google_apis/gaia/oauth2_token_service.h
+++ b/google_apis/gaia/oauth2_token_service.h
@@ -180,11 +180,11 @@ class OAuth2TokenService {
 
   // Lists account IDs of all accounts with a refresh token maintained by this
   // instance.
-  std::vector<std::string> GetAccounts() const;
+  virtual std::vector<std::string> GetAccounts() const;
 
   // Returns true if a refresh token exists for |account_id|. If false, calls to
   // |StartRequest| will result in a Consumer::OnGetTokenFailure callback.
-  bool RefreshTokenIsAvailable(const std::string& account_id) const;
+  virtual bool RefreshTokenIsAvailable(const std::string& account_id) const;
 
   // Returns true if a refresh token exists for |account_id| and it is in an
   // error state.
@@ -321,10 +321,12 @@ class OAuth2TokenService {
     ScopeSet scopes;
   };
 
+ protected:
   // Provide a request context used for fetching access tokens with the
   // |StartRequest| method.
-  net::URLRequestContextGetter* GetRequestContext() const;
+  virtual net::URLRequestContextGetter* GetRequestContext() const;
 
+ private:
   // Struct that contains the information of an OAuth2 access token.
   struct CacheEntry {
     std::string access_token;
diff --git a/ipc/ipc_channel.h b/ipc/ipc_channel.h
index ec7b7c963c6c1b1e3cd97d326ece57877287114b..acc8feaaf592cc684eb3fd647c9e7d680d7c63e3 100644
--- a/ipc/ipc_channel.h
+++ b/ipc/ipc_channel.h
@@ -140,7 +140,7 @@ class IPC_EXPORT Channel : public Sender {
 
   // The maximum message size in bytes. Attempting to receive a message of this
   // size or bigger results in a channel error.
-  static constexpr size_t kMaximumMessageSize = 128 * 1024 * 1024;
+  static constexpr size_t kMaximumMessageSize = 256 * 1024 * 1024;
 
   // Amount of data to read at once from the pipe.
   static const size_t kReadBufferSize = 4 * 1024;
diff --git a/ipc/ipc_message_start.h b/ipc/ipc_message_start.h
index e24a858ed0ba340ba7a3ada0b20aaefb490ae2fb..7291afa727e1613688b146e2195613a630dc45ea 100644
--- a/ipc/ipc_message_start.h
+++ b/ipc/ipc_message_start.h
@@ -55,6 +55,9 @@ enum IPCMessageStart {
   WebRtcLoggingMsgStart,
   TtsMsgStart,
   NaClHostMsgStart,
+#if defined(USE_SYSTEM_PROPRIETARY_CODECS)
+  MediaPipelineMsgStart,
+#endif  // defined(USE_SYSTEM_PROPRIETARY_CODECS)
   EncryptedMediaMsgStart,
   CacheStorageMsgStart,
   ServiceWorkerMsgStart,
@@ -84,6 +87,7 @@ enum IPCMessageStart {
   SurfaceViewManagerMsgStart,
   ExtensionWorkerMsgStart,
   SubresourceFilterMsgStart,
+  VivaldiMsgStart,
   LastIPCMsgStart  // Must come last.
 };
 
diff --git a/ipc/ipc_message_utils.cc b/ipc/ipc_message_utils.cc
index ebadd934f9250574533763c8a7bfc84e4c44308a..b092460c2fff8d941028e320a543a61e3cf4b216 100644
--- a/ipc/ipc_message_utils.cc
+++ b/ipc/ipc_message_utils.cc
@@ -344,7 +344,8 @@ void ParamTraits<unsigned int>::Log(const param_type& p, std::string* l) {
 }
 
 #if defined(OS_WIN) || defined(OS_LINUX) || defined(OS_FUCHSIA) || \
-    (defined(OS_ANDROID) && defined(ARCH_CPU_64_BITS))
+    (defined(OS_ANDROID) && defined(ARCH_CPU_64_BITS)) || \
+    defined(USE_SYSTEM_PROPRIETARY_CODECS)
 void ParamTraits<long>::Log(const param_type& p, std::string* l) {
   l->append(base::NumberToString(p));
 }
diff --git a/ipc/ipc_message_utils.h b/ipc/ipc_message_utils.h
index 54fee405190b5ce9ae147722f642af751aba4bf6..ed4824f02f507443fbfde660b664da7fbfe5a060 100644
--- a/ipc/ipc_message_utils.h
+++ b/ipc/ipc_message_utils.h
@@ -196,7 +196,8 @@ struct ParamTraits<unsigned int> {
 // Since we want to support Android 32<>64 bit IPC, as long as we don't have
 // these traits for 32 bit ARM then that'll catch any errors.
 #if defined(OS_WIN) || defined(OS_LINUX) || defined(OS_FUCHSIA) || \
-    (defined(OS_ANDROID) && defined(ARCH_CPU_64_BITS))
+    (defined(OS_ANDROID) && defined(ARCH_CPU_64_BITS)) || \
+    defined(USE_SYSTEM_PROPRIETARY_CODECS)
 template <>
 struct ParamTraits<long> {
   typedef long param_type;
diff --git a/ipc/ipc_mojo_bootstrap.cc b/ipc/ipc_mojo_bootstrap.cc
index 8248b6025c9d627fb8cafd9f8a06029e0183e76a..6b0fd3b85af738eba2e75fd7219073ad228cd5ca 100644
--- a/ipc/ipc_mojo_bootstrap.cc
+++ b/ipc/ipc_mojo_bootstrap.cc
@@ -123,9 +123,11 @@ class ChannelAssociatedGroupController
 
   void ShutDown() {
     DCHECK(thread_checker_.CalledOnValidThread());
+    shut_down_ = true;
     connector_->CloseMessagePipe();
     OnPipeError();
     connector_.reset();
+    outgoing_messages_.clear();
   }
 
   // mojo::AssociatedGroupController:
@@ -585,7 +587,8 @@ class ChannelAssociatedGroupController
     if (task_runner_->BelongsToCurrentThread()) {
       DCHECK(thread_checker_.CalledOnValidThread());
       if (!connector_ || paused_) {
-        outgoing_messages_.emplace_back(std::move(*message));
+        if (!shut_down_)
+          outgoing_messages_.emplace_back(std::move(*message));
         return true;
       }
       return connector_->Accept(message);
@@ -869,6 +872,7 @@ class ChannelAssociatedGroupController
   base::Lock lock_;
 
   bool encountered_error_ = false;
+  bool shut_down_ = false;
 
   // ID #1 is reserved for the mojom::Channel interface.
   uint32_t next_interface_id_ = 2;
diff --git a/jingle/notifier/base/notifier_options_util.cc b/jingle/notifier/base/notifier_options_util.cc
index 9754b2c7be8043cfddfdf6e6d80b941d7f9a7bb8..fb2fe153bfacfbdf41f70665575d555cadc179b0 100644
--- a/jingle/notifier/base/notifier_options_util.cc
+++ b/jingle/notifier/base/notifier_options_util.cc
@@ -10,6 +10,8 @@
 #include "third_party/libjingle_xmpp/xmpp/constants.h"
 #include "third_party/libjingle_xmpp/xmpp/jid.h"
 
+#include "sync/vivaldi_sync_urls.h"
+
 namespace notifier {
 
 buzz::XmppClientSettings MakeXmppClientSettings(
@@ -51,14 +53,14 @@ ServerList GetServerList(
     // The default servers support SSLTCP.
     servers.push_back(
         ServerInformation(
-            net::HostPortPair("talk.google.com",
+            net::HostPortPair(SYNC_NOTIFIER_SERVER1,
                               notifier::kDefaultXmppPort),
-            SUPPORTS_SSLTCP));
+            SYNC_SECURE_NOTIFIER_SERVER));
     servers.push_back(
         ServerInformation(
-            net::HostPortPair("talkx.l.google.com",
+            net::HostPortPair(SYNC_NOTIFIER_SERVER2,
                               notifier::kDefaultXmppPort),
-            SUPPORTS_SSLTCP));
+            SYNC_SECURE_NOTIFIER_SERVER));
   }
   return servers;
 }
diff --git a/pdf/pdfium/pdfium_engine.cc b/pdf/pdfium/pdfium_engine.cc
index cf056745b335918b3d7f3f540f62ee1af4caed7e..32ddfc20e4f4c114b53732b537419ce4f30d24e9 100644
--- a/pdf/pdfium/pdfium_engine.cc
+++ b/pdf/pdfium/pdfium_engine.cc
@@ -3706,7 +3706,10 @@ void PDFiumEngine::DeviceToPage(int page_index,
 }
 
 int PDFiumEngine::GetVisiblePageIndex(FPDF_PAGE page) {
-  for (int page_index : visible_pages_) {
+  // Copy visible_pages_ since it can change as a result of loading the page in
+  // GetPage(). See https://crbug.com/822091.
+  std::vector<int> visible_pages_copy(visible_pages_);
+  for (int page_index : visible_pages_copy) {
     if (pages_[page_index]->GetPage() == page)
       return page_index;
   }
diff --git a/ppapi/shared_impl/media_stream_buffer_manager_unittest.cc b/ppapi/shared_impl/media_stream_buffer_manager_unittest.cc
index 97cf90a17b42db364b10bfd7ca2cddda5661d8c8..9e988ae40d4336203d75736951532998fb2a7a04 100644
--- a/ppapi/shared_impl/media_stream_buffer_manager_unittest.cc
+++ b/ppapi/shared_impl/media_stream_buffer_manager_unittest.cc
@@ -87,6 +87,8 @@ TEST(MediaStreamBufferManager, General) {
     EXPECT_EQ(NULL, manager.GetBufferPointer(-1));
     EXPECT_EQ(NULL, manager.GetBufferPointer(kNumberOfBuffers));
 
+    // Disabling death test for Official Vivaldi builds, since the test fails for some reason
+#if !(defined(OFFICIAL_BUILD) && defined(VIVALDI_BUILD))
     // Test crash for passing invalid index to EnqueueBuffer().
     // String arguments aren't passed to CHECK() in official builds.
 #if defined(OFFICIAL_BUILD) && defined(NDEBUG)
@@ -96,6 +98,7 @@ TEST(MediaStreamBufferManager, General) {
     EXPECT_DEATH(manager.EnqueueBuffer(-1), ".*Check failed: index >= 0.*");
     EXPECT_DEATH(manager.EnqueueBuffer(kNumberOfBuffers),
                  ".*Check failed: index < number_of_buffers_.*");
+#endif
 #endif
   }
 
diff --git a/rlz/win/lib/rlz_value_store_registry.cc b/rlz/win/lib/rlz_value_store_registry.cc
index d8b93a5f7322fa33f0f63d3939e71a66fe59f800..0157cd0f9eea5fc4dae1849b36934560b326c0fd 100644
--- a/rlz/win/lib/rlz_value_store_registry.cc
+++ b/rlz/win/lib/rlz_value_store_registry.cc
@@ -39,9 +39,9 @@ namespace {
 //
 // The server does not care about any of these constants.
 //
-const char kLibKeyName[]               = "Software\\Google\\Common\\Rlz";
-const wchar_t kGoogleKeyName[]         = L"Software\\Google";
-const wchar_t kGoogleCommonKeyName[]   = L"Software\\Google\\Common";
+const char kLibKeyName[]               = "Software\\Vivaldi\\Common\\Rlz";
+const wchar_t kGoogleKeyName[]         = L"Software\\Vivaldi";
+const wchar_t kGoogleCommonKeyName[]   = L"Software\\Vivaldi\\Common";
 const char kRlzsSubkeyName[]           = "RLZs";
 const char kEventsSubkeyName[]         = "Events";
 const char kStatefulEventsSubkeyName[] = "StatefulEvents";
diff --git a/sandbox/linux/suid/client/setuid_sandbox_host.cc b/sandbox/linux/suid/client/setuid_sandbox_host.cc
index 89836aaabb3dfc072cacb4c7edd62eced59e730c..5e6bc1c2a77f6ce7d1a6c7f8abb498b7933267fe 100644
--- a/sandbox/linux/suid/client/setuid_sandbox_host.cc
+++ b/sandbox/linux/suid/client/setuid_sandbox_host.cc
@@ -121,7 +121,7 @@ base::FilePath SetuidSandboxHost::GetSandboxBinaryPath() {
   base::FilePath sandbox_binary;
   base::FilePath exe_dir;
   if (PathService::Get(base::DIR_EXE, &exe_dir)) {
-    base::FilePath sandbox_candidate = exe_dir.AppendASCII("chrome-sandbox");
+    base::FilePath sandbox_candidate = exe_dir.AppendASCII("vivaldi-sandbox");
     if (base::PathExists(sandbox_candidate))
       sandbox_binary = sandbox_candidate;
   }
diff --git a/testing/buildbot/filters/mojo.fyi.network_browser_tests.filter b/testing/buildbot/filters/mojo.fyi.network_browser_tests.filter
index 669d00a7988c17643ae716c68d984d19ef0eab4b..c3a730bb0f4e7d5cb8ed30835f5357f3acfc9364 100644
--- a/testing/buildbot/filters/mojo.fyi.network_browser_tests.filter
+++ b/testing/buildbot/filters/mojo.fyi.network_browser_tests.filter
@@ -539,15 +539,6 @@
 
 # Finish ServiceWorker networking glue rewrite.
 # http://crbug.com/715640
--ChromeServiceWorkerFetchPPAPIPrivateTest.OtherOrigin
--ChromeServiceWorkerFetchPPAPIPrivateTest.OtherOriginCORS
--ChromeServiceWorkerFetchPPAPIPrivateTest.OtherOriginCORSCredentials
--ChromeServiceWorkerFetchPPAPIPrivateTest.OtherOriginCredentials
--ChromeServiceWorkerFetchPPAPIPrivateTest.SameOrigin
--ChromeServiceWorkerFetchPPAPIPrivateTest.SameOriginCORS
--ChromeServiceWorkerFetchPPAPIPrivateTest.SameOriginCORSCredentials
--ChromeServiceWorkerFetchPPAPIPrivateTest.SameOriginCredentials
--ChromeServiceWorkerFetchPPAPITest.SameOrigin
 -ServiceWorkerAppTest.RegisterAndPostMessage
 -ServiceWorkerBackgroundSyncTestWithJSBindings/ServiceWorkerBackgroundSyncTest.Sync/0
 -ServiceWorkerBackgroundSyncTestWithNativeBindings/ServiceWorkerBackgroundSyncTest.Sync/0
@@ -565,6 +556,7 @@
 -ServiceWorkerTestWithJSBindings/ServiceWorkerTest.EventsToStoppedWorker/0
 -ServiceWorkerTestWithJSBindings/ServiceWorkerTest.FetchArbitraryPaths/0
 -ServiceWorkerTestWithJSBindings/ServiceWorkerTest.FilteredEvents/0
+-ServiceWorkerTestWithJSBindings/ServiceWorkerTest.MimeHandlerView/0
 -ServiceWorkerTestWithJSBindings/ServiceWorkerTest.GetBackgroundClientFailsWithNoBackgroundPage/0
 -ServiceWorkerTestWithJSBindings/ServiceWorkerTest.LoadingBackgroundPageBypassesServiceWorker/0
 -ServiceWorkerTestWithJSBindings/ServiceWorkerTest.NotificationAPI/0
@@ -583,6 +575,7 @@
 -ServiceWorkerTestWithNativeBindings/ServiceWorkerTest.EventsToStoppedWorker/0
 -ServiceWorkerTestWithNativeBindings/ServiceWorkerTest.FetchArbitraryPaths/0
 -ServiceWorkerTestWithNativeBindings/ServiceWorkerTest.FilteredEvents/0
+-ServiceWorkerTestWithNativeBindings/ServiceWorkerTest.MimeHandlerView/0
 -ServiceWorkerTestWithNativeBindings/ServiceWorkerTest.GetBackgroundClientFailsWithNoBackgroundPage/0
 -ServiceWorkerTestWithNativeBindings/ServiceWorkerTest.LoadingBackgroundPageBypassesServiceWorker/0
 -ServiceWorkerTestWithNativeBindings/ServiceWorkerTest.NotificationAPI/0
